// @cloud/tiny3/esm2015/components/loading/TiLoadingComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * Loading组件用于加载场景，提供了两种类型
 *
 * 10.0.3新增
 *
 * <example-url>../tiny3demo/#/loading/loading-all</example-url>
 */
export class TiLoadingComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * @ignore
         * 弱类型有4个圈
         */
        this.items = [];
        this.type = 'default'; // 加载类型
        this.size = 'small'; // 加载尺寸
    }
    // tslint:disable-next-line:use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        this.items.length = 4;
    }
    /**
     * @ignore
     * 根据用户传入尺寸获取border的宽度
     */
    getBorderWidth(size) {
        return parseInt(size, 10) * 0.3 / 2 + 'px';
    }
}
TiLoadingComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-loading',
                template: "<div *ngIf=\"type === 'default'\"\r\n    class=\"ti3-loading-default\"\r\n    [ngStyle]=\"{'width': size, 'height': size, 'border-width': getBorderWidth(size)}\"\r\n    [ngClass]=\"{'ti3-loading-default-small': size === 'small',\r\n                'ti3-loading-default-middle': size === 'middle',\r\n                'ti3-loading-default-large': size === 'large'}\"></div>\r\n\r\n<div  *ngIf=\"type === 'weak'\">\r\n    <span *ngFor=\"let item of items\" class=\"ti3-loading-weak\"></span>\r\n</div>",
                styles: ["::ng-deep :root{--ti-loading-default-large-size:32px;--ti-loading-default-middle-size:24px;--ti-loading-default-small-size:12px;--ti-loading-weak-size:5px}.ti3-loading-default{animation:default-animation .8s linear 0s infinite;border-color:var(--ti-common-color-line-active);border-radius:50%;border-right-color:transparent;border-style:solid;box-sizing:border-box;display:inline-block}.ti3-loading-default.ti3-loading-default-small{border-width:calc(var(--ti-loading-default-small-size)*0.3/2);height:var(--ti-loading-default-small-size);width:var(--ti-loading-default-small-size)}.ti3-loading-default.ti3-loading-default-middle{border-width:calc(var(--ti-loading-default-middle-size)*0.3/2);height:var(--ti-loading-default-middle-size);width:var(--ti-loading-default-middle-size)}.ti3-loading-default.ti3-loading-default-large{border-width:calc(var(--ti-loading-default-large-size)*0.3/2);height:var(--ti-loading-default-large-size);width:var(--ti-loading-default-large-size)}.ti3-loading-weak{animation:weak-animation 1.2s ease-in-out infinite;background-color:var(--ti-common-color-line-active);border-radius:50%;display:inline-block;height:var(--ti-loading-weak-size);margin-right:var(--ti-common-space-2x);opacity:.7;width:var(--ti-loading-weak-size)}.ti3-loading-weak:first-child{animation-delay:0s}.ti3-loading-weak:nth-child(2){animation-delay:.2s}.ti3-loading-weak:nth-child(3){animation-delay:.4s}.ti3-loading-weak:nth-child(4){animation-delay:.6s}@keyframes default-animation{0%{transform:rotate(0deg)}to{transform:rotate(1turn)}}@keyframes weak-animation{0%{opacity:.2}25%{opacity:.5}50%{opacity:.7}to{opacity:.2}}"]
            },] }
];
TiLoadingComponent.propDecorators = {
    type: [{ type: Input }],
    size: [{ type: Input }]
};
