// @cloud/tiny3/esm2015/services/message/TiMessageService.js
import { __rest } from "tslib";
/**
 * 该类对外暴露，以服务的方式提供方法生成message组件
 * open({
 *  id:message id,string类型,用于生成唯一的按钮可测试id，并且防止弹框多次打开（同modal组件）
 *  type:message类型，支持'prompt' | 'warn' | 'error' | 'confirm'，默认为confirm
 *  title:头部title,string类型，可支持string形式的DOM片段
 *  content:内容区域DOM,支持类型为：TemplateRef<any> | Function | string
 *  context:内容区域上下文，类型为Object
 *  okButton:{
 *    text：string，按钮文本显示
 *    click: Function，点击回调
 *    primary: boolean,默认为true,程序中保证了两个按钮中只有一个可为true
 *    autofocus: boolean,默认为true,程序中保证了两个按钮中只有一个可为true
 *    show: boolean，标识按钮是否显示,默认为true
 * }
 *  cancleBtn:{
 *    text：string，按钮文本显示
 *    click: Function，点击回调
 *    primary: boolean,默认为false,程序中保证了两个按钮中只有一个可为true
 *    autofocus: boolean,默认为false,程序中保证了两个按钮中只有一个可为true
 *    show: boolean，标识按钮是否显示,默认为true
 * }
 * })
 */
import { Injectable } from '@angular/core';
import { TiModalService } from '../modal/TiModalService';
import { TiMessageComponent } from './TiMessageComponent';
import { TiMessageModule } from './TiMessageModule';
import * as i0 from "@angular/core";
import * as i1 from "../modal/TiModalService";
import * as i2 from "./TiMessageModule";
/**
 * 消息框组件，使用服务方式生成，使用该服务时需要引入模块TiMessageModule，通过调用TiMessageService.open(config)方法生成弹出框
 *
 * <p><span style="color: red">使用此组件时需要开发者在项目模块(建议在根模块)
 * 中引入BrowserAnimationsModule。</span>这是因为此组件中使用了Angular动画，需要引入BrowserAnimationsModule，
 * 但是 BrowserAnimationsModule 不能在懒加载模块被重复引入，所以需要开发者来引入BrowserAnimationsModule，保证其引入一次。</p>
 *
 * <example-url>../tiny3demo/#/message/message-all</example-url>
 */
export class TiMessageService {
    constructor(tiModal) {
        this.tiModal = tiModal;
    }
    /**
     * 生成消息弹框方法
     * [config] 弹框配置信息
     * 返回 弹框实例信息，使用该实例信息可调用弹框的close等方法；内容为component类型时，也可通过content属性获取内容结果数据
     */
    open(config) {
        const { type, title, content, context, okButton, cancelButton } = config, modalConfig = __rest(config, ["type", "title", "content", "context", "okButton", "cancelButton"]);
        return this.tiModal.open(TiMessageComponent, Object.assign({ context: {
                id: config.id,
                type,
                title,
                content,
                context,
                okButton,
                cancelButton
            }, modalClass: 'ti3-msg-default-width' }, modalConfig));
    }
}
TiMessageService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TiMessageService_Factory() { return new TiMessageService(i0.ɵɵinject(i1.TiModalService)); }, token: TiMessageService, providedIn: i2.TiMessageModule });
TiMessageService.decorators = [
    { type: Injectable, args: [{
                providedIn: TiMessageModule
            },] }
];
TiMessageService.ctorParameters = () => [
    { type: TiModalService }
];
