// @cloud/tiny3/esm2015/services/message/TiTranscludeDirective.js
import { Directive, ElementRef, Input } from '@angular/core';
import { TiPopupService } from '../popup/TiPopupService';
import { TiContentWrapperComponent } from './TiContentWrapperComponent';
/**
 * @ignore
 * 生成并且compile内容为宿主元素的子元素
 */
export class TiTranscludeDirective {
    constructor(eleRef, popService) {
        this.eleRef = eleRef;
        this.popService = popService;
    }
    ngOnInit() {
        this.wrapperRef = this.popService.create(TiContentWrapperComponent);
        this.wrapperRef.show({
            content: this.tiTransclude.content,
            contentContext: this.tiTransclude.context,
            container: this.eleRef
        });
    }
    ngOnDestroy() {
        this.wrapperRef.hide();
    }
}
TiTranscludeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiTransclude]'
            },] }
];
TiTranscludeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: TiPopupService }
];
TiTranscludeDirective.propDecorators = {
    tiTransclude: [{ type: Input }]
};
