// @cloud/cfdata/service/AccountService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountService = void 0;
var HttpService_1 = require("./HttpService");
var ApiService_1 = require("./ApiService");
var FrameData_1 = require("../FrameData");
var AccountService = /** @class */ (function (_super) {
    __extends(AccountService, _super);
    function AccountService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * 貌似内部外部都没有用
     * Querys besum info
     * @returns
     */
    AccountService.queryBESumInfo = function () {
        return HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/BSS/OpenApi/v1/customers/get-beinfos?excludeUselessBe=1',
            timeout: 10000,
            targetUrl: 'cbc=/BSS/OpenApi/v1/customers/get-beinfos',
        });
    };
    AccountService.queryAccountBalance = function (customerId, beId) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() +
                    '/rest/csb-financial-service/v1/account/balances?customerId={customer_id}&accountType=-1&beId={be_id}',
                o: {
                    customer_id: customerId,
                    be_id: beId,
                },
            },
            timeout: 10000,
            targetUrl: 'cbc=/csb-financial-service/v1/account/balances',
        });
    };
    AccountService.queryRealNameAuth = function (params) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/BSS/OpenApi/v1/extention/customer/customerverified/status',
            },
            params: params,
            timeout: 60000,
            targetUrl: 'cbc=/BSS/OpenApi/v1/extention/customer/customerverified/status',
        });
        return promise;
    };
    /**
     * @internal
     * Querys custome restricted
     * @param params
     * @returns
     */
    AccountService.queryCustomeRestricted = function (params) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/cbc/cbccustmgrservice/v1/customer/restricted',
            },
            params: params,
            timeout: 30000,
            targetUrl: 'cbc=/rest/cbc/cbccustmgrservice/v1/customer/restricted',
        });
        return promise;
    };
    /**
     * 查询账户冻结场景
     * Querys custom frozen
     * @param params
     * @returns
     */
    AccountService.queryCustomFrozen = function (params) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/cbc/cbccustmgrservice/v1/customer/frozen',
            },
            params: params,
            timeout: 30000,
            targetUrl: 'cbc=/rest/cbc/cbccustmgrservice/v1/customer/frozen',
        });
        return promise;
    };
    /**
     * 需要延时调用，因为依赖selectRegionId
     * Querys custom maintenances
     * @returns
     */
    AccountService.queryCustomMaintenances = function (selectRegionId) {
        var regionId = selectRegionId || '';
        if (!regionId)
            return Promise.resolve();
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() +
                    '/rest/silvan/rest/v1.0/customMaintenances/{region_id}/{service_id}?locale={locale}',
                o: {
                    region_id: regionId,
                    service_id: window.location.pathname.split('/')[1],
                    locale: FrameData_1.FrameData.getLanguage() || '',
                },
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/customMaintenances/' +
                regionId +
                '/' +
                window.location.pathname.split('/')[1],
        });
        return promise;
    };
    /**
     * @internal
     * Querys customer manangment
     * @param options
     * @returns
     */
    AccountService.queryCustomerManangment = function (options) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/BSS/OpenApi/v1/customers/queryCusBrief/{userId}',
                o: {
                    userId: options.domainId,
                },
            },
            timeout: 10000,
            targetUrl: 'cbc=/BSS/OpenApi/v1/customers/queryCusBrief/' + options.domainId,
        });
        return promise;
    };
    /**
     * @internal
     * Gets seller and partner
     * @param customerOptions
     * @returns
     */
    AccountService.getSellerAndPartner = function (customerOptions) {
        customerOptions = customerOptions || {};
        var promise = HttpService_1.HttpService.post({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/cbc/csbsrmservice/v1/isv-mgr/isv-info/query',
            params: customerOptions,
            timeout: 10000,
            targetUrl: 'cbc=/rest/cbc/csbsrmservice/v1/isv-mgr/isv-info/query',
        });
        return promise;
    };
    /**
     * @internal
     * Gets alliance account info
     * @param options
     * @returns
     */
    AccountService.getAllianceAccountInfo = function (options) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/cbc/cbccustmgrservice/v1/site/customer/xdomains',
            },
            params: {
                customer_id: options.domainId,
                site_type: options.siteType,
            },
            timeout: 10000,
            targetUrl: 'cbc=/rest/cbc/cbccustmgrservice/v1/site/customer/xdomains',
        });
        return promise;
    };
    /**
     * @intternal
     * Gets samlresponse
     * @param url
     * @returns
     */
    AccountService.getSAMLResponse = function (url) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, HttpService_1.HttpService.get({ url: url, timeout: 30000 })];
            });
        });
    };
    AccountService.getOCBToken = function (samlResponse, url) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, HttpService_1.HttpService.post({
                        url: url,
                        timeout: 30000,
                        params: { SAMLResponse: samlResponse, 'X-Idp-Id': '@ALLY_HUAWEICLOUD' },
                        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
                    })];
            });
        });
    };
    /**
     * @internal
     * Gets user waiting renew num
     * @returns
     */
    AccountService.getUserWaitingRenewNum = function () {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/csb/csbordersubscriptionservice/v2/subscriptions/waitingrenew',
            },
            timeout: 60000,
            targetUrl: 'cbc=/rest/csb/csbordersubscriptionservice/v2/subscriptions/waitingrenew',
        });
        return promise;
    };
    AccountService.getFederatedUserInfo = function (domainId) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/iam/v3-ext/domain/{domain_id}',
                o: {
                    domain_id: domainId,
                },
            },
            mask: true,
            timeout: 10000,
            targetUrl: 'iam=/v3-ext/domain/' + domainId,
        });
    };
    /**
     * @internal
     * Gets cbc resource statistic
     * @param options
     * @returns
     */
    AccountService.getCbcResourceStatistic = function (options) {
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/cbc/csbsubscriptionservice/v1/statistic/{customerId}',
                o: {
                    customerId: options.projectId,
                },
            },
            params: {
                regionId: options.regionId,
                expireDay: options.expireDay,
            },
            timeout: 20000,
            targetUrl: 'cbc=/rest/cbc/csbsubscriptionservice/v1/statistic/' + options.projectId,
        });
        return promise;
    };
    return AccountService;
}(ApiService_1.ApiService));
exports.AccountService = AccountService;
