// @cloud/cfdata/service/Analysis.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Analysis = void 0;
var ConfigData_1 = require("../ConfigData");
var UserService_1 = require("./UserService");
var cloudBIScriptLoading = false;
var globalCloudBIKey = null;
var Analysis = /** @class */ (function () {
    function Analysis() {
    }
    // 调用start方法，封装 onEvent 为 cfCloudBiOnEvent
    Analysis.initCloudBi = function (domainId) {
        // 埋点文件加载
        if (!domainId)
            return;
        this.localizationConfig = ConfigData_1.ConfigData.getUiSwitchConfig();
        var cloudBIKey = '8f1bb48d165bb0ab3b8d69128fdc3317';
        if (this.localizationConfig.analysis && domainId) {
            this.initUBA(cloudBIKey);
        }
        // js加埋点公用方法调用onEvent事件 --- godzilla多处方法中调用，是否需同步
        window.cfCloudBiOnEvent = function (EventLabel) {
            var domainPre = (window.location.host || '').split('.')[0] || '';
            var eventCategory = 'www_framework'.replace('www', domainPre);
            var AppKey = '8f1bb48d165bb0ab3b8d69128fdc3317';
            window.onEvent &&
                window.onEvent(eventCategory, 'click', EventLabel, 1, {
                    C1: domainPre,
                    D1: window.location.href,
                    AppKey: AppKey,
                });
        };
    };
    Analysis.start = function (cloudBIKey, GAKey, domain, service, regionId) {
        // globalCloudBIkey为null或者默认值，把传入的cloudBIKey赋给它
        if (globalCloudBIKey === null || globalCloudBIKey === '8f1bb48d165bb0ab3b8d69128fdc3317') {
            globalCloudBIKey = cloudBIKey;
        }
        this.localizationConfig = ConfigData_1.ConfigData.getUiSwitchConfig();
        if (!this.checkRegionAuth(this.localizationConfig.analysis, this.localizationConfig.ubaRegions, regionId)) {
            window.DATrack = function () {
                return false;
            };
            window.onEvent = function () {
                return false;
            };
            window.init = function () {
                return false;
            };
            /* eslint-disable */
            return Promise.reject();
        }
        var jsonParamView = {};
        var jsonParamEvent = {};
        var jsonCFParamEvent = {};
        var jsonParamInit = {};
        var domainPre = (window.location.host || '').split('.')[0] || '';
        var title = document.title;
        var reg = /(^|&)channelId=([^&]*)(&|$)/;
        var r = window.location.search.substr(1).match(reg);
        var href = window.location.href;
        var ref = '';
        var consoleName = href.split('/')[3] || '';
        var urlCloudbi = this.localizationConfig.cfAnalysisUrlCloudbi ? this.localizationConfig.cfAnalysisUrlCloudbi : '';
        if (document.referrer.length > 0) {
            ref = document.referrer;
        }
        try {
            if (ref.length === 0 && opener.location.href.length > 0) {
                /* eslint-disable  no-unused-vars */
                ref = opener.location.href;
            }
        }
        catch (e) {
            /* eslint-disable no-empty */
        }
        // init
        jsonParamInit['UserAccount'] = domain || 'host';
        jsonParamInit['isdebug'] = true;
        // pageView C1页面名称 D15服务console
        jsonParamView['C1'] = title;
        jsonParamView['D15'] = consoleName;
        if (r !== null) {
            jsonParamView['D16'] = unescape(r[2]);
        }
        // event
        jsonParamEvent['C1'] = domainPre;
        jsonParamEvent['D1'] = href;
        jsonCFParamEvent['C1'] = domainPre;
        jsonCFParamEvent['D1'] = href;
        jsonCFParamEvent['AppKey'] = '8f1bb48d165bb0ab3b8d69128fdc3317';
        // 在 window 对象上绑定一个方法，用于 actionMenu 等不能通过直接增加属性的埋点元素
        // 使用手动触发的方式，进行埋点
        if (window.DATrack) {
            delete window.DATrack;
        }
        // metaData参照格式：www_v1_home.click.按钮名称
        window.DATrack = function (metaData) {
            eventTrack(metaData);
        };
        var promise = new Promise(function (resolve, reject) {
            if (cloudBIScriptLoading || typeof window.onEvent === 'function')
                return;
            cloudBIScriptLoading = true;
            try {
                getScript(urlCloudbi, function () {
                    // 加载cloudbi.js,成功后，并执行回调函数
                    cloudBIScriptLoading = false;
                    window.init(2, globalCloudBIKey, GAKey, jsonParamInit);
                    if (service === 'home') {
                        window.onPageView(window.location.pathname + window.location.hash + window.location.search, jsonParamView);
                    }
                    document.addEventListener('click', function (event) {
                        var eventTarget = event.target;
                        var item;
                        if (typeof eventTarget.closest === 'function') {
                            item = eventTarget.closest('[meta-data-uba]');
                        }
                        else {
                            item = getMetaDataUbaElem(eventTarget);
                        }
                        if (!item)
                            return;
                        var metaData = item.innerText;
                        eventTrack && eventTrack(metaData);
                    });
                    resolve({
                        jsonParamInit: jsonParamInit,
                        jsonParamView: jsonParamView,
                    });
                });
            }
            catch (e) {
                /* eslint-disable */
                reject();
            }
        });
        return promise;
        function eventTrack(metaData, eventValue) {
            if (eventValue === void 0) { eventValue = 1; }
            if (!metaData) {
                return;
            }
            try {
                window.onEvent('www_framework'.replace('www', domainPre), 'click', metaData, eventValue, jsonCFParamEvent);
            }
            catch (e) {
                /* eslint-disable no-empty */
            }
        }
        function getMetaDataUbaElem(target) {
            if (target.classList && target.classList.contains('[meta-data-uba]')) {
                return target;
            }
            else {
                return target.parentElement && getMetaDataUbaElem(target.parentElement);
            }
        }
        // ajax getScript的原生实现: 加载script并执行回调
        function getScript(source, callback) {
            var script = document.createElement('script');
            var prior = document.getElementsByTagName('script')[0];
            script.async = 1;
            script.onload = script.onreadystatechange = function (_, isAbort) {
                if (isAbort || !script.readyState || /loaded|complete/.test(script.readyState)) {
                    script.onload = script.onreadystatechange = null;
                    script = undefined;
                    if (!isAbort && callback)
                        setTimeout(callback, 0);
                }
            };
            script.src = source;
            prior.parentNode.insertBefore(script, prior);
        }
    };
    ;
    /**
    * 添加页面埋点
    */
    Analysis.addAutoPV = function (cloudBIKey, params) {
        if (!cloudBIKey) {
            return;
        }
        var onPageViewLocal = window.onPageView;
        window.onPageView = function () { }; // 让uba.js的onPageView API失效，所有的PV上报统一由框架控制，避免云服务重复上报
        var BIKeyMap = {}; // 需要兼容cloudBIKey为string的场景，统一数据结构
        if (typeof cloudBIKey === 'string') {
            BIKeyMap['default'] = cloudBIKey;
        }
        else {
            BIKeyMap = cloudBIKey;
        }
        pageViewListener(); // uba加载成功后，也需要针对当前的页面出发一次PV上报，此场景下fromState无法获取，为undefined
        // 监听hash变化
        window.addEventListener('hashchange', function (e) {
            pageViewListener({ PageName: e.newURL, FromPageName: e.oldURL });
        });
        function getCurrentBIKey(PageName) {
            var result = BIKeyMap['default'] || '8f1bb48d165bb0ab3b8d69128fdc3317';
            var hash = location.hash;
            if (PageName) {
                hash = PageName.split('#')[1];
            }
            var keyPath = hash.split('/')[1];
            if (keyPath in BIKeyMap) {
                result = BIKeyMap[keyPath];
            }
            return result;
        }
        function pageViewListener(opts) {
            var options = opts || {};
            /**
                * init()是uba的接口
                * 参数1：对于控制台服务都是1
                * 参数2：cloudbi给各服务分配的id
                * 参数3：GA给各服务分配的id
                * 参数4：参数集合
                * */
            var BIKey = getCurrentBIKey(options.PageName);
            window.init(1, BIKey, undefined, params.jsonParamInit);
            if (options.PageName) {
                params.jsonParamView['pageName'] = options.PageName;
            }
            if (options.FromPageName) {
                params.jsonParamView['FromPageName'] = options.FromPageName;
            }
            onPageViewLocal('', params.jsonParamView);
        }
    };
    ;
    Analysis.addPageActivity = function () {
        var _this = this;
        document.addEventListener('click', function (event) {
            var target = document.querySelector(event.target[0]); // $(event.target)
            if (target && (_this.closest(target, '#J_header').length > 0 || _this.closest(target, '#J_sidebar').length > 0 || _this.closest(target, '#cf-consult').length > 0 || _this.closest(target, '.ti3-leftmenu-container').length > 0)) {
                return;
            }
            // 高度过大的元素，认为是空白区域
            var MAX_HEIGHT = 32;
            if (target && target.outerHeight() >= MAX_HEIGHT) {
                return;
            }
            var loc = window.location;
            var path = loc.pathname.slice(1); // 从第一位截取，如"/ecm/"
            var hash = loc.hash.slice(2); // 从第二位截取，如"#/ecs/manager/vmList"
            var eventCategory = 'console_' + path + hash;
            window.onEvent(eventCategory, 'click', target && target.innerText() || '', 'console_active');
        });
    };
    ;
    Analysis.initUBA = function (cloudBIKey) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                try {
                    UserService_1.UserSerivce.getLoginUser().then(function (user) { return __awaiter(_this, void 0, void 0, function () {
                        var _a, GAKey, domainId, serviceName, regionId, promiseBUALoaded;
                        return __generator(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    _a = [undefined, user.domainId, '', user.region], GAKey = _a[0], domainId = _a[1], serviceName = _a[2], regionId = _a[3];
                                    return [4 /*yield*/, this.start(cloudBIKey, GAKey, domainId, serviceName, regionId)];
                                case 1:
                                    promiseBUALoaded = _b.sent();
                                    this.addAutoPV(cloudBIKey, promiseBUALoaded); // initUBA是新增接口，服务调用的时候，自动上报PV埋点
                                    this.addPageActivity();
                                    return [2 /*return*/];
                            }
                        });
                    }); });
                }
                catch (e) { }
                return [2 /*return*/];
            });
        });
    };
    ;
    Analysis.init = function (cloudBIKey) {
        var consoleContext = window.getConsoleContext && window.getConsoleContext();
        var consoleConfig = consoleContext && consoleContext.get && consoleContext.get({ name: 'config' });
        if (consoleConfig && consoleConfig.set) {
            consoleConfig.set('ubaAppKeyMap', typeof cloudBIKey === 'string' ? { default: cloudBIKey } : cloudBIKey);
        }
    };
    // 构造closest方法
    Analysis.closest = function (el, selector) {
        var matchesSelector = el.matches || el.webkitMatchesSelector || el.mozMatchesSelector || el.msMatchesSelector;
        while (el) {
            if (matchesSelector.call(el, selector)) {
                return el;
            }
            else {
                /* eslint-disable */
                el = el.parentElement;
            }
        }
        return '';
    };
    ;
    Analysis.checkRegionAuth = function (authName, regions, regionId) {
        if (!authName || !Array.isArray(regions) || regions.length === 0) {
            return false;
        }
        var projectNameRegExp = /^(DeC)_([^_]+)_(.+)$/;
        // 有‘_’说明是project，继续获取其对应的实际regionId
        if (regionId.indexOf('_') > -1) {
            if (projectNameRegExp.test(regionId)) {
                var projectElms = regionId.match(projectNameRegExp);
                regionId = projectElms[2];
            }
            else { // 'regionId_projectId'场景
                var tempIndex = regionId.indexOf('_');
                regionId = regionId.slice(0, tempIndex);
            }
        }
        if (regions.indexOf(regionId) === -1) {
            return false;
        }
        return true;
    };
    ;
    return Analysis;
}());
exports.Analysis = Analysis;
