// @cloud/cfdata/service/EndpointService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointService = void 0;
var ApiService_1 = require("./ApiService");
var HttpService_1 = require("./HttpService");
var FrameData_1 = require("../FrameData");
var EndpointService = /** @class */ (function (_super) {
    __extends(EndpointService, _super);
    function EndpointService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EndpointService.queryEndpoints = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/endpoints',
            timeout: 60000,
            params: {
                region: options.region,
                include: options.include,
            },
            targetUrl: 'silvan=/silvan/rest/v1.0/endpoints',
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
            cacheKey: {
                locale: FrameData_1.FrameData.getLanguage(),
                region: options.region,
                cfApiName: 'endpoints',
            },
            cacheTimeout: 604800,
            cacheResult: true,
        });
        return promise;
    };
    EndpointService.favoriteEndpoints = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/users/{user_id}/endpoints',
                o: {
                    user_id: options.user_id,
                },
            },
            mask: options.maskFlag || false,
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' + options.user_id + '/endpoints',
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
        });
        return promise;
    };
    /**
     * @internal
     * Adds favorite endpoint
     * @param options
     * @returns
     */
    EndpointService.addFavoriteEndpoint = function (options) {
        options = options || {};
        var reqOptions = {
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/users/{user_id}/endpoints/{id}',
                o: {
                    user_id: options.user_id,
                    id: options.id,
                },
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' + options.user_id + '/endpoints/' + options.id,
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
        };
        // 清空收藏后刷新展示默认的收藏，之后首次点击添加收藏
        if (options.id === 'all') {
            Object.assign(reqOptions, { params: options.params });
        }
        var promise = HttpService_1.HttpService.post(reqOptions);
        return promise;
    };
    /**
     * @internal
     * Deletes favorite endpoint
     * @param options
     * @returns
     */
    EndpointService.deleteFavoriteEndpoint = function (options) {
        options = options || {};
        var promise;
        // 清空收藏后刷新展示默认的收藏，之后首次点击删除
        var reqOptions = {
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/users/{user_id}/endpoints/{id}',
                o: {
                    user_id: options.user_id,
                    id: options.id,
                },
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' + options.user_id + '/endpoints/' + options.id,
        };
        if (options.id === 'all') {
            Object.assign(reqOptions, {
                params: options.params,
            });
            promise = HttpService_1.HttpService.post(reqOptions);
        }
        else {
            promise = HttpService_1.HttpService.delete(reqOptions);
        }
        return promise;
    };
    /**
     * @internal
     * Sorts favorite endpoint
     * @param options
     * @returns
     */
    EndpointService.sortFavoriteEndpoint = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.post({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/users/{user_id}/endpoints/{id}',
                o: {
                    user_id: options.user_id,
                    id: 'all',
                },
            },
            timeout: 60000,
            params: {
                endpointResources: options.endpointResources,
            },
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' + options.user_id + '/endpoints/' + options.id,
        });
        return promise;
    };
    return EndpointService;
}(ApiService_1.ApiService));
exports.EndpointService = EndpointService;
