// @cloud/cfdata/service/FrameworkService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameworkService = void 0;
var ApiService_1 = require("./ApiService");
var HttpService_1 = require("./HttpService");
var FrameData_1 = require("../FrameData");
var FrameworkService = /** @class */ (function (_super) {
    __extends(FrameworkService, _super);
    function FrameworkService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FrameworkService.getLinks = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/links',
            timeout: 60000,
            params: {
                key: options.key || '',
                type: options.type || '',
                region: options.region || '',
            },
            targetUrl: 'silvan=/silvan/rest/v1.0/links',
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
            cacheKey: {
                cfApiName: 'links',
            },
            cacheTimeout: 604800,
            cacheResult: true,
        });
        return promise;
    };
    FrameworkService.getNewGuideNotNextPrompt = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/users/{user_id}/endpointslogged/{service_id}',
                o: {
                    user_id: options.user_id,
                    service_id: options.service_id,
                },
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' + options.user_id + '/endpointslogged/' + options.service_id,
        });
        return promise;
    };
    FrameworkService.setNewGuideNotNextPrompt = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.post({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() +
                    '/rest/silvan/rest/v1.0/users/{user_id}/endpointslogged/{service_id}/{version}',
                o: {
                    user_id: options.user_id,
                    service_id: options.service_id,
                    version: options.version,
                },
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/users/' +
                options.user_id +
                '/endpointslogged/' +
                options.service_id +
                '/' +
                options.version,
        });
        return promise;
    };
    FrameworkService.heartbeat = function () {
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/heartbeat',
            timeout: 60000,
        });
        return promise;
    };
    FrameworkService.queryAssociationValueToSearch = function (url, queryValue) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, HttpService_1.HttpService.get({
                        url: url,
                        timeout: 30000,
                        query: { q: queryValue, from: 'nav', language: FrameData_1.FrameData.getCfFrame().lang },
                    })];
            });
        });
    };
    return FrameworkService;
}(ApiService_1.ApiService));
exports.FrameworkService = FrameworkService;
