// @cloud/cfdata/service/IncidentService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.IncidentService = void 0;
var ApiService_1 = require("./ApiService");
var HttpService_1 = require("./HttpService");
var FrameData_1 = require("../FrameData");
var IncidentService = /** @class */ (function (_super) {
    __extends(IncidentService, _super);
    function IncidentService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IncidentService.createIncidentInfo = function (createFeedback) {
        createFeedback = createFeedback || {};
        var promise = HttpService_1.HttpService.post({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/osm/incidentservice/api/v1/feedback',
            params: {
                createFeedback: createFeedback,
            },
            timeout: 60000,
            targetUrl: 'osm=/osm/incidentservice/api/v1/feedback',
        });
        return promise;
    };
    // 传进来的这个参数没有用？
    IncidentService.queryCusBrief = function ( /* userId */) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/osm/incidentservice/api/v1/queryincident',
            },
            params: {
                status: 1,
            },
            timeout: 10000,
            targetUrl: 'osm=/osm/incidentservice/api/v1/queryincident',
        });
    };
    return IncidentService;
}(ApiService_1.ApiService));
exports.IncidentService = IncidentService;
