// @cloud/cfdata/service/MessageService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageService = void 0;
var HttpService_1 = require("./HttpService");
var ApiService_1 = require("./ApiService");
var FrameData_1 = require("../FrameData");
var MessageService = /** @class */ (function (_super) {
    __extends(MessageService, _super);
    function MessageService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MessageService.getMessages = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/messages',
            timeout: 60000,
            params: {
                start: options.start || 0,
                limit: options.limit || 0,
            },
        });
        return promise;
    };
    MessageService.getMessagesTypes = function () {
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/messageTypes',
            timeout: 60000,
        });
        return promise;
    };
    MessageService.messageStatus = function (options) {
        return HttpService_1.HttpService.put({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/mc/v1/{domain_id}/messages/status',
                o: {
                    domain_id: options.domainId,
                },
            },
            timeout: 10000,
            params: {
                message_ids: options.messageIds,
                category_id: options.categoryId,
            },
            targetUrl: 'mc=/mc/v1/' + options.domainId + '/messages/status',
        });
    };
    MessageService.queryMcUnreadCount = function (domainId) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/mc/v1/{domain_id}/messages/envelope/unread',
                o: {
                    domain_id: domainId,
                },
            },
            timeout: 10000,
            targetUrl: 'mc=/mc/v1/' + domainId + '/messages/envelope/unread',
        });
    };
    MessageService.queryMcMessages = function (domainId, params) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/mc/v1/{domain_id}/messages',
                o: {
                    domain_id: domainId,
                },
            },
            params: params,
            timeout: 10000,
            targetUrl: 'mc=/mc/v1/' + domainId + '/messages',
        });
    };
    return MessageService;
}(ApiService_1.ApiService));
exports.MessageService = MessageService;
