// @cloud/cfdata/service/RegionService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegionService = void 0;
var HttpService_1 = require("./HttpService");
var ApiService_1 = require("./ApiService");
var FrameData_1 = require("../FrameData");
var UserData_1 = require("../UserData");
var RegionService = /** @class */ (function (_super) {
    __extends(RegionService, _super);
    function RegionService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RegionService.queryRegions = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/regions',
            timeout: 60000,
            params: {
                selfDevelop: options.selfDevelop,
                alliance: options.alliance,
            },
            targetUrl: 'silvan=/silvan/rest/v1.0/regions',
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
            cacheKey: {
                locale: FrameData_1.FrameData.getLanguage(),
                cfApiName: 'regions',
                userId: UserData_1.UserData.getCfUserSync().userId,
            },
            cacheTimeout: 604800,
            cacheResult: true,
        });
        return promise;
    };
    /**
     * @internal
     * Querys regions from rm
     * @param options
     * @returns
     */
    RegionService.queryRegionsFromRm = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/lcs/v1/regionMgr/domains/{domain_id}/regions',
                o: {
                    domain_id: options.domainId,
                },
            },
            timeout: 60000,
            params: {
                selfDevelop: options.selfDevelop,
                alliance: options.alliance,
                link: true,
                locale: FrameData_1.FrameData.getLanguage(),
            },
            targetUrl: 'lcs=/v1/regionMgr/internal/domains/' + options.domainId + '/regions',
            headers: {
                'Frame-Domain-Type': ApiService_1.ApiService.getDomainType(),
            },
            cacheKey: {
                locale: FrameData_1.FrameData.getLanguage(),
                cfApiName: 'regions',
                userId: UserData_1.UserData.getCfUserSync().userId,
            },
            cacheTimeout: 604800,
            cacheResult: true,
        });
        return promise;
    };
    RegionService.changeRegion = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/changeRegion',
            timeout: 60000,
            mask: true,
            params: {
                project: options.project,
            },
        });
        return promise;
    };
    RegionService.queryMaintenanceRegions = function () {
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/maintenanceRegion',
            timeout: 60000,
            headers: {
                'x-request-from': 'Framework',
            },
        });
        return promise;
    };
    return RegionService;
}(ApiService_1.ApiService));
exports.RegionService = RegionService;
