// @cloud/cfdata/service/TraceService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceService = void 0;
var ApiService_1 = require("./ApiService");
var HttpService_1 = require("./HttpService");
var FrameData_1 = require("../FrameData");
var TraceService = /** @class */ (function (_super) {
    __extends(TraceService, _super);
    function TraceService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TraceService.queryTracker = function (projectId) {
        // TODO:这是从window上读取么？
        var params = /* params || */ {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/v1.0/{projectId}/tracker',
                o: {
                    projectId: projectId,
                },
            },
            params: params,
            timeout: 60000,
            targetUrl: 'cts=/v1.0/' + projectId + '/tracker',
        });
        return promise;
    };
    TraceService.queryTraceList = function (projectId, trackerName, params /* , regionId */) {
        params = params || {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/v2.0/{projectId}/{trackername}/trace',
                o: {
                    trackername: trackerName,
                    projectId: projectId,
                },
            },
            params: params,
            timeout: 60000,
            // "region": regionId,
            targetUrl: 'cts=/v2.0/' + projectId + '/' + trackerName + '/trace',
        });
        return promise;
    };
    return TraceService;
}(ApiService_1.ApiService));
exports.TraceService = TraceService;
