// @cloud/cfdata/service/UserService.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSerivce = void 0;
var ApiService_1 = require("./ApiService");
var HttpService_1 = require("./HttpService");
var ConfigData_1 = require("../ConfigData");
var FrameData_1 = require("../FrameData");
/**
 *
 *
 * @export
 * @class UserSerivce
 * @extends {ApiService}
 */
var UserSerivce = /** @class */ (function (_super) {
    __extends(UserSerivce, _super);
    function UserSerivce() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UserSerivce.getLoginUser = function (isRefreshToken) {
        var requestConfig = {
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/me',
            timeout: 60000,
            headers: {
                'x-request-from': 'Framework',
            },
        };
        if (isRefreshToken) {
            requestConfig.headers = __assign(__assign({}, requestConfig.headers), { 'Token-Refresh': String(isRefreshToken) });
        }
        var promise = new Promise(function (resolve, reject) {
            HttpService_1.HttpService.get(requestConfig)
                .then(function (data) {
                UserSerivce.meDataBak = data;
                resolve(data);
            })
                .catch(function (error) {
                if (UserSerivce.meDataBak) {
                    resolve(UserSerivce.meDataBak);
                }
                else {
                    reject(error);
                }
            });
        });
        return promise;
    };
    /**
     * @internal
     * 貌似内外都没有用到
     * Gets user choices
     * @param options
     * @returns user choices
     */
    UserSerivce.getUserChoices = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/user_choices',
            params: {
                userId: options.userId || '',
                preferName: options.type || '',
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/user_choices',
        });
        return promise;
    };
    /**
     * @internal
     * 貌似内外都没有用到
     * Updates user choices
     * @param options
     * @returns user choices
     */
    UserSerivce.updateUserChoices = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.post({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/silvan/rest/v1.0/user_choices',
            params: {
                userId: options.userId || '',
                preferName: options.type || '',
                preferDisplayName: options.value || '',
                preferDesc: options.preferDesc || '',
            },
            timeout: 60000,
            targetUrl: 'silvan=/silvan/rest/v1.0/user_choices',
        });
        return promise;
    };
    /**
     * 查询AssumeRoles
     *
     * @param  options
     * @returns
     * @memberof UserSerivce
     */
    UserSerivce.getAssumeRoles = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/iam/assumeRoles/{user_id}',
                o: {
                    user_id: options.userId,
                },
            },
            params: {
                limit: 5,
            },
            timeout: 60000,
            targetUrl: 'iam=/v3-huawei/xrole/history/' + options.userId,
        });
        return promise;
    };
    /**
     * @internal
     * 貌似内外都没有用到
     * Assumes role
     * @param options
     * @returns role
     */
    UserSerivce.assumeRole = function (options) {
        options = options || {};
        var promise = HttpService_1.HttpService.get({
            url: FrameData_1.FrameData.getAppWebPath() + '/rest/changeRole',
            timeout: 60000,
            mask: true,
            params: {
                agencyId: options.agencyId,
            },
        });
        return promise;
    };
    UserSerivce.getUserProjects = function (options) {
        options = options || {};
        var promise;
        if (ConfigData_1.ConfigData.getUiSwitchConfig().x_cloud_type === 'private') {
            promise = HttpService_1.HttpService.get({
                url: {
                    s: FrameData_1.FrameData.getAppWebPath() + '/rest/iam/v3/users/{user_id}/projects',
                    o: {
                        user_id: options.userId,
                    },
                },
                timeout: 60000,
                targetUrl: 'iam=/v3/users/' + options.userId + '/projects',
                cacheKey: {
                    locale: FrameData_1.FrameData.getLanguage(),
                    userId: options.userId,
                    cfApiName: 'projects',
                },
                cacheTimeout: 604800,
                cacheResult: true,
            });
        }
        else {
            promise = HttpService_1.HttpService.get({
                url: FrameData_1.FrameData.getAppWebPath() + '/rest/iam/v3/projects',
                timeout: 60000,
                targetUrl: 'iam=/v3/auth/projects',
                cacheKey: {
                    locale: FrameData_1.FrameData.getLanguage(),
                    userId: options.userId,
                    cfApiName: 'projects',
                },
                cacheTimeout: 604800,
                cacheResult: true,
            });
        }
        return promise;
    };
    UserSerivce.queryIamUser = function (userId) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/iam/service/user/{userId}',
                o: {
                    userId: userId,
                },
            },
            timeout: 10000,
            params: {
                scope: 'imgpaths',
            },
            targetUrl: 'iam=/v3-huawei/service/user/' + userId + '?scope=imgpaths',
        });
    };
    /**
     * @internal
     * 貌似内外都没有用到
     * Querys mc user role
     * @param userId
     * @returns mc user role
     */
    UserSerivce.queryMcUserRole = function (userId) {
        return HttpService_1.HttpService.get({
            url: {
                s: FrameData_1.FrameData.getAppWebPath() + '/rest/mc/v1/{user_id}/role',
                o: {
                    user_id: userId,
                },
            },
            timeout: 10000,
            targetUrl: 'mc=/mc/v1/' + userId + '/role',
        });
    };
    return UserSerivce;
}(ApiService_1.ApiService));
exports.UserSerivce = UserSerivce;
