// @cloud/cfdata/service/UtilService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtilService = void 0;
var StorageService_1 = require("./StorageService");
var UtilService = /** @class */ (function () {
    function UtilService() {
    }
    UtilService.i18nReplace = function (s, o) {
        return s.replace
            ? s.replace(this.i18nSubRegRex, function (match, key) {
                return ['Number', 'String', 'Boolean'].includes(UtilService.getTypeOfData(o[key])) ? o[key] : match;
            })
            : s;
    };
    UtilService.i18nReplaceFull = function (s, o) {
        if (!s || !o) {
            return;
        }
        return s.replace
            ? s.replace(this.i18nSubRegRex, function (match, key) {
                return o[key] ? o[key] : match;
            })
            : s;
    };
    UtilService.trimEmpty = function (value) {
        if (!value) {
            return '';
        }
        return value.replace(/(^\s*)|(\s*$)/g, '');
    };
    /**
     * 输入数字大于999，显示为999+
     * @param num
     * @returns {*}
     */
    UtilService.dealMaxNumber = function (num) {
        if (isNaN(num))
            return 0;
        if (num >= 1000)
            return '999+';
        return num;
    };
    UtilService.getRecentServices = function (services) {
        var _recentServices = [];
        // 优先使用传入的serviceslist、未传入时同步获取endpointsdata
        var serviceList = services;
        var length = serviceList && serviceList.length;
        if (length) {
            var rencentServices = StorageService_1.StorageService.getCookie('recentServices');
            var rencentServicesLen = rencentServices && rencentServices.length;
            if (!rencentServicesLen) {
                return _recentServices;
            }
            for (var j = 0; j < rencentServicesLen; j++) {
                for (var k = 0; k < length; k++) {
                    var item = serviceList[k];
                    if (item.id === rencentServices[j]) {
                        _recentServices.push(item);
                        break;
                    }
                }
            }
        }
        return _recentServices;
    };
    UtilService.setRecentService = function (serviceId) {
        var recentServiceList = StorageService_1.StorageService.getCookie('recentServices') || [];
        var hasService = recentServiceList.indexOf(serviceId);
        if (hasService > -1) {
            recentServiceList.splice(hasService, 1);
        }
        recentServiceList.unshift(serviceId);
        if (recentServiceList.length > 6) {
            recentServiceList = recentServiceList.slice(0, 6);
        }
        StorageService_1.StorageService.setCookie('recentServices', recentServiceList);
    };
    UtilService.dateFormat = function (fmt, date) {
        var ret;
        var opt = {
            'Y+': date.getFullYear().toString(),
            'm+': (date.getMonth() + 1).toString(),
            'd+': date.getDate().toString(),
            'H+': date.getHours().toString(),
            'M+': date.getMinutes().toString(),
            'S+': date.getSeconds().toString(),
        };
        for (var k in opt) {
            ret = new RegExp('(' + k + ')').exec(fmt);
            if (ret) {
                fmt = fmt.replace(ret[1], ret[1].length === 1 ? opt[k] : opt[k].padStart(ret[1].length, '0'));
            }
        }
        return fmt;
    };
    // "Undefined"  "Null"  "Boolean" "Number"  "String"  "Object"  "Array"  "Function"
    UtilService.getTypeOfData = function (data) {
        var res = Object.prototype.toString.call(data);
        return res.substr(0, res.length - 1).split(' ')[1];
    };
    UtilService.getUrlParameter = function (paramKey, scopeFlag) {
        var href = window.location.href;
        var indexHash = href.indexOf('#');
        var indexQuest = href.indexOf('?');
        var sPageURL = '';
        if (indexHash > indexQuest) {
            sPageURL = window.location.search.substring(1); // hash在参数之后时，能通过search获取参数
        }
        else {
            sPageURL = href.split('?')[1] || ''; // hash在前，参数在后，则无法通过search获取参数
        }
        if (sPageURL) {
            var sURLVariables = sPageURL.split('&');
            for (var i = 0; i < sURLVariables.length; i++) {
                var sParameterName = sURLVariables[i].split('=');
                if (sParameterName[0] === paramKey) {
                    return sParameterName[1];
                }
            }
        }
        if (scopeFlag) {
            var consoleContext = window.getConsoleContext && window.getConsoleContext();
            var consoleDataService = (consoleContext === null || consoleContext === void 0 ? void 0 : consoleContext.get) && consoleContext.get({ name: 'config' });
            if (paramKey === 'agencyId') {
                return ((consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.get) && (consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.get('cfUserId'))) || '';
            }
            else if (paramKey === 'region') {
                var projectName = ((consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.get) && (consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.get('cfProjectName'))) || '';
                return encodeURIComponent(projectName);
            }
        }
        else {
            return '';
        }
        return '';
    };
    UtilService.convertSearchString2Obj = function (search) {
        var res = {};
        var searchParams = UtilService.startsWith(search, '?') ? search.substring(1) : search;
        if (!searchParams)
            return res;
        var sURLVariables = searchParams.split('&');
        for (var i = 0; i < sURLVariables.length; i++) {
            var param = sURLVariables[i].split('=');
            var key = decodeURIComponent(param[0]);
            if (res[key]) {
                res[key] = [].concat(res[key]).concat(decodeURIComponent(param[1] || ''));
            }
            else
                res[key] = decodeURIComponent(param[1] || '');
        }
        return res;
    };
    UtilService.convertObj2SearchString = function (params) {
        if (params === void 0) { params = {}; }
        var _result = [];
        var _loop_1 = function (key) {
            var type = UtilService.getTypeOfData(params[key]);
            if (type === 'Array') {
                params[key].forEach(function (_value) {
                    var _type = UtilService.getTypeOfData(_value);
                    var value = ['Number', 'String', 'Boolean'].includes(_type) ? _value : '';
                    _result.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
                });
            }
            else {
                var value = ['Number', 'String', 'Boolean'].includes(type) ? params[key] : '';
                _result.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
            }
        };
        for (var key in params) {
            _loop_1(key);
        }
        return _result.join('&');
    };
    UtilService.reArrangeSearchParam = function (curUrl) {
        var origin = curUrl.origin, pathname = curUrl.pathname, search = curUrl.search, hash = curUrl.hash;
        var hashRoute = hash.split('?')[0];
        var suffixStr = hash.split('?')[1] || '';
        var prefixStr = search.substring(1);
        var prefixObj = UtilService.convertSearchString2Obj(prefixStr);
        var defaultPrefixParams = [
            'locale',
            'region',
            'agencyId',
            'app_id',
            'source',
            'enterprise_project_id',
            'iscros',
            'cfModuleHide',
        ];
        Object.keys(prefixObj).forEach(function (param) {
            if (!defaultPrefixParams.includes(param)) {
                suffixStr = suffixStr ? suffixStr + "&" + param + "=" + prefixObj[param] : param + "=" + prefixObj[param];
                delete prefixObj[param];
            }
        });
        var newPrefixStr = UtilService.convertObj2SearchString(prefixObj);
        var newSearch = newPrefixStr ? "?" + newPrefixStr : '';
        var newHash = hashRoute + (suffixStr ? "?" + suffixStr : '');
        return "" + origin + pathname + newSearch + newHash;
    };
    UtilService.initHistoryInterceptor = function () {
        var historyWrapper = function (type) {
            var originHistoryEvent = history[type];
            return function () {
                var fullUrl;
                var origin = location.origin, pathname = location.pathname, search = location.search;
                if (UtilService.startsWith(arguments[2], '#')) {
                    fullUrl = new URL("" + origin + pathname + search + arguments[2]);
                }
                else
                    fullUrl = new URL(arguments[2]);
                if (search)
                    arguments[2] = UtilService.reArrangeSearchParam(fullUrl);
                var res = originHistoryEvent.apply(this, arguments);
                window.dispatchEvent(new Event(type));
                return res;
            };
        };
        history.pushState = historyWrapper('pushState');
        history.replaceState = historyWrapper('replaceState');
    };
    UtilService.delUrlParameter = function (url, name) {
        return url
            .replace(new RegExp('[?&]' + name + '=[^&#]*(#.*)?$'), '$1')
            .replace(new RegExp('([?&])' + name + '=[^&]*&'), '$1');
    };
    UtilService.isArray = function (o) {
        return Object.prototype.toString.call(o);
    };
    /* eslint-disable no-unused-expressions */
    UtilService.browserDetect = function () {
        // 浏览器判断
        var browser = {};
        var userAgent = navigator.userAgent.toLowerCase();
        var aAgentInfo;
        /* eslint-disable no-cond-assign */
        (aAgentInfo = userAgent.match(/rv:([\d.]+)\) like gecko/))
            ? (browser['ie'] = true)
            : (aAgentInfo = userAgent.match(/msie ([\d.]+)/))
                ? (browser['ie'] = true)
                : (aAgentInfo = userAgent.match(/firefox\/([\d.]+)/))
                    ? (browser['firefox'] = true)
                    : (aAgentInfo = userAgent.match(/chrome\/([\d.]+)/))
                        ? (browser['chrome'] = true)
                        : (aAgentInfo = userAgent.match(/opera.([\d.]+)/))
                            ? (browser['opera'] = true)
                            : (aAgentInfo = userAgent.match(/version\/([\d.]+).*safari/))
                                ? (browser['safari'] = true)
                                : 0;
        this.isArray(aAgentInfo) && (browser['version'] = aAgentInfo[1]);
        return browser;
    };
    UtilService.matchSupportLanDesc = function () {
        // 所有描述匹配
        var allMatchDescArray = [
            ['en-us', 'English'],
            ['zh-cn', '中文（简体）'],
            ['es-es', 'Español'],
            ['es-us', 'Español'],
            ['pt-br', 'Português'],
            ['fr-fr', 'Français'],
        ];
        // 如果是香港或者华为云默认中文描述改成“简体中文”
        if (window['theme'] === 'hec' || window['theme'] === 'hec-hk') {
            allMatchDescArray = allMatchDescArray.map(function (item) {
                if (item[0] === 'zh-cn') {
                    item = ['zh-cn', '简体中文'];
                }
                return item;
            });
        }
        // 根据 支持的语言返回对应的描述
        allMatchDescArray = allMatchDescArray.filter(function (item) {
            return window['cfThemeJs'].lanConfig.supportLan.indexOf(item[0]) > -1;
        });
        return allMatchDescArray;
    };
    /**
     *
     * @param href
     * @param flag 默认添加agencyId、region、locale urlparameter
     * @param param
     * @param refrence 添加refrence urlparameter
     */
    UtilService.genHWSHref = function (_href, flag, param, refrence) {
        var href = _href;
        if (!param) {
            param = {};
        }
        if (!href || href === '#' || href === 'javascript:void(0)') {
            return href;
        }
        if (!flag) {
            href = UtilService.addOrReplaceUrlParameter(href, 'agencyId', UtilService.getUrlParameter('agencyId', true));
            var region = param['region'] || UtilService.getUrlParameter('region', true);
            if (region && region !== '' && region !== 'null') {
                href = UtilService.addOrReplaceUrlParameter(href, 'region', region);
            }
            // 处理国际化
            href = UtilService.addOrReplaceUrlParameter(href, 'locale', StorageService_1.StorageService.getCookie('locale'));
        }
        else if (flag === 'locale') {
            // 处理国际化
            href = UtilService.addOrReplaceUrlParameter(href, 'locale', StorageService_1.StorageService.getCookie('locale'));
        }
        if (refrence) {
            href = UtilService.addOrReplaceUrlParameter(href, 'refrence', refrence);
        }
        var thirdPartyAccessValue = StorageService_1.StorageService.getCookie('third-party-access');
        if (thirdPartyAccessValue && thirdPartyAccessValue === 'agc') {
            var enterpriseProjectId = param['enterprise_project_id'] || UtilService.getUrlParameter('enterprise_project_id', true);
            var appId = param['app_id'] || UtilService.getUrlParameter('app_id', true);
            var source = param['source'] || UtilService.getUrlParameter('source', true);
            if (enterpriseProjectId && enterpriseProjectId !== '' && enterpriseProjectId !== 'null') {
                href = UtilService.addOrReplaceUrlParameter(href, 'enterprise_project_id', enterpriseProjectId);
            }
            if (appId && appId !== '' && appId !== 'null') {
                href = UtilService.addOrReplaceUrlParameter(href, 'app_id', appId);
            }
            if (source && source !== '' && source !== 'null') {
                href = UtilService.addOrReplaceUrlParameter(href, 'source', source);
            }
        }
        return href;
    };
    // !value为true时删除参数
    UtilService.addOrReplaceUrlParameter = function (href, key, value) {
        if (!href || !key) {
            return href;
        }
        var hrefs = href.split('#/');
        var hrefPostfix = '';
        // 将#后面的路由信息拼接到URL最后
        if (hrefs.length > 1) {
            hrefPostfix = '#/' + hrefs[1];
        }
        hrefs[0] = UtilService.delUrlParameter(hrefs[0], key);
        if (value) {
            if (hrefs[0].indexOf('?') !== -1) {
                hrefs[0] = hrefs[0] + '&' + key + '=' + value;
            }
            else {
                hrefs[0] = hrefs[0] + '?' + key + '=' + value;
            }
        }
        return hrefs[0] + hrefPostfix;
    };
    UtilService.getRegion = function () {
        var projectName = UtilService.getProjectName();
        var projectNameRegExp = /^(DeC)_([^_]+)_(.+)$/;
        if (projectNameRegExp.test(projectName)) {
            return projectNameRegExp.exec(projectName)[2];
        }
        else
            return projectName.split('_')[0];
    };
    UtilService.getProjectName = function () {
        return UtilService.getUrlParameter('region', true) || StorageService_1.StorageService.getSessionStorage('cfCurrentRegion') || '';
    };
    // Time33 哈希算法
    UtilService.time33 = function (text) {
        var hash = 5381; // magic number
        var index = text.length;
        while (index) {
            hash = (hash * 33) ^ text.charCodeAt(--index);
        }
        return hash >>> 0;
    };
    UtilService.hashCode = function (text) {
        return String(UtilService.time33(text));
    };
    UtilService.startsWith = function (str, prefix) {
        if (typeof str !== 'string' || typeof prefix !== 'string' || str === '' || prefix === '')
            return false;
        return str.substr(0, prefix.length).indexOf(prefix) !== -1;
    };
    UtilService.onload = function (callback, timer) {
        var run = function () {
            setTimeout(callback, timer);
        };
        if (document.readyState === 'complete') {
            run();
        }
        else {
            window.addEventListener('load', run);
        }
    };
    /* eslint-disable no-useless-escape */
    UtilService.i18nSubRegRex = /\{\s*([^\|\}]+?)\s*(?:\|([^\}]*))?\s*\}/g;
    return UtilService;
}());
exports.UtilService = UtilService;
