// @cloud/tinyplus3/esm2015/components/resources/TpResourceItemComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * @ignore
 */
export class TpResourceItemComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 选中某项资源异常数时，触发事件
         */
        this.errorSelect = new EventEmitter();
    }
    /**
     * @ignore
     * 获取显示异常资源数
     */
    getErrorNumber(error, errorMax) {
        const baseValue = errorMax && errorMax > 0 ? errorMax : 99;
        return error > baseValue ? `${baseValue}+` : error;
    }
    /**
     * @ignore
     */
    onClickError(item, event) {
        // 阻止事件冒泡，触发父级事件
        event.stopPropagation();
        this.errorSelect.emit(item);
    }
}
TpResourceItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-resource-item',
                template: "<section class=\"tp-resource-body\">\r\n    <section class=\"tp-resource-card-title\">\r\n        <section class=\"tp-resource-card-name\"\r\n        tiOverflow\r\n        [tiTipContent]=\"(item.nameTip&&item.nameTip.content) ? item.nameTip.content : item.name\"\r\n        [tiTipPosition]=\"(item.nameTip&&item.nameTip.position) ? item.nameTip.position : ''\">{{item.name}}</section>\r\n    </section>\r\n    <section class=\"tp-resource-total\"\r\n    [tiTip]=\"(item.totalTip&&item.totalTip.content) ? item.totalTip.content : ''\"\r\n    [tiTipPosition]=\"(item.totalTip&&item.totalTip.position) ? item.totalTip.position : 'bottom-left'\">{{item.total}}</section>\r\n</section>\r\n\r\n<!--\u5F02\u5E38\u6570\u636E\u663E\u793A-->\r\n<section class=\"tp-resource-error\" tiOutline [tabindex]=\"0\" (keydown.enter)=\"onClickError(item, $event)\" (click)=\"onClickError(item, $event)\"\r\n*ngIf=\"item.error > 0\"\r\n[tiTip]=\"(item.errorTip&&item.errorTip.content) ? item.errorTip.content : ''\"\r\n[tiTipPosition]=\"(item.errorTip&&item.errorTip.position) ? item.errorTip.position : 'bottom-left'\">\r\n{{getErrorNumber(item.error, item.errorMax)}}</section>\r\n",
                styles: [":host{display:block;position:relative}.tp-resource-body{cursor:pointer;display:flex;justify-content:space-between;overflow:hidden}.tp-resource-card-title{color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-base);min-width:21px}.tp-resource-card-name{display:inline-block;line-height:22px;width:100%}.tp-resource-total{color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-3);font-weight:700;line-height:22px;padding-left:var(--ti-common-space-4x)}.tp-resource-error{background:var(--ti-common-color-error);border-radius:8px;color:var(--ti-common-color-text-white);cursor:pointer;font-size:var(--ti-common-font-size-base);height:16px;line-height:16px;min-width:16px;padding:0 var(--ti-common-space-base);position:absolute;right:-24px;text-align:center;top:-16px}"]
            },] }
];
TpResourceItemComponent.propDecorators = {
    item: [{ type: Input }],
    errorSelect: [{ type: Output }]
};
