// @cloud/tinyplus3/esm2015/components/resources/TpResourcesComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 资源列表 resources
 *
 * <example-url>../tinyplus3demo/#/resources/resources-all</example-url>
 */
export class TpResourcesComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 选中某项资源时，触发事件
         */
        this.select = new EventEmitter();
        /**
         * 选中某项资源异常数时，触发事件
         */
        this.errorSelect = new EventEmitter();
    }
    /**
     * @ignore
     */
    onClick(item) {
        this.select.emit(item);
    }
    /**
     * @ignore
     */
    onErrorSelect(item) {
        this.errorSelect.emit(item);
    }
}
TpResourcesComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-resources',
                template: "<div class=\"tp-resources-title\">{{title}}</div>\r\n<div class=\"tp-resources-items\">\r\n    <ng-container *ngFor=\"let item of items; index as i\">\r\n        <tp-card class=\"tp-resources-item\" *ngIf=\"!item.hide\" tiOutline [tabindex]=\"0\" (keydown.enter)=\"onClick(item)\" (click)=\"onClick(item)\">\r\n            <tp-resource-item [item]=\"item\" (errorSelect)=\"onErrorSelect(item)\" [id]=\"appendId('item_' + i)\"></tp-resource-item>\r\n        </tp-card>\r\n    </ng-container>\r\n</div>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host{display:block}.tp-resources-title{font-size:var(--ti-common-font-size-1);font-weight:700;line-height:20px;margin-bottom:var(--ti-common-space-4x)}.tp-resources-items{display:flex;flex-wrap:wrap;margin-bottom:calc(-1*var(--ti-common-space-3x));margin-right:calc(-1*var(--ti-common-space-3x))}.tp-resources-item{-ms-box-sizing:border-box;background:var(--tp-resources-card-default-bgColor);border:1px solid var(--tp-resources-card-border-color);box-sizing:border-box;cursor:pointer;height:40px;margin:0 var(--ti-common-space-3x) var(--ti-common-space-3x) 0}:host tp-card:not(.tp-card-disabled){padding:9px var(--ti-common-space-4x);width:calc(33.33333% - var(--ti-common-space-3x))}:host tp-card:not(.tp-card-disabled):hover{transition:.2s}:host tp-card:not(.tp-card-disabled):focus,:host tp-card:not(.tp-card-disabled):hover{background:var(--tp-resources-card-hover-bgColor);border-color:var(--tp-resources-card-border-color-hover)}@media (min-width:1760px){:host tp-card:not(.tp-card-disabled){width:calc(25% - var(--ti-common-space-3x))}}@keyframes tp-resource-scale-up-keyframes{0%{transform:scale(50%)}to{transform:scale(1)}}@keyframes tp-fade-in-keyframes{0%{opacity:0}to{opacity:1}}@keyframes tp-resource-translate-keyframes{0%{transform:translateY(10px)}to{transform:translate(0)}}.tp-resources-item{animation:tp-resource-translate-keyframes .25s var(--ti-timing-function-standard),tp-fade-in-keyframes .25s var(--ti-timing-function-standard) both}.tp-resource-error{animation:tp-resource-scale-up-keyframes .25s var(--ti-timing-function-standard) .15s,tp-fade-in-keyframes .25s var(--ti-timing-function-standard) .15s both}"]
            },] }
];
TpResourcesComponent.propDecorators = {
    title: [{ type: Input }],
    items: [{ type: Input }],
    select: [{ type: Output }],
    errorSelect: [{ type: Output }]
};
