// @cloud/tinyplus3/esm2015/components/resources/TpResourcesModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TpResourcesComponent } from './TpResourcesComponent';
import { TpResourceItemComponent } from './TpResourceItemComponent';
import { TiOverflowModule, TiTipModule, TiOutlineModule } from '@cloud/tiny3';
import { TpCardModule } from '../card/TpCardModule';
export class TpResourcesModule {
}
TpResourcesModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    TpCardModule,
                    TiOverflowModule,
                    TiTipModule,
                    TiOutlineModule
                ],
                exports: [TpResourcesComponent, TpResourceItemComponent],
                declarations: [TpResourcesComponent, TpResourceItemComponent]
            },] }
];
export { TpResourcesComponent } from './TpResourcesComponent';
export { TpResourceItemComponent } from './TpResourceItemComponent';
