// @cloud/tinyplus3/esm2015/components/progressbar/TpProgressbarComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 *  进度条组件
 *
 *  本组件是ti-progressbar 的业务封装，有2种使用场景：
 *
 *  1.无预警：进度条[必选]+百分比或分数[必选]
 *
 *  2.有预警：进度条[必选]+百分比或分数[必选]
 *
 *  场景2设置了warn 和 alarm 接口后生效
 *
 *  10.0.1 新增
 *
 * <example-url>../tinyplus3demo/#/progressbar/progressbar-all</example-url>
 */
export class TpProgressbarComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * label 显示类型，
         */
        this.labelType = 'percent';
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        this.updateLabel();
        this.updateWarnLevel();
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        if (changes['value'] && !changes['value'].firstChange ||
            (changes['warn'] && !changes['warn'].firstChange) ||
            changes['alarm'] && !changes['alarm'].firstChange) {
            this.updateLabel();
            this.updateWarnLevel();
        }
        if (changes['labelType'] && !changes['labelType'].firstChange) {
            this.updateLabel();
        }
    }
    updateWarnLevel() {
        if (this.warn && this.alarm) {
            if (this.value > this.alarm) {
                this.warnLevel = 'alarm';
            }
            else if (this.value > this.warn) {
                this.warnLevel = 'warn';
            }
            else {
                this.warnLevel = 'normal';
            }
        }
    }
    updateLabel() {
        // 百分比保留两位小数
        this.label = this.labelType === 'percent' ? `${Math.round(this.value / this.max * 10000) / 100}%` : `${this.value}/${this.max}`;
    }
}
TpProgressbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-progressbar',
                template: "<ti-progressbar\r\n    [ngClass]=\"{'tp-progressbar-alarm': warnLevel === 'alarm','tp-progressbar-warn': warnLevel === 'warn','tp-progressbar-normal': warnLevel === 'normal'}\"\r\n    [value]=\"value\" [max]=\"max\">\r\n    </ti-progressbar>\r\n<span class=\"tp-progressbar-label\" [id]=\"appendId('label')\">\r\n    {{label}}<ng-container *ngIf=\"unit\">({{unit}})</ng-container>\r\n</span>",
                styles: [":host ::ng-deep{align-items:center;display:flex;white-space:nowrap;width:100%}:host ::ng-deep ti-progressbar.ti3-progress{flex:1;width:auto}:host ::ng-deep .tp-progressbar-normal>.ti3-progress-bar{background-color:var(--ti-common-color-success)}:host ::ng-deep .tp-progressbar-warn>.ti3-progress-bar{background-color:var(--ti-common-color-warn-secondary)}:host ::ng-deep .tp-progressbar-alarm>.ti3-progress-bar{background-color:var(--ti-common-color-error)}.tp-progressbar-label{display:inline-block;margin-left:var(--ti-common-space-2x);vertical-align:top}"]
            },] }
];
TpProgressbarComponent.propDecorators = {
    labelType: [{ type: Input }],
    value: [{ type: Input }],
    max: [{ type: Input }],
    warn: [{ type: Input }],
    alarm: [{ type: Input }],
    unit: [{ type: Input }]
};
