// @cloud/tinyplus3/esm2015/components/linklist/TpLinklistComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 链接列表组件 linklist
 *
 * <example-url>../tinyplus3demo/#/linklist/linklist-all</example-url>
 */
export class TpLinklistComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 点击某个链接文本触发此事件
         */
        this.select = new EventEmitter();
    }
    /**
     * @ignore
     */
    onClick(item) {
        this.select.emit(item);
        // 不设置链接时，阻止跳转（a标签在click事件返回值为false时，会阻止默认行为）
        if (!item.href) {
            return false;
        }
    }
}
TpLinklistComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-linklist',
                template: "<ng-container *ngFor=\"let option of options; index as index\">\r\n    <div class=\"tp-linklist\" *ngIf=\"!option.hide\">\r\n        <div class=\"tp-linklist-title\" [id]=\"appendId('title_' + index)\">{{option.title}}</div>\r\n        <ul class=\"tp-link-list\" [id]=\"appendId('list_' + index)\">\r\n            <ng-container *ngFor=\"let item of option.items; index as i\">\r\n                <li class=\"tp-link-item\" *ngIf=\"!item.hide\">\r\n                    <a *ngIf=\"option.isNavigation else richTextTemplate\"\r\n                        rel=\"noopener noreferrer\"\r\n                        tiNavigation\r\n                       [href]=\"item.href\"\r\n                       [target]=\"item.target || '_blank'\"\r\n                       (click)=\"onClick(item)\"\r\n                       [id]=\"appendId('label_' + index + '_' + i)\">{{item.label}}</a>\r\n                    <ng-template #richTextTemplate>\r\n                        <a tiRichText\r\n                            rel=\"noopener noreferrer\"\r\n                           [href]=\"item.href\"\r\n                           [target]=\"item.target || '_blank'\"\r\n                           (click)=\"onClick(item)\"\r\n                           [id]=\"appendId('label_' + index + '_' + i)\">{{item.label}}</a>\r\n                    </ng-template>\r\n                </li>\r\n            </ng-container>\r\n        </ul>\r\n    </div>\r\n</ng-container>",
                styles: [":host{display:block}.tp-linklist{border-bottom:1px solid var(--ti-common-color-line-normal);margin-bottom:30px;padding-bottom:30px}.tp-linklist:last-child{border-bottom:0;margin-bottom:0;padding-bottom:0}.tp-linklist-title{color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-1);font-weight:700;line-height:20px;margin-bottom:var(--ti-common-space-4x)}.tp-link-list .tp-link-item{line-height:var(--ti-common-line-height-number);margin-bottom:var(--ti-common-space-3x)}.tp-link-list .tp-link-item:last-child{margin-bottom:0}"]
            },] }
];
TpLinklistComponent.propDecorators = {
    options: [{ type: Input }],
    select: [{ type: Output }]
};
