// @cloud/tinyplus3/esm2015/components/selectsearchbox/TpSelectsearchboxComponent.js
import { Component, ViewEncapsulation } from '@angular/core';
/**
 * Selectsearchbox组件由下拉选择组件和搜索框组件构成。
 *
 * Selectsearchbox组件使用的时候需要和下拉框（TiSelect）及搜索框（TiSearchBox）一起使用。
 *
 * API详情请参考下拉框（TiSelect）及搜索框（TiSearchBox）。
 *
 * <example-url>../tinyplus3demo/#/selectsearchbox/selectsearchbox-all</example-url>
 */
export class TpSelectsearchboxComponent {
}
TpSelectsearchboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-selectsearchbox',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.tp-selectsearchbox-container]': 'true'
                },
                styles: ["tp-selectsearchbox.tp-selectsearchbox-container{display:inline-block}tp-selectsearchbox.tp-selectsearchbox-container .ti3-searchbox-container{border-left:none}tp-selectsearchbox.tp-selectsearchbox-container .ti3-searchbox-container[tiFocused]:before{background-color:var(--ti-input-border-color-focus)!important;bottom:-1px;top:-1px}tp-selectsearchbox.tp-selectsearchbox-container ti-select{border-bottom-right-radius:0;border-right:none;border-top-right-radius:0;position:relative}tp-selectsearchbox.tp-selectsearchbox-container ti-select[tiFocused]+.ti3-searchbox-container:before{background-color:var(--ti-input-border-color-focus);bottom:-1px;top:-1px}tp-selectsearchbox.tp-selectsearchbox-container ti-searchbox-notsearch.ti3-searchbox-container,tp-selectsearchbox.tp-selectsearchbox-container ti-searchbox.ti3-searchbox-container{border-bottom-left-radius:0;border-top-left-radius:0}tp-selectsearchbox.tp-selectsearchbox-container .ti3-searchbox-container:before{background-color:var(--ti-common-color-line-dividing);bottom:5px;content:\"\";left:0;position:absolute;top:5px;width:1px}tp-selectsearchbox.tp-selectsearchbox-container:hover .ti3-searchbox-container:before{background-color:var(--ti-input-border-color-hover);bottom:-1px;top:-1px}"]
            },] }
];
