// @cloud/tinyplus3/esm2015/components/copy/TpCopyComponent.js
import { Component, ElementRef, Input, Renderer2 } from '@angular/core';
import { TiBaseComponent, TiLocale, TiPopupService, TiPosition } from '@cloud/tiny3';
import { Subject } from 'rxjs';
import { delay, throttleTime } from 'rxjs/operators';
import { TpToastComponent } from './TpToastComponent';
/**
 * 复制组件，用于复制文本
 *
 * <example-url>../tinyplus3demo/#/copy/copy-all</example-url>
 */
export class TpCopyComponent extends TiBaseComponent {
    constructor(hostRef, Renderer, tiPopup) {
        super(hostRef, Renderer);
        this.hostRef = hostRef;
        this.Renderer = Renderer;
        this.tiPopup = tiPopup;
        this.clickObserve = new Subject();
        this.hideToast = new Subject();
        this.deleteToast = new Subject();
        this.hostSpace = 4; // toast距copy图标的距离
    }
    ngOnInit() {
        super.ngOnInit();
        // 创建点击事件的observe
        this.createClickObserve();
        // 创建hideToast的observe
        this.createHideToastObserve();
        // 创建deleteToast的observe
        this.createDeleteToastObserve();
    }
    /**
     * @ignore
     * 图标点击事件
     */
    onClick() {
        if (!this.content) {
            return;
        }
        this.clickObserve.next();
    }
    createClickObserve() {
        this.clickSub = this.clickObserve.pipe(throttleTime(2000))
            .subscribe(() => {
            // 创建textarea
            const textarea = this.Renderer.createElement('textarea');
            this.Renderer.appendChild(document.body, textarea);
            this.Renderer.setStyle(textarea, 'postion', 'absolute');
            this.Renderer.setStyle(textarea, 'top', '-9999px');
            this.Renderer.setStyle(textarea, 'left', '-9999px');
            this.Renderer.setProperty(textarea, 'value', this.content);
            textarea.select();
            document.execCommand('Copy');
            this.Renderer.removeChild(document.body, textarea);
            // 复制成功后，显示即时提示信息
            this.showToast();
            // 显示后隐藏
            this.hideToast.next();
        });
    }
    createHideToastObserve() {
        this.hideToastSub = this.hideToast.pipe(delay(1000))
            .subscribe(() => {
            this.Renderer.addClass(this.toastRef.location.nativeElement, 'fadeout');
            // 隐藏后删除
            this.deleteToast.next();
        });
    }
    createDeleteToastObserve() {
        this.deleteToastSub = this.deleteToast.pipe(delay(300))
            .subscribe(() => {
            this.toast.hide();
        });
    }
    /**
     * 复制成功后，显示toast提示
     */
    showToast() {
        this.toast = this.tiPopup.create(TpToastComponent);
        this.toastRef = this.toast.show({
            content: this.successTip || TiLocale.getLocaleWords().tpCopy.successTip,
            container: 'body'
        });
        TiPosition.setPosition({
            hostEle: this.hostRef.nativeElement,
            targetEle: this.toastRef.location.nativeElement,
            hostSpace: this.hostSpace,
            fixMaxHeight: true,
            determinPositionFn: (layout) => {
                if (layout.avilableLayout.right >= layout.targetLayout.width + this.hostSpace) {
                    return 'right';
                }
                else {
                    return 'left';
                }
            }
        });
        this.Renderer.addClass(this.toastRef.location.nativeElement, 'fadein');
    }
    ngOnDestroy() {
        if (this.toast) {
            this.toast.hide();
        }
        this.clickSub.unsubscribe();
        this.hideToastSub.unsubscribe();
        this.deleteToastSub.unsubscribe();
    }
}
TpCopyComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-copy',
                template: "<tp-icon local tiOutline\r\n    name=\"cloud-action-copy\"\r\n    (click)=\"onClick()\"\r\n    [tabindex]=\"0\"\r\n    (keydown.enter)=\"onClick()\"\r\n    [id]=\"appendId('copyicon')\">\r\n</tp-icon>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host{color:var(--ti-common-color-icon-normal);cursor:pointer;margin-left:var(--ti-common-space-2x)}:host:hover{color:var(--ti-common-color-icon-hover)}:host[dark]{color:var(--ti-common-color-icon-darkbg-normal)}:host[dark]:hover{color:var(--ti-common-color-icon-darkbg-hover)}ti-table td :host{visibility:hidden}ti-table td:hover :host{visibility:visible}"]
            },] }
];
TpCopyComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TiPopupService }
];
TpCopyComponent.propDecorators = {
    content: [{ type: Input }],
    successTip: [{ type: Input }]
};
