// @cloud/tinyplus3/esm2015/components/halfmodal/TpHalfmodalComponent.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { TiBaseComponent, Util, TiKeymap } from '@cloud/tiny3';
/**
 * 右侧半屏弹窗组件
 *
 * 半屏弹窗分为两种，1.不带灰色背景层，点击弹窗外会关闭弹窗；2.带灰色背景层，点击弹窗外不能关闭弹窗。
 *
 * 两种弹窗都可以通过点击右上角“X”图标或者hide()方法关闭弹窗
 *
 * 注：第一类弹窗，点击弹窗外关闭的原理是在document上注册点击事件触发`TpHalfmodalComponent.hide()`，如果不想关闭弹窗可以阻止事件冒泡。
 *
 * <example-url>../tinyplus3demo/#/halfmodal/halfmodal-all</example-url>
 */
export class TpHalfmodalComponent extends TiBaseComponent {
    constructor(elementRef, renderer2) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        /**
         * 10.0.2新增
         *
         * 是否含有遮罩层，默认没有遮罩层；
         */
        this.backdrop = false;
        /**
         * 10.1.3新增
         * 弹窗顶部距窗口顶部的距离
         */
        this.clientRectTop = '50px';
        /**
         * 10.1.3新增
         * 是否显示关闭图标
         */
        this.closeIcon = true;
        /**
         * 10.1.3新增
         * 关闭前的事件接口，如果监听beforeHide，则由业务关闭弹窗
         */
        this.beforeHide = new EventEmitter();
        /**
         * @ignore
         */
        this.isShow = false;
        this.isWantShow = false;
        // 可聚焦元素
        this.focusableElementsString = `a[href], area[href], input:not([disabled]):not([tabindex=\'-1\']),
    button:not([disabled]):not([tabindex=\'-1\']),select:not([disabled]):not([tabindex=\'-1\']),
    textarea:not([disabled]):not([tabindex=\'-1\']),
    iframe, object, embed, *[tabindex]:not([tabindex=\'-1\']), *[contenteditable=true]`;
    }
    /**
     * @description 向上查找最近的指定祖先元素
     * @param ele 子元素
     * @param parentNode 祖先元素标签名
     * @return 找到的祖先元素或者undefined
     */
    static findClosest(ele, parentNodeName) {
        if (ele.nodeName === 'HTML' || Util.isNull(ele.parentNode)) {
            return undefined;
        }
        else if (ele.parentNode.nodeName === parentNodeName) {
            return ele.parentNode;
        }
        else {
            return this.findClosest(ele.parentNode, parentNodeName);
        }
    }
    /**
     * 弹窗显示事件
     * @param element 通过事件触发halfModal显示的元素
     *
     * 10.1.3版本前，如果完全是通过JS调用打开半屏弹窗，则不需该参数，点击弹窗外任意位置会关闭弹窗
     * 10.1.3及以上不需要参数
     */
    show(element) {
        this.isShow = true;
        this.isWantShow = true;
        this.renderer2.addClass(this.nativeElement, 'tp-halfmodal-animation');
        if (this.backdrop && !this.backdropEle) {
            this.backdropEle = this.renderer2.createElement('div');
            this.renderer2.addClass(this.backdropEle, 'tp-halfmodal-backdrop');
            this.renderer2.appendChild(document.body, this.backdropEle);
            // 防止多次打开半屏弹窗的时候，事件注册多次。
            if (!this.unListenDocumentKeydown) {
                this.unListenDocumentKeydown = this.renderer2.listen(document, 'keydown', (event) => {
                    switch (event.which) {
                        case TiKeymap.KEY_TAB: // tab键用于处理在提示框内循环获取焦点
                            this.clickTab(event);
                            break;
                        default:
                            break;
                    }
                });
            }
        }
        else {
            if (!this.unlistenClick) {
                this.unlistenClick = this.renderer2.listen(document, 'click', ($event) => {
                    if (element && element.contains($event.target) || this.isWantShow) {
                        this.isWantShow = false;
                        return;
                    }
                    if (TpHalfmodalComponent.findClosest($event.target, 'TI-MESSAGE') || TpHalfmodalComponent.findClosest($event.target, 'TI-MODAL-WRAPPER')) {
                        return;
                    }
                    // 判断鼠标是否在弹窗范围内
                    if (document.documentElement.clientWidth - $event.clientX > this.nativeElement.offsetWidth
                        || document.documentElement.clientHeight - $event.clientY > this.nativeElement.offsetHeight) {
                        this.wantHide();
                    }
                });
            }
        }
    }
    clickTab(event) {
        const introModal = document.querySelector('.tp-halfmodal-animation');
        const focusableElements = introModal === null || introModal === void 0 ? void 0 : introModal.querySelectorAll(this.focusableElementsString);
        Util.focusInDialogOnTabchange(event, focusableElements);
    }
    /**
     * @ignore
     * 即将关闭，如果监听beforeHide，则由业务关闭弹窗
     */
    wantHide() {
        if (this.beforeHide.observers.length > 0) {
            this.beforeHide.emit(this);
        }
        else {
            this.hide();
        }
    }
    /**
     * 10.0.2对外提供
     *
     * 弹窗隐藏事件
     */
    hide() {
        this.hideBackdrop();
        this.isShow = false;
        if (this.unlistenClick) {
            this.unlistenClick();
            this.unlistenClick = undefined;
        }
    }
    hideBackdrop() {
        if (this.backdropEle) {
            this.renderer2.removeChild(document.body, this.backdropEle);
            this.backdropEle = undefined;
        }
    }
    ngOnDestroy() {
        this.hideBackdrop();
        if (this.unlistenClick) {
            this.unlistenClick();
            this.unlistenClick = undefined;
        }
        if (this.unListenDocumentKeydown) {
            this.unListenDocumentKeydown();
            this.unListenDocumentKeydown = undefined;
        }
    }
}
TpHalfmodalComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-halfmodal',
                template: "\r\n<div *ngIf=\"isShow\" style=\"height: 100%\">\r\n    <tp-icon local tiOutline name=\"cloud-close\"\r\n        *ngIf=\"closeIcon\"\r\n        class=\"tp-halfmodal-close\"\r\n        (click)=\"wantHide()\" [id]=\"appendId('halfmodal_close')\" [attr.tabindex]=\"0\" (keydown.enter)=\"hide()\"></tp-icon>\r\n    <div class=\"tp-halfmodal-cotent\">\r\n        <ng-content></ng-content>\r\n    </div>\r\n</div>",
                host: {
                    '[style.width]': 'width',
                    '[style.top]': 'clientRectTop',
                    '[style.right]': 'width ? "-" + width : "-600px"',
                    '[hidden]': '!isShow'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-halfmodal-content-space:var(--ti-common-space-2x);--tp-halfmodal-space:var(--ti-common-space-8x);--tp-halfmodal-width:600px}:host{-ms-box-sizing:border-box;background-color:var(--ti-common-color-bg-white-normal);bottom:0;box-shadow:var(--ti-common-shadow-4-left);box-sizing:border-box;overflow-y:auto;padding:var(--tp-halfmodal-space);position:fixed;width:var(--tp-halfmodal-width);z-index:999}:host .tp-halfmodal-close{cursor:pointer;position:absolute;right:20px;top:20px}:host .tp-halfmodal-close:hover{color:var(--ti-common-color-icon-hover)}:host .tp-halfmodal-cotent{height:100%;padding-right:var(--tp-halfmodal-content-space)}::ng-deep.tp-halfmodal-backdrop{background-color:var(--ti-modal-backdrop-bg-color);bottom:0;box-sizing:border-box;left:0;opacity:.2;position:fixed;right:0;top:0;z-index:998}::ng-deep.tp-halfmodal-animation{animation:show .25s forwards}@keyframes show{to{right:0}}"]
            },] }
];
TpHalfmodalComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TpHalfmodalComponent.propDecorators = {
    width: [{ type: Input }],
    backdrop: [{ type: Input }],
    clientRectTop: [{ type: Input }],
    closeIcon: [{ type: Input }],
    beforeHide: [{ type: Output }]
};
