// @cloud/tinyplus3/esm2015/components/count/TpCountPatternComponent.js
// tslint:disable-next-line: no-implicit-dependencies
import { Component, Input } from '@angular/core';
// tslint:disable-next-line: no-implicit-dependencies
import { TiBaseComponent, TiLocale } from '@cloud/tiny3';
/**
 * 表格倒计时组件
 *
 * <example-url>../tinyplus3demo/#/count/count-all</example-url>
 */
export class TpCountPatternComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 计费模式
         */
        this.type = 'period';
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        this.updateText();
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        if (changes['type'] && !changes['type'].firstChange) {
            this.updateText();
        }
    }
    updateText() {
        this.text = this.type === 'period' ?
            TiLocale.getLocaleWords().tpCount.periodCharging : TiLocale.getLocaleWords().tpCount.ondemandCharging;
    }
}
TpCountPatternComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-count-pattern',
                template: `
      {{ text }}
    `
            },] }
];
TpCountPatternComponent.propDecorators = {
    type: [{ type: Input }]
};
