// @cloud/tinyplus3/esm2015/components/count/TpCountModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiLocaleModule, TiOverflowModule, TiTipModule } from '@cloud/tiny3';
import { TpLocaleModule } from '../../pipes/locale/TpLocaleModule';
import { TpCountComponent } from './TpCountComponent';
import { TpCountTimeComponent } from './TpCountTimeComponent';
import { TpCountStateComponent } from './TpCountStateComponent';
import { TpCountPatternComponent } from './TpCountPatternComponent';
export class TpCountModule {
}
TpCountModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiTipModule,
                    TpLocaleModule,
                    TiOverflowModule,
                    TiLocaleModule
                ],
                exports: [TpCountComponent, TpCountTimeComponent, TpCountStateComponent, TpCountPatternComponent],
                declarations: [TpCountComponent, TpCountTimeComponent, TpCountStateComponent, TpCountPatternComponent]
            },] }
];
export { TpCountComponent } from './TpCountComponent';
export { TpCountPatternComponent } from './TpCountPatternComponent';
export { TpCountStateComponent } from './TpCountStateComponent';
export { TpCountTimeComponent } from './TpCountTimeComponent';
