// @cloud/tinyplus3/esm2015/components/count/TpCountTimeComponent.js
import { Component, Input } from '@angular/core';
// tslint:disable-next-line: no-implicit-dependencies
import { TiBaseComponent, TiLocale } from '@cloud/tiny3';
import { TpDatePipe } from '../../pipes/locale/TpDatePipe';
/**
 * 表格倒计时组件
 *
 * <example-url>../tinyplus3demo/#/count/count-all</example-url>
 */
export class TpCountTimeComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * @ignore
         * 内部使用，only 单独使用，combine 和tp-count-state 一起使用
         */
        this.pattern = 'only';
        /**
         * @ignore
         * 倒计时最后显示的文本
         */
        this.countdownText = '';
        /**
         * @ignore
         * tip 的内容
         */
        this.tipContext = [];
        /**
         * @ignore
         * tip 的最大宽度
         */
        this.tipMaxWidth = '500px';
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        // 生成倒计时内容
        this.generateContent();
        // 生成tip内容
        if (this.showTip()) {
            this.generateTipContent();
        }
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        // 倒计时内容跟countDownCode和countDownInfos有关
        // 倒计时的颜色提示类别跟countDownCode和countDownInfos有关
        // 倒计时tip提示内容跟 countDownCode和countDownTips有关
        if (changes['countDownCode'] && !changes['countDownCode'].firstChange) {
            this.generateContent();
            this.updateStateLevel();
            this.generateTipContent();
            return;
        }
        if (changes['countDownInfos'] && !changes['countDownInfos'].firstChange) {
            this.generateContent();
            this.updateStateLevel();
        }
        if (changes['countDownTips'] && !changes['countDownTips'].firstChange) {
            this.generateTipContent();
        }
    }
    /**
     * @ignore
     * 是否需要tip
     */
    showTip() {
        // tp-count-time 单独使用且资源倒计时不是按需使用状态
        return this.countDownTips && this.countDownCode !== 'hws_countdown_ondemand_using';
    }
    /**
     * 生成倒计时内容
     */
    generateContent() {
        // 时间间隔
        let timespan = '';
        // 时间信息
        let hourInfo = '';
        let minuteInfo = '';
        //  根据返回时间拼接出时间间隔  {d} 天数 {h} 小时数 {m} 分钟数
        if (this.countDownInfos) {
            if (this.countDownInfos.nextOperationRemainingDay) {
                timespan = this.countDownInfos.nextOperationRemainingDay > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.day
                        .replace(/{d}/, this.countDownInfos.nextOperationRemainingDay) :
                    TiLocale.getLocaleWords().tpCountdown.day
                        .replace(/s/, '')
                        .replace(/{d}/, this.countDownInfos.nextOperationRemainingDay);
            }
            else if (this.countDownInfos.nextOperationRemainingHour) {
                hourInfo = this.countDownInfos.nextOperationRemainingHour > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.hour
                        .replace(/{h}/, this.countDownInfos.nextOperationRemainingHour) :
                    TiLocale.getLocaleWords().tpCountdown.hour
                        .replace(/s/, '')
                        .replace(/{h}/, this.countDownInfos.nextOperationRemainingHour);
                minuteInfo = this.countDownInfos.nextOperationRemainingMinute > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute) :
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/s/, '')
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute);
                timespan = `${hourInfo} ${minuteInfo}`;
            }
            else if (this.countDownInfos.nextOperationRemainingMinute) {
                timespan = this.countDownInfos.nextOperationRemainingMinute > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute) :
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/s/, '')
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute);
            }
            else if (this.countDownInfos.nextOperationRemainingSecond) {
                timespan = TiLocale.getLocaleWords().tpCountdown.second;
            }
        }
        // 根据状态码，得到倒计时整体字符串。 {TS} 时间间隔
        const statusKey = this.countDownCode ? this.countDownCode.slice(14) : '';
        if (TiLocale.getLocaleWords().tpCount[statusKey]) {
            this.countdownText = TiLocale.getLocaleWords().tpCount[statusKey]
                .replace(/{TS}/, timespan);
        }
        // 如果是按需正常使用，没有倒计时，只提示创建时间，单独处理 {ymdhms} 是 格式为yyyy/mm/dd  hh:mm:ss 日期时间
        if (this.countDownCode === 'hws_countdown_ondemand_using') {
            // tslint:disable-next-line: max-line-length
            this.countdownText = this.countdownText.replace(/{ymdhms}/, new TpDatePipe().transform(this.countDownInfos.createTime, 'dateTimeZone'));
        }
        // 更新样式提示
        this.updateStateLevel();
    }
    updateStateLevel() {
        // 添加样式类 宽限期，冻结期，违规冻结 提示红色，正常使用，如果小于七天，提示黄色
        if (this.countDownInfos.status > 2) {
            this.stateLevel = 'alarm';
        }
        else if ((this.countDownCode === 'hws_countdown_period_using' || this.countDownCode === 'hws_countdown_period_toondemand')
            && this.isDayLess7()) {
            this.stateLevel = 'warn';
        }
    }
    // 倒计时是否小于七天
    isDayLess7() {
        return !(this.countDownInfos.nextOperationRemainingDay && this.countDownInfos.nextOperationRemainingDay > 7);
    }
    /**
     * @ignore
     * 生成tips 内容
     */
    generateTipContent() {
        // 当tp-count-time 作为 tp-count 子元素时不传countDownTips
        if (!this.countDownTips) {
            return;
        }
        this.tipContext = [];
        switch (this.countDownCode) {
            case 'hws_countdown_period_using':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.create_time}`,
                    time: `${this.countDownTips.effTime}`
                });
                this.tipContext.push({
                    title: ` ${TiLocale.getLocaleWords().tpCountdown.expire_time} `,
                    time: `${this.countDownTips.expTime}`
                });
                break;
            case 'hws_countdown_period_gracing':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.expire_time}`,
                    time: `${this.countDownTips.expTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.countDownTips.graceTime}`
                });
                break;
            case 'hws_countdown_period_frozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.countDownTips.frozenTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.countDownTips.releaseTime}`
                });
                break;
            case 'hws_countdown_period_illegalfrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.countDownTips.frozenTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.countDownTips.releaseTime}`
                });
                break;
            case 'hws_countdown_period_policefrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.countDownTips.frozenTime}`
                });
                break;
            case 'hws_countdown_period_toondemand':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.create_time}`,
                    time: `${this.countDownTips.effTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.toonneed_time}`,
                    time: `${this.countDownTips.expTime}`
                });
                break;
            case 'hws_countdown_ondemand_using':
                // 按需正在使用不添加tip
                break;
            case 'hws_countdown_ondemand_gracing':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.arrears_time}`,
                    time: `${this.countDownTips.expTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.countDownTips.graceTime}`
                });
                break;
            case 'hws_countdown_ondemand_frozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.countDownTips.frozenTime}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time} `,
                    time: `${this.countDownTips.releaseTime}`
                });
                break;
            case 'hws_countdown_ondemand_illegalfrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.countDownTips.frozenTime}`
                });
                break;
            case 'hws_countdown_ondemand_illegalfrozen_arrear':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: ` ${this.countDownTips.frozenTime} `
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.countDownTips.releaseTime}`
                });
                break;
            case 'hws_countdown_ondemand_policefrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.countDownTips.frozenTime}`
                });
                break;
            default:
                break;
        }
    }
}
TpCountTimeComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-count-time',
                template: "<!-- \u5355\u72EC\u4F7F\u7528\uFF0C\u5904\u7406\u6DFB\u52A0tip -->\r\n<ng-container *ngIf=\"pattern === 'only'\">\r\n    <!-- \u65E0\u8BBA\u6587\u672C\u662F\u5426\u6EA2\u51FA\uFF0C\u90FD\u8981\u6709tip \uFF0C\u4E14tip\u5185\u5BB9\u4E0D\u662F\u5143\u7D20\u5185\u6587\u672C -->\r\n    <span class=\"tp-count-time-text tp-count-time-underline\" *ngIf=\"showTip()\" [ngClass]=\"{'tp-count-warn': stateLevel === 'warn',\r\n        'tp-count-alarm' : stateLevel === 'alarm'}\" [tiTip]=\"tipContent\" [tiTipContext]='tipContext'\r\n        [tiTipMaxWidth]='tipMaxWidth' [id]=\"appendId('countdown_text')\">{{countdownText}}</span>\r\n    <!-- \u6587\u672C\u6EA2\u51FA\u65F6\u624D\u51FAtip,\u4E14tip\u5185\u5BB9\u4E3A\u5143\u7D20\u6587\u672C -->\r\n    <span class=\"tp-count-time-text\" *ngIf=\"!showTip()\" [id]=\"appendId('countdown_text')\"  tiOverflow>\r\n        {{countdownText}}\r\n    </span>\r\n</ng-container>\r\n<!-- \u5728tp-count \u4E2D\u548Ctp-count-state \u4E00\u8D77\u4F7F\u7528\u4E0D\u9700\u8981\u5904\u7406tip -->\r\n<ng-container *ngIf=\"pattern === 'combine'\">\r\n    <span  class=\"tp-count-time-text\" [id]=\"appendId('countdown_text')\">\r\n        {{countdownText}}\r\n    </span>\r\n</ng-container>\r\n<ng-template #tipContent let-context='context'>\r\n    <ng-container *ngFor=\"let item of context\">{{item.title}} {{item.time| tpDate: 'dateTimeZone'}}<br></ng-container>\r\n</ng-template>",
                host: {
                    // 标记 tp-count-time 是单独使用还是在tp-count中和tp-count-state一起使用
                    '[class.tp-count-time-only]': 'pattern === "only" '
                },
                styles: [":host{display:inline}:host.tp-count-time-only .tp-count-time-text{box-sizing:border-box;display:inline-block;max-width:100%}:host.tp-count-time-only .tp-count-time-text.tp-count-time-underline{border-bottom:1px dashed var(--ti-common-color-dash-line-normal);overflow:hidden;text-overflow:ellipsis;white-space:nowrap}"]
            },] }
];
TpCountTimeComponent.propDecorators = {
    countDownCode: [{ type: Input }],
    countDownInfos: [{ type: Input }],
    countDownTips: [{ type: Input }],
    pattern: [{ type: Input }]
};
