// @cloud/tinyplus3/esm2015/components/collapsebutton/TpCollapsebuttonModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiButtonModule, TiLocaleModule } from '@cloud/tiny3';
import { TpIconModule } from '../icon/TpIconModule';
import { TpCollapsebuttonComponent } from './TpCollapsebuttonComponent';
import { TpCollapsepanelComponent } from './TpCollapsepanelComponent';
export class TpCollapsebuttonModule {
}
TpCollapsebuttonModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiButtonModule,
                    TiLocaleModule,
                    TpIconModule
                ],
                exports: [
                    TpCollapsebuttonComponent,
                    TpCollapsepanelComponent
                ],
                declarations: [
                    TpCollapsebuttonComponent,
                    TpCollapsepanelComponent
                ]
            },] }
];
export { TpCollapsebuttonComponent } from './TpCollapsebuttonComponent';
export { TpCollapsepanelComponent } from './TpCollapsepanelComponent';
