// @cloud/tinyplus3/esm2015/components/countdown/TpCountdownComponent.js
// tslint:disable-next-line: no-implicit-dependencies
import { Component, Input } from '@angular/core';
import { TiLocale, TiBaseComponent } from '@cloud/tiny3';
import { CfUtil } from '../../cfutils/CfUtil';
import { TpDatePipe } from '../../pipes/locale/TpLocaleModule';
/**
 * 资源倒计时组件
 * <example-url>../tinyplus3demo/#/countdown/countdown-all</example-url>
 */
export class TpCountdownComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 是否显示资源的状态
         */
        this.showStatus = true;
        /**
         * @ignore
         * 倒计时最后显示的文本
         */
        this.countdownText = '';
        /**
         * @ignore
         * tip 的内容
         */
        this.tipContext = [];
        /**
         * @ignore
         * tip 的最大宽度
         */
        this.tipMaxWidth = '500px';
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        // 生成倒计时内容
        this.generateContent();
        // 生成tip内容
        this.generateTipContent();
    }
    /**
     * 生成倒计时内容
     */
    generateContent() {
        // 时间间隔
        let timespan = '';
        // 时间信息
        let hourInfo = '';
        let minuteInfo = '';
        //  根据返回时间拼接出时间间隔  {d} 天数 {h} 小时数 {m} 分钟数
        if (this.countDownInfos) {
            if (this.countDownInfos.nextOperationRemainingDay) {
                timespan = this.countDownInfos.nextOperationRemainingDay > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.day
                        .replace(/{d}/, this.countDownInfos.nextOperationRemainingDay) :
                    TiLocale.getLocaleWords().tpCountdown.day
                        .replace(/s/, '')
                        .replace(/{d}/, this.countDownInfos.nextOperationRemainingDay);
            }
            else if (this.countDownInfos.nextOperationRemainingSecond) {
                timespan = TiLocale.getLocaleWords().tpCountdown.second;
            }
            else if (this.countDownInfos.nextOperationRemainingHour) {
                hourInfo = this.countDownInfos.nextOperationRemainingHour > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.hour
                        .replace(/{h}/, this.countDownInfos.nextOperationRemainingHour) :
                    TiLocale.getLocaleWords().tpCountdown.hour
                        .replace(/s/, '')
                        .replace(/{h}/, this.countDownInfos.nextOperationRemainingHour);
                minuteInfo = this.countDownInfos.nextOperationRemainingMinute > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute) :
                    TiLocale.getLocaleWords().tpCountdown.minute
                        .replace(/s/, '')
                        .replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute);
                timespan = hourInfo + ' ' + minuteInfo;
            }
            else if (this.countDownInfos.nextOperationRemainingMinute) {
                timespan = this.countDownInfos.nextOperationRemainingMinute > 1 ?
                    TiLocale.getLocaleWords().tpCountdown.minute.replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute) :
                    TiLocale.getLocaleWords().tpCountdown.minute.replace(/s/, '').replace(/{m}/, this.countDownInfos.nextOperationRemainingMinute);
            }
        }
        // 根据状态码，得到倒计时整体字符串。 {TS} 时间间隔
        const statusKey = this.countDownCode ? this.countDownCode.slice(14) : '';
        if (TiLocale.getLocaleWords().tpCountdown[statusKey]) {
            this.countdownText = TiLocale.getLocaleWords().tpCountdown[statusKey]
                .replace(/{TS}/, timespan);
        }
        // 如果是按需正常使用，没有倒计时，只提示创建时间，单独处理 {ymdhms} 是 格式为本地化日期时间
        if (this.countDownCode === 'hws_countdown_ondemand_using') {
            this.countdownText = this.countdownText.replace(/{ymdhms}/, new TpDatePipe().transform(this.countDownTips.effTime, 'dateTime'));
        }
        // 是否展示状态
        if (!this.showStatus) {
            this.countdownText = this.countdownText.split(TiLocale.getLocaleWords().tpCountdown.spliter)
                .pop();
        }
        // 添加样式类 宽限期，冻结期，违规冻结 提示红色，正常使用，如果小于七天，提示黄色
        if (this.countDownInfos.status > 2) {
            this.nativeElement.classList.add('tp-countdown-alarm');
        }
        else if ((this.countDownCode === 'hws_countdown_period_using' || this.countDownCode === 'hws_countdown_period_toondemand')
            && this.isDayLess7()) {
            this.nativeElement.classList.add('tp-countdown-warn');
        }
    }
    // 倒计时是否小于七天
    isDayLess7() {
        return !(this.countDownInfos.nextOperationRemainingDay && this.countDownInfos.nextOperationRemainingDay > 7);
    }
    /**
     * 生成tips 内容
     */
    generateTipContent() {
        switch (this.countDownCode) {
            case 'hws_countdown_period_using':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.create_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.effTime)}`
                });
                this.tipContext.push({
                    title: ` ${TiLocale.getLocaleWords().tpCountdown.expire_time} `,
                    time: `${this.formatLocaleDateTime(this.countDownTips.expTime)}`
                });
                break;
            case 'hws_countdown_period_gracing':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.expire_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.expTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.formatLocaleDateTime(this.countDownTips.graceTime)}`
                });
                break;
            case 'hws_countdown_period_frozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.releaseTime)}`
                });
                break;
            case 'hws_countdown_period_illegalfrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.releaseTime)}`
                });
                break;
            case 'hws_countdown_period_policefrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time} `,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                break;
            case 'hws_countdown_period_toondemand':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.create_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.effTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.toonneed_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.expTime)}`
                });
                break;
            case 'hws_countdown_ondemand_using':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.create_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.effTime)}`
                });
                break;
            case 'hws_countdown_ondemand_gracing':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.arrears_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.expTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.graceTime)}`
                });
                break;
            case 'hws_countdown_ondemand_frozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time} `,
                    time: `${this.formatLocaleDateTime(this.countDownTips.releaseTime)}`
                });
                break;
            case 'hws_countdown_ondemand_illegalfrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                break;
            case 'hws_countdown_ondemand_illegalfrozen_arrear':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: ` ${this.formatLocaleDateTime(this.countDownTips.frozenTime)} `
                });
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.release_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.releaseTime)}`
                });
                break;
            case 'hws_countdown_ondemand_policefrozen':
                this.tipContext.push({
                    title: `${TiLocale.getLocaleWords().tpCountdown.frozen_time}`,
                    time: `${this.formatLocaleDateTime(this.countDownTips.frozenTime)}`
                });
                break;
            default:
                break;
        }
    }
    /**
     * 日期时间戳转换成带时区的
     */
    formatLocaleDateTime(req) {
        return CfUtil.formatLocaleDateTime(req);
    }
}
TpCountdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-countdown',
                template: "<span [tiTip]=\"tipContent\" [tiTipContext]='tipContext' [tiTipMaxWidth]='tipMaxWidth' [id]=\"appendId('countdown_text')\">{{countdownText}}</span>\r\n<ng-template #tipContent let-context='context'>\r\n    <ng-container *ngFor=\"let item of context\">{{item.title}} {{item.time}}<br></ng-container>\r\n</ng-template>\r\n",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host{border-bottom:1px dashed var(--ti-common-color-dash-line-normal);box-sizing:border-box;color:var(--ti-common-color-text-primary);cursor:default;display:inline;font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number)}:host.tp-countdown-warn{color:var(--ti-common-color-warn-text)}:host.tp-countdown-alarm{color:var(--ti-common-color-error-text)}"]
            },] }
];
TpCountdownComponent.propDecorators = {
    showStatus: [{ type: Input }],
    countDownCode: [{ type: Input }],
    countDownInfos: [{ type: Input }],
    countDownTips: [{ type: Input }]
};
