// @cloud/tiny3/esm2015/directives/popconfirm/TiPopconfirmModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiPopconfirmComponent } from './TiPopconfirmComponent';
import { TiTipServiceModule } from '../../services/tip/TiTipServiceModule';
import { TiLocaleModule } from '../../locale/TiLocaleModule';
import { TiIconModule } from '../../components/icon/TiIconModule';
import { TiButtonModule } from '../../components/button/TiButtonModule';
import { TiPopconfirmDirective } from './TiPopconfirmDirective';
export class TiPopconfirmModule {
}
TiPopconfirmModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiTipServiceModule,
                    TiLocaleModule,
                    TiIconModule,
                    TiButtonModule
                ],
                exports: [TiPopconfirmDirective],
                declarations: [TiPopconfirmComponent, TiPopconfirmDirective],
                entryComponents: [TiPopconfirmComponent]
            },] }
];
export { TiPopconfirmComponent } from './TiPopconfirmComponent';
export { TiPopconfirmDirective } from './TiPopconfirmDirective';
