// @cloud/tiny3/esm2015/directives/popconfirm/TiPopconfirmComponent.js
import { Component, Input } from '@angular/core';
import { TiLocale } from '../../locale/TiLocaleModule';
import { TiBaseComponent } from '../../components/base/TiBaseModule';
export class TiPopconfirmComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        this.tiPopconfirm = TiLocale.getLocaleWords().tiPopconfirm;
    }
    ngOnInit() {
        this.config.yesText = this.config.yesText || this.tiPopconfirm.yesLabel;
        this.config.noText = this.config.noText || this.tiPopconfirm.noLabel;
    }
    ngAfterViewInit() {
        // 轻量级弹窗阴影定制
        this.renderer.addClass(this.nativeElement.offsetParent, 'ti3-popconfirm-tip'); // 在init中设置会影响button样式 TODO: IE下还存在背景闪烁问题
    }
}
TiPopconfirmComponent.decorators = [
    { type: Component, args: [{
                template: "<div class=\"ti3-popconfirm-header\" >\r\n    <ti-icon class=\"ti3-popconfirm-warn-icon\" name=\"warn\"></ti-icon>\r\n    <span *ngIf='config.content' class=\"ti3-popconfirm-title\" [innerHtml]='config.content'></span>\r\n</div>\r\n<div class=\"ti3-popconfirm-footer\">\r\n    <button [id]=\"appendId('yesbutton')\"\r\n            type=\"button\"\r\n            tiButton\r\n            color=\"{{config.yesPrimary ? 'danger' : ''}}\"\r\n            (click)=\"destroyPopover(true)\">{{ config.yesText }}</button><!--\r\n    --><button [id]=\"appendId('nobutton')\"\r\n               type=\"button\"\r\n               tiButton\r\n               color=\"{{!config.yesPrimary ? 'danger' : ''}}\"\r\n               (click)=\"destroyPopover(false)\">{{ config.noText }}</button>\r\n</div>",
                host: {
                    '[class.ti3-popconfirm-container]': 'true',
                    '[tabindex]': '0'
                },
                styles: [".ti3-popconfirm-tip{box-shadow:var(--ti-common-shadow-4-down)!important}:host.ti3-popconfirm-container{display:inline-block;outline:none}.ti3-popconfirm-header{align-content:center;display:flex}.ti3-popconfirm-warn-icon{color:var(--ti-common-color-warn);font-size:var(--ti-common-font-size-2);line-height:18px}.ti3-popconfirm-title{color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number);margin-left:var(--ti-common-space-2x)}.ti3-popconfirm-footer{margin-top:12px;text-align:right}.ti3-popconfirm-footer button:nth-of-type(2){margin-left:var(--ti-common-space-2x)}"]
            },] }
];
TiPopconfirmComponent.propDecorators = {
    config: [{ type: Input }],
    destroyPopover: [{ type: Input }]
};
