// @cloud/tinyplus3/esm2015/components/frozen/TpFrozenComponent.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, TemplateRef, ViewChild
// tslint:disable-next-line: no-implicit-dependencies
 } from '@angular/core';
// tslint:disable-next-line: no-implicit-dependencies
import { TiBaseComponent, TiTipService, Util } from '@cloud/tiny3';
// tslint:disable-next-line: no-implicit-dependencies
import { CfUtil } from '../../cfutils/CfUtil';
/**
 * 账号冻结组件，组件查询CBC查询账号冻结场景接口，获取冻结信息，展示块提示，给目标元素添加tip提示。
 *
 * 9.0.3版本新增
 *
 * <example-url>../tinyplus3demo/#/frozen/frozen-all</example-url>
 */
export class TpFrozenComponent extends TiBaseComponent {
    constructor(hostRef, renderer2, tipService) {
        super(hostRef, renderer2);
        this.hostRef = hostRef;
        this.renderer2 = renderer2;
        this.tipService = tipService;
        /**
         * 获取数据成功事件
         */
        this.response = new EventEmitter();
        /**
         * @ignore
         * 内部alert 打开关闭
         */
        this.open = false;
        /**
         * @ignore
         * alert 类型
         */
        this.alertType = 'error';
        /**
         * @ignore
         * 内部alert类型图标
         */
        this.typeIcon = true;
        /**
         * @ignore
         * 内部alert关闭图标
         */
        this.closeIcon = false;
        /**
         * @ignore
         * alert 内容
         */
        this.label = '';
    }
    /**
     * @ignore
     * 从http返回数据中获取内容字符串
     */
    static getMsgContent(data, msgType) {
        if (data && Array.isArray(data.frozenInfos) && data.frozenInfos.length > 0) {
            // 返回数据中 frozenInfos按优先级排序，所以取第一个
            const frozenInfo = data.frozenInfos[0];
            // status 1 冻结 2 解冻
            if (frozenInfo.status !== 1 || !Util.isString(frozenInfo.statusMsg)) {
                return;
            }
            try {
                frozenInfo.statusMsgs = JSON.parse(frozenInfo.statusMsg);
            }
            catch (error) {
                console.warn(error);
                return;
            }
            const msg = frozenInfo.statusMsgs.find((item) => {
                return item.msgType === msgType;
            });
            if (msg && Util.isString(msg.msgContent)) {
                return msg.msgContent;
            }
        }
    }
    static getUserInfo(config) {
        // config 本地测试使用，实际使用时请忽略此参数
        if (config) {
            return new Promise((resolve, reject) => {
                resolve(config.domainId);
            });
        }
        else {
            return CfUtil.getUser().then((user) => user.domainId);
        }
    }
    /**
     * @ignore
     * 请求账号冻结场景接口
     * config 本地测试使用，实际使用时请忽略此参数
     */
    static queryFrozenInfo(config) {
        const queryFrozenInfoPromise = TpFrozenComponent.getUserInfo(config)
            .then((domainId) => {
            const camel = config ? config.camel : CfUtil.getHttpService();
            const customerId = domainId;
            if (customerId && camel) {
                let url = '/rest/cbc/cbccustmgrservice/v1/customer/frozen';
                // AppWebPath 为console.home框架的设置,例如'/console'路径
                const appWebPath = CfUtil.getAppWebPath();
                if (Util.isString(appWebPath)) {
                    url = appWebPath + url;
                }
                return camel.get({
                    url: {
                        s: url
                    },
                    params: { customerId }
                });
            }
            else {
                return Promise.reject(new Error('frozen 依赖camel服务，使用前请先注入'));
            }
        });
        return queryFrozenInfoPromise;
    }
    /**
     * 查询冻结
     * config 为方便测试设置，实际使用时请忽略
     */
    queryFrozen(config) {
        TpFrozenComponent.queryFrozenInfo(config)
            .then((response) => {
            this.label = TpFrozenComponent.getMsgContent(response, 'POP');
            // 是字符串显示
            this.open = Util.isString(this.label);
            // 给按钮添加tip
            this.tipContent = TpFrozenComponent.getMsgContent(response, 'TIP');
            if (Util.isString(this.tipContent) && this.buttonElem) {
                this.addTip(this.buttonElem);
            }
            if (!this.label && !this.tipContent) {
                this.unfrozen();
            }
            // 获取数据成功，触发response事件，将后台原始响应数据传递出去，方便使用者自行处理相关逻辑
            this.response.emit(response);
        }, (error) => {
            this.response.emit(error);
            this.open = false;
        });
    }
    /**
     * 隐藏提示块,销毁tip
     */
    unfrozen() {
        this.open = false;
        // @ts-ignore:忽略destroy检查
        if (this.tipInstance && Util.isFunction(this.tipInstance.destroy)) {
            // @ts-ignore:外层有判断
            this.tipInstance.destroy();
            this.tipInstance = undefined;
        }
    }
    /**
     *  给目标元素添加tip
     *  v9.0.4 版本添加
     */
    addTip(...targets) {
        if (this.tipContentTemplate && this.tipContent) {
            targets.forEach((element) => {
                this.tipInstance = this.tipService.create(element, {
                    trigger: 'mouse',
                    showFn: () => {
                        return {
                            content: this.tipContentTemplate
                        };
                    }
                });
            });
        }
    }
}
TpFrozenComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-frozen',
                template: "<ti-alert\r\n    [type]=\"alertType\"\r\n    [closeIcon]=\"closeIcon\"\r\n    [typeIcon]=\"typeIcon\"\r\n    [open]=\"open\"\r\n    [id]=\"appendId('frozen_alert')\"\r\n    >\r\n        <section [innerHTML]=\"label\" [id]=\"appendId('label')\"></section>\r\n</ti-alert>\r\n\r\n<!-- tip \u63D0\u793A\u5185\u5BB9\uFF0C\u56E0\u4E3A\u5B57\u7B26\u4E32\u4E2D\u6709a\u94FE\u63A5\u6240\u4EE5\u4F7F\u7528\u4E86\u6A21\u677F -->\r\n<ng-template #tipContentTemplate >\r\n    <section [innerHTML]=\"tipContent\"></section>\r\n</ng-template>\r\n\r\n\r\n",
                styles: [".tp-link-no-decoration{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-link-no-decoration:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}:host{display:block}:host ::ng-deep ti-alert{width:100%}:host ::ng-deep ti-alert a{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}:host ::ng-deep ti-alert a:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}"]
            },] }
];
TpFrozenComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TiTipService }
];
TpFrozenComponent.propDecorators = {
    buttonElem: [{ type: Input }],
    response: [{ type: Output }],
    tipContentTemplate: [{ type: ViewChild, args: ['tipContentTemplate', { static: false },] }]
};
