// @cloud/tiny3/esm2015/utils/ObservableMap.js
// TODO:未来看typescript如何继承Map
/**
 * @ignore
 */
export class ObservableMap {
    constructor() {
        // 代理模式
        this.map = new Map();
        // 观察者模式
        this.handlerSet = new Set();
        this.map = new Map();
    }
    get size() {
        return this.map.size;
    }
    addObserver(handlerFn) {
        this.handlerSet.add(handlerFn);
    }
    removeObserver(handlerFn) {
        this.handlerSet.delete(handlerFn);
    }
    // 通知函数，不会通知到发起源
    notify(key, value, add, from) {
        this.handlerSet.forEach((handlerFn) => {
            handlerFn(key, value, add, from);
        });
    }
    //  设置，可以添加发起源，那么消息不会通知到发起源。
    set(key, value, from) {
        if (!this.map.has(key) || (this.map.has(key) && this.map.get(key) !== value)) {
            this.map.set(key, value);
            this.notify(key, value, true, from);
        }
        return this;
    }
    //  删除，可以添加发起源，那么消息不会通知到发起源。
    delete(key, from) {
        if (this.map.has(key)) {
            this.map.delete(key);
            this.notify(key, undefined, false, from);
            return true;
        }
        return false;
    }
    has(key) {
        return this.map.has(key);
    }
    get(key) {
        return this.map.get(key);
    }
    getMap() {
        return this.map;
    }
}
