// @cloud/tiny3/esm2015/utils/TiBrowser.js
/* eslint-disable no-cond-assign */
// 另，angular也有brower-util.ts，但是没有对浏览器版本号的判断。
/*
 * 浏览器类型的枚举定义。字符串枚举，提供了运行时可读性。
 */
/* export enum Browser {  // export是为了不成为全局定义。此文件存在2个export
    EDGE = 'EDGE',
    IE = 'IE',
    FIREFOX = 'FIREFOX',
    CHROME = 'CHROME',
    OPERA = 'OPERA',
    SAFARI = 'SAFARI',
    OTHER = 'OTHER'
} */
/**
 * BrowserInfo浏览器工具类，提供浏览器类别和版本信息
 *
 * 浏览器的类型及版本信息
 *
 * <h3>示例</h3>
 * <iframe src="../tiny3demo/#/utils/browser-basic" frameborder='0' width="100%" onload="this.height = document.documentElement.clientHeight"></iframe>
 */
export class TiBrowser {
    /**
     * 浏览器类型
     * @returns 返回枚举型浏览器类型字符串
     */
    static browser() {
        TiBrowser.init();
        return TiBrowser._browser;
    }
    /**
     * 浏览器版本
     */
    static version() {
        TiBrowser.init();
        return TiBrowser._version;
    }
    /**
     * 是否Edge
     */
    static isEdge() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.EDGE;
    }
    /**
     * 是否IE
     */
    static isIE() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.IE;
    }
    /**
     * 是否Firefox
     */
    static isFirefox() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.FIREFOX;
    }
    /**
     * 是否Chrome
     */
    static isChrome() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.CHROME;
    }
    /**
     * 是否Opera
     */
    static isOpera() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.OPERA;
    }
    /**
     * 是否Safari
     */
    static isSafari() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.SAFARI;
    }
    /**
     * 是否Other
     */
    static isOther() {
        TiBrowser.init();
        return TiBrowser.browser() === TiBrowser.OTHER;
    }
    static init() {
        if (TiBrowser.isInit) { // 如果已经初始化，那么返回
            return;
        }
        // 浏览器判断，已支持SSR
        const userAgent = typeof navigator !== 'undefined' ? navigator.userAgent.toLowerCase() : '';
        let aAgentInfo;
        // tslint:disable-next-line:no-conditional-assignment
        (aAgentInfo = userAgent.match(/edge\/([\d.]+)/)) ? TiBrowser._browser = TiBrowser.EDGE :
            // tslint:disable-next-line:no-conditional-assignment
            (aAgentInfo = userAgent.match(/rv:([\d.]+)\) like gecko/)) ? TiBrowser._browser = TiBrowser.IE :
                // tslint:disable-next-line:no-conditional-assignment
                (aAgentInfo = userAgent.match(/msie ([\d.]+)/)) ? TiBrowser._browser = TiBrowser.IE :
                    // tslint:disable-next-line:no-conditional-assignment
                    (aAgentInfo = userAgent.match(/firefox\/([\d.]+)/)) ? TiBrowser._browser = TiBrowser.FIREFOX :
                        // tslint:disable-next-line:no-conditional-assignment
                        (aAgentInfo = userAgent.match(/chrome\/([\d.]+)/)) ? TiBrowser._browser = TiBrowser.CHROME :
                            // tslint:disable-next-line:no-conditional-assignment
                            (aAgentInfo = userAgent.match(/opera.([\d.]+)/)) ? TiBrowser._browser = TiBrowser.OPERA :
                                // tslint:disable-next-line:no-conditional-assignment
                                (aAgentInfo = userAgent.match(/version\/([\d.]+).*safari/)) ? TiBrowser._browser = TiBrowser.SAFARI :
                                    TiBrowser._browser = TiBrowser.OTHER;
        // 版本号取整
        TiBrowser._version = (aAgentInfo !== null) ? parseInt(aAgentInfo[1].split('.')[0], 10) : 0;
        // 改变初始化值
        TiBrowser.isInit = true;
    }
}
/**
 * EDGE浏览器字符串
 */
TiBrowser.EDGE = 'EDGE';
/**
 * IE浏览器字符串
 */
TiBrowser.IE = 'IE';
/**
 * FIREFOX浏览器字符串
 */
TiBrowser.FIREFOX = 'FIREFOX';
/**
 * CHROME浏览器字符串
 */
TiBrowser.CHROME = 'CHROME';
/**
 * OPERA浏览器字符串
 */
TiBrowser.OPERA = 'OPERA';
/**
 * SAFARI浏览器字符串
 */
TiBrowser.SAFARI = 'SAFARI';
/**
 * OTHER浏览器字符串
 */
TiBrowser.OTHER = 'OTHER';
TiBrowser._browser = TiBrowser.OTHER;
TiBrowser._version = 0; // 浏览器的大版本号，整数。取第一个小数点前
TiBrowser.isInit = false;
