// @cloud/tiny3/esm2015/utils/TiLog.js
/*
 * Logger工具类提供全局日志输出级别控制。
 *
 * 提供三个静态日志方法，log()/warn()/error()
 *
 * log()是调试期日志，产品时一定关闭。组件对打包对外提供时也关闭，不干扰用户开发。
 *
 * info()是产品期也可以输出的重要日志。暂不提供。
 *
 * warn()/error()产品期，可酌情打开。后期可以考虑，error日志回传给服务器。
 *
 */
/**
 * Logger工具类, 提供全局日志输出和级别控制。
 *
 * <h3>示例</h3>
 * <iframe src="../tiny3demo/#/utils/log-basic" frameborder='0' width="100%" onload="this.height = document.documentElement.clientHeight"></iframe>
 */
export class TiLog {
    /**
     * 设置输出日志级别：LOG/WARN/ERROR/OFF. 默认是关闭
     */
    static setLevel(level) {
        TiLog.level = level;
    }
    /**
     *  一般的调试和运行信息
     * @param [message] 信息
     * @param optionalParams 参数
     */
    static log(message, ...optionalParams) {
        if (TiLog.LEVEL_LOG >= TiLog.level && console.log) {
            console.log(message, ...optionalParams);
        }
    }
    // public static info(message?: any, ...optionalParams: Array<any>): void {
    //     if (Logger.infoable && console.log) {
    //         // 在Chrome, console.info()只是log()的别名，tslint不建议使用。
    //         console.log(message, ...optionalParams);
    //     }
    // }
    /**
     *  会出现潜在错误的提示
     * @param [message] 信息
     * @param optionalParams 参数
     */
    static warn(message, ...optionalParams) {
        if (TiLog.LEVEL_WARN >= TiLog.level && console.warn) {
            console.warn(message, ...optionalParams);
        }
    }
    /**
     * 发生错误事件，已经影响系统的正常运行
     * @param [message] 信息
     * @param optionalParams 参数
     */
    static error(message, ...optionalParams) {
        if (TiLog.LEVEL_ERROR >= TiLog.level && console.error) {
            console.error(message, ...optionalParams);
        }
    }
}
/**
 * 可以输出Log以上日志，包含Log/Warn/Error，log表示一般的调试和运行信息。
 */
TiLog.LEVEL_LOG = 1;
/**
 * 可以输出Warn以上日志，包含Warn/Error日志，Warn表示会出现潜在错误的提示。
 */
TiLog.LEVEL_WARN = 2;
/**
 * 可以输出Error日志，Error表示发生错误事件，已经影响系统的正常运行。
 */
TiLog.LEVEL_ERROR = 3;
/**
 * 关闭输出所有日志
 */
TiLog.LEVEL_OFF = 10;
TiLog.level = TiLog.LEVEL_OFF;
