// @cloud/tinyplus3/esm2015/components/layout/TpLayoutModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TpLayoutHeaderComponent } from './TpLayoutHeaderComponent';
import { TpLayoutContentComponent } from './TpLayoutContentComponent';
import { TpLayoutSectionComponent } from './TpLayoutSectionComponent';
import { TpLayoutContentBodyComponent } from './TpLayoutContentBodyComponent';
import { TpLayoutContentHeaderComponent } from './TpLayoutContentHeaderComponent';
import { TpLayoutColumnComponent } from './TpLayoutColumnComponent';
export class TpLayoutModule {
}
TpLayoutModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [TpLayoutContentComponent,
                    TpLayoutContentBodyComponent,
                    TpLayoutHeaderComponent,
                    TpLayoutSectionComponent,
                    TpLayoutContentHeaderComponent,
                    TpLayoutColumnComponent],
                declarations: [TpLayoutContentComponent,
                    TpLayoutContentBodyComponent,
                    TpLayoutHeaderComponent,
                    TpLayoutSectionComponent,
                    TpLayoutContentHeaderComponent,
                    TpLayoutColumnComponent]
            },] }
];
export { TpLayoutContentComponent } from './TpLayoutContentComponent';
export { TpLayoutContentBodyComponent } from './TpLayoutContentBodyComponent';
export { TpLayoutHeaderComponent } from './TpLayoutHeaderComponent';
export { TpLayoutSectionComponent } from './TpLayoutSectionComponent';
export { TpLayoutContentHeaderComponent } from './TpLayoutContentHeaderComponent';
export { TpLayoutColumnComponent } from './TpLayoutColumnComponent';
