// @cloud/tinyplus3/esm2015/components/info/TpInfoComponent.js
import { Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
// import { TpUtil } from '../../utils/TpUtil';
import { TiLocale, TiBaseComponent } from '@cloud/tiny3';
/**
 * @ignore
 * <example-url>../tinyplus3demo/#/info/info-all</example-url>
 */
export class TpInfoComponent extends TiBaseComponent {
    constructor(element, renderer) {
        super(element, renderer);
        this.element = element;
        this.renderer = renderer;
        /**
         * 提示信息类型，有两种可选择类型：'restricted'，'unverified'
         */
        this.infoType = 'unverified';
        this.config = {
            restricted: {
                cls: 'tp-info-restricted',
                content: TiLocale.getLocaleWords().tpInfo.restricted
            },
            unverified: {
                cls: 'tp-info-unverified',
                content: TiLocale.getLocaleWords().tpInfo.unverified
            }
        };
    }
    ngOnInit() {
        super.ngOnInit();
        const typeConfig = this.config[this.infoType];
        this.renderer.addClass(this.element.nativeElement, typeConfig.cls);
        this.content = typeConfig.content;
    }
}
TpInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-info',
                template: "<tp-icon local name='cloud-action-state-abnormal' class='tp-info-icon'></tp-icon>\r\n<div class='tp-info-content' [innerHTML]='content' #contentEle [id]=\"appendId('content')\"></div>\r\n",
                styles: [".tp-link-no-decoration{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-link-no-decoration:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}:host.tp-info-restricted{color:var(--ti-common-color-error-text)}:host.tp-info-restricted .tp-info-icon{color:var(--ti-common-color-error)}:host.tp-info-unverified{color:var(--ti-common-color-warn-text)}:host.tp-info-unverified .tp-info-icon{color:var(--ti-common-color-warn)}.tp-info-icon{font-size:16px;margin-right:10px;vertical-align:middle}.tp-info-content{display:inline;vertical-align:middle}:host ::ng-deep a{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}:host ::ng-deep a:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}"]
            },] }
];
TpInfoComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TpInfoComponent.propDecorators = {
    infoType: [{ type: Input }],
    contentEle: [{ type: ViewChild, args: ['contentEle', { static: true },] }]
};
