// @cloud/tiny3/esm2015/components/dominator/TiDominatorComponent.js
import { Component, ContentChild, ElementRef, EventEmitter, Input, NgZone, Output, Renderer2, TemplateRef } from '@angular/core';
import { TiFormComponent } from '../base/TiBaseModule';
import { Util } from '../../utils/Util';
/**
 * @ignore
 */
export class TiDominatorComponent extends TiFormComponent {
    constructor(render, hostRef, zone) {
        super(hostRef, render);
        this.zone = zone;
        this.buttonStyle = 'dropdown'; // 按钮样式 三角or加号 'dropdown' 'plus'
        this.multiple = false; // 单选多选
        this.selectedTipPosition = 'auto'; // 选中项文本超出时tip展开方向
        this.dropshow = false; // 在select组件中使用，记录下拉面板展开状态
        /**
         * 设置dominator最多显示几行，默认2行, multiple为true时生效
         */
        this.maxLine = 2;
        /**
         * 设置是否显示已选项的个数，默认不显示，multiple为true时生效
         *
         * 10.1.0新增
         */
        this.showSelectedNumber = false;
        /**
         * 单选下拉选择框是否开启清除已选项功能
         *
         * 3.0.8新增
         */
        this.clearable = false;
        this.delete = new EventEmitter();
        /**
         * 当单选下拉选择框开启清除已选项功能时，点击清除按钮时触发
         *
         * 3.0.8新增
         */
        this.clear = new EventEmitter();
        this.isMouseIn = false;
    }
    ngOnInit() {
        super.ngOnInit();
        // 每行tag的高度24px, +2 是因为 .ti3-select-dominator-text 元素上下padding
        // tslint:disable-next-line: binary-expression-operand-order
        this.maxHeight = !this.multiple ? '' : this.maxHeight ? this.maxHeight : (this.maxLine * 24 + 2) + 'px';
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.setFocusableElems([this.nativeElement]);
        this.zone.runOutsideAngular(() => {
            // 内部mousedown事件冒泡至根节点
            this.renderer.listen(this.nativeElement, 'mousedown', (event) => {
                if (this.disabled) {
                    event.preventDefault(); // 不能聚焦，不能click
                }
            });
            if (this.clearable && !this.multiple) {
                this.renderer.listen(this.nativeElement, 'mouseenter', () => {
                    if (this.disabled || Util.isEmptyString(this.model)) {
                        return;
                    }
                    this.zone.run(() => {
                        this.isMouseIn = true;
                    });
                });
                this.renderer.listen(this.nativeElement, 'mouseleave', () => {
                    if (this.isMouseIn) {
                        this.zone.run(() => {
                            this.isMouseIn = false;
                        });
                    }
                });
            }
        });
    }
    /**
     * 兼容旧版：
     * 3.1.2版select只能内嵌一个模板，无命名。
     * 新版可以内嵌2个模板，示例书写要求都命名。
     * 但需要兼容旧版无命名测试用例。
     */
    ngAfterContentInit() {
        super.ngAfterContentInit();
        // 如果item模板为空，那么把第一个出现的无标签（无footer等标签）模板作为item
        if (!this.itemTemplate && this.firstTemplate !== this.placeholderTemplate) {
            this.itemTemplate = this.firstTemplate;
        }
    }
    /**
     * 用户点击叉号，删除一个项目
     * @param item
     * @returns
     */
    onDelete(item) {
        if (item.disabled === true) {
            return;
        }
        // 删除选中
        const index = this.model.indexOf(item);
        if (index !== -1) {
            this.model.splice(index, 1);
        }
        this.delete.emit({
            item: item,
            model: this.model
        });
        // 强行向外通知model改变。
        this.model = this.model.concat();
    }
    /**
     * @ignore
     */
    onClickClear(event) {
        event.stopPropagation();
        this.model = null;
        this.clear.emit();
    }
    showClear() {
        return this.clearable && !this.multiple && !this.disabled && (this.isMouseIn || this.focused) && this.model;
    }
}
TiDominatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-dominator',
                template: "<!--\u6A21\u677F\u4F18\u5148\uFF0Clabel\u5176\u6B21-->\r\n<section class=\"ti3-select-dominator-text\"\r\n    [ngStyle]=\"{'max-height': maxHeight}\"\r\n    [ngClass]=\"{\r\n        'ti3-multiselect-container': multiple,\r\n        'ti3-tags-input-container': type === 'tagsinput'\r\n        }\"\r\n    [id]=\"appendId('input')\">\r\n    <!-- \u5355\u9009 -->\r\n    <ng-container *ngIf=\"!multiple else multipleTemplate\">\r\n        <section class=\"ti3-overflow-padding\"\r\n            [tiTip]=\" model?model.tip : ''\"\r\n            [tiTipContext]=\"model\"\r\n            [tiTipPosition]='selectedTipPosition'\r\n            [tiTipHostEleX]=\"nativeElement\"\r\n            ><ng-container  *ngIf=\"model else placeholderTemplate?placeholderTemplate:defaultPlaceholderTemplate\">\r\n                <ng-container *ngIf=\"itemTemplate else defaultItemTemplate\">\r\n                  <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: model}\">\r\n                  </ng-container>\r\n                </ng-container>\r\n                <ng-template #defaultItemTemplate><span [title]=\"model[labelKey]\">{{model[labelKey]}}</span></ng-template>\r\n            </ng-container>\r\n            <ng-template #defaultPlaceholderTemplate><span class=\"ti3-placeholder\" [title]=\"placeholder\">{{placeholder}}</span></ng-template>\r\n        </section>\r\n    </ng-container>\r\n    <!-- \u591A\u9009 -->\r\n    <ng-template #multipleTemplate>\r\n        <ng-container *ngIf=\"model && model.length else placeholderTemplate?placeholderTemplate:defaultPlaceholderMultiTemplate\">\r\n            <ti-tag *ngFor=\"let item of model; let i = index\" [maxWidth]=\"tagMaxWidth\" [id]=\"appendId('tag'+i)\"\r\n                [disabled]=\"disabled||item.disabled\"\r\n                [tiTip]=\"item.tip\"\r\n                [tiTipPosition]='selectedTipPosition'\r\n                [tiTipContext]=\"item\"\r\n                (delete)=\"onDelete(item)\"\r\n                ><ng-container *ngIf=\"itemTemplate else defaultItemTemplate\">\r\n                    <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: item, index: i}\">\r\n                    </ng-container>\r\n                </ng-container>\r\n                <ng-template #defaultItemTemplate><span [title]=\"item[labelKey]\">{{item[labelKey]}}</span></ng-template>\r\n            </ti-tag>\r\n        </ng-container>\r\n\r\n        <!-- tagsinput\u7EC4\u4EF6\u4E2Ddominator\u7684\u6700\u540E\u4E00\u4E2A\u5143\u7D20\u4E3Ainput\u8F93\u5165\u6846 -->\r\n        <ng-container *ngIf=\"type === 'tagsinput'\">\r\n            <ng-content select=\"input\"></ng-content>\r\n        </ng-container>\r\n        <!-- tag\u8F93\u5165\u533A\u57DF\uFF0C\u7EFC\u5408\u641C\u7D22\u4F7F\u7528-->\r\n        <ng-container>\r\n            <ng-content select=\"[tag-editor]\"></ng-content>\r\n        </ng-container>\r\n        <ng-template #defaultPlaceholderMultiTemplate>\r\n            <section *ngIf=\"type !== 'tagsinput'\" class=\"ti3-overflow-padding ti3-placeholder\" [title]=\"placeholder\">{{placeholder}}</section>\r\n        </ng-template>\r\n    </ng-template>\r\n</section>\r\n\r\n<!-- tagsinput\u7EC4\u4EF6\u6CA1\u6709\u4E0B\u62C9\u6309\u94AE -->\r\n<section *ngIf=\"type !== 'tagsinput' && !showClear()\"\r\n         [ngClass]=\"{'ti3-select-dominator-dropdown-btn':(buttonStyle==='dropdown'),\r\n                     'ti3-icon ti3-icon-add ti3-select-dominator-plus-btn':(buttonStyle==='plus'),\r\n                    'ti3-select-dominator-dropdown-btn-transform-up': dropshow}\"\r\n         [id]=\"appendId('btn')\">\r\n</section>\r\n\r\n<!-- \u591A\u9009\u65F6\uFF1A\u663E\u793A\u5F53\u524D\u9009\u4E2D\u4E86\u51E0\u9879 -->\r\n<section *ngIf=\"showSelectedNumber && multiple && model && model.length > 0\" class=\"ti3-multiselect-selected\">{{ 'tiSelect.selected' | tiTranslate }}<span\r\n    style=\"padding-left: 4px\">{{model.length}}</span>\r\n</section>\r\n\r\n<!-- \u5355\u9009\u6709\u53EF\u6E05\u9664\u7684\u573A\u666F -->\r\n<section *ngIf=\"showClear()\"\r\n         class=\"ti3-icon ti3-icon-close ti3-select-dominator-clear-btn\"\r\n        [id]=\"appendId('clear')\"\r\n        (click)=\"onClickClear($event)\">\r\n</section>\r\n<!-- dominator\u53F3\u4FA7\u56FE\u6807\u6309\u94AE\u7EC4, \u7EFC\u5408\u641C\u7D22\u7EC4\u4EF6\u4F7F\u7528 -->\r\n<ng-content select=\"[actions]\"></ng-content>\r\n<!--IE\u6D4F\u89C8\u5668\u517C\u5BB9\u6027\u95EE\u9898\uFF1A\u5728IE\u4E0B\uFF0C\u6240\u6709div\uFF0C\u90E8\u5206span\uFF08\u8BBE\u7F6E\u4E3Ainline-block\u3001block\u7684span\uFF09\u662F\u4EA4\u4E92\u5143\u7D20\uFF0C\u53EF\u4EE5\u843D\u7126\u70B9\u3002\r\n\u800C\u5176\u4ED6\u6D4F\u89C8\u5668\uFF0Cdiv\u548Cspan\u9ED8\u8BA4\u662F\u4E0D\u53EF\u4EA4\u4E92\u5143\u7D20\uFF0C\u4E0D\u80FD\u843D\u7126\u70B9\u3002\r\n\u89E3\u51B3\u65B9\u6848\uFF1Asection\u6807\u7B7E\uFF0C\u53EF\u4EE5\u7B49\u540C\u4E8Ediv\u4F7F\u7528\uFF0CIE\u548C\u5176\u4ED6\u6D4F\u89C8\u5668\u884C\u4E3A\u4E00\u81F4\u3002\r\n-->",
                host: {
                    '[class.ti3-select-dominator-container]': 'true',
                    '[class.ti3-multiselect-container]': 'multiple'
                },
                providers: [TiFormComponent.getValueAccessor(TiDominatorComponent)],
                styles: ["::ng-deep :root{--ti-dominator-button-type-dropdown-width:35px;--ti-dominator-select-box-width:calc(var(--ti-input-height) - 2px)}:host{display:inline-block}:host.ti3-select-dominator-container{-ms-box-sizing:border-box;border:1px solid transparent;box-sizing:border-box;cursor:pointer;display:inline-block;height:var(--ti-input-height);line-height:calc(var(--ti-input-height) - 2px);margin:-1px;position:relative;text-align:left;vertical-align:middle;white-space:normal;width:calc(100% + 2px)}:host.ti3-select-dominator-container:focus{outline:none}:host[disabled].ti3-select-dominator-container{cursor:not-allowed!important}:host[disabled] .ti3-select-dominator-dropdown-btn:after{border-top-color:var(--ti-dominator-select-arrow-color-disabled)!important}:host[disabled] .ti3-select-dominator-text{color:var(--ti-common-color-text-disabled)}:host[disabled] .ti3-multiselect-box-cell{border:1px solid var(--ti-dominator-multiselect-box-cell-border-color-disabled)}.ti3-overflow-padding{height:100%;overflow:hidden;padding:0 10px;text-overflow:ellipsis;white-space:nowrap}.ti3-select-dominator-text{-ms-box-sizing:border-box;border:none;border-radius:var(--ti-input-border-radius);box-sizing:border-box;color:var(--ti-input-text-color);display:inline-block;height:100%;text-decoration:none;width:calc(100% - var(--ti-dominator-select-box-width))}.ti3-placeholder{color:var(--ti-input-placeholder-color)}.ti3-select-dominator-clear-btn,.ti3-select-dominator-dropdown-btn{-ms-box-sizing:border-box;border-radius:0 var(--ti-input-border-radius) var(--ti-input-border-radius) 0;box-sizing:border-box;float:right;height:calc(var(--ti-input-height) - 2px);width:var(--ti-dominator-select-box-width)}.ti3-select-dominator-dropdown-btn:after{border-left:4px solid transparent;border-right:4px solid transparent;border-top:5px solid var(--ti-dominator-select-arrow-color);content:\"\";height:0;position:absolute;right:calc(var(--ti-dominator-select-box-width)/2 - 4px);top:calc((var(--ti-input-height) - 7px)/2);width:0}.ti3-select-dominator-dropdown-btn-transform-up:after{transform:rotate(180deg)}.ti3-select-dominator-clear-btn{color:var(--ti-dominator-select-clear-color);font-size:12px;text-align:center}.ti3-select-dominator-clear-btn:hover{color:var(--ti-common-color-icon-hover)}:host.ti3-multiselect-container{height:auto;line-height:1;min-height:var(--ti-input-height);overflow:hidden}:host.ti3-multiselect-container>.ti3-select-dominator-text{line-height:calc(var(--ti-input-height) - 4px);max-height:calc(2*var(--ti-input-height) - 4px);overflow-y:auto;padding:1px;vertical-align:middle}:host.ti3-multiselect-container>.ti3-select-dominator-text::-webkit-scrollbar-track{margin:0}:host.ti3-multiselect-container>.ti3-select-dominator-text:after{content:\"\";display:block;overflow:hidden}:host.ti3-multiselect-container>.ti3-select-dominator-text.ti3-tags-input-container{width:100%}:host.ti3-multiselect-container .ti3-multiselect-selected{display:inline-block;padding:var(--ti-common-space-base) var(--ti-common-space-3x)}"]
            },] }
];
TiDominatorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone }
];
TiDominatorComponent.propDecorators = {
    labelKey: [{ type: Input }],
    placeholder: [{ type: Input }],
    buttonStyle: [{ type: Input }],
    multiple: [{ type: Input }],
    selectedTipPosition: [{ type: Input }],
    dropshow: [{ type: Input }],
    tagMaxWidth: [{ type: Input }],
    maxLine: [{ type: Input }],
    showSelectedNumber: [{ type: Input }],
    maxHeight: [{ type: Input }],
    type: [{ type: Input }],
    clearable: [{ type: Input }],
    delete: [{ type: Output }],
    clear: [{ type: Output }],
    firstTemplate: [{ type: ContentChild, args: [TemplateRef, /* TODO: add static flag */ { static: true },] }],
    itemTemplate: [{ type: ContentChild, args: ['item', /* TODO: add static flag */ { static: true },] }],
    placeholderTemplate: [{ type: ContentChild, args: ['placeholder',] }]
};
