// @cloud/cfdata/component/MessageComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComponent = void 0;
var StorageService_1 = require("../service/StorageService");
var MAX_MSG_NUMBER = 3; // 限制最多3条
var msgID = 1;
var ERROR_DURATION = 10000;
var INFO_DURATION = 5000;
var MessageComponent = /** @class */ (function () {
    function MessageComponent() {
    }
    MessageComponent.showError = function (message, duration, wrapperClass) {
        this.alertMsg('error', message, duration, wrapperClass);
    };
    MessageComponent.showSuccess = function (message, duration, wrapperClass) {
        this.alertMsg('success', message, duration, wrapperClass);
    };
    MessageComponent.showPrompt = function (message, duration, wrapperClass) {
        this.alertMsg('prompt', message, duration, wrapperClass);
    };
    MessageComponent.showWarn = function (message, duration, wrapperClass) {
        this.alertMsg('warn', message, duration, wrapperClass);
    };
    /*
    添加消息缓存， ctc、dt等局点适用
    */
    MessageComponent.storageMsg = function (type, msg) {
        var msgData = {};
        msgData['content'] = '[' + type + ']' + msg;
        msgData['time'] = new Date();
        var agencyID = StorageService_1.StorageService.getCookie('agencyID');
        var tipsMessages = StorageService_1.StorageService.getSessionStorage('framework_tips_msg' + agencyID);
        tipsMessages = tipsMessages ? tipsMessages : [];
        if (tipsMessages.unshift(msgData) > 50) {
            tipsMessages.pop();
        }
        StorageService_1.StorageService.setSessionStorage('framework_tips_msg' + agencyID, tipsMessages);
        StorageService_1.StorageService.setSessionStorage('framework_tips_new_msg' + agencyID, true);
        // 广播事件‘cfStorageMsg’
        window.dispatchEvent(new CustomEvent('cfStorageMsg'));
    };
    /*
    绘制消息弹窗方法
    */
    MessageComponent.alertMsg = function (type, message, duration, wrapperClass, noTypeIcon, noCloseIcon) {
        if (!duration) {
            if (type === 'error') {
                duration = ERROR_DURATION;
            }
            else {
                duration = INFO_DURATION;
            }
        }
        // 若服务未传入wrapperClass，设置默认值'body'
        if (!wrapperClass) {
            wrapperClass = 'body';
        }
        var MsgItemObj = {
            type: type,
            message: message,
            id: 'msg_' + msgID++,
            typeIcon: !noTypeIcon,
            closeIcon: !noCloseIcon,
            duration: duration,
        };
        var timerMsg;
        if (!document.querySelector(wrapperClass + '> .cf-message-wrapper')) {
            var messageWrapper_1 = document.createElement('div');
            messageWrapper_1.setAttribute('class', 'cf-message-wrapper');
            var bodyNode = document.querySelector(wrapperClass);
            bodyNode.appendChild(messageWrapper_1);
            // 头部高度变化带来的message位置调整不适用于弹窗组件内部的提示
            if (wrapperClass === 'body') {
                // 当头部高度变化改变message的位置
                var consoleContext = window.getConsoleContext && window.getConsoleContext();
                var consoleDataService = (consoleContext === null || consoleContext === void 0 ? void 0 : consoleContext.get) && consoleContext.get({ name: 'safearea' });
                if (consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.getSafeArea) {
                    var safeArea = consoleDataService.getSafeArea();
                    messageWrapper_1.style.top = safeArea.top + 10 + "px";
                }
                if (consoleDataService === null || consoleDataService === void 0 ? void 0 : consoleDataService.onChange) {
                    consoleDataService.onChange(function (e) {
                        messageWrapper_1.style.top = e.detail.top + 10 + "px";
                    });
                }
            }
        }
        var messageWraper = document.querySelector(wrapperClass + ' > .cf-message-wrapper');
        var messageItem = document.createElement('div');
        var messageItemCon = "<span class='cf-message-icon cf-icon-status-" + MsgItemObj.type + "'></span><p class='cf-message-p'>" + MsgItemObj.message + "</p><span class='cf-message-close hwsicon-frame-image-close'></span>";
        messageItem.innerHTML = messageItemCon;
        messageItem.setAttribute('class', String('cf-message-' + MsgItemObj.type));
        if (messageWraper.hasChildNodes()) {
            // 当前容器含有消息子节点，新增弹窗添加在现有首个子节点之前
            var firstNode = messageWraper.firstChild;
            messageWraper.insertBefore(messageItem, firstNode);
            // 新增消息如果与现有消息一致，同样的消息仅显示一条，duration时间同步最新的一条
            if (firstNode.isEqualNode(messageItem)) {
                messageWraper.removeChild(firstNode); // remove()方法在IE下存在兼容性问题
            }
            if (messageWraper.childNodes.length > MAX_MSG_NUMBER) {
                messageWraper.removeChild(messageWraper.lastChild);
            }
        }
        else {
            // 不包含消息子节点，则直接插入到容器中
            messageWraper.appendChild(messageItem);
        }
        timerMsg = setTimeout(function () {
            // 清空消息Dom节点时，需先判断parentNode是否存在
            if (messageItem.parentNode) {
                messageItem.outerHTML = '';
            }
        }, duration);
        messageItem.addEventListener('click', function (event) {
            // 点击关闭按钮，消息弹窗隐藏
            if (event.target.classList.contains('hwsicon-frame-image-close')) {
                messageItem.outerHTML = '';
                clearTimeout(timerMsg);
            }
        });
        if (message) {
            this.storageMsg(type, message);
        }
    };
    return MessageComponent;
}());
exports.MessageComponent = MessageComponent;
