// @cloud/tiny3/esm2015/components/leftmenu/TiLeftmenuFootComponent.js
import { TiLeftmenuComponent } from './TiLeftmenuComponent';
import { Component, ElementRef, Renderer2 } from '@angular/core';
export class TiLeftmenuFootComponent {
    constructor(renderer2, elementRef, tiLeftmenu) {
        this.renderer2 = renderer2;
        this.elementRef = elementRef;
        this.tiLeftmenu = tiLeftmenu;
    }
    ngAfterViewInit() {
        const templateDom = this.renderer2.createElement('div');
        const panelDom = this.tiLeftmenu.nativeElement.querySelector('.ti3-leftmenu-panel');
        const hostElement = this.elementRef.nativeElement;
        // 修复SSR错误：ERROR ReferenceError: getComputedStyle is not defined
        const spaceRight = typeof getComputedStyle === 'function' ? getComputedStyle(hostElement).getPropertyValue('stroke-width') : '0';
        this.renderer2.setStyle(templateDom, 'height', hostElement.offsetHeight + 'px');
        this.renderer2.appendChild(panelDom, templateDom);
        this.renderer2.setStyle(hostElement, 'width', `calc(100% - ${this.tiLeftmenu.scrollWidth}px)`);
        this.renderer2.setStyle(hostElement, 'marginRight', `${this.tiLeftmenu.scrollWidth}px`);
        this.renderer2.setStyle(hostElement, 'paddingRight', `calc(${spaceRight} - ${this.tiLeftmenu.scrollWidth}px)`);
    }
}
TiLeftmenuFootComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-leftmenu-foot',
                template: "<ng-content></ng-content>",
                host: {
                    '[class.ti3-leftmenu-foot]': 'true'
                }
            },] }
];
TiLeftmenuFootComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: TiLeftmenuComponent }
];
