// @cloud/tiny3/esm2015/components/leftmenu/TiLeftmenuComponent.js
import { Component, ElementRef, EventEmitter, Input, NgZone, Optional, Output, Renderer2, ViewEncapsulation, Inject } from '@angular/core';
import { ActivatedRoute, NavigationEnd, Router } from '@angular/router';
import { DOCUMENT } from '@angular/common';
import { Util } from '../../utils/Util';
import { TiBaseComponent } from '../base/TiBaseModule';
import { TiRenderer } from '../../services/renderer/TiRenderer';
/**
 * TiLeftmenu组件提供了一种方便灵活实现页面布局左侧菜单的方式，支持两级菜单。
 *
 * 其内部包含5个组件:
 *
 * TiLeftmenu : 左侧菜单最外层的指令；
 *
 * TiLeftmenuGroup: 菜单分组指令；
 *
 * TileftmenuGroupItem：分组文本内容指令；
 *
 * TiLeftmenuHead : 菜单头部内容指令；
 *
 * TiLeftmenuLevel1 ：一级菜单指令；
 *
 * TiLeftmenuItem ：一级菜单文本内容指令；
 *
 * TiLeftmenuLevel2 ：二级菜单指令。
 *
 * <example-url>../tiny3demo/#/leftmenu/leftmenu-all</example-url>
 */
export class TiLeftmenuComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, router, activeRoute, tiRenderer, zone, document) {
        super(elementRef, renderer2);
        this.renderer2 = renderer2;
        this.router = router;
        this.activeRoute = activeRoute;
        this.tiRenderer = tiRenderer;
        this.zone = zone;
        this.document = document;
        /**
         * 激活菜单项改变时向外通知
         */
        this.activeChange = new EventEmitter();
        /**
         * 控制右侧内容区的刷新
         *
         * 注意：当重复点击当前激活菜单项时，需要刷新右侧内容区，为了实现此功能设置了该接口；
         * 该接口是双向绑定的，当重复点击当前激活菜单项时，reloadState的值会经历 true(上次的值) -> false -> true
         * 的变化过程，开发者给改接口与其右侧区域容器上ngIf接口绑定的是同一个变量，所以ngIf的值也会经历同样的变化，
         * 从而做到右侧内容区的刷新。
         */
        this.reloadState = true;
        /**
         * 右侧内容区是否需要刷新的状态值改变时向外通知
         */
        this.reloadStateChange = new EventEmitter();
        // routable接口在3.0.3版本起对外隐藏，10.0.1重新开放
        /**
         * 10.0.1新增
         *
         * 设置是否使用路由来控制菜单对应内容显示/隐藏，默认使用路由。
         */
        this.routable = true;
        /**
         * 是否开启左侧菜单面板可折叠功能，默认开启。
         */
        this.toggleable = true;
        /**
         * 左侧菜单面板是否为收起状态，用于设置面板收起/展开状态，默认展开。
         */
        this.collapsed = false;
        /**
         * 左侧菜单面板收起/展开状态改变时向外通知
         */
        this.collapsedChange = new EventEmitter();
        /**
         * @ignore
         * tiLeftmenu中包含的 TiLeftmenuLevel1Component 实例集合
         */
        this.level1Items = [];
        /**
         * @ignore
         */
        this.resolvedPromise = Promise.resolve(null);
        /**
         * 标志量，是否为url改变(需要去改变active)
         */
        this.isUrlChange = false;
    }
    // 判断item的router是否是父路由下默认激活的子路由（即子路由与父路由路径一致）
    static isDefaultActiveChildRoute(item) {
        // 默认激活的子路由的router可设为[]或['']
        return !item.router || item.router.length === 0 || item.router[0] === '';
    }
    ngOnInit() {
        super.ngOnInit();
        // 初始时右侧内容区默认显示，需使外部用户的reloadState值为true。
        // 双向绑定数据时，对初始传入的值立即修改并传出时会报错，
        // 此处参考ngModel源码setValue的处理，使用promise延后执行时序
        this.resolvedPromise.then(() => {
            this.reloadStateChange.emit(true);
        });
        this.scrollWidth = this.getScrollbarSize('Y');
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        if (!this.routable) {
            return;
        }
        // 直接输入url(刷新页面)改变状态时，leftmenu能同步修改激活状态
        const urlActiveItem = this.getActiveItemByUrl(this.level1Items);
        if (urlActiveItem) {
            // 双向绑定数据时，对初始传入的值立即修改并传出时会报错，
            // 此处参考ngModel源码setValue的处理，使用promise延后执行时序
            this.resolvedPromise.then(() => {
                if (urlActiveItem !== this.active) {
                    this.isUrlChange = true;
                    this.active = urlActiveItem;
                    this.activeChange.emit(this.active);
                }
            });
        }
        else {
            this.link();
        }
        // 从非leftMenu处跳转改变状态时，leftmenu能同步修改激活状态
        this.navigationSubscription = this.router.events.subscribe((evt) => {
            if (evt instanceof NavigationEnd) {
                const newActiveItem = this.getActiveItemByUrl(this.level1Items);
                if (newActiveItem && newActiveItem !== this.active) {
                    this.active = newActiveItem;
                    this.activeChange.emit(this.active);
                    this.isUrlChange = true;
                }
            }
        });
        this.panelContainer = this.nativeElement.querySelector('.ti3-leftmenu-panel-container');
        const headElement = this.nativeElement.querySelector('.ti3-leftmenu-head');
        let headCloneElement;
        this.zone.runOutsideAngular(() => {
            this.renderer2.listen(this.panelContainer, 'scroll', () => {
                const scrollTop = this.panelContainer.scrollTop;
                if (scrollTop > 0) {
                    if (!headCloneElement) {
                        headCloneElement = headElement.cloneNode(true);
                        this.renderer2.addClass(headCloneElement, 'ti3-leftmenu-head-clone');
                        this.renderer2.setStyle(headCloneElement, 'width', `calc(100% - ${this.scrollWidth}px)`);
                        this.nativeElement.appendChild(headCloneElement);
                    }
                }
                else if (headCloneElement) {
                    this.nativeElement.removeChild(headCloneElement);
                    headCloneElement = undefined;
                }
            });
        });
    }
    ngOnChanges(changes) {
        const activeObj = changes['active'];
        if (activeObj && !activeObj.firstChange) {
            for (const level1Item of this.level1Items) {
                level1Item.setShowChildren();
                level1Item.setActiveState();
            }
            if (this.routable && !this.isUrlChange) {
                this.link();
            }
            this.isUrlChange = false;
        }
    }
    ngOnDestroy() {
        if (this.navigationSubscription) {
            this.navigationSubscription.unsubscribe();
        }
    }
    /**
     * @ignore
     * 模板中使用
     */
    toggleClickFn() {
        this.collapsed = !this.collapsed;
        this.collapsedChange.emit(this.collapsed);
    }
    getActiveItemByUrl(arr) {
        let resultItem;
        for (const component of arr) {
            if (this.isActived(component.item)) {
                resultItem = component.item;
                if (TiLeftmenuComponent.isDefaultActiveChildRoute(component.item)) {
                    continue;
                }
                else {
                    return resultItem;
                }
            }
            if (component.children && component.children.length > 0) {
                const result = this.getActiveItemByUrl(component.children);
                if (result) {
                    return result;
                }
            }
        }
        return resultItem;
    }
    isActived(item) {
        const relativeRoute = this.getRelativeRoute(item.routerExtras);
        if (this.isMatchCurrentUrl(item.router, relativeRoute)) {
            return true;
        }
        if (Util.isArray(item.routerList) && item.routerList.length > 0) {
            for (const router of item.routerList) {
                if (this.isMatchCurrentUrl(router, relativeRoute)) {
                    return true;
                }
            }
        }
        return false;
    }
    isMatchCurrentUrl(router, relativeRoute) {
        if (!Util.isArray(router)) {
            return false;
        }
        const itemUrlTree = this.router.createUrlTree(router, relativeRoute);
        return this.router.isActive(itemUrlTree, false);
    }
    link() {
        if (this.active) {
            this.navigate(this.active);
        }
    }
    /**
     * @ignore
     */
    navigate(item) {
        if (TiLeftmenuComponent.isDefaultActiveChildRoute(item)) {
            //  routeConfig 可能为null
            if (!this.activeRoute.routeConfig) {
                return;
            }
            const path = this.activeRoute.routeConfig.path;
            const url = this.router.routerState.snapshot.url;
            const rootUrl = url.split(path)[0];
            // 绝对路径跳转
            this.router.navigate([rootUrl + path], item.routerExtras);
        }
        else {
            this.router.navigate(item.router, this.getNavigationExtras(item.routerExtras));
        }
    }
    /**
     * @ignore
     * 点击当前选中项时，触发内容区重载
     * @param item 当前点击项
     */
    triggerReload(item) {
        // 使reloadState由false变为true,配合内容区的ngIf做到内容区的重载
        this.reloadStateChange.emit(false);
        setTimeout(() => {
            this.reloadStateChange.emit(true);
            // 如果当前路由就是item.router对应路由，则此导航不作用(路由相同)；
            // 如果当前路由是item.routerList中的路由，则需要跳转到item.router对应路由
            this.navigate(item);
        }, 0);
    }
    getNavigationExtras(routerExtras) {
        return Object.assign({ relativeTo: this.activeRoute }, routerExtras);
    }
    getRelativeRoute(routerExtras) {
        if (!routerExtras || !routerExtras['relativeTo']) {
            return { relativeTo: this.activeRoute };
        }
        return { relativeTo: routerExtras['relativeTo'] };
    }
    /**
 * 得到浏览器滚动条的宽度和高度
 * 代表滚动条的方向，取值为X or Y
 */
    getScrollbarSize(direction) {
        const outerDiv = this.renderer.createElement('div');
        const innerDiv = this.renderer.createElement('div');
        this.renderer.setStyle(innerDiv, 'width', '100%');
        this.renderer.setStyle(innerDiv, 'height', '100%');
        this.renderer.appendChild(outerDiv, innerDiv);
        this.tiRenderer.setStyles(outerDiv, {
            position: 'absolute',
            top: '-9999px',
            left: '-9999px',
            width: '100px',
            height: '100px',
            overflow: 'hidden'
        });
        this.renderer.appendChild(this.document.body, outerDiv);
        const field = direction === 'X' ? 'height' : 'width';
        const noScrollSize = typeof innerDiv.getBoundingClientRect === 'function' ? innerDiv.getBoundingClientRect()[field] : 0;
        this.renderer.setStyle(outerDiv, 'overflow', 'scroll');
        const withScrollSize = typeof innerDiv.getBoundingClientRect === 'function' ? innerDiv.getBoundingClientRect()[field] : 0;
        this.document.body.removeChild(outerDiv);
        return parseFloat((noScrollSize - withScrollSize).toFixed(1));
    }
}
TiLeftmenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-leftmenu',
                template: "<div  class=\"ti3-leftmenu-panel-container\">\r\n    <div class=\"ti3-leftmenu-panel\">\r\n        <ng-content></ng-content>\r\n    </div>\r\n</div>\r\n<ng-content *ngIf='!collapsed' select=\"ti-leftmenu-foot\"></ng-content>\r\n<div class=\"ti3-leftmenu-toggle\"\r\n     *ngIf=\"toggleable\"\r\n     [ngClass]=\"{'ti3-leftmenu-toggle-show' : !collapsed,\r\n                 'ti3-leftmenu-toggle-hide' : collapsed }\"\r\n    [tabindex]=\"0\"\r\n    tiOutline\r\n    (keydown.enter)=\"toggleClickFn()\"\r\n    (click)=\"toggleClickFn()\"\r\n    [id]=\"appendId('toggle')\">\r\n</div>\r\n",
                host: {
                    '[class.ti3-leftmenu-container]': 'true',
                    '[class.ti3-leftmenu-hide]': 'collapsed'
                },
                encapsulation: ViewEncapsulation.None,
                styles: [":root{--ti-leftmenu-level2-left-line:30px;--ti-leftmenu-toggle-width:12px}.ti3-leftmenu-container{background-color:var(--ti-leftmenu-container-bg-color);bottom:0;box-shadow:var(--ti-common-shadow-1-right);display:block;left:0;position:fixed;top:0;width:var(--ti-leftmenu-width);z-index:990}.ti3-leftmenu-container.ti3-leftmenu-hide{width:0!important}.ti3-leftmenu-container.ti3-leftmenu-hide .ti3-leftmenu-panel-container{height:0!important}.ti3-leftmenu-container.ti3-leftmenu-hide .ti3-leftmenu-foot,.ti3-leftmenu-container.ti3-leftmenu-hide .ti3-leftmenu-head{margin:0;padding:0}.ti3-leftmenu-panel-container{overflow-x:hidden;overflow-y:auto}.ti3-leftmenu-panel,.ti3-leftmenu-panel-container{-ms-box-sizing:border-box;bottom:0;box-sizing:border-box;left:0;position:absolute;top:0;width:100%}.ti3-leftmenu-panel{color:var(--ti-common-color-text-secondary);font-size:var(--ti-leftmenu-font-size);overflow-x:visible;overflow-y:visible;padding:0 2px}.ti3-leftmenu-toggle{background-color:var(--ti-leftmenu-container-bg-color);border-radius:0 var(--ti-leftmenu-toggle-width) var(--ti-leftmenu-toggle-width) 0;cursor:pointer;height:80px;position:absolute;top:50%;transform:translateY(-50%);width:var(--ti-leftmenu-toggle-width)}.ti3-leftmenu-toggle:after{border-left:calc(var(--ti-leftmenu-angle-normal-width)/2) solid transparent;border-right:calc(var(--ti-leftmenu-angle-normal-width)/2) solid transparent;border-top:var(--ti-leftmenu-angle-normal-height) solid var(--ti-leftmenu-icon-color-nomal);content:\"\";display:block;height:0;left:calc(50% - var(--ti-leftmenu-angle-normal-width)/2);position:relative;top:calc(50% - var(--ti-leftmenu-angle-normal-height)/2);transform:rotate(90deg);width:0}.ti3-leftmenu-toggle.ti3-leftmenu-toggle-show{left:auto;right:calc(-1*var(--ti-leftmenu-toggle-width))}.ti3-leftmenu-toggle.ti3-leftmenu-toggle-hide{left:0;right:auto}.ti3-leftmenu-toggle.ti3-leftmenu-toggle-hide:after{transform:rotate(270deg)}.ti3-leftmenu-head{-ms-box-sizing:border-box;box-sizing:border-box;color:var(--ti-leftmenu-head-text-color);display:block;font-size:var(--ti-leftmenu-head-font-size);line-height:20px;margin-bottom:10px;overflow:hidden;padding:var(--ti-leftmenu-head-padding);text-align:var(--ti-leftment-leftmenu-head-text-align)}.ti3-leftmenu-head:after{border-bottom:1px solid var(--ti-leftmenu-head-line-dividing-color);bottom:-20px;content:\"\";display:block;position:relative;width:100%}.ti3-leftmenu-head-clone{background:var(--ti-leftmenu-container-bg-color);left:0;position:absolute;top:0;z-index:3}.ti3-leftmenu-level1{display:block;margin:0;padding:0}.ti3-leftmenu-level1+.ti3-leftmenu-group-container:before{border:1px dashed var(--ti-common-color-line-dividing);bottom:auto;content:\"\";left:var(--ti-leftmenu-group-padding-left);position:absolute;top:-2px;width:calc(100% - var(--ti-leftmenu-group-padding-left) - var(--ti-leftmenu-group-padding-right))}.ti3-leftmenu-level1-dividing{position:relative}.ti3-leftmenu-level1-dividing:after{border:1px dashed var(--ti-common-color-line-dividing);bottom:0;content:\"\";left:var(--ti-leftmenu-group-padding-left);position:absolute;top:auto;width:calc(100% - var(--ti-leftmenu-group-padding-left) - var(--ti-leftmenu-group-padding-right))}.ti3-leftmenu-level2-container{position:relative}.ti3-leftmenu-level2-container:after{border:1px solid var(--ti-common-color-line-dividing);bottom:var(--ti-common-space-10);content:\"\";display:block;left:var(--ti-leftmenu-level2-left-line);position:absolute;top:var(--ti-common-space-10)}.ti3-leftmenu-level1-item{-ms-box-sizing:border-box;box-sizing:border-box;cursor:pointer;display:block;line-height:var(--ti-common-line-height-number);position:relative}.ti3-leftmenu-level1-item .ti3-leftmenu-triangle.ti3-leftmenu-triangle-down{transform:rotate(0)}.ti3-leftmenu-level1-item .ti3-leftmenu-triangle.ti3-leftmenu-triangle-up{transform:rotate(180deg)}.ti3-leftmenu-level1-item:hover{background-color:var(--ti-leftmenu-item-bg-color-hover);color:var(--ti-leftmenu-item-text-color-hover)}.ti3-leftmenu-level1-item:hover .ti3-leftmenu-triangle.ti3-leftmenu-triangle-down,.ti3-leftmenu-level1-item:hover .ti3-leftmenu-triangle.ti3-leftmenu-triangle-up{border-top-color:var(--ti-leftmenu-icon-color-hover)}.ti3-leftmenu-level1-item.ti3-leftmenu-level1-active-item{color:var(--ti-leftmenu-item-selected-text-color)}.ti3-leftmenu-level1-active-item,.ti3-leftmenu-level2-active{position:relative}.ti3-leftmenu-level1-active-item .ti3-leftmenu-triangle.ti3-leftmenu-triangle-down,.ti3-leftmenu-level2-active .ti3-leftmenu-triangle.ti3-leftmenu-triangle-down{border-top-color:var(--ti-leftmenu-icon-color-hover)}.ti3-leftmenu-level1-active-item:before,.ti3-leftmenu-level2-active:before{background:var(--ti-leftmenu-item-selected-border-left-color);bottom:var(--ti-common-space-10);content:\"\";position:absolute;top:var(--ti-common-space-10);width:var(--ti-leftmenu-border-left-width);z-index:1}.ti3-leftmenu-level1-active-item:before{left:var(--ti-common-space-10)}.ti3-leftmenu-level2-active:before{left:var(--ti-leftmenu-level2-left-line)}.ti3-leftmenu-level1-label{display:block;padding:var(--ti-leftmenu-level1-label-padding)}.ti3-leftmenu-level1-state-icon{font-size:12px;position:absolute;right:9px;top:calc(50% - 11px)}.ti3-leftmenu-level2{-ms-box-sizing:border-box;box-sizing:border-box;cursor:pointer;display:block;line-height:var(--ti-common-line-height-number);padding:var(--ti-leftmenu-second-level-padding);position:relative}.ti3-leftmenu-level2:hover{background-color:var(--ti-leftmenu-item-bg-color-hover);color:var(--ti-leftmenu-item-text-color-hover)}.ti3-leftmenu-level2.ti3-leftmenu-level2-active{color:var(--ti-leftmenu-item-selected-text-color)}.ti3-leftmenu-triangle{border-left:calc(var(--ti-leftmenu-angle-normal-width)/2) solid transparent;border-right:calc(var(--ti-leftmenu-angle-normal-width)/2) solid transparent;border-top:var(--ti-leftmenu-angle-normal-height) solid var(--ti-leftmenu-icon-color-nomal);content:\"\";height:0;position:absolute;right:14px;top:calc(50% - var(--ti-leftmenu-angle-normal-height)/2);width:0}.ti3-leftmenu-href{font-size:16px;position:absolute;right:10px;top:50%;transform:translateY(-50%)}.ti3-leftmenu-logo{height:60px}.ti3-leftmenu-logo .ti3-leftmenu-logo-ficon{color:var(--ti-leftmenu-head-text-color);font-size:60px;line-height:60px;vertical-align:middle}.ti3-leftmenu-item-group:after{border-top:1px dashed var(--ti-common-color-line-dividing);content:\"\";display:block;margin:var(--ti-common-space-10) var(--ti-leftmenu-group-padding-right) var(--ti-common-space-10) var(--ti-leftmenu-group-padding-left)}.ti3-leftmenu-content{background:#fff;bottom:0;left:240px;overflow:auto;position:absolute;right:0;top:0}.ti3-leftmenu-group-container{display:block;position:relative}.ti3-leftmenu-group-container:after{border:1px dashed var(--ti-common-color-line-dividing);bottom:0;content:\"\";left:var(--ti-leftmenu-group-padding-left);position:absolute;top:auto;width:calc(100% - var(--ti-leftmenu-group-padding-left) - var(--ti-leftmenu-group-padding-right))}.ti3-leftmenu-group-container:last-child:after{display:none}.ti3-leftmenu-group-wrapper{color:var(--ti-common-color-text-weaken);cursor:pointer;line-height:var(--ti-common-line-height-number);padding:var(--ti-leftmenu-group-padding-top) var(--ti-leftmenu-group-padding-right) var(--ti-leftmenu-group-padding-bottom) var(--ti-leftmenu-group-padding-left);position:relative}.ti3-leftmenu-group-wrapper .ti3-leftmenu-triangle-down{transform:rotate(0)}.ti3-leftmenu-group-wrapper .ti3-leftmenu-triangle-up{display:none}.ti3-leftmenu-group-wrapper:hover{background:var(--ti-leftmenu-item-bg-color-hover);color:var(--ti-leftmenu-item-text-color-hover)}.ti3-leftmenu-group-wrapper:hover .ti3-leftmenu-triangle-down{border-top-color:var(--ti-leftmenu-icon-color-hover)}.ti3-leftmenu-group-wrapper:hover .ti3-leftmenu-triangle-up{border-top-color:var(--ti-leftmenu-icon-color-hover);display:block;transform:rotate(180deg)}.ti3-leftmenu-group-active{color:var(--ti-leftmenu-item-text-color-hover)}.ti3-leftmenu-group-active .ti3-leftmenu-triangle-down{border-top-color:var(--ti-leftmenu-icon-color-hover)}.ti3-leftmenu-group-active:before{background:var(--ti-leftmenu-icon-color-hover);content:\"\";height:calc(100% - var(--ti-leftmenu-group-padding-top) - var(--ti-leftmenu-group-padding-bottom));left:10px;position:absolute;top:var(--ti-leftmenu-group-padding-top);width:2px}.ti3-leftmenu-foot{-ms-box-sizing:border-box;background:var(--ti-leftmenu-container-bg-color);bottom:0;box-sizing:border-box;color:var(--ti-leftmenu-foot-text-color);line-height:var(--ti-common-line-height-number);overflow:hidden;padding:var(--ti-leftmenu-foot-padding-top) 0 var(--ti-leftmenu-foot-padding-bottom) var(--ti-leftmenu-foot-padding-left);position:absolute;stroke-width:var(--ti-leftmenu-foot-space-right);text-align:var(--ti-leftment-leftmenu-head-text-align);z-index:3}.ti3-leftmenu-foot:hover{color:var(--ti-leftmenu-foot-text-color-hover);cursor:pointer}.ti3-leftmenu-foot:before{border-bottom:1px solid var(--ti-common-color-line-dividing);content:\"\";display:block;left:0;position:relative;top:calc(-1*var(--ti-leftmenu-foot-padding-top))}"]
            },] }
];
TiLeftmenuComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: Router, decorators: [{ type: Optional }] },
    { type: ActivatedRoute, decorators: [{ type: Optional }] },
    { type: TiRenderer },
    { type: NgZone },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
TiLeftmenuComponent.propDecorators = {
    active: [{ type: Input }],
    activeChange: [{ type: Output }],
    reloadState: [{ type: Input }],
    reloadStateChange: [{ type: Output }],
    routable: [{ type: Input }],
    toggleable: [{ type: Input }],
    collapsed: [{ type: Input }],
    collapsedChange: [{ type: Output }]
};
