// @cloud/tiny3/esm2015/components/leftmenu/TiLeftmenuGroupComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * TiLeftmenuGroup 是分组菜单组件，嵌在 TiLeftmenu组件中使用。
 *
 * 3.1.3新增
 *
 */
export class TiLeftmenuGroupComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 设置当前分组菜单项是否展开
         */
        this.expanded = false;
        /**
         * @ignore
         * 当前TiLeftmenuGroup下的所有TiLeftmenuLevel1的集合
         */
        this.children = [];
    }
    /**
     * @ignore
     * 在模板中使用，实现分组的展开与收起
     */
    clickGroup() {
        this.expanded = !this.expanded;
    }
    /**
     * @ignore
     * 在模板中使用，当前分组中的子项是否有激活项
     */
    hasActivedChild() {
        if (!this.children || this.children.length <= 0) {
            return false;
        }
        for (const level1 of this.children) {
            if (level1.isActived || level1.hasActivedChildren()) {
                return true;
            }
        }
        return false;
    }
}
TiLeftmenuGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-leftmenu-group',
                template: "<div class=\"ti3-leftmenu-group-wrapper\"\r\n    [ngClass]=\"{'ti3-leftmenu-group-active': !expanded && hasActivedChild()}\"\r\n    [tabindex]=\"0\"\r\n    tiOutline\r\n    (click)=\"clickGroup()\"\r\n    (keydown.enter)=\"clickGroup()\"\r\n    [id]=\"appendId('leftmenu_group')\">\r\n    <ng-content select='ti-leftmenu-group-item'></ng-content>\r\n    <span class=\"ti3-leftmenu-triangle\" [ngClass]=\"{'ti3-leftmenu-triangle-down' : !expanded ,\r\n    'ti3-leftmenu-triangle-up' : expanded}\"></span>\r\n</div>\r\n<div [hidden]=\"!expanded\">\r\n    <ng-content></ng-content>\r\n</div>",
                host: {
                    '[class.ti3-leftmenu-group-container]': 'true'
                }
            },] }
];
TiLeftmenuGroupComponent.propDecorators = {
    expanded: [{ type: Input }]
};
