// @cloud/tiny3/esm2015/components/leftmenu/TiLeftmenuModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiIconModule } from '../icon/TiIconModule';
import { TiRendererModule } from '../../services/renderer/TiRendererModule';
import { TiLeftmenuComponent } from './TiLeftmenuComponent';
import { TiLeftmenuHeadComponent } from './TiLeftmenuHeadComponent';
import { TiLeftmenuGroupItemComponent } from './TiLeftmenuGroupItemComponent';
import { TiLeftmenuGroupComponent } from './TiLeftmenuGroupComponent';
import { TiLeftmenuLevel1Component } from './TiLeftmenuLevel1Component';
import { TiLeftmenuItemComponent } from './TiLeftmenuItemComponent';
import { TiLeftmenuLevel2Component } from './TiLeftmenuLevel2Component';
import { TiLeftmenuFootComponent } from './TiLeftmenuFootComponent';
import { TiOutlineModule } from '../../directives/outline/TiOutlineModule';
export class TiLeftmenuModule {
}
TiLeftmenuModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiIconModule,
                    TiRendererModule,
                    TiOutlineModule
                ],
                exports: [TiLeftmenuComponent,
                    TiLeftmenuHeadComponent,
                    TiLeftmenuLevel1Component,
                    TiLeftmenuItemComponent,
                    TiLeftmenuLevel2Component,
                    TiLeftmenuGroupComponent,
                    TiLeftmenuGroupItemComponent,
                    TiLeftmenuFootComponent],
                declarations: [TiLeftmenuComponent,
                    TiLeftmenuHeadComponent,
                    TiLeftmenuLevel1Component,
                    TiLeftmenuItemComponent,
                    TiLeftmenuLevel2Component,
                    TiLeftmenuGroupComponent,
                    TiLeftmenuGroupItemComponent,
                    TiLeftmenuFootComponent]
            },] }
];
export * from './TiLeftmenuComponent';
export { TiLeftmenuComponent } from './TiLeftmenuComponent';
export { TiLeftmenuHeadComponent } from './TiLeftmenuHeadComponent';
export { TiLeftmenuLevel1Component } from './TiLeftmenuLevel1Component';
export { TiLeftmenuItemComponent } from './TiLeftmenuItemComponent';
export { TiLeftmenuLevel2Component } from './TiLeftmenuLevel2Component';
export { TiLeftmenuGroupComponent } from './TiLeftmenuGroupComponent';
export { TiLeftmenuGroupItemComponent } from './TiLeftmenuGroupItemComponent';
export { TiLeftmenuFootComponent } from './TiLeftmenuFootComponent';
