// @cloud/tiny3/esm2015/components/tab/TiTabModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TiTabsComponent } from './TiTabsComponent';
import { TiTabComponent } from './TiTabComponent';
import { TiTabHeaderComponent } from './TiTabHeaderComponent';
import { TiIncludeModule } from '../../directives/include/TiIncludeModule';
import { TiRendererModule } from '../../services/renderer/TiRendererModule';
import { TiIconModule } from '../icon/TiIconModule';
import { TiDroplistModule } from '../droplist/TiDroplistModule';
export class TiTabModule {
}
TiTabModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    TiDroplistModule,
                    TiIconModule,
                    TiIncludeModule,
                    TiRendererModule
                ],
                exports: [TiTabsComponent, TiTabComponent, TiTabHeaderComponent],
                declarations: [TiTabsComponent, TiTabComponent, TiTabHeaderComponent]
            },] }
];
export { TiTabComponent } from './TiTabComponent';
export { TiTabsComponent } from './TiTabsComponent';
export { TiTabHeaderComponent } from './TiTabHeaderComponent';
