// @cloud/tiny3/esm2015/components/tab/TiTabHeaderComponent.js
import { Component, ElementRef } from '@angular/core';
import { TiTabComponent } from './TiTabComponent';
import { Util } from '../../utils/Util';
/**
 * TiTabHeaderComponent 是单个页签的标题头组件，其包裹的元素会作为当前页签的标题头
 *
 */
export class TiTabHeaderComponent {
    constructor(tab, elementRef) {
        this.tab = tab;
        this.elementRef = elementRef;
    }
    ngAfterContentInit() {
        // 由于此处的AfterContentInit会比TiTabComponent组件的OnInit
        // 执行得晚，所以ti-tab-header中的内容会覆盖header接口传入的字符串值
        this.tab.headNode = this.elementRef.nativeElement;
        if (!this.tab.header) {
            // 处理自定义dom没有header接口时对应下拉面板展示
            this.tab.header = this.elementRef.nativeElement.innerText;
        }
    }
    ngAfterViewChecked() {
        if (!this.tab.tabset.hasSlider) {
            return;
        }
        if (Util.isUndefined(this.oldHeaderWidth)) {
            this.oldHeaderWidth = this.elementRef.nativeElement.offsetWidth;
            return;
        }
        const headerWidth = this.elementRef.nativeElement.offsetWidth;
        if (headerWidth !== this.oldHeaderWidth) {
            this.tab.tabset.headerWidthChange = true;
            this.oldHeaderWidth = headerWidth;
        }
    }
}
TiTabHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-tab-header',
                template: "<ng-content></ng-content>",
                host: {
                    '[class.ti3-tab-head-no-outline]': 'true'
                }
            },] }
];
TiTabHeaderComponent.ctorParameters = () => [
    { type: TiTabComponent },
    { type: ElementRef }
];
