// @cloud/tiny3/esm2015/components/tab/TiTabComponent.js
import { Component, ContentChild, ElementRef, EventEmitter, HostBinding, Input, Output, Renderer2, TemplateRef } from '@angular/core';
import { TiTabsComponent } from './TiTabsComponent';
import { Util } from '../../utils/Util';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * TiTabComponent 是单个页签组件，包含了页签标题头指令TiTabHeader 和 该页签对应的内容部分
 *
 */
export class TiTabComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, tabset) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.tabset = tabset;
        /**
         * 当前页签的激活状态发生改变时向外进行事件通知,传递的参数为 boolean 类型，表示当前页签是否是激活状态
         */
        this.activeChange = new EventEmitter();
        /**
         * @ignore
         * 记录此页签的内容区是否已经加载过
         */
        this.loaded = false;
    }
    ngOnInit() {
        super.ngOnInit();
        this.createHeadNode();
        this.tabset.addTab(this);
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        const activeObj = changes['active'];
        if (activeObj && activeObj.currentValue && this.tabset.selectedTab !== this && !this.disabled) {
            this.tabset.activeTab(this, !activeObj.firstChange);
        }
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        // 父容器TiTabs的OnDestroy执行晚于TiTab的OnDestroy，所以此处需要setTimeout来延时
        setTimeout(() => {
            if (this.tabset.destroyed) {
                return;
            }
            this.tabset.removeTab(this);
            // 如果没有tab，那么下划线隐藏
            if (this.tabset.tabs.length === 0) {
                this.tabset.tiRenderer.setStyles(this.tabset.slider.nativeElement, {
                    width: 0,
                    left: 0
                });
            }
        }, 0);
    }
    createHeadNode() {
        if (Util.isString(this.header)) {
            this.headNode = this.renderer2.createText(this.header);
        }
    }
}
TiTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-tab',
                template: "<ng-content></ng-content>\r\n<ng-container *ngIf=\"contentTemplate && loaded\">\r\n    <ng-container *ngTemplateOutlet=\"contentTemplate\"></ng-container>\r\n</ng-container>\r\n",
                host: {
                    '[class.ti3-tab-pane]': 'true'
                },
                styles: [":host.ti3-tab-pane{animation-duration:.3s;animation-fill-mode:backwards;animation-timing-function:ease;display:none;padding-top:var(--ti-common-space-3x)}:host.ti3-tab-pane.ti3-tab-active{display:block}:host.ti3-tab-pane.active-remove{display:none!important}:host.ti3-tab-pane.active-add{animation-name:tiFadeIn}"]
            },] }
];
TiTabComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TiTabsComponent }
];
TiTabComponent.propDecorators = {
    header: [{ type: Input }],
    disabled: [{ type: Input }],
    active: [{ type: Input }, { type: HostBinding, args: ['class.ti3-tab-active',] }],
    activeChange: [{ type: Output }],
    contentTemplate: [{ type: ContentChild, args: ['tiTabContent', /* TODO: add static flag */ { static: true },] }]
};
