// @cloud/tiny3/esm2015/services/modal/TiBackdropComponent.js
import { Component, HostListener, Renderer2 } from '@angular/core';
import { animate, state, style, transition, trigger } from '@angular/animations';
/**
 * @ignore
 */
export const animateStyle = '0.3s cubic-bezier(0.40, 0.00, 0.20, 1.00)';
/**
 * @ignore
 */
export class TiBackdropComponent {
    constructor(renderer) {
        this.renderer = renderer;
        this.index = 0; // 多个弹出框情况下，zindex层级设置辅助
    }
    set showState(value) {
        this._showState = value;
    }
    get showState() {
        return this._showState;
    }
    // 在微前端，多个子应用之间url路径跳转,等到弹窗隐藏结束后，this.destroy() 已经无效。
    onhashchange() {
        this.destroy();
    }
    onAnimationDone($event) {
        // 弹框隐藏动画结束后,销毁弹框DOM
        if ($event.toState === 'hide') {
            this.destroy();
            return;
        }
    }
}
TiBackdropComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-backdrop',
                template: `<div class='ti3-modal-backdrop'
                [@.disabled]='!animation' [@backdropAnimate] = 'showState'
                (@backdropAnimate.done)='onAnimationDone($event)'></div>`,
                animations: [
                    trigger('backdropAnimate', [
                        state('show', style({
                            opacity: 0.2
                        })),
                        state('hide, void', style({
                            opacity: 0
                        })),
                        transition('show => hide', animate(animateStyle)),
                        transition('hide => show', animate(animateStyle))
                    ])
                ],
                styles: [".ti3-modal-backdrop{background-color:var(--ti-modal-backdrop-bg-color);bottom:0;box-sizing:border-box;left:0;opacity:.2;position:fixed;right:0;top:0;z-index:1040}"]
            },] }
];
TiBackdropComponent.ctorParameters = () => [
    { type: Renderer2 }
];
TiBackdropComponent.propDecorators = {
    onhashchange: [{ type: HostListener, args: ['window:hashchange',] }]
};
