// @cloud/tiny3/esm2015/services/modal/TiModalModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiPopupModule } from '../popup/TiPopupModule';
import { TiModalComponent } from './TiModalComponent';
import { TiModalNoAnimationComponent } from './TiModalNoAnimationComponent';
import { TiBackdropComponent } from './TiBackdropComponent';
import { TiBackdropNoAnimationComponent } from './TiBackdropNoAnimationComponent';
import { TiModalHeaderComponent } from './TiModalHeaderComponent';
import { TiModalBodyComponent } from './TiModalBodyComponent';
import { TiModalFooterComponent } from './TiModalFooterComponent';
import { TiRendererModule } from '../renderer/TiRendererModule';
import { TiDraggableModule } from '../../directives/drag/TiDraggableModule';
import { TiIconModule } from '../../components/icon/TiIconModule';
import { TiOutlineModule } from '../../directives/outline/TiOutlineModule';
export class TiModalModule {
}
TiModalModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule,
                    TiDraggableModule,
                    TiIconModule,
                    TiPopupModule,
                    TiOutlineModule,
                    TiRendererModule],
                exports: [TiModalComponent, TiModalNoAnimationComponent,
                    TiBackdropNoAnimationComponent, TiModalHeaderComponent,
                    TiModalFooterComponent, TiModalBodyComponent],
                declarations: [TiModalComponent, TiModalNoAnimationComponent,
                    TiBackdropNoAnimationComponent, TiModalHeaderComponent,
                    TiModalFooterComponent, TiModalBodyComponent, TiBackdropComponent],
                entryComponents: [TiModalComponent, TiModalNoAnimationComponent, TiBackdropNoAnimationComponent, TiBackdropComponent]
            },] }
];
export * from './TiModalInterface';
export { TiModalHeaderComponent } from './TiModalHeaderComponent';
export { TiModalFooterComponent } from './TiModalFooterComponent';
export { TiModalBodyComponent } from './TiModalBodyComponent';
