// @cloud/tiny3/esm2015/services/modal/TiModalService.js
import { __rest } from "tslib";
import { Injectable } from '@angular/core';
import { TiPopupService } from '../popup/TiPopupService';
import { TiBackdropComponent } from './TiBackdropComponent';
import { TiBackdropNoAnimationComponent } from './TiBackdropNoAnimationComponent';
import { TiModalComponent } from './TiModalComponent';
import { TiModalNoAnimationComponent } from './TiModalNoAnimationComponent';
import { TiBrowser, Util } from '../../utils/Util';
import { TiModalModule } from './TiModalModule';
import * as i0 from "@angular/core";
import * as i1 from "../popup/TiPopupService";
import * as i2 from "./TiModalModule";
/**
 * 弹框组件提供服务方式供业务使用，使用该服务时需要引入模块TiModalModule，开发者通过调用TiModalService.open方法生成弹出框
 *
 * <p><span style="color: red">使用此组件时需要开发者在项目模块(建议在根模块)
 * 中引入BrowserAnimationsModule。</span>这是因为此组件中使用了Angular动画，需要引入BrowserAnimationsModule，
 * 但是 BrowserAnimationsModule 不能在懒加载模块被重复引入，所以需要开发者来引入BrowserAnimationsModule，保证其引入一次。</p>
 *
 * <example-url>../tiny3demo/#/modal/modal-all</example-url>
 */
export class TiModalService {
    constructor(popService) {
        this.popService = popService;
        /**
         * 页面中存在的所有弹框实例，用于处理业务中无法明确当前有多少有效弹窗场景
         */
        this.openedModals = []; // 所有当前处于打开状态弹框缓存数组
        /* tslint:disable:no-null-keyword */
        this.backdropComponentRef = null; // 单个页面中多个弹框出现时,只允许存在一个遮罩层,因此使用该标志位用来缓存遮罩实例
        this.defaultConfig = {
            backdrop: true,
            draggable: true,
            animation: true,
            closeOnEsc: true,
            closeIcon: true
        };
    }
    /**
     * 生成弹框方法
     *
     * @param content 弹框内容，支持TemplateRef及组件形式，内容由ti-modal-header、ti-modal-body及ti-modal-footer组件构成弹框内容整体。
     *                不支持字符串形式，如果误传入字符串形式，不仅会有报错，还会存在XSS攻击风险，不过XSS攻击风险已在10.1.4版本已处理。
     * @param [config] 弹框配置信息
     * @returns 弹框实例信息，使用该实例信息可调用弹框的close等方法；弹框内容为component类型时，也可通过该实例信息的content属性获取弹框内容数据
     */
    open(content, config) {
        let modalInstance; // 弹框最终返回的实例对象
        let modalComponentRef; // 弹框组件实例,使用该实例获取弹框DOM元素及做属性赋值操作
        const modalConfig = Object.assign(Object.assign({}, this.defaultConfig), config); // 弹框配置合并
        let modal; // generateModal方法返回值
        // 通过id唯一标识防止重复打开相同弹框,id重复情况下返回先前实例
        if (modalConfig && !Util.isUndefined(modalConfig.id)) {
            const index = this.openedModals.findIndex((item) => {
                return item._id === modalConfig.id;
            });
            if (index !== -1) {
                return this.openedModals[index];
            }
        }
        // 根据beforeClose返回处理弹窗的销毁：业务在beforeClose中调用弹框销毁
        const destroy = (reason) => {
            if (typeof modalConfig.beforeClose === 'function') {
                modalConfig.beforeClose(modalInstance, reason);
            }
            else {
                destroyModal(reason);
            }
        };
        // 销毁弹框：通过调用组件的动画状态方式实现动画处理
        const destroyModal = (reason) => {
            if (reason) {
                if (typeof modalConfig.close === 'function') {
                    modalConfig.close(modalInstance);
                }
            }
            else {
                if (typeof modalConfig.dismiss === 'function') {
                    modalConfig.dismiss(modalInstance);
                }
            }
            modalComponentRef.instance.showState = 'hide';
        };
        // 销毁弹框实体：弹框动画执行完成后调用该方法
        const removeModalEle = () => {
            // 销毁弹框实体DOM
            modal.hide();
            // 移除缓存实例
            const index = this.openedModals.indexOf(modalInstance);
            this.openedModals.splice(index, 1);
            modalInstance = null;
            // 销毁backdrop
            destroyBackdrop();
        };
        const destroyBackdrop = () => {
            // backdrop定义为false时，不进行处理
            // backdropComponentRef 已经销毁
            if (!modalConfig.backdrop || !this.backdropComponentRef) {
                return;
            }
            // 修改backdrop zIndex，确保有多个弹窗情况下，最外层弹窗不被遮挡
            this.backdropComponentRef.instance.index = this.openedModals.length - 1;
            // 当前还有其它弹框情况下,不用销毁弹框
            if (this.openedModals.length !== 0) {
                return;
            }
            this.backdropComponentRef.instance.showState = 'hide';
        };
        // 将当前要打开的modal放到openedModals列表中
        modalInstance = {
            _id: modalConfig && modalConfig.id,
            close() {
                destroy(true);
            },
            dismiss() {
                destroy(false);
            },
            destroy(reason) {
                destroyModal(reason);
            },
            _remove: removeModalEle,
            content: null
        };
        modal = this.generateModal(content, modalConfig, modalInstance);
        // 生成弹框window,并返回对应的组件实例
        modalComponentRef = modal.modalComponentRef;
        // 生成模态背景backdrop
        this.backdropComponentRef = this.generateBackdrop(modalConfig);
        return modalInstance;
    }
    generateModal(content, config, modalInstance) {
        const { context } = config, modalConfig = __rest(config, ["context"]);
        // tiModalComponent生成需要使用到的上下文
        const modalContext = Object.assign({ modalInstance, index: this.openedModals.length }, modalConfig);
        // 内容部分需要使用到的上下文
        const contentContext = Object.assign({ close: modalInstance.close, dismiss: modalInstance.dismiss }, context);
        let modalComponent;
        modalComponent = (TiBrowser.isIE() && (TiBrowser.version() === 9))
            ? this.popService.create(TiModalNoAnimationComponent)
            : this.popService.create(TiModalComponent);
        const modalComponentRef = modalComponent.show({
            content,
            context: modalContext,
            contentContext,
            container: 'body'
        });
        // 控制元素动画呈现，元素呈现后修改动画状态才会有动画效果，因此在此处处理
        modalComponentRef.instance.showState = 'show';
        modalInstance.content = modalComponentRef.tiContentRef.componentRef;
        this.openedModals.push(modalInstance);
        return { modalComponentRef, hide: modalComponent.hide };
    }
    generateBackdrop(config) {
        // 配置中不需要backdrop,不生成
        if (!config.backdrop) {
            return this.backdropComponentRef;
        }
        if (this.backdropComponentRef) {
            // backdrop已存在的情况下,修改其zIndex,并确保不生成backdrop
            if (this.backdropComponentRef.instance.showState === 'show') {
                this.backdropComponentRef.instance.index = this.openedModals.length - 1;
                return this.backdropComponentRef;
                // 如果backdrop正在销毁(动画执行期间)时再新生成一个backdrop，这时应该将旧的backdrop立即销毁
            }
            else {
                this.backdropComponentRef.destroy();
            }
        }
        const backdropComponent = (TiBrowser.isIE() && (TiBrowser.version() === 9))
            ? this.popService.create(TiBackdropNoAnimationComponent)
            : this.popService.create(TiBackdropComponent);
        const backdropRef = backdropComponent.show({
            context: {
                animation: config.animation,
                destroy: () => {
                    backdropComponent.hide();
                    this.backdropComponentRef = null;
                }
            },
            container: 'body'
        });
        // 控制元素动画呈现，元素呈现后修改动画状态才会有动画效果，因此在此处处理
        backdropRef.instance.showState = 'show';
        return backdropRef;
    }
}
TiModalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TiModalService_Factory() { return new TiModalService(i0.ɵɵinject(i1.TiPopupService)); }, token: TiModalService, providedIn: i2.TiModalModule });
TiModalService.decorators = [
    { type: Injectable, args: [{
                providedIn: TiModalModule
            },] }
];
TiModalService.ctorParameters = () => [
    { type: TiPopupService }
];
