// @cloud/tinyplus3/fesm2015/cloud-tinyplus3-locale-ocb.js
import { TiLocale } from '@cloud/tiny3';
import { defaultzh_CN, dateHyphen, currencyEUR, defaulten_US, defaultfr_FR } from '@cloud/tinyplus3/locale/i18n';

const siteConfig = {
    tpRegionName: 'ocb',
    tpSupportMFA: true,
    tpSupportEmail: true,
    tpRegionSmallItem: false,
    tpRegionSelectStyle: true
};

class TpConfig {
    static init(language = 'zh-CN') {
        const words = TiLocale.getWords();
        // 设置所有语言包
        TiLocale.setWords({
            'zh-CN': Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, words['zh-CN']), defaultzh_CN), dateHyphen), currencyEUR), siteConfig),
            'en-US': Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, words['en-US']), defaulten_US), dateHyphen), currencyEUR), siteConfig),
            'fr-FR': Object.assign(Object.assign(Object.assign(Object.assign(Object.assign({}, words['fr-FR']), defaultfr_FR), dateHyphen), currencyEUR), siteConfig)
        });
        // 设置默认语言
        TiLocale.setLocale(language);
    }
}

/**
 * Generated bundle index. Do not edit.
 */

export { TpConfig };
