// @cloud/tinyplus3/esm2015/components/score/TpScoreComponent.js
import { Component, ElementRef, Input, Renderer2 } from '@angular/core';
import { TiFormComponent, TiLocale } from '@cloud/tiny3';
/**
 * NPS评分组件
 *
 * <example-url>../tinyplus3demo/#/score/score-all</example-url>
 */
export class TpScoreComponent extends TiFormComponent {
    constructor(hostRef, renderer) {
        super(hostRef, renderer);
        this.hostRef = hostRef;
        this.renderer = renderer;
        /**
         * @ignore
         * 定义含有11项成员的数组
         */
        this.scoreArray = new Array(11);
        this.minText = this.minText || TiLocale.getLocaleWords().tpScore.minText;
        this.maxText = this.maxText || TiLocale.getLocaleWords().tpScore.maxText;
    }
    /**
     * @ignore
     * 选择评分时，触发事件
     */
    onClick(value) {
        this.model = value;
    }
}
TpScoreComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-score',
                template: "<span *ngFor=\"let item of scoreArray index as i; let last=last;\"\r\n      class=\"tp-score-item\"\r\n      [ngClass]=\"{'tp-score-item-active': model === i,\r\n                  'tp-score-item-last-child': last}\"\r\n      tiOutline\r\n      [tabindex]=\"0\" (keydown.enter)=\"onClick(i)\"\r\n      (click)=\"onClick(i)\"\r\n      [id]=\"appendId(i)\">{{i}}</span>\r\n<br>\r\n\r\n<span class=\"tp-score-text\">\r\n    <span [id]=\"appendId('min')\">{{minText}}</span>\r\n    <span [id]=\"appendId('max')\">{{maxText}}</span>\r\n</span>",
                providers: [TiFormComponent.getValueAccessor(TpScoreComponent)],
                styles: ["::ng-deep :root{--tp-score-item-height:28px;--tp-score-item-width:36px}:host{display:inline-block;font-size:var(--ti-common-font-size-base);font-weight:400;line-height:var(--ti-common-line-height-number)}.tp-score-item{background-color:var(--ti-common-color-bg-light-normal);border-radius:var(--ti-common-border-radius-normal) 0 var(--ti-common-border-radius-normal);cursor:pointer;display:inline-block;height:var(--tp-score-item-height);line-height:var(--tp-score-item-height);margin-right:var(--ti-common-space-base);text-align:center;width:var(--tp-score-item-width)}.tp-score-item:hover{background-color:var(--ti-common-color-bg-light-emphasize)}.tp-score-item.tp-score-item-active{background-color:var(--ti-common-color-bg-emphasize);color:var(--ti-common-color-text-white)}.tp-score-item-last-child{margin-right:0}.tp-score-text{color:var(--ti-common-color-text-weaken);display:flex;justify-content:space-between;margin-top:var(--ti-common-space-2x)}"]
            },] }
];
TpScoreComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TpScoreComponent.propDecorators = {
    minText: [{ type: Input }],
    maxText: [{ type: Input }]
};
