// @cloud/tinyplus3/esm2015/components/priceline/TpPricelineComponent.js
// tslint:disable-next-line: no-implicit-dependencies
import { Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
// tslint:disable-next-line: no-implicit-dependencies
import { TiButtongroupComponent, TiFormComponent, TiLocale, Util } from '@cloud/tiny3';
/**
 *  价格曲线组件, 对应tiny2中btnSlider组件price-option接口配置后的功能
 *
 *  9.0.4版本新增
 *
 * <example-url>../tinyplus3demo/#/priceline/priceline-all</example-url>
 */
export class TpPricelineComponent extends TiFormComponent {
    constructor(hostRef, renderer) {
        super(hostRef, renderer);
        /**
         * 按钮配置数据
         */
        this.items = [];
        /**
         * 设置单选按钮组是否可以取消选中
         */
        this.deselectable = false;
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngAfterViewInit() {
        super.ngAfterViewInit();
        this.initCanvas();
        if (this.canvasOption) {
            // 视图渲染后修改了视图相关数据，所以使用setTimeout
            setTimeout(() => {
                this.draw(this.items);
            }, 0);
        }
        else {
            this.timer = setInterval(() => {
                this.initCanvas();
                if (this.canvasOption) {
                    this.draw(this.items);
                    clearInterval(this.timer);
                }
                // 测试IE11刷新需要40ms左右完成，URL访问需要80ms左右完成，真实业务场景比较复杂取100
            }, 100);
        }
        // tslint:disable-next-line: newline-per-chained-call
        this.setFocusableElems(this.buttonGroupCom.getFocusableElems().concat(this.nativeElement));
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnChanges(changes) {
        // 价格变化
        const items = changes['items'];
        if (items && !items.firstChange) {
            this.step = 10;
            this.previousItems = items.previousValue;
            this.changeDraw();
        }
    }
    onNgModelChange(data) {
        // 选中项变化，折线的标题会变化，重新绘制
        if (this.model) {
            this.draw(this.items);
        }
    }
    ngOnDestroy() {
        super.ngOnDestroy();
        clearInterval(this.timer);
    }
    // 价格变动时重绘动画
    // 实现思路，动画时间是200ms，每一次绘制间隔设置20ms，分10步绘制，
    // todo 1.绘制间隔设置为16.66接近 60帧动效，2.使用 requestAnimationFrame 代替 setTimeout
    changeDraw() {
        // 动画过程中变化的items
        const currentItems = this.items.map((item, index) => {
            // 重新生成一个新对象，避免修改原始数据
            const newItem = Object.assign({}, item);
            newItem.price = newItem.price - (newItem.price - this.previousItems[index].price) * (this.step / 10);
            return newItem;
        });
        this.draw(currentItems);
        this.step--;
        // 20 每一帧间隔时间20ms
        setTimeout(() => {
            if (this.step > -1) {
                this.changeDraw();
            }
        }, 20);
    }
    // 初始化价格曲线 canvas
    initCanvas() {
        let lineColor;
        let color;
        let activeColor;
        let startColor;
        let endColor;
        if (Util.supportsCssVars()) {
            // Chrome等支持CSSVar的浏览器，正常操作,  从CSSVar变量中取出颜色。
            const cSSStyleDeclaration = getComputedStyle(document.documentElement);
            lineColor = cSSStyleDeclaration.getPropertyValue('--ti-common-color-line-active');
            color = cSSStyleDeclaration.getPropertyValue('--ti-common-color-text-weaken');
            activeColor = cSSStyleDeclaration.getPropertyValue('--ti-common-color-text-primary');
            startColor = cSSStyleDeclaration.getPropertyValue('--ti-common-color-bg-emphasize');
            endColor = startColor;
        }
        else {
            // IE和低版Edge不支持CSSVar的浏览器，从一个CSS样式中取出颜色。
            const cSSStyleDeclaration = getComputedStyle(this.canvasRef.nativeElement);
            color = cSSStyleDeclaration.getPropertyValue('flood-color');
            activeColor = cSSStyleDeclaration.getPropertyValue('lighting-color');
            const cSSStyleDeclaration2 = getComputedStyle(this.canvasWrapRef.nativeElement);
            lineColor = cSSStyleDeclaration2.getPropertyValue('flood-color');
            startColor = cSSStyleDeclaration2.getPropertyValue('lighting-color');
            endColor = cSSStyleDeclaration2.getPropertyValue('stop-color');
            // tslint:disable-next-line: max-line-length
        }
        if (!lineColor || !color || !activeColor || !startColor || !endColor) {
            return;
        }
        // canvas曲线配置项
        this.canvasOption = {
            // 整个画板的高度
            height: 60,
            carveHeight: 40,
            // 经验值，设置为1折线不平滑
            lineWidth: 1.4,
            // chrome  拿到的颜色有前后空格
            lineColor: lineColor.trim(),
            title: TiLocale.getLocaleWords().tpPriceline.carveTitle,
            color: color.trim(),
            titleX: 0,
            titleY: 14,
            fontSize: '12px',
            fontFamily: 'Roboto,Microsoft YaHei,微软雅黑,Sans-serif',
            activeColor: activeColor.trim(),
            tagOffset: 6,
            tagUnit: '/' + TiLocale.getLocaleWords().tpPriceline.tagUnit,
            tagPrefix: TiLocale.getLocaleWords().tpPriceline.tagPrefix,
            currency: TiLocale.getLocaleWords().tpPriceline.tagCurrency,
            startColor: this.color2rgba(startColor.trim(), 0.26),
            endColor: this.color2rgba(endColor.trim(), 0)
        };
        // 设置canvas的高度
        this.renderer.setAttribute(this.canvasRef.nativeElement, 'height', this.canvasOption.height);
        // 设置canvas 宽度和canvas上下文，在IE11环境 ngAfterViewInit中设置，获取宽度有误差，所以放在此处
        const wrapStyles = getComputedStyle(this.canvasWrapRef.nativeElement);
        this.canvasWidth = parseInt(wrapStyles.width, 10);
        this.renderer.setAttribute(this.canvasRef.nativeElement, 'width', String(this.canvasWidth));
        this.ctx = this.canvasRef.nativeElement.getContext('2d');
    }
    // ie11 canvas 渐变透明不支持# 开头的八位颜色写法，所有需要转化成rgba表示形式
    // tslint:disable-next-line: prefer-function-over-method
    color2rgba(hex, alpha) {
        const colorRegex = /^#?[a-fA-F0-9]+$/;
        if (!colorRegex.test(hex) || hex.length !== 7) {
            return hex;
        }
        const values = [
            parseInt(hex.slice(1, 3), 16),
            parseInt(hex.slice(3, 5), 16),
            parseInt(hex.slice(5), 16)
        ];
        if (alpha >= 0 && alpha <= 1) {
            values.push(alpha);
        }
        else {
            values.push(1);
        }
        return `rgba(${values.join(',')})`;
    }
    // 画图实现思路：1 获取按钮的位置和宽度，2.根据按钮的位置宽度和items中的price数据，在每个价格变化时生成一个记录对象放在rects中，
    // 3. 根据记录的对象数组rects, 先画矩形背景和价格标签，4. 再根据rects画折线
    draw(items) {
        if (!items || items.length < 2 || !this.canvasOption) {
            return;
        }
        // 每次重绘前清空画板
        this.ctx.clearRect(0, 0, this.canvasWidth, this.canvasOption.height);
        this.rects = [];
        this.x = 0;
        this.lastX = 0;
        this.maxPrice = items[0].price;
        // tslint:disable-next-line: prefer-for-of
        for (let index = 1; index < this.items.length; index++) {
            if (items[index].price > this.maxPrice) {
                this.maxPrice = items[index].price;
            }
        }
        this.y = this.canvasOption.height - this.canvasOption.carveHeight * items[0].price / this.maxPrice;
        const buttons = this.nativeElement.querySelectorAll('.ti3-btn-item-radio');
        //  曲线图的背景和选择按钮的的
        this.lastButton = buttons[0].getBoundingClientRect();
        // 画标题
        this.drawText({
            x: this.canvasOption.titleX,
            y: this.canvasOption.titleY,
            text: this.canvasOption.title
        });
        // 遍历按钮数据
        const len = this.items.length;
        for (let index = 1; index < len; index++) {
            // 不是最后一条并且价格没有变化，不需要处理
            if (index !== len - 1 && this.items[index].price === this.items[index - 1].price) {
                continue;
            }
            const button = buttons[index].getBoundingClientRect();
            // 最后一条数据，有价格变化，需要画2次
            if (index === len - 1 && items[index].price !== items[index - 1].price) {
                this.x = this.x + (button.left - this.lastButton.left);
                this.record(items, index);
                // 最后一次
                this.lastX = this.x;
                this.y = this.canvasOption.height - this.getCurrentHeight(items[index].price, items);
                this.lastButton = button;
                this.x = this.x + button.width;
                this.record(items, index, true);
                // 最后一个数据没有价格变化
            }
            else if (index === len - 1 && items[index].price === items[index - 1].price) {
                this.x = this.x + ((button.left + button.width) - this.lastButton.left);
                this.record(items, index, true);
                // 中间的按钮数据
            }
            else {
                this.x = this.x + button.left - this.lastButton.left;
                this.record(items, index);
                this.lastX = this.x;
                this.y = this.canvasOption.height - this.getCurrentHeight(items[index].price, items);
                this.lastButton = button;
            }
        }
        // 画柱状背景和标题
        this.rects.forEach((rect) => {
            this.drawRect(rect);
        });
        // 画折线
        this.drawLine(this.ctx, { width: this.canvasOption.lineWidth, color: this.canvasOption.lineColor });
    }
    // 根据第一个按钮数据的价格计算当前价格柱状高度
    getCurrentHeight(curPrice, items) {
        return this.canvasOption.carveHeight * curPrice / this.maxPrice;
    }
    // 每一步遍历数据，记录一个配置点
    record(items, currrentIndex, isEnd) {
        const price = this.items[isEnd ? currrentIndex : currrentIndex - 1].price;
        // 柱状背景
        const rectOption = {
            x: this.lastX,
            y: this.y,
            width: isEnd ? (this.x - this.lastX) : (this.x - this.lastX - 2),
            height: this.canvasOption.height - this.y,
            // tslint:disable-next-line: max-line-length
            text: `${this.rects.length === 0 ? this.canvasOption.tagPrefix : ''}${this.canvasOption.currency}${price.toFixed(2)}${this.canvasOption.tagUnit}`,
            active: price === this.model.price
        };
        this.rects.push(rectOption);
    }
    // canvas 画文字
    drawText(config) {
        const option = Object.assign({ offset: this.canvasOption.tagOffset, color: this.canvasOption.color, fontSize: this.canvasOption.fontSize, fontFamily: this.canvasOption.fontFamily }, config);
        this.ctx.fillStyle = option.active ? this.canvasOption.activeColor : this.canvasOption.color;
        // font 同时设置字体大小和字体时，需用空格隔开
        this.ctx.font = `${option.fontSize} ${option.fontFamily}`;
        if (option.alignRight) {
            const textMetrics = this.ctx.measureText(option.text);
            option.x = (option.x + option.width) - textMetrics.width;
            option.y = option.y - 6;
        }
        this.ctx.fillText(option.text, option.x, option.y);
    }
    // 画线
    drawLine(ctx, option) {
        ctx.lineJoin = 'round';
        ctx.lineWidth = option.width;
        ctx.strokeStyle = option.color;
        ctx.beginPath();
        for (let index = 0; index < this.rects.length; index++) {
            const rect = this.rects[index];
            // 第一个矩形，从起点开始
            if (index === 0) {
                ctx.moveTo(Math.round(rect.x), Math.round(rect.y));
            }
            else {
                // 非第一个，竖线部分
                ctx.lineTo(Math.round(rect.x - 1), Math.round(rect.y));
            }
            // 矩形的左边到右边
            ctx.lineTo(Math.round(rect.x + rect.width + 1), Math.round(rect.y));
        }
        ctx.stroke();
    }
    // 画背景和文字
    drawRect(option) {
        // tslint:disable-next-line: max-line-length
        const canvasGradient = this.ctx.createLinearGradient(0, this.canvasOption.height - this.canvasOption.carveHeight, 0, this.canvasOption.height);
        canvasGradient.addColorStop(0, this.canvasOption.startColor);
        canvasGradient.addColorStop(1, this.canvasOption.endColor);
        this.ctx.fillStyle = canvasGradient;
        this.ctx.fillRect(option.x, option.y, option.width, option.height);
        this.drawText(Object.assign(Object.assign({}, option), { alignRight: true }));
    }
}
TpPricelineComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-priceline',
                template: "\r\n <section #canvasWrap class=\"tp-priceline-container\" >\r\n        <canvas #canvas ></canvas>\r\n</section>\r\n<ti-button-group [items]=\"items\"\r\n    [deselectable]=\"deselectable\"\r\n    [(ngModel)]=\"model\"\r\n    [id]=\"appendId('buttongroup')\"\r\n    (ngModelChange)=\"onNgModelChange($event)\"\r\n>\r\n</ti-button-group>",
                providers: [TiFormComponent.getValueAccessor(TpPricelineComponent)],
                styles: ["::ng-deep :root{--tp-priceline-canvas-height:60px}:host{display:inline-block;outline:none}.tp-priceline-container{flood-color:var(--ti-common-color-line-active);height:var(--tp-priceline-canvas-height);lighting-color:var(--ti-common-color-bg-emphasize);padding-bottom:var(--ti-common-space-2x);stop-color:var(--ti-common-color-bg-emphasize);width:100%}.tp-priceline-container canvas{display:block;flood-color:var(--ti-common-color-text-weaken);lighting-color:var(--ti-common-color-text-primary)}::ng-deep .tp-priceline-item-sup-discount{background-color:var(--ti-common-color-warn);border-radius:4px 0 4px 0}"]
            },] }
];
TpPricelineComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TpPricelineComponent.propDecorators = {
    items: [{ type: Input }],
    deselectable: [{ type: Input }],
    canvasRef: [{ type: ViewChild, args: ['canvas',] }],
    canvasWrapRef: [{ type: ViewChild, args: ['canvasWrap',] }],
    buttonGroupCom: [{ type: ViewChild, args: [TiButtongroupComponent,] }]
};
