// @cloud/tiny3/esm2015/components/pagination/TiPaginationComponent.js
import { Component, ElementRef, EventEmitter, Input, Optional, Output, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { Util } from '../../utils/Util';
import { TiLocale, TiLocaleFormat } from '../../locale/TiLocaleModule'; // 获取词条
import { TiFormComponent } from '../base/TiBaseModule';
import { TiTableComponent } from '../table/TiTableModule';
import { TiSelectComponent } from '../select/TiSelectModule';
/**
 * Pagination分页组件
 *
 * 分页组件一般情况下与表格组件配合使用，实现对表格数据的分页显示。
 *
 * <example-url>../tiny3demo/#/pagination/pagination-all</example-url>
 */
export class TiPaginationComponent extends TiFormComponent {
    constructor(render, hostRef, table, select) {
        super(hostRef, render);
        this.render = render;
        this.hostRef = hostRef;
        this.table = table;
        this.select = select;
        /**
         * 设置组件的类型：三种类型default(默认)、simple、mini
         */
        this.type = 'default';
        /**
         * 组件内部改变总条数时会触发该事件，向外通知当前总条数。
         */
        this.totalNumberChange = new EventEmitter();
        /**
         * 设置当前页码
         */
        this.currentPage = 1;
        /**
         * 组件内部改变当前页码时会触发该事件，向外通知当前页码数。
         */
        this.currentPageChange = new EventEmitter();
        /**
         * 设置是否总条数的显示与隐藏
         */
        this.showTotalNumber = true;
        /**
         * 设置是否显示页码跳转按钮（默认不显示)
         */
        this.showGotoLink = false;
        /**
         * 组件是否开启自动隐藏功能。开启后，当总条数totalNumber小于每页显示条数选项列表中的最小值时，隐藏分页
         */
        this.autoHide = false;
        /**
         * 组件是否禁用状态
         */
        this.disabled = false;
        /**
         * 设置最后一页的显示与隐藏（一般数据量比较大的场景隐藏最后一页）：默认显示
         */
        this.showLastPage = true;
        /**
         * 当操作选择框改变每页显示条数时会触发该事件，向外通知当前页码、每页显示条数和总条数。
         */
        this.pageNumChange = new EventEmitter();
        /**
         * 当操作选择框改变每页显示条数或操作页码改变当前页码时会触发该事件，向外通知当前页码、每页显示条数和总条数。
         */
        this.pageUpdate = new EventEmitter();
        /**
         * @ignore
         */
        this.mini = undefined;
        /**
         * @ignore
         * 用户输入即将跳转的页码
         */
        this.gotoPage = {};
        this.hidePage = false;
        this.defaultPageSize = {
            options: [10, 20, 50],
            size: 10,
            width: '60px',
            hide: false // 设置是否在页面上隐藏每页显示条数选项：下拉列表
        };
    }
    /**
     * @description 创建每一个页码对应的数据
     * @params key 页码显示，为数字或者是"···"
     * @params isActive 是否为当前激活页
     */
    static makePage(key, isActive) {
        return {
            key,
            active: isActive
        };
    }
    ngOnInit() {
        super.ngOnInit();
        // 组件对外各个接口的初始化处理
        this.initData();
        this.renderPage();
        if (this.isWithTable()) {
            this.subscription = this.table.paginationSubject.subscribe((value) => {
                if (!Util.isUndefined(value.totalNumber) && value.totalNumber !== this.totalNumber) {
                    this.totalNumber = value.totalNumber;
                    this.verifyCurrentPage(this.itemsPerPage.selected, true);
                    this.totalNumberChange.emit(this.totalNumber);
                }
                if (!Util.isUndefined(value.currentPage) && value.currentPage !== this.currentPage) {
                    this.currentPage = value.currentPage;
                    this.currentPageChange.emit(this.currentPage);
                }
            });
            // 表格记忆
            this.updateFromStorage();
        }
        // select下拉面板（dropsearch）中使用分页
        if (this.select) {
            this.select.pagination = this;
            this.select.dropsearchCom.pagination = this;
        }
    }
    ngOnChanges(changes) {
        const currentPageObj = changes['currentPage'];
        const totalNumberObj = changes['totalNumber'];
        if ((currentPageObj && !currentPageObj.firstChange) || (totalNumberObj && !totalNumberObj.firstChange)) {
            this.verifyCurrentPage(this.itemsPerPage.selected, true);
        }
    }
    ngDoCheck() {
        // 监听开发者设置的pageSize.size的改变
        if (!Util.isUndefined(this.pageSize) && this.pageSize.size !== this.oldSize) {
            const index = this.itemsPerPage.options.indexOf(this.pageSize.size);
            if (index !== -1) {
                this.itemsPerPage.selected = this.itemsPerPage.options[index];
                this.verifyCurrentPage(this.itemsPerPage.selected, true);
                this.oldSize = this.pageSize.size;
            }
        }
        if (this.oldTotalNumber !== this.totalNumber || this.itemsPerPage.selected !== this.oldSelected ||
            this.oldCurrentPage !== this.currentPage) {
            this.renderPage();
            this.oldTotalNumber = this.totalNumber;
            this.oldSelected = this.itemsPerPage.selected;
            this.oldCurrentPage = this.currentPage;
            this.updateFocusableElems = true; // select使用过这个变量后，会重置为false
        }
    }
    ngAfterViewChecked() {
        // select中使用的分页为default类型且不带左边select，所以此处限定type === 'default'
        if (this.type === 'default' && this.updateFocusableElems) {
            // 修复SSR报错：ERROR TypeError: this.nativeElement.querySelectorAll(...) is not a function or its return value is not iterable
            // this.setFocusableElems([...(this.nativeElement as HTMLElement).querySelectorAll('a')]);
            this.setFocusableElems(Array.prototype.slice.call(this.nativeElement.querySelectorAll('a')));
        }
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    // 数据初始化
    initData() {
        // 总条数数字国际化处理，拼接成 ---> 总条数: 条数
        this.setTotalNumberFormat();
        // 初始化获取itemsPerPage：select组件所需的属性
        this.itemsPerPage = this.getFormatedSize();
        this.verifyCurrentPage(this.itemsPerPage.selected, true);
        this.oldCurrentPage = this.currentPage;
        this.oldTotalNumber = this.totalNumber;
        this.oldSize = this.itemsPerPage.selected;
        this.oldSelected = this.itemsPerPage.selected;
    }
    // 是否有表格当前页记忆
    isStorageCurrentPage() {
        // tslint:disable-next-line: max-line-length
        return this.table.storageId && (this.table.storageConfig.pagination === true || (this.table.storageConfig.pagination && this.table.storageConfig.pagination['currentPage'] === true));
    }
    // 是否有表格每页个数记忆
    isStorageItemsPerPage() {
        // tslint:disable-next-line: max-line-length
        return this.table.storageId && (this.table.storageConfig.pagination === true || (this.table.storageConfig.pagination && this.table.storageConfig.pagination['itemsPerPage'] === true));
    }
    // 处理表格记忆
    updateFromStorage() {
        // 修复SSR报错：ERROR TypeError: Cannot read property 'currentPage' of undefined
        // 当前页
        if (this.isStorageCurrentPage()) {
            const currentPage = this.table.sessionStorageState && this.table.sessionStorageState.currentPage;
            if (currentPage > 0) {
                Promise.resolve(currentPage).then(() => {
                    this.currentPage = currentPage;
                    this.currentPageChange.emit(currentPage);
                });
            }
        }
        // 修复SSR错误：ERROR TypeError: Cannot read property 'itemsPerPage' of undefined
        // 每页条数
        if (this.isStorageItemsPerPage()) {
            const itemsPerPage = this.table.localStorageState && this.table.localStorageState.itemsPerPage;
            if (itemsPerPage > 0) {
                Promise.resolve(itemsPerPage).then(() => {
                    this.pageSize.size = itemsPerPage;
                    const pagination = {
                        currentPage: this.currentPage,
                        size: itemsPerPage,
                        totalNumber: this.totalNumber
                    };
                    this.pageNumChange.emit(pagination);
                });
            }
        }
    }
    /**
     * 根据接口配置来渲染页面
     */
    renderPage() {
        if (this.type === 'mini') {
            this.configMini();
        }
        else {
            // goto页码与当前页码保持一致
            this.gotoPage = {
                page: this.currentPage
            };
            // 页码列表
            this.pages = this.getPages();
        }
        // 用户开启自动隐藏功能且无需显示分页(若总条数 =< 最小的每页显示条数)时，则不需要显示分页时，隐藏分页
        // 初始化totalNumber undefined
        this.hidePage = this.autoHide &&
            (Util.isUndefined(this.totalNumber) || Math.min.apply(null, this.itemsPerPage.options) >= this.totalNumber);
        if (this.hidePage) {
            this.render.setStyle(this.hostRef.nativeElement, 'display', 'none');
        }
        else {
            this.render.setStyle(this.hostRef.nativeElement, 'display', 'inline-block');
        }
        // 若分页与ti-table配合使用，则当分页更新时，触发ti-table的updatePagination
        // 进而触发表格更新回调
        if (this.isWithTable()) {
            this.table.updatePagination(this.currentPage, this.itemsPerPage.selected);
        }
    }
    /**
     * @ignore
     * 对总页数进行数字国际化处理
     */
    setTotalNumberFormat() {
        // 数字国际化格式：整数位保留最小位数.小数位保留最小位数-小数位最大保留位置
        const formatTotalNumber = Util.isUndefined(this.totalNumber) ? '0' : TiLocaleFormat.formatNumber(this.totalNumber, '1.0-0');
        return TiLocale.getLocaleWords().tiPagination['totalLabel'] + formatTotalNumber;
    }
    getFormatedSize() {
        const sizeConfig = Object.assign(Object.assign({}, this.defaultPageSize), this.pageSize);
        let index = sizeConfig.options.indexOf(sizeConfig.size);
        if (index === -1) {
            index = 0;
            this.syncPageSize(sizeConfig.options[index]);
        }
        return {
            selected: sizeConfig.options[index],
            options: sizeConfig.options,
            hide: sizeConfig.hide,
            style: {
                width: sizeConfig.width // 宽度设置
            }
        };
    }
    /**
     * 同步变化开发者的pageSize数据
     */
    syncPageSize(pageSizeNum) {
        if (this.pageSize instanceof Object) {
            this.pageSize.size = pageSizeNum;
        }
    }
    /**
     * @ignore
     * 当前页码是否为最小页码
     */
    noPrevious() {
        return this.currentPage === TiPaginationComponent.MIN_PAGE;
    }
    /**
     * @ignore
     * 当前页码是否为最大页码
     */
    noNext() {
        return this.currentPage === this.totalPageNum;
    }
    /**
     * @ignore
     * 当前页码是否为'···' 不可被聚焦
     */
    isEllipse(page) {
        return page.key === '···';
    }
    /**
     * @ignore
     * 操作pageSize选择框,每页显示条数改变时触发
     */
    onSizeChange(pageSizeNum) {
        this.syncPageSize(pageSizeNum);
        this.verifyCurrentPage(pageSizeNum, false);
        const pagination = {
            currentPage: this.currentPage,
            size: pageSizeNum,
            totalNumber: this.totalNumber
        };
        this.pageNumChange.emit(pagination);
        this.pageUpdate.emit(pagination);
    }
    // 通过操作选择某一页
    selectPage(page, evt) {
        // tiny2中此处阻止a标签浏览器默认事件
        if (!Util.isUndefined(evt)) {
            evt.preventDefault();
        }
        if (this.disabled) {
            return;
        }
        const selectedPage = parseInt(page, 10);
        if (selectedPage !== this.currentPage) {
            this.currentPage = selectedPage;
            this.currentPageChange.emit(this.currentPage);
            this.pageUpdate.emit({
                currentPage: this.currentPage,
                size: this.itemsPerPage.selected,
                totalNumber: this.totalNumber
            });
        }
    }
    /**
     * 计算总页数
     */
    calculateTotalPages(pageSizeNum) {
        const totalPages = Math.ceil(this.totalNumber / pageSizeNum);
        return Math.max(totalPages || 0, 1); // 当totalPages是undefined时，取0
    }
    /**
     * 组装mini类型时模板需要的数据
     */
    configMini() {
        const options = [];
        for (let i = 1; i <= this.totalPageNum; i++) {
            options.push({
                id: `${i}`,
                label: `${i}/${this.totalPageNum}`
            });
        }
        this.itemsMini = {
            options,
            selected: options[this.currentPage - 1],
            change: (option) => {
                this.selectPage(option.id);
            }
        };
    }
    /**
     * @description 根据每页显示条数及数据总条数获取页码列表
     */
    getPages() {
        const startAndEndPage = this.getInterval(); // 获得显示的起始和结束页
        const startPagNum = startAndEndPage[0];
        const endPagNum = startAndEndPage[1];
        const pages = [];
        const restPageNum = this.totalPageNum - this.currentPage; // 剩余页数
        // 只有1页的情况
        if (this.totalPageNum === 1) {
            pages.push(TiPaginationComponent.makePage(1, true));
            return pages;
        }
        // 当总页数大于8，并且中间连续按钮的起始位置没有和第1页相连时，创建第1页按钮和省略号
        if (startPagNum > 2 && this.totalPageNum > 8) {
            pages.push(TiPaginationComponent.makePage(1, this.currentPage === 1));
            pages.push(TiPaginationComponent.makePage('···', false));
        }
        // 创建中间页按钮
        for (let i = startPagNum; i <= endPagNum; i++) {
            pages.push(TiPaginationComponent.makePage(i, this.currentPage === i));
        }
        // 当总页数大于8，并且中间连续按钮的结束位置没有和最后一页按钮相连时，创建省略号
        // 当总页数大于8，并且最后一页隐藏且不是后四页时，创建省略号
        if (this.totalPageNum > 8 && (this.showLastPage && endPagNum < (this.totalPageNum - 1) ||
            !this.showLastPage && restPageNum > 3)) {
            pages.push(TiPaginationComponent.makePage('···', false));
        }
        // 最后一页隐藏：如果中间连续按钮的结束位置为总页数的前一页且当前页是后四页时，创建最后一页
        // 最后一页不隐藏：如果中间连续按钮的结束位置，不是最后一页按钮时，创建最后一页按钮
        if ((!this.showLastPage && endPagNum === this.totalPageNum - 1 && restPageNum <= 3) ||
            (this.showLastPage && endPagNum < this.totalPageNum)) {
            pages.push(TiPaginationComponent.makePage(this.totalPageNum, this.currentPage === this.totalPageNum));
        }
        return pages;
    }
    /**
     * @description 根据当前页码和要显示的数目，计算分页链接的起始页
     */
    getInterval() {
        let start;
        let end;
        // 1.根据Agile规范：页数小于等于8，起、始按钮值分别设为1和最大页数：1 - 总数
        if (this.totalPageNum <= 8) {
            start = 1;
            end = this.totalPageNum;
            return [start, end];
        }
        // 2.总页数大于8，当前项为前三页：1 - 5 / 1 - 6
        if (this.currentPage <= 3) {
            start = 1;
            end = 5;
            if (!this.showLastPage) {
                end = 6;
            }
            return [start, end];
        }
        // 3.总页数大于8，当前项是后4页：总数-4 - 总数
        if (this.currentPage > this.totalPageNum - 3) {
            start = this.totalPageNum - 4;
            end = this.totalPageNum;
            return [start, end];
        }
        // 4.总页数大于8，当前项不在前三页，后三页：显示中间4个按钮
        start = this.currentPage - 1;
        end = this.currentPage + 2;
        // 一般数据量比较大的场景隐藏最后一页时，显示中间5个按钮
        if (!this.showLastPage) {
            end = this.currentPage + 3;
        }
        return [start, end];
    }
    /**
     * 设置合法的新页码
     */
    verifyCurrentPage(pageSizeNum, triggerPageUpdate) {
        this.totalPageNum = this.calculateTotalPages(pageSizeNum);
        if (this.currentPage > this.totalPageNum || this.currentPage < TiPaginationComponent.MIN_PAGE) {
            this.currentPage = this.currentPage > this.totalPageNum ? this.totalPageNum : TiPaginationComponent.MIN_PAGE;
            if (triggerPageUpdate) {
                this.pageUpdate.emit({
                    currentPage: this.currentPage,
                    size: pageSizeNum,
                    totalNumber: this.totalNumber
                });
                Promise.resolve(undefined)
                    .then(() => {
                    this.currentPageChange.emit(this.currentPage);
                });
            }
            else {
                this.currentPageChange.emit(this.currentPage);
            }
        }
    }
    /**
     * @ignore
     * 输入跳转页码enter快捷键处理
     */
    enterPageHandler(e) {
        this.gotoPageHandler(e);
    }
    /**
     * @ignore
     * 根据要跳转的页码，同步修改当前页码
     */
    gotoPageHandler(event) {
        // 当输入值为不合法时处理
        if (this.gotoPage.page < TiPaginationComponent.MIN_PAGE || this.gotoPage.page > this.totalPageNum) {
            this.gotoPage.page = this.currentPage;
            return;
        }
        if (this.gotoPage.page !== this.currentPage) {
            this.currentPage = this.gotoPage.page;
            this.currentPageChange.emit(this.currentPage);
            this.pageUpdate.emit({
                currentPage: this.currentPage,
                size: this.itemsPerPage.selected,
                totalNumber: this.totalNumber
            });
        }
    }
    /**
     * @ignore
     * 上一页按钮的点击事件
     */
    onPrevClick(event) {
        return !this.noPrevious() && this.selectPage(this.currentPage - 1, event);
    }
    /**
     * @ignore
     * 页码点击事件
     */
    onPageClick(page, event) {
        return !this.isEllipse(page) && this.selectPage(page.key, event);
    }
    /**
     * @ignore
     * 下一页按钮的点击事件
     */
    onNextClick(event) {
        return !this.noNext() && this.selectPage(this.currentPage + 1, event);
    }
    /**
     * @ignore
     * input输入框modelchange事件
     */
    ngModelChange(value) {
        if (value === '') {
            this.gotoPage.page = '';
            return;
        }
        const newPage = parseInt(value, 10);
        if (isNaN(newPage)) {
            this.render.setProperty(this.inputRef.nativeElement, 'value', this.gotoPage.page);
            return;
        }
        this.gotoPage.page = newPage;
        this.render.setProperty(this.inputRef.nativeElement, 'value', this.gotoPage.page);
    }
    /**
     * @ignore
     */
    trackByFn(index, item) {
        return item.key;
    }
    isWithTable() {
        return this.table && this.render.parentNode(this.nativeElement).tagName === 'TI-TABLE';
    }
}
TiPaginationComponent.MIN_PAGE = 1;
TiPaginationComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-pagination',
                template: "<ng-container *ngIf=\"type !== 'mini' else mini\">\r\n    <!-- 1.select\u4E0B\u62C9\u7EC4\u4EF6 -->\r\n    <ti-select\r\n            class=\"ti3-page-size-option\"\r\n            size=\"small\"\r\n            [id]=\"appendId('select')\"\r\n            [options]='itemsPerPage.options'\r\n            *ngIf=\"itemsPerPage.hide !== true\"\r\n            [disabled]='disabled'\r\n            [ngStyle]='itemsPerPage.style'\r\n            (ngModelChange)=\"onSizeChange($event)\"\r\n            [(ngModel)]=\"itemsPerPage.selected\">\r\n            <ng-template let-item>{{item}}</ng-template>\r\n    </ti-select>\r\n    <!-- 2.\u603B\u6761\u6570 -->\r\n    <label class=\"ti3-pag-total-items\"\r\n            [ngClass]=\"{'ti3-pag-total-disable': disabled}\"\r\n            *ngIf=\"showTotalNumber\"\r\n            [id]=\"appendId('total_items')\">{{setTotalNumberFormat()}}</label>\r\n    <!-- 3.\u9875\u7801-->\r\n    <div class=\"ti3-pag-pages\" [ngClass]=\"{'ti3-pag-pages-disable': disabled}\">\r\n        <a href=\"javascript:void(0);\"\r\n            oncontextmenu=\"return false\"\r\n            class=\"ti3-pag-prev\"\r\n            [id]=\"appendId('prev_page')\"\r\n            [ngClass]=\"{disabled: noPrevious() || disabled}\"\r\n            [attr.title]=\"noPrevious() ? '' : 'tiPagination.prevTitle' | tiTranslate\"\r\n            [attr.tabindex]=\"(noPrevious() || disabled) ? -1 : ''\"\r\n            (click)=\"onPrevClick($event)\"> <i class=\"ti3-icon ti3-icon-angle-left\"></i> </a>\r\n        <ng-container *ngFor=\"let page of pages; trackBy: trackByFn\">\r\n            <a href=\"javascript:void(0);\"\r\n            oncontextmenu=\"return false\"\r\n            class=\"ti3-pag-page\"\r\n            *ngIf=\"page.active || type !== 'simple'\"\r\n            [id]=\"type !== 'simple' ? appendId('go_') + page.key : 0\"\r\n            [attr.tabindex]=\"(disabled || isEllipse(page) || type === 'simple') ? -1 : ''\"\r\n            tiOutline\r\n            [class.active]=\"page.active && type !== 'simple'\"\r\n            [class.simple]=\"type === 'simple'\"\r\n            [class.ellipse]=\"isEllipse(page)\"\r\n            [class.ti3-page-disable]=\"disabled\"\r\n            (click)=\"onPageClick(page, $event)\">{{page.key}}</a>\r\n        </ng-container>\r\n        <a href=\"javascript:void(0);\"\r\n            oncontextmenu=\"return false\"\r\n            class=\"ti3-pag-next\"\r\n            [id]=\"appendId('next_page')\"\r\n            [ngClass]=\"{disabled: noNext() || disabled}\"\r\n            [attr.title]=\"noNext() ? '' : 'tiPagination.nextTitle' | tiTranslate\"\r\n            [attr.tabindex]=\"(noNext() || disabled) ? -1 : ''\"\r\n            (click)=\"onNextClick($event)\"> <i class=\"ti3-icon ti3-icon-angle-right\"></i> </a>\r\n    </div>\r\n    <!-- 4.\u8DF3\u8F6C\u6309\u94AE -->\r\n    <div class=\"ti3-pag-goto\" *ngIf=\"type === 'default' && showGotoLink === true\"\r\n        [ngClass]=\"{'ti3-pag-goto-disable': disabled}\">\r\n        <span class=\"ti3-pag-goto-text\">{{ 'tiPagination.gotoLabel' | tiTranslate }}</span>\r\n        <input #input\r\n                class=\"ti3-pag-goto-input\"\r\n                tiText\r\n                type=\"text\"\r\n                [ngModel]=\"gotoPage.page\"\r\n                [disabled]=\"disabled\"\r\n                (ngModelChange)=\"ngModelChange($event)\"\r\n                (keydown.enter)=\"enterPageHandler($event)\"\r\n                [id]=\"appendId('input')\">\r\n        <a class=\"ti3-pag-goto-icon-container\"\r\n            (click)=\"gotoPageHandler($event)\"\r\n            [id]=\"appendId('go')\"> <div class=\"ti3-pag-goto-icon\"></div> </a>\r\n    </div>\r\n</ng-container>\r\n<ng-template #mini>\r\n    <div class=\"ti3-pag-mini-pages\">\r\n            <!-- 1.\u603B\u6761\u6570 -->\r\n        <label class=\"ti3-pag-total-items\"\r\n                [ngClass]=\"{'ti3-pag-total-disable': disabled}\"\r\n                *ngIf=\"showTotalNumber\"\r\n                [id]=\"appendId('total_items')\">{{setTotalNumberFormat()}}</label>\r\n        <!-- 2.\u4E0A\u4E00\u9875 -->\r\n        <a href=\"javascript:void(0);\"\r\n            oncontextmenu=\"return false\"\r\n            class=\"ti3-pag-prev\"\r\n            [id]=\"appendId('mini_prev_page')\"\r\n            [ngClass]=\"{disabled: noPrevious() || disabled}\"\r\n            [attr.title]=\"noPrevious() ? '' : 'tiPagination.prevTitle' | tiTranslate\"\r\n            [attr.tabindex]=\"(noPrevious() || disabled) ? -1 : ''\"\r\n            (click)=\"onPrevClick($event)\"> <i class=\"ti3-icon ti3-icon-angle-left\"></i> </a>\r\n        <!-- 3.\u4E0B\u62C9\u6846 -->\r\n        <ti-select\r\n                size=\"small\"\r\n                class=\"ti3-page-mini-option\"\r\n                [options]=\"itemsMini.options\"\r\n                [id]=\"appendId('mini_select')\"\r\n                [(ngModel)]=\"itemsMini.selected\"\r\n                [disabled]='disabled'\r\n                (ngModelChange)=\"itemsMini.change($event)\">\r\n        </ti-select>\r\n        <!-- 4.\u4E0B\u4E00\u9875 -->\r\n        <a href=\"javascript:void(0);\"\r\n            oncontextmenu=\"return false\"\r\n            class=\"ti3-pag-next\"\r\n            [id]=\"appendId('mini_next_page')\"\r\n            [ngClass]=\"{disabled: noNext() || disabled}\"\r\n            [attr.title]=\"noNext() ? '' : 'tiPagination.nextTitle' | tiTranslate\"\r\n            [attr.tabindex]=\"(noNext() || disabled) ? -1 : ''\"\r\n            (click)=\"onNextClick($event)\"> <i class=\"ti3-icon ti3-icon-angle-right\"></i> </a>\r\n    </div>\r\n</ng-template>\r\n",
                host: {
                    '[class.ti3-page-container]': 'true',
                    '[class.ti3-pag-container-disable]': 'this.disabled',
                    '[attr.unselectable]': '"on"'
                },
                styles: ["::ng-deep :root{--ti-pagination-box-model:border-box;--ti-pagination-mini-select-margin:0 var(--ti-common-space-2x);--ti-pagination-page-padding-horizon:6px;--ti-pagination-page-space:var(--ti-common-space-base);--ti-pagination-total-item-pages-space:var(--ti-common-space-4x)}::ng-deep .ti3-table>.ti3-page-container{margin-top:var(--ti-pagination-margin-top-table)}:host.ti3-page-container{display:inline-block}:host.ti3-page-container .ti3-pag-container{-ms-user-select:none;-webkit-user-select:none;display:inline-block;float:left;line-height:normal;user-select:none}:host.ti3-page-container .ti3-pag-total-items{color:var(--ti-pagination-text-color);float:left;font-weight:400;height:var(--ti-pagination-height);line-height:var(--ti-pagination-height);margin:0 var(--ti-pagination-total-item-pages-space);vertical-align:middle}:host.ti3-page-container .ti3-pag-pages{float:left;height:var(--ti-pagination-height)}:host.ti3-page-container .ti3-pag-pages>.ti3-pag-prev{margin-right:var(--ti-pagination-page-space)}:host.ti3-page-container .ti3-pag-next,:host.ti3-page-container .ti3-pag-page,:host.ti3-page-container .ti3-pag-prev{-ms-box-sizing:var(--ti-pagination-box-model);border-radius:var(--ti-pagination-border-radius);box-sizing:var(--ti-pagination-box-model);cursor:pointer;display:inline-block;height:var(--ti-pagination-height);line-height:var(--ti-pagination-height);text-align:center;text-decoration:none;vertical-align:middle}:host.ti3-page-container .ti3-pag-next:hover,:host.ti3-page-container .ti3-pag-page:hover,:host.ti3-page-container .ti3-pag-prev:hover{text-decoration:none}:host.ti3-page-container .ti3-pag-page{color:var(--ti-pagination-text-color);margin-right:var(--ti-pagination-page-space);padding:0 var(--ti-pagination-page-padding-horizon)}:host.ti3-page-container .ti3-pag-page.active{background-color:var(--ti-pagination-bg-color-active);color:var(--ti-pagination-text-color-active)}:host.ti3-page-container .ti3-pag-page.ellipse{background:transparent;border:none;cursor:default;outline:none}:host.ti3-page-container .ti3-pag-page:not(.ellipse):hover{background-color:var(--ti-pagination-bg-color-hover);color:var(--ti-pagination-text-color-hover)}:host.ti3-page-container .ti3-pag-page.active:hover,:host.ti3-page-container .ti3-pag-page.simple:hover{background-color:var(--ti-pagination-bg-color-active-hover);color:var(--ti-pagination-text-color-active-hover)}:host.ti3-page-container .ti3-pag-page:not(.ellipse):active{background-color:var(--ti-pagination-bg-color-press);color:var(--ti-pagination-text-color-active-active)}:host.ti3-page-container .ti3-pag-pages-disable .ti3-page-disable{background-color:var(--ti-pagination-bg-color);color:var(--ti-pagination-text-color-disabled);cursor:not-allowed!important;outline:none}:host.ti3-page-container .ti3-pag-pages-disable .ti3-page-disable:hover{background-color:var(--ti-pagination-bg-color);color:var(--ti-pagination-text-color-disabled)!important}:host.ti3-page-container .ti3-pag-next,:host.ti3-page-container .ti3-pag-prev{color:var(--ti-pagination-prev-next-color);height:var(--ti-pagination-prev-next-width);line-height:var(--ti-pagination-prev-next-width);width:var(--ti-pagination-prev-next-width)}:host.ti3-page-container .ti3-pag-next.disabled,:host.ti3-page-container .ti3-pag-prev.disabled{background-color:var(--ti-pagination-bg-color-disabled);color:var(--ti-pagination-text-color-disabled);cursor:not-allowed;outline:none}:host.ti3-page-container .ti3-pag-next.disabled>.ti3-icon-angle-left,:host.ti3-page-container .ti3-pag-next.disabled>.ti3-icon-angle-right,:host.ti3-page-container .ti3-pag-prev.disabled>.ti3-icon-angle-left,:host.ti3-page-container .ti3-pag-prev.disabled>.ti3-icon-angle-right{color:var(--ti-pagination-icon-color-disabled)}:host.ti3-page-container .ti3-pag-next:not(.disabled):hover,:host.ti3-page-container .ti3-pag-prev:not(.disabled):hover{background-color:var(--ti-pagination-bg-color-hover);color:var(--ti-pagination-text-color-hover)}:host.ti3-page-container .ti3-pag-next:not(.disabled):active,:host.ti3-page-container .ti3-pag-prev:not(.disabled):active{background-color:var(--ti-pagination-button-bg-active);color:var(--ti-pagination-prev-next-color-active)}:host.ti3-page-container .ti3-pag-mini-pages .ti3-pag-next,:host.ti3-page-container .ti3-pag-mini-pages .ti3-pag-prev,:host.ti3-page-container .ti3-page-mini-option,:host.ti3-page-container .ti3-page-size-option{float:left}:host.ti3-page-container .ti3-pag-mini-pages .ti3-page-mini-option{float:left;margin:var(--ti-pagination-mini-select-margin);width:60px}:host.ti3-page-container .ti3-pag-mini-pages .ti3-pag-total-items{margin-left:0}:host.ti3-page-container .ti3-pag-container-disable{cursor:not-allowed}:host.ti3-page-container .ti3-pag-total-disable{color:var(--ti-pagination-total-disable);cursor:not-allowed}:host.ti3-page-container .ti3-pag-goto,:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-text{display:inline-block;float:left;height:var(--ti-pagination-height)}:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-text{color:var(--ti-pagination-text-color);line-height:var(--ti-pagination-height);margin:0 15px;vertical-align:middle}:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-input{-ms-box-sizing:border-box;background-color:var(--ti-input-bg-color);border:1px solid var(--ti-pagination-border-color);border-radius:2px!important;box-sizing:border-box;color:var(--ti-pagination-goto-input-color);float:left;height:var(--ti-pagination-height);outline:none;padding:0 5px;vertical-align:middle;width:var(--ti-pagination-goto-input-width)}:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-input:focus{border:1px solid var(--ti-input-border-color-focus)}:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-icon-container{-ms-box-sizing:var(--ti-pagination-box-model);background-color:var(--ti-pagination-bg-color);border:1px solid var(--ti-pagination-border-color);border-radius:3px;box-sizing:var(--ti-pagination-box-model);cursor:pointer;display:inline-block;height:var(--ti-pagination-height);margin-left:15px;text-decoration:none;vertical-align:middle;width:var(--ti-pagination-height)}:host.ti3-page-container .ti3-pag-goto .ti3-pag-goto-icon-container .ti3-pag-goto-icon{border-bottom:6px solid transparent;border-left:6px solid var(--ti-pagination-goto-icon-color);border-top:7px solid transparent;height:0;margin:5px 38%;width:0}:host.ti3-page-container .ti3-pag-goto-disable{color:var(--ti-pagination-total-disable);cursor:not-allowed}:host.ti3-page-container .ti3-pag-goto-disable .ti3-pag-goto-icon{border-left:6px solid #999!important;cursor:not-allowed}", "::ng-deep :root{--ti-text-clear-width:26px;--ti-text-password-width:36px}:host[tiText]{-ms-box-sizing:border-box;background-color:var(--ti-input-bg-color);border:1px solid var(--ti-input-border-color);border-radius:var(--ti-input-border-radius);box-sizing:border-box;color:var(--ti-input-text-color);flood-color:var(--ti-input-clear-color);font-size:var(--ti-input-font-size);height:var(--ti-input-height);lighting-color:var(--ti-input-clear-color-hover);line-height:normal;outline:none;padding:0 var(--ti-input-padding-horizontal);stop-color:var(--ti-input-clear-color-disabled);vertical-align:middle}:host[tiText].ti3-text-input-show-icon{background-repeat:no-repeat;background-size:16px}:host[tiText].ti3-text-input-show-clear{background-position:center right 5px;padding-right:var(--ti-text-clear-width)!important}:host[tiText].ti3-text-input-show-password{background-position:center right 10px;padding-right:var(--ti-text-password-width)!important}:host[tiText].ti3-text-input-noborder{border:0;box-shadow:none;outline:0}:host[tiText]:not([disabled]):not([noborder]).ti3-text-input-show-icon.ti3-text-clear-active:hover{cursor:pointer}:host[tiText]:not([disabled]):not([noborder]):hover{border:1px solid var(--ti-input-border-color-hover)}:host[tiText]:not([disabled]):not([noborder]):focus{border:1px solid var(--ti-input-border-color-focus)}:host[tiText][disabled]:disabled{background-color:var(--ti-input-bg-color-disabled);border:1px solid var(--ti-input-border-color-disabled);color:var(--ti-input-text-color-disabled);cursor:not-allowed!important}:host[tiText][disabled]:disabled.ti3-text-input-noborder{border:0}.ti3-password-hack-input{height:0;left:-9999px;position:absolute;top:-9999px;width:0}:host[tiText]{transition:border-color .15s}:host[tiText].ti3-text-input-show-icon:focus{transition:transform .15s cubic-bezier(.4,0,.2,1),font-size .15s cubic-bezier(.4,0,.2,1)}:host[tiText]:hover:not(:focus){transition:border-color .2s}"]
            },] }
];
TiPaginationComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: TiTableComponent, decorators: [{ type: Optional }] },
    { type: TiSelectComponent, decorators: [{ type: Optional }] }
];
TiPaginationComponent.propDecorators = {
    type: [{ type: Input }],
    totalNumber: [{ type: Input }],
    totalNumberChange: [{ type: Output }],
    currentPage: [{ type: Input }],
    currentPageChange: [{ type: Output }],
    pageSize: [{ type: Input }],
    showTotalNumber: [{ type: Input }],
    showGotoLink: [{ type: Input }],
    autoHide: [{ type: Input }],
    disabled: [{ type: Input }],
    showLastPage: [{ type: Input }],
    pageNumChange: [{ type: Output }],
    pageUpdate: [{ type: Output }],
    mini: [{ type: ViewChild, args: ['mini', { static: true },] }],
    inputRef: [{ type: ViewChild, args: ['input',] }]
};
