// @cloud/tiny3/esm2015/components/alert/TiAlertComponent.js
import { Component, EventEmitter, Input, Output, ElementRef, Renderer2, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { TiBaseComponent } from '../base/TiBaseModule';
import { Util } from '../../utils/Util';
/**
 * Alert组件用于消息提示，提供了四种类型
 *
 * <example-url>../tiny3demo/#/alert/alert-all</example-url>
 */
export class TiAlertComponent extends TiBaseComponent {
    constructor(hostRef, renderer, document) {
        super(hostRef, renderer);
        this.hostRef = hostRef;
        this.renderer = renderer;
        this.document = document;
        /**
         * 是否显示左侧提示类型图标
         */
        this.typeIcon = true;
        /**
         * 是否显示关闭按钮
         */
        this.closeIcon = true;
        /**
         * 是否触发下拉类组件消失的事件，适用于alert的呈现状态影响页面下拉类组件位置的场景
         */
        this.triggerScroll = false;
        /**
         * 弹框类型
         */
        this.type = 'success';
        /**
         * 样式大小
         */
        this.size = 'normal';
        this.isOpen = true;
        /**
         * 是否为深色背景提示框
         */
        this.darkTheme = false;
        /**
         * @ignore
         * 是否设置阴影，组件规范定义无阴影（页面级及卡片内）
         */
        this.boxShadow = false;
        /**
         * open状态变化事件，点击关闭按钮或延时消失时都会触发
         */
        this.openChange = new EventEmitter();
    }
    /* tslint:disable:no-redundant-jsdoc */
    /**
     * 用于设置弹框是否弹出，可动态修改，默认值为true
     *
     * **请注意：动态修改情况下，需要设置为双向绑定，以确保组件延迟消失时open状态的正确感知**
     *
     * @type：boolean
     *
     */
    get open() {
        return this.isOpen;
    }
    /**
     * @ignore
     */
    set open(value) {
        if (value === this.open) {
            return;
        }
        if (value) { // 值为true时，才会延迟关闭
            this.dismissOnTime();
        }
        else { // 值为false时，关闭时清除定时器
            this.clearTimeout();
        }
        // open动态修改均涉及tiScroll事件
        this.triggerScrollEvent();
        this.isOpen = value;
    }
    ngOnInit() {
        super.ngOnInit();
        // 初始open值为true时，设置延时关闭
        if (this.open) {
            this.dismissOnTime();
        }
    }
    ngOnDestroy() {
        if (this.timeoutId !== undefined) {
            clearTimeout(this.timeoutId);
            if (this.open) {
                this.close();
            }
        }
    }
    /**
     * @ignore
     * alert呈现后，处理多长时间后消失
     */
    dismissOnTime() {
        // 未配置消失时间或当前已处于隐藏状态，则不做处理
        if (isNaN(this.dismissOnTimeout)) {
            return;
        }
        this.timeoutId = setTimeout(() => {
            this.close();
        }, this.dismissOnTimeout);
    }
    /**
     * @ignore
     * 关闭alert方法，供内部关闭调用
     */
    close() {
        this.open = false;
        this.openChange.emit(this.open); // 及时通知open状态，确保open状态双向绑定功能正常
    }
    /**
     * @ignore
     * alert出现或者消失时，可能会触发页面滚动，此时需要触发tiScroll事件通知相关组件
     */
    triggerScrollEvent() {
        if (!this.triggerScroll) {
            return;
        }
        setTimeout(() => {
            Util.trigger(this.document, 'tiScroll');
        }, 0);
    }
    clearTimeout() {
        if (this.timeoutId !== undefined) {
            clearTimeout(this.timeoutId);
        }
    }
}
TiAlertComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-alert',
                template: "<div class=\"ti3-alert-container\"\r\n     *ngIf=\"open\"\r\n     [ngClass]=\"{'ti3-alert-prompt-container': type === 'prompt',\r\n             'ti3-alert-error-container': type === 'error',\r\n             'ti3-alert-warn-container': type === 'warn',\r\n             'ti3-alert-success-container': type === 'success',\r\n             'ti3-alert-simple-container': type === 'simple',\r\n             'ti3-alert-box-shadow': boxShadow,\r\n             'ti3-alert-dark-theme': darkTheme,\r\n             'ti3-alert-size-small': size === 'small'}\">\r\n    <span class=\"ti3-alert-circle\" *ngIf=\"(type === 'prompt' || type === 'error' || type === 'success') && typeIcon\"><!--\r\n --><span class=\"ti3-icon\"\r\n          [ngClass]=\"{'ti3-icon-alert-prompt': type === 'prompt',\r\n                     'ti3-icon-alert-warn': type === 'error',\r\n                     'ti3-icon-alert-success': type === 'success'}\"></span><!--\r\n --></span><!--\r\n --><span class=\"ti3-icon ti3-icon-alert-warn-bg ti3-alert-warn-bg\" *ngIf=\"type === 'warn' && typeIcon\"><!--\r\n        --><span class=\"ti3-icon ti3-icon-alert-warn ti3-alert-warn-icon\" ></span>\r\n     </span><!--\r\n --><div class=\"ti3-alert-label\"\r\n        [ngClass]=\"{'ti3-alert-label-with-closeIcon': closeIcon && (!typeIcon || type === 'simple'),\r\n                    'ti3-alert-label-with-typeIcon': !closeIcon && typeIcon && type !== 'simple',\r\n                    'ti3-alert-only-label': !closeIcon && (!typeIcon || type === 'simple')}\"\r\n                    [id]=\"appendId('label')\">\r\n        <ng-content></ng-content>\r\n    </div><!--\r\n --><span class=\"ti3-alert-close-icon ti3-icon ti3-icon-close\"\r\n         *ngIf=\"closeIcon\"\r\n         (click)=\"close()\"\r\n         [tabindex]=\"0\"\r\n         tiOutline\r\n         (keydown.enter)=\"close()\"\r\n          [id]=\"appendId('alert_close_icon')\"></span>\r\n</div>",
                styles: ["::ng-deep :root{--ti-alert-label-line-height-value:calc(var(--ti-common-line-height-number)*var(--ti-common-font-size-base));--ti-alert-padding-horizon:15px;--ti-alert-padding-vertical:11px;--ti-alert-small-padding-vertical:7px;--ti-alert-type-icon-vertical-align:top}:host{display:block;width:var(--ti-alert-default-width)}.ti3-alert-container{-ms-box-sizing:border-box;border:1px solid;border-radius:var(--ti-alert-box-radius);box-sizing:border-box;font-weight:400;padding:var(--ti-alert-padding-vertical) var(--ti-alert-padding-horizon);width:100%}.ti3-alert-container.ti3-alert-dark-theme{background:var(--ti-alert-dark-bg-color);border:none}.ti3-alert-container.ti3-alert-dark-theme .ti3-alert-label{color:var(--ti-common-color-text-gray)}.ti3-alert-container.ti3-alert-dark-theme .ti3-alert-close-icon{color:var(--ti-common-color-icon-darkbg-normal)}.ti3-alert-container.ti3-alert-size-small{padding:var(--ti-alert-small-padding-vertical) var(--ti-alert-padding-horizon)}.ti3-alert-error-container{background:var(--ti-alert-error-bg-color);border-color:var(--ti-alert-error-border-color)}.ti3-alert-error-container .ti3-alert-circle{background:var(--ti-alert-error-icon-color)}.ti3-alert-prompt-container{background:var(--ti-alert-prompt-bg-color);border-color:var(--ti-alert-prompt-border-color)}.ti3-alert-prompt-container .ti3-alert-circle{background:var(--ti-alert-prompt-icon-color)}.ti3-alert-warn-container{background:var(--ti-alert-warn-bg-color);border-color:var(--ti-alert-warn-border-color)}.ti3-alert-success-container{background:var(--ti-alert-success-bg-color);border-color:var(--ti-alert-success-border-color)}.ti3-alert-success-container .ti3-alert-circle{background:var(--ti-alert-success-close-color)}.ti3-alert-simple-container{border-color:var(--ti-alert-simple-border-color)}.ti3-alert-success-container.ti3-alert-box-shadow{box-shadow:var(--ti-alert-success-box-shadow)}.ti3-alert-error-container.ti3-alert-box-shadow{box-shadow:var(--ti-alert-error-box-shadow)}.ti3-alert-prompt-container.ti3-alert-box-shadow{box-shadow:var(--ti-alert-prompt-box-shadow)}.ti3-alert-warn-container.ti3-alert-box-shadow{box-shadow:var(--ti-alert-warn-box-shadow)}.ti3-alert-dark-theme.ti3-alert-box-shadow{box-shadow:var(--ti-alert-dark-box-shadow)}.ti3-alert-circle{border-radius:50%;color:var(--ti-common-color-icon-white);height:var(--ti-alert-type-icon-width);line-height:var(--ti-alert-type-icon-width);text-align:center;width:var(--ti-alert-type-icon-width)}.ti3-alert-circle,.ti3-alert-warn-bg{display:inline-block;margin:calc((var(--ti-alert-label-line-height-value) - var(--ti-alert-type-icon-width))/2) var(--ti-alert-type-icon-margin-right) 0 0;vertical-align:var(--ti-alert-type-icon-vertical-align)}.ti3-alert-warn-bg{color:var(--ti-alert-warn-icon-color);font-size:var(--ti-alert-type-icon-width);position:relative}.ti3-alert-warn-icon{bottom:1px;color:var(--ti-common-color-icon-white);font-size:var(--ti-common-font-size-base);left:calc((var(--ti-alert-type-icon-width) - var(--ti-common-font-size-base))/2);position:absolute}.ti3-alert-label{color:var(--ti-alert-label-color);display:inline-block;line-height:var(--ti-common-line-height-number);overflow-wrap:break-word;width:calc(100% - var(--ti-alert-type-icon-width) - var(--ti-alert-close-icon-width) - var(--ti-alert-type-icon-margin-right) - var(--ti-alert-close-icon-margin-left));word-wrap:break-word}.ti3-alert-label.ti3-alert-label-with-closeIcon{width:calc(100% - var(--ti-alert-close-icon-width) - var(--ti-alert-close-icon-margin-left))}.ti3-alert-label.ti3-alert-label-with-typeIcon{width:calc(100% - var(--ti-alert-type-icon-width) - var(--ti-alert-type-icon-margin-right))}.ti3-alert-label.ti3-alert-only-label{width:100%}.ti3-alert-close-icon{cursor:pointer;display:inline-block;line-height:var(--ti-common-line-height-number);margin-left:var(--ti-alert-close-icon-margin-left);vertical-align:top;width:var(--ti-alert-close-icon-width)}.ti3-alert-error-container .ti3-alert-close-icon{color:var(--ti-alert-error-close-color)}.ti3-alert-prompt-container .ti3-alert-close-icon{color:var(--ti-alert-prompt-close-color)}.ti3-alert-success-container .ti3-alert-close-icon{color:var(--ti-alert-success-close-color)}.ti3-alert-warn-container .ti3-alert-close-icon{color:var(--ti-alert-warn-close-color)}.ti3-alert-simple-container .ti3-alert-close-icon{color:var(--ti-alert-simple-close-color)}"]
            },] }
];
TiAlertComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
TiAlertComponent.propDecorators = {
    typeIcon: [{ type: Input }],
    closeIcon: [{ type: Input }],
    dismissOnTimeout: [{ type: Input }],
    triggerScroll: [{ type: Input }],
    type: [{ type: Input }],
    size: [{ type: Input }],
    darkTheme: [{ type: Input }],
    boxShadow: [{ type: Input }],
    open: [{ type: Input }],
    openChange: [{ type: Output }]
};
