// @cloud/tiny3/esm2015/components/imagepreview/TiImagepreviewComponent.js
import { Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * @ignore
 *
 * 10.0.2版本新增组件
 *
 * 该组件主要用于图片放大预览
 *
 * <example-url>../tiny3demo/#/imagepreview/imagepreview-all</example-url>
 */
export class TiImagepreviewComponent extends TiBaseComponent {
    constructor(sanitizer, renderer, elementRef) {
        super(elementRef, renderer);
        this.sanitizer = sanitizer;
        this.renderer = renderer;
        this.elementRef = elementRef;
        /**
         * 需要预览的文件列表
         */
        this.fileList = [];
        /**
         * 是否展示文件名称，默认展示
         */
        this.showFileName = true;
        /**
         * 默认展示的图片的索引
         */
        this.index = 0;
    }
    ngAfterViewInit() {
        this.addImgLoadListener();
    }
    /**
     * @ignore 翻到上一张图片
     */
    toNext() {
        if (this.index === (this.fileList.length - 1)) {
            return;
        }
        this.addImgLoadListener();
        this.index++;
    }
    /**
     * @ignore 翻到下一张图片
     */
    toPrevious() {
        if (this.index === 0) {
            return;
        }
        this.addImgLoadListener();
        this.index--;
    }
    // 由于图片高度不确定，需动态设置top值，所以增加load事件。
    addImgLoadListener() {
        this.imgTemplate.nativeElement.onload = () => {
            const imgHeight = this.imgTemplate.nativeElement.height;
            if (imgHeight !== 0) {
                const top = (imgHeight - 48) / 2;
                this.renderer.setStyle(this.switchBtnTemplate.nativeElement, 'top', top + 'px');
            }
            // 防止循环触发load事件
            // tslint:disable-next-line: no-null-keyword
            this.imgTemplate.nativeElement.onload = null;
        };
    }
}
TiImagepreviewComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-imagepreview',
                template: "<ti-modal-header class=\"ti3-image-preview-modal-header\">\r\n    <span *ngIf=\"showFileName\" [id]=\"appendId('name')\">{{fileList[index].name}}</span>\r\n</ti-modal-header>\r\n<ti-modal-body class='ti3-image-preview-modal-body'>\r\n    <div class='ti3-image-preview-container'>\r\n        <img #imgTemplate [src]='sanitizer.bypassSecurityTrustUrl(fileList[index].previewUrl)' />\r\n    </div>\r\n    <div #switchBtn class=\"ti3-image-preview-icon-container\">\r\n        <ti-icon tiOutline [tabindex]=\"index === 0 ? -1 : 0\" name='angle-left' class=\"ti3-image-preview-icon ti3-image-preview-icon-left\"\r\n            [ngClass]=\"{'ti3-image-preview-icon-disabled': index === 0 ? true: false }\" (click)=\"toPrevious()\" (keydown.enter)=\"toPrevious()\" [id]=\"appendId('left')\"></ti-icon>\r\n        <ti-icon tiOutline [tabindex]=\"index === fileList.length - 1 ? -1 : 0\" name=\"angle-right\" class=\"ti3-image-preview-icon ti3-image-preview-icon-right\"\r\n            [ngClass]=\"{'ti3-image-preview-icon-disabled': index === fileList.length - 1 ? true : false}\" (click)=\"toNext()\" (keydown.enter)=\"toNext()\" [id]=\"appendId('right')\"></ti-icon>\r\n    </div>\r\n</ti-modal-body>",
                styles: ["::ng-deep :root{--ti-image-preview-container-margin:var(--ti-common-space-5x);--ti-image-preview-container-width:600px;--ti-image-preview-icon-width:24px;--ti-image-preview-modal-header-font-size:14px;--ti-image-preview-modal-header-line-height:20px;--ti-image-preview-modal-header-padding-left:var(--ti-common-space-5x);--ti-image-preview-modal-header-padding-top:12px}.ti3-image-preview-modal-header{font-size:var(--ti-image-preview-modal-header-font-size)!important;line-height:var(--ti-image-preview-modal-header-line-height)!important;padding:var(--ti-image-preview-modal-header-padding-top) var(--ti-image-preview-modal-header-padding-left)!important}.ti3-image-preview-modal-body{margin:0 var(--ti-image-preview-container-margin)!important;max-height:none!important;padding:0}.ti3-image-preview-container{margin:0 auto var(--ti-image-preview-container-margin);width:var(--ti-image-preview-container-width)}.ti3-image-preview-container img{border-style:none;display:inline-block;width:100%}.ti3-image-preview-icon-container{position:absolute;width:100%}.ti3-image-preview-icon-container .ti3-image-preview-icon{background:rgba(0,0,0,.2);border-radius:2px;color:#fff;cursor:pointer;display:inline-block;font-size:calc(var(--ti-common-font-size-base)*2);height:calc(var(--ti-image-preview-icon-width)*2);line-height:calc(var(--ti-image-preview-icon-width)*2);width:var(--ti-image-preview-icon-width)}.ti3-image-preview-icon-container .ti3-image-preview-icon:focus,.ti3-image-preview-icon-container .ti3-image-preview-icon:hover{background:rgba(0,0,0,.3)}.ti3-image-preview-icon-container .ti3-image-preview-icon-disabled{background:rgba(0,0,0,.2)!important;color:hsla(0,0%,100%,.4);cursor:not-allowed;outline:none}.ti3-image-preview-icon-container .ti3-image-preview-icon-left{float:left}.ti3-image-preview-icon-container .ti3-image-preview-icon-right{float:right}"]
            },] }
];
TiImagepreviewComponent.ctorParameters = () => [
    { type: DomSanitizer },
    { type: Renderer2 },
    { type: ElementRef }
];
TiImagepreviewComponent.propDecorators = {
    fileList: [{ type: Input }],
    showFileName: [{ type: Input }],
    index: [{ type: Input }],
    imgTemplate: [{ type: ViewChild, args: ['imgTemplate', { static: true },] }],
    switchBtnTemplate: [{ type: ViewChild, args: ['switchBtn', { static: true },] }]
};
