// @cloud/tinyplus3/esm2015/components/countdownalert/TpCountdownalertComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent, TiLocale } from '@cloud/tiny3';
/**
 * 倒计时提示组件
 *
 * <example-url>../tinyplus3demo/#/countdownalert/countdownalert-all</example-url>
 */
export class TpCountdownalertComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 告警类型， 默认为error
         */
        this.type = 'error';
        /**
         * 是否显示左侧提示类型图标
         */
        this.typeIcon = true;
        /**
         * 是否显示关闭按钮
         */
        this.closeIcon = true;
        /**
         * 是否显示countdownalert
         */
        this.open = false;
        /**
         * 是否触发下拉类组件消失的事件，适用于countdownalert的呈现状态影响页面下拉类组件位置的场景
         */
        this.triggerScroll = false;
        /**
         * @ignore
         */
        this.innerLabel = '';
    }
    ngOnInit() {
        super.ngOnInit();
        if (this.label) {
            return;
        }
        // 处理label
        const localeWords = TiLocale.getLocaleWords().tpCountdownalert;
        let timespan = '';
        if (this.infos) {
            timespan = this.getTimespan(this.infos, localeWords);
        }
        if (this.status) {
            const statusKey = this.status ? this.status.slice(14) : '';
            if (Object.keys(localeWords).indexOf(statusKey) >= 0) {
                this.innerLabel = localeWords[statusKey].replace(/{TS}/, String(timespan))
                    .replace(/{RN}/, this.resource || '');
            }
            else {
                // 状态正常，不需要显示alert
                this.open = false;
            }
        }
        // 处理充值续费链接：包周期为续费，服务需要传自定义链接；按需为充值，充值为固定链接
        if (this.status === 'hws_countdown_period_gracing' || this.status === 'hws_countdown_period_frozen') {
            this.showBuyLink = true;
            this.linkText = localeWords.renew;
        }
        else if (this.status === 'hws_countdown_ondemand_gracing' || this.status === 'hws_countdown_ondemand_frozen') {
            this.showBuyLink = true;
            this.linkText = localeWords.recharge;
            this.linkHref = 'https://account.huaweicloud.com/usercenter/#/userindex/balanceRecharge';
        }
        // 国际站 联系方式处理
        const intlContactPath = '/ticket/?#/ticketindex/createfeedback'; // 国际站联系客服路径
        const url = `${window.location.protocol}//${window.location.host}${intlContactPath}`;
        this.innerLabel = this.innerLabel.replace(/{contactLink}/, `<a class="tp-countdownalert-link" target="_blank" rel="noopener noreferrer"  href="${url}">${localeWords.contact_customer_service}</a>`);
        // 自定义客服联系电话号码处理
        this.innerLabel = this.innerLabel.replace(/{phoneNumber}/, this.phoneNumber || localeWords.phoneNumber);
    }
    getTimespan(infos, localeWords) {
        // tslint:disable-next-line: typedef
        const { day, hour, minute, second } = localeWords;
        if (infos.nextOperationRemainingDay) {
            return day.replace(/{d}/, infos.nextOperationRemainingDay);
        }
        else if (infos.nextOperationRemainingHour) {
            return hour.replace(/{h}/, infos.nextOperationRemainingHour)
                .replace(/{m}/, infos.nextOperationRemainingMinute);
        }
        else if (infos.nextOperationRemainingMinute) {
            return minute.replace(/{m}/, infos.nextOperationRemainingMinute);
        }
        else if (infos.nextOperationRemainingSecond) {
            return second;
        }
        return '';
    }
}
TpCountdownalertComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-countdownalert',
                template: "<ti-alert style=\"width: 100%\"\r\n    [type]=\"type\"\r\n    [closeIcon]=\"closeIcon\"\r\n    [typeIcon]=\"typeIcon\"\r\n    [open]=\"open\"\r\n    [dismissOnTimeout]=\"dismissOnTimeout\"\r\n    [triggerScroll]=\"triggerScroll\"\r\n    [id]=\"appendId('alert')\">\r\n        <span [innerHTML]=\"label || innerLabel\" [id]=\"appendId('label')\"></span>\r\n        <a class=\"tp-countdownalert-link\"\r\n            rel=\"noopener noreferrer\"\r\n            *ngIf=\"showBuyLink\"\r\n            target=\"_blank\"\r\n            [href]=\"linkHref\"\r\n            [id]=\"appendId('link_text')\">{{linkText}}</a>\r\n</ti-alert>",
                styles: ["a.tp-countdownalert-link{color:var(--ti-common-color-text-link)}a.tp-countdownalert-link:hover{cursor:pointer}a.tp-countdownalert-link:link,a.tp-countdownalert-link:visited{color:var(--ti-common-color-text-link);text-decoration:none}"]
            },] }
];
TpCountdownalertComponent.propDecorators = {
    type: [{ type: Input }],
    typeIcon: [{ type: Input }],
    closeIcon: [{ type: Input }],
    open: [{ type: Input }],
    dismissOnTimeout: [{ type: Input }],
    triggerScroll: [{ type: Input }],
    label: [{ type: Input }],
    infos: [{ type: Input }],
    status: [{ type: Input }],
    resource: [{ type: Input }],
    linkHref: [{ type: Input }],
    phoneNumber: [{ type: Input }]
};
