// @cloud/tiny3/esm2015/components/tree/TiTreeModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiTreeComponent } from './TiTreeComponent';
import { TiCheckboxModule } from '../checkbox/TiCheckboxModule';
import { FormsModule } from '@angular/forms';
import { TiIconModule } from '../icon/TiIconModule';
import { TiLoadingModule } from '../loading/TiLoadingModule';
import { TiHighlightPipe } from './TiHighlightPipe';
import { TiRendererModule } from '../../services/renderer/TiRendererModule';
import { TiTextModule } from '../text/TiTextModule';
import { TiValidationModule } from '../../directives/validation/TiValidationModule';
import { TiAutoSelectDirective } from './TiAutoSelectDirective';
export class TiTreeModule {
}
TiTreeModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiCheckboxModule,
                    TiIconModule,
                    FormsModule,
                    TiLoadingModule,
                    TiRendererModule,
                    TiTextModule,
                    TiValidationModule
                ],
                exports: [TiTreeComponent],
                declarations: [TiTreeComponent, TiHighlightPipe, TiAutoSelectDirective]
            },] }
];
export * from './TiTreeComponent';
export { TiTreeComponent } from './TiTreeComponent';
export { TiTreeUtil } from './TiTreeUtil';
