// @cloud/tinyplus3/esm2015/components/descandbuy/TpDescandbuyComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
export class TpDescandbuyComponent extends TiBaseComponent {
    /**
     * @ignore
     */
    onButtonClick(button) {
        if (typeof button.click === 'function') {
            button.click();
        }
    }
    /**
     * @ignore
     */
    onPricecalcClick() {
        if (typeof this.config.pricecalc.beforeClick === 'function') {
            this.config.pricecalc.beforeClick();
        }
    }
}
TpDescandbuyComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-descandbuy',
                template: "<tp-helptip class=\"tp-descandbuy-title\"\r\n    [label]=\"config.helptip.label\"\r\n    [iconTip]=\"config.helptip.iconTip\"\r\n    [iconTipPosition]=\"config.helptip.iconTipPosition\"\r\n    [iconTipMaxWidth]=\"config.helptip.iconTipMaxWidth\"\r\n    [id]=\"appendId('helptip')\">\r\n</tp-helptip>\r\n<div style=\"flex: 1;\">\r\n    <ng-content></ng-content>\r\n</div>\r\n<div>\r\n    <ng-container *ngFor=\"let button of config.buttons;let i = index\">\r\n        <span [tiTip]=\"button.tip\" [tiTipPosition]=\"button.tipPosition\" style=\"display: inline-block;\"\r\n            *ngIf=\"button.show !== false\">\r\n            <button tiButton [id]=\"button.id\" [disabled]=\"button.disabled\" size=\"large\" [color]=\"button.color\"\r\n                [ngClass]=\"{'tp-descandbuy-button-left-space':i > 0}\" [icon]=\"button.iconName\"\r\n                (click)=\"onButtonClick(button)\">\r\n                <tp-icon *ngIf=\"button.iconName\" [name]=\"button.iconName\" style=\"vertical-align: top;\"></tp-icon>\r\n                {{button.label}}\r\n            </button>\r\n        </span>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"config.pricecalc\">\r\n        <tp-pricecalc [ngClass]=\"{'tp-descandbuy-pricecalc-top-space': config.buttons.length > 0}\" style=\"display: block;\"\r\n            [values]=\"config.pricecalc.values\"\r\n            [url]=\"config.pricecalc.url\"\r\n            [detailHref]=\"config.pricecalc.detailHref\"\r\n            (beforeClick)=\"onPricecalcClick()\"\r\n            [id]=\"appendId('pricecalc')\">\r\n        </tp-pricecalc>\r\n    </ng-container>\r\n</div>\r\n",
                styles: [":host{align-items:center;display:flex;justify-content:space-between;min-height:32px}:host .tp-descandbuy-title{font-size:var(--ti-common-font-size-2);font-weight:700}:host .tp-descandbuy-button-left-space{margin-left:var(--ti-common-space-2x)}:host .tp-descandbuy-pricecalc-top-space{margin-top:var(--ti-common-space-4x)}"]
            },] }
];
TpDescandbuyComponent.propDecorators = {
    config: [{ type: Input }]
};
