// @cloud/tiny3/esm2015/components/formfield/TiButtonItemComponent.js
import { Component, ElementRef, Input } from '@angular/core';
import { TiFormfieldComponent } from './TiFormfieldComponent';
/**
 * 本组件嵌在 tiFormfield 中使用，包裹着表单按钮
 */
export class TiButtonItemComponent {
    constructor(formfield, elementRef) {
        this.formfield = formfield;
        this.elementRef = elementRef;
        /**
         * 设置表单按钮是否显示
         *
         * tiButtonItem 不支持 ngIf 和 hidden，由该接口来控制表单按钮的显示与否
         */
        this.show = true;
        /**
         * @ignore
         */
        this.verticalAlignObj = {};
    }
    ngOnInit() {
        this.btnContent = this.elementRef.nativeElement;
        this.btnColspan = (this.formfield.colsNum - 1) * TiButtonItemComponent.TD_NUM_PER_ITEM + 2;
        if (this.verticalAlign) {
            this.verticalAlignObj['vertical-align'] = this.verticalAlign;
        }
        if (this.show) {
            this.formfield.addBtnItem(this);
        }
    }
    ngOnChanges(changes) {
        const showObj = changes['show'];
        if (showObj && !showObj.firstChange) {
            if (showObj.currentValue) {
                this.formfield.addBtnItem(this);
            }
            else {
                this.formfield.removeBtnItem(this);
            }
        }
    }
    ngOnDestroy() {
        this.formfield.removeBtnItem(this);
    }
}
/**
 * 每一个ti-item占4个td。第一个td用来显示“*”，第二个td用来显示label，
 * 第三个td用来显示content, 第四个td是item间距。
 */
TiButtonItemComponent.TD_NUM_PER_ITEM = 4;
TiButtonItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-button-item',
                template: "<ng-content></ng-content>"
            },] }
];
TiButtonItemComponent.ctorParameters = () => [
    { type: TiFormfieldComponent },
    { type: ElementRef }
];
TiButtonItemComponent.propDecorators = {
    show: [{ type: Input }],
    verticalAlign: [{ type: Input }]
};
