// @cloud/tiny3/esm2015/components/formfield/TiFormfieldModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiIncludeModule } from '../../directives/include/TiIncludeModule';
import { TiIconModule } from '../icon/TiIconModule';
import { TiFormfieldComponent } from './TiFormfieldComponent';
import { TiItemComponent } from './TiItemComponent';
import { TiItemLabelComponent } from './TiItemLabelComponent';
import { TiButtonItemComponent } from './TiButtonItemComponent';
export class TiFormfieldModule {
}
TiFormfieldModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiIconModule,
                    TiIncludeModule
                ],
                exports: [TiFormfieldComponent,
                    TiItemComponent,
                    TiItemLabelComponent,
                    TiButtonItemComponent],
                declarations: [TiFormfieldComponent,
                    TiItemComponent,
                    TiItemLabelComponent,
                    TiButtonItemComponent]
            },] }
];
export { TiFormfieldComponent } from './TiFormfieldComponent';
export { TiItemComponent } from './TiItemComponent';
export { TiItemLabelComponent } from './TiItemLabelComponent';
export { TiButtonItemComponent } from './TiButtonItemComponent';
