// @cloud/tiny3/esm2015/components/dropsearch/TiDropsearchComponent.js
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { TiFormComponent } from '../base/TiBaseModule';
import { TiSearchboxNotsearchComponent } from '../searchbox/TiSearchboxModule';
import { Util } from '../../utils/Util';
import { TiDropComponent } from '../drop/TiDropModule';
import { TiDroplistComponent } from '../droplist/TiDroplistModule';
import { TiLocale } from '../../locale/TiLocaleModule';
/**
 * @ignore
 * 下拉面板带有可搜索的数组组件，主要用于select，表格列设置。
 */
export class TiDropsearchComponent extends TiDroplistComponent {
    constructor() {
        super(...arguments);
        this.searchable = false;
        this.selectAll = false; // 多选模式下是否有全选功能
        /**
         * 面板对齐方式, drop中的panelAlign接口透传出去
         */
        this.panelAlign = 'left';
        /**
         * @ignore
         * dominator 到drop的距离
         */
        this.dominatorSpace = TiDropComponent.DOMINATOR_SPACE + 'px';
        /**
         * @ignore
         * 10.0.2新增
         *
         * 后台搜索，传出搜索框的值
         */
        this.beforeSearch = new EventEmitter();
        this.placeholder = TiLocale.getLocaleWords().tiSelect.search;
        /**
         * 搜索结果数组
         */
        this.searchResult = new Array();
        this.oldSearchable = null;
        this.focusableElemsInSearch = [];
        this.focusableElemsInPagination = [];
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        // 外部修改show属性的处理，初始定义在ngOnInit中处理
        if (this.searchable) {
            if (changes['options'] && this.options && !changes['options'].firstChange) {
                this.searchWordChange(this.searchboxCom.model);
            }
        }
        if (changes['searchable']) {
            this.heightExcludeList = changes['searchable'].currentValue ? TiDroplistComponent.SEARCHBOX_EXCLUDE_HEIGHT : 2;
        }
    }
    ngOnInit() {
        super.ngOnInit();
        this.searchResult = this.options;
    }
    ngAfterViewChecked() {
        super.ngAfterViewChecked();
        let focusableElemsChange;
        if (this.searchable !== this.oldSearchable) {
            super.unlistenKendown();
            if (this.searchable) {
                this.focusableElemsInSearch = this.searchboxCom.getFocusableElems();
                focusableElemsChange = true;
                super.listenKeydown(this.searchboxCom.getFocusableElems()[0]);
            }
            else {
                this.focusableElemsInSearch = [];
                super.listenKeydown(this.dominatorElem);
            }
            this.oldSearchable = this.searchable;
        }
        if (this.pagination && this.pagination.updateFocusableElems) {
            this.focusableElemsInPagination = this.pagination.getFocusableElems();
            focusableElemsChange = true;
        }
        if (focusableElemsChange) {
            this.setFocusableElems(this.focusableElemsInSearch.concat(this.focusableElemsInPagination));
        }
    }
    /**
     * 覆盖父类show()
     */
    show() {
        if (this.isShow) {
            return;
        }
        // 清空搜索框并处理下拉列表内容
        if (this.searchable) {
            this.searchResult = this.options;
            setTimeout(() => {
                super.show();
                // 焦点转移至搜索框
                this.searchboxCom.focus();
            }, 0);
        }
        else {
            super.show();
        }
    }
    /**
     * 覆盖父类hide(), 有搜索框时焦点转移
     */
    hide() {
        if (this.isShow) {
            super.hide();
            if (this.searchable) {
                this.dominatorElem.focus();
                this.searchWord = '';
            }
        }
    }
    /**
     * 仅hide，不带焦点转移。
     */
    hideWithoutFocus() {
        super.hide();
        this.searchWord = '';
    }
    searchWordChange(searchWord) {
        this.beforeSearch.emit();
        if (this.hasBeforeSearchObservers) {
            return;
        }
        if (this.options && this.options.length >= 0) {
            // 搜索结果临时值。结果默认值，是原数据
            let searchResult = this.options;
            // 如果搜索词存在
            if (!Util.isEmptyString(searchWord)) {
                // 如果是分组，则找出第二级item集合
                if (this.options[0] && this.listCom.isGroup(this.options[0])) {
                    searchResult = [];
                    this.options.forEach((group) => {
                        searchResult = searchResult.concat(group.children);
                    });
                }
                // 在集合中搜索
                searchResult = searchResult.filter((option) => {
                    if (!this.searchKeys) {
                        // 没有定义searchKeys时，取labelKey
                        return Util.isString(option[this.labelKey]) && option[this.labelKey].toLowerCase()
                            .indexOf(searchWord.toLowerCase()) >= 0;
                    }
                    else {
                        //  已定义searchKeys，任一条目匹配即可
                        for (const searchKey of this.searchKeys) {
                            if (Util.isString(option[searchKey]) && option[searchKey].toLowerCase()
                                .indexOf(searchWord.toLowerCase()) >= 0) {
                                return true;
                            }
                        }
                        return false;
                    }
                });
            }
            this.setSearchResult(searchResult);
        }
    }
    /**
     * 搜索后，按需重新定位、设置hoverOption
     */
    setSearchResult(searchResult) {
        var _a;
        // list个数不同时，重定位下拉位置
        if (((_a = this.searchResult) === null || _a === void 0 ? void 0 : _a.length) !== searchResult.length) {
            setTimeout(() => {
                this.rePosition(true);
            }, 0);
        }
        this.searchResult = searchResult;
    }
}
TiDropsearchComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-dropsearch',
                template: "<ti-drop [dominatorElem]=\"dominatorElem\"\r\n    [panelWidth]='panelWidth' [class.ti3-select-small]=\"size === 'small'\"\r\n    [panelAlign]=\"panelAlign\"\r\n    [dominatorSpace]=\"dominatorSpace\">\r\n    <ng-content select=\"[custom]\"></ng-content>\r\n    <div *ngIf=\"searchable\">\r\n        <ti-searchbox-notsearch #searchboxCom noBorder\r\n                class =\"ti3-dropdown-searchbox\"\r\n                [placeholder]=\"placeholder\"\r\n                [(ngModel)]=\"searchWord\"\r\n                (ngModelChange)=\"searchWordChange($event)\"\r\n                [id]=\"appendId('searchbox')\">\r\n        </ti-searchbox-notsearch>\r\n    </div>\r\n    <ti-list [multiple]=\"multiple\"\r\n        [options]=\"((type !== 'suggestion' && searchable) || type === 'suggestion')?searchResult:options\"\r\n        [(ngModel)]=\"model\"\r\n        [selectAll]=\"selectAll\"\r\n        style=\"display:block;\"\r\n        (select)=\"onSelect($event)\"\r\n        [id]=\"appendId('list')\"\r\n        [tipPosition]=\"tipPosition\"\r\n        [noDataText]=\"noDataText\">\r\n        <ng-template let-item let-i=index let-last=last>\r\n                <ng-container *ngIf=\"itemTemplate else listLabelTemplate\">\r\n                        <ng-container *ngTemplateOutlet=\"itemTemplate; context: {$implicit: item,index: i,last: last}\">\r\n                        </ng-container>\r\n                </ng-container>\r\n                <ng-template #listLabelTemplate>{{item[labelKey]}}</ng-template>\r\n        </ng-template>\r\n    </ti-list>\r\n    <ng-container *ngIf=\"footerTemplateRef\">\r\n        <section #footer>\r\n            <ng-container *ngTemplateOutlet=\"footerTemplateRef\"></ng-container>\r\n        </section>\r\n    </ng-container>\r\n</ti-drop>",
                providers: [TiFormComponent.getValueAccessor(TiDropsearchComponent)],
                styles: [".ti3-dropdown-searchbox{width:100%!important}"]
            },] }
];
TiDropsearchComponent.propDecorators = {
    searchable: [{ type: Input }],
    searchKeys: [{ type: Input }],
    selectAll: [{ type: Input }],
    panelAlign: [{ type: Input }],
    dominatorSpace: [{ type: Input }],
    beforeSearch: [{ type: Output }],
    searchboxCom: [{ type: ViewChild, args: [TiSearchboxNotsearchComponent,] }],
    dropCom: [{ type: ViewChild, args: [TiDropComponent,] }],
    footerTemplateRef: [{ type: ContentChild, args: ['footer', { static: true },] }]
};
