// @cloud/tinyplus3/esm2015/components/subtitle/TpSubtitleComponent.js
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { TiBaseComponent, TiSelectComponent, Util } from '@cloud/tiny3';
/**
 * 返回标题组件
 *
 * <example-url>../tinyplus3demo/#/subtitle/subtitle-all</example-url>
 */
export class TpSubtitleComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 后退按钮链接
         */
        this.href = '';
        /**
         * 数据选项，如果选项大于1个，标题是下拉的形式
         */
        this.items = [];
        /**
         * items 大于1项时，展开面板中是否有搜索
         */
        this.searchable = false;
        /**
         * 指定在何处显示链接的资源, 即a标签的target属性
         */
        this.target = '_blank';
        /**
         * 设置最大宽度
         *  10.1.2 新增
         */
        this.maxWidth = '250px';
        /**
         * 设置标题溢出提示的方向
         *  10.1.3 新增
         */
        this.tipPosition = 'top';
        /**
         * 选中项，items 的条数大于1时才生效，双向数据绑定
         */
        this.selectedChange = new EventEmitter();
        /**
         * 后退按钮点击回调
         */
        this.back = new EventEmitter();
    }
    ngOnInit() {
        super.ngOnInit();
        // 如果没有传入选中项，且items 条目数大于1 默认选中第一项，
        if (Util.isUndefined(this.selectedItem) && this.items.length > 1) {
            this.selectedItem = this.items[0];
        }
    }
    ngAfterViewInit() {
        super.ngAfterViewInit();
        // 选项数大于等于2 ，才出现select
        if (this.items.length < 2) {
            return;
        }
        // 暗色背景，给ti-select 添加dark属性
        const attrs = this.hostRef.nativeElement.attributes;
        if (!Util.isUndefined(attrs['dark'])) {
            this.renderer.setAttribute(this.selectComRef.nativeElement, 'dark', '');
        }
    }
    /**
     * @ignore
     * 触发back 事件
     */
    onClick($event) {
        this.back.emit($event);
    }
    /**
     * @ignore
     * select选中项改变，触发subtitle selectedChange事件
     */
    onSelectedOptionChange(option) {
        this.selectedChange.emit(option);
    }
}
TpSubtitleComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-subtitle',
                template: "<!-- \u540E\u9000\u6309\u94AE -->\r\n<a class=\"tp-subtitle-icon-back\"\r\n    rel=\"noopener noreferrer\"\r\n    *ngIf=\"!backRouterLink\"\r\n    [href]=\"href\"\r\n    [target]=\"target\"\r\n    (click)=\"onClick($event)\"\r\n    [id]=\"appendId('prev')\">\r\n  <tp-icon local name=\"cloud-action-prev\"></tp-icon>\r\n  <!-- \u652F\u6301\u8DEF\u7531 -->\r\n</a><a class=\"tp-subtitle-icon-back\"\r\n    *ngIf=\"backRouterLink\"\r\n    [routerLink]=\"backRouterLink\"\r\n    [queryParams]=\"queryParams\"\r\n    (click)=\"onClick($event)\"\r\n    [id]=\"appendId('prev')\">\r\n  <tp-icon local name=\"cloud-action-prev\"></tp-icon>\r\n</a><section #textRef  class=\"tp-subtitle-text\"\r\n  *ngIf=\"items?.length === 1\" tiOverflow\r\n  [tiTipHostEleX]=\"textRef.nativeElement\"\r\n  [tiTipPosition]=\"tipPosition\"\r\n  [id]=\"appendId('label')\">{{items[0]?.label}}</section><!--\r\n--><ti-select\r\n      *ngIf=\"items?.length > 1\"\r\n      dominatorSpace=\"5px\"\r\n      noborder\r\n      [searchable]=\"searchable\"\r\n      [options]=\"items\"\r\n      [(ngModel)]=\"selectedItem\"\r\n      [panelWidth]=\"panelWidth\"\r\n      (ngModelChange)=\"onSelectedOptionChange($event)\"\r\n      [id]=\"appendId('select')\">\r\n      <ng-template #selected let-item>\r\n        <span class=\"tp-subtitle-selected-text\">\r\n          {{item?.label}}\r\n        </span>\r\n      </ng-template>\r\n  </ti-select>\r\n",
                host: {
                    '[style.max-width]': 'maxWidth'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-subtitle-font-weight:bold;--tp-subtitle-height:24px;--tp-subtitle-icon-width:30px;--tp-subtitle-line-dividing-height:16px;--tp-subtitle-text-maxwidth:210px}:host{display:inline-block;height:var(--tp-subtitle-height);line-height:var(--tp-subtitle-height);width:100%}.tp-subtitle-icon-back{-ms-box-sizing:border-box;box-sizing:border-box;color:var(--ti-common-color-icon-normal);cursor:pointer;display:inline-block;padding-left:var(--ti-common-space-10);position:relative;text-decoration:none;vertical-align:top;width:var(--tp-subtitle-icon-width)}.tp-subtitle-icon-back,.tp-subtitle-icon-back tp-icon{font-weight:var(--tp-subtitle-font-weight)}.tp-subtitle-icon-back:hover{color:var(--ti-common-color-icon-hover)}.tp-subtitle-icon-back:after{border-right:1px solid var(--tp-subtitle-line-color);content:\"\";height:var(--tp-subtitle-line-dividing-height);position:absolute;right:0;top:var(--ti-common-space-base)}.tp-subtitle-text{-ms-box-sizing:border-box;box-sizing:border-box;color:var(--ti-common-color-text-primary);cursor:default;display:inline-block;font-size:var(--ti-common-font-size-2);font-weight:var(--tp-subtitle-font-weight);max-width:calc(100% - var(--tp-subtitle-icon-width) - 20px);padding-left:var(--ti-common-space-10)}:host ::ng-deep ti-select{height:var(--tp-subtitle-height);line-height:var(--tp-subtitle-height);max-width:var(--tp-subtitle-text-maxwidth);width:auto}:host ::ng-deep ti-select .ti3-overflow-padding{padding-right:0}:host ::ng-deep ti-select ::ng-deep ti-dominator.ti3-select-dominator-container{height:var(--tp-subtitle-height);line-height:var(--tp-subtitle-height);vertical-align:top}:host .tp-subtitle-selected-text{font-size:var(--ti-common-font-size-2);font-weight:var(--tp-subtitle-font-weight)}:host[dark] .tp-subtitle-icon-back{color:var(--tp-subtitle-dark-icon-color)}:host[dark] .tp-subtitle-icon-back:after{opacity:.2}:host[dark] .tp-subtitle-icon-back:hover{color:var(--ti-common-color-icon-hover)}:host[dark] .tp-subtitle-text{color:var(--tp-subtitle-dark-text-color)}:host[dark] ::ng-deep ti-select{background-color:transparent;border-color:transparent}:host[dark] ::ng-deep ti-select:hover,:host[dark] ::ng-deep ti-select[tiFocused]{border-color:transparent}:host[dark] .tp-subtitle-selected-text{color:var(--tp-subtitle-dark-selected-text-color)}"]
            },] }
];
TpSubtitleComponent.propDecorators = {
    href: [{ type: Input }],
    items: [{ type: Input }],
    searchable: [{ type: Input }],
    target: [{ type: Input }],
    selectedItem: [{ type: Input, args: ['selected',] }],
    backRouterLink: [{ type: Input }],
    queryParams: [{ type: Input }],
    panelWidth: [{ type: Input }],
    maxWidth: [{ type: Input }],
    tipPosition: [{ type: Input }],
    selectedChange: [{ type: Output }],
    back: [{ type: Output }],
    selectComRef: [{ type: ViewChild, args: [TiSelectComponent,] }]
};
