// @cloud/tiny3/esm2015/directives/validation/validators/directives/BaseValidator.js
import { Directive, forwardRef } from '@angular/core';
/* tslint:disable:no-implicit-dependencies */
import { NG_VALIDATORS, Validators } from '@angular/forms';
import { TiValidators } from '../TiValidators';
import { Util } from '../../../../utils/Util';
// 下面注释，是为了避免lib编译错误。具体原理未知，副作用未知。
// @dynamic
/**
 * @ignore
 */
export class BaseValidator {
    constructor() {
        this.enabled = true; // 无需传参的校验规则是否开启校验
        this._validator = Validators.nullValidator;
    }
    /**
     * 获取各directive中的参数配置
     * @param validatorClass 校验实现类
     * @param name 校验指令名：由于该方法在类外部调用，无法使用BaseValidator中的参数，因此需要传入
     */
    static getDirectiveConfig(validatorClass, name) {
        return {
            selector: `[${name}][formControlName],[${name}][formControl],[${name}][ngModel]`,
            providers: [{
                    provide: NG_VALIDATORS,
                    /* tslint:disable:no-forward-ref */
                    useExisting: forwardRef(() => validatorClass),
                    multi: true
                }]
        };
    }
    static getValidatorStr(name) {
        return name.replace(new RegExp('ti'), '')
            .replace(new RegExp('^\\S'), (s) => {
            return s.toLowerCase();
        });
    }
    ngOnChanges(changes) {
        if (changes['baseValue'] || changes['enabled']) {
            this._createValidator();
            if (this._onChange) {
                this._onChange();
            }
        }
    }
    validate(c) {
        return (this.enabled === false || this.enabled === 'false') ? null : this._validator(c);
    }
    registerOnValidatorChange(fn) {
        this._onChange = fn;
    }
    _createValidator() {
        const validatorStr = this.validatorStr;
        const validatorParam = this.baseValue;
        /* tslint:disable:prefer-conditional-expression */
        if (Util.isArray(validatorParam)) {
            /* tslint:disable:no-null-keyword */
            this._validator = TiValidators[validatorStr].apply(null, validatorParam);
        }
        else if (!Util.isUndefined(validatorParam)) {
            this._validator = TiValidators[validatorStr](validatorParam);
        }
        else {
            this._validator = TiValidators[validatorStr];
        }
    }
}
BaseValidator.decorators = [
    { type: Directive, args: [{ selector: '[BaseValidator]' },] }
];
