// @cloud/tiny3/esm2015/directives/validation/validators/directives/ContainsValidatorDirective.js
// tslint:disable-next-line: ordered-imports
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(ContainsValidatorDirective, ContainsValidatorDirective.NAME))
export class ContainsValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(ContainsValidatorDirective.NAME);
    }
}
ContainsValidatorDirective.NAME = 'tiContains';
ContainsValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiContains][formControlName],[tiContains][formControl],[tiContains][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => ContainsValidatorDirective),
                        multi: true
                    }]
            },] }
];
ContainsValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [ContainsValidatorDirective.NAME,] }]
};
