// @cloud/tiny3/esm2015/directives/validation/validators/directives/DateValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(DateValidatorDirective, DateValidatorDirective.NAME))
export class DateValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(DateValidatorDirective.NAME);
    }
}
DateValidatorDirective.NAME = 'tiDate';
DateValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiDate][formControlName],[tiDate][formControl],[tiDate][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => DateValidatorDirective),
                        multi: true
                    }]
            },] }
];
DateValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [DateValidatorDirective.NAME,] }]
};
