// @cloud/tiny3/esm2015/directives/validation/validators/directives/DigitsValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(DigitsValidatorDirective, DigitsValidatorDirective.NAME))
export class DigitsValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(DigitsValidatorDirective.NAME);
    }
}
DigitsValidatorDirective.NAME = 'tiDigits';
DigitsValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiDigits][formControlName],[tiDigits][formControl],[tiDigits][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => DigitsValidatorDirective),
                        multi: true
                    }]
            },] }
];
DigitsValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [DigitsValidatorDirective.NAME,] }]
};
