// @cloud/tiny3/esm2015/directives/validation/validators/directives/EmailValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(EmailValidatorDirective, EmailValidatorDirective.NAME))
export class EmailValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(EmailValidatorDirective.NAME);
    }
}
EmailValidatorDirective.NAME = 'tiEmail';
EmailValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiEmail][formControlName],[tiEmail][formControl],[tiEmail][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => EmailValidatorDirective),
                        multi: true
                    }]
            },] }
];
EmailValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [EmailValidatorDirective.NAME,] }]
};
