// @cloud/tiny3/esm2015/directives/validation/validators/directives/EqualValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(EqualValidatorDirective, EqualValidatorDirective.NAME))
export class EqualValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(EqualValidatorDirective.NAME);
    }
}
EqualValidatorDirective.NAME = 'tiEqual';
EqualValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiEqual][formControlName],[tiEqual][formControl],[tiEqual][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => EqualValidatorDirective),
                        multi: true
                    }]
            },] }
];
EqualValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [EqualValidatorDirective.NAME,] }]
};
