// @cloud/tiny3/esm2015/directives/validation/validators/directives/IntegerValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(IntegerValidatorDirective, IntegerValidatorDirective.NAME))
export class IntegerValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(IntegerValidatorDirective.NAME);
    }
}
IntegerValidatorDirective.NAME = 'tiInteger';
IntegerValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiInteger][formControlName],[tiInteger][formControl],[tiInteger][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => IntegerValidatorDirective),
                        multi: true
                    }]
            },] }
];
IntegerValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [IntegerValidatorDirective.NAME,] }]
};
