// @cloud/tiny3/esm2015/directives/validation/validators/directives/Ipv4ValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(Ipv4ValidatorDirective, Ipv4ValidatorDirective.NAME))
export class Ipv4ValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(Ipv4ValidatorDirective.NAME);
    }
}
Ipv4ValidatorDirective.NAME = 'tiIpv4';
Ipv4ValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiIpv4][formControlName],[tiIpv4][formControl],[tiIpv4][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => Ipv4ValidatorDirective),
                        multi: true
                    }]
            },] }
];
Ipv4ValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [Ipv4ValidatorDirective.NAME,] }]
};
