// @cloud/tiny3/esm2015/directives/validation/validators/directives/Ipv6ValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(Ipv6ValidatorDirective, Ipv6ValidatorDirective.NAME))
export class Ipv6ValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(Ipv6ValidatorDirective.NAME);
    }
}
Ipv6ValidatorDirective.NAME = 'tiIpv6';
Ipv6ValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiIpv6][formControlName],[tiIpv6][formControl],[tiIpv6][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => Ipv6ValidatorDirective),
                        multi: true
                    }]
            },] }
];
Ipv6ValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [Ipv6ValidatorDirective.NAME,] }]
};
