// @cloud/tiny3/esm2015/directives/validation/validators/directives/MaxLengthValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(MaxLengthValidatorDirective, MaxLengthValidatorDirective.NAME))
export class MaxLengthValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(MaxLengthValidatorDirective.NAME);
    }
}
MaxLengthValidatorDirective.NAME = 'tiMaxLength';
MaxLengthValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiMaxLength][formControlName],[tiMaxLength][formControl],[tiMaxLength][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => MaxLengthValidatorDirective),
                        multi: true
                    }]
            },] }
];
MaxLengthValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [MaxLengthValidatorDirective.NAME,] }]
};
