// @cloud/tiny3/esm2015/directives/validation/validators/directives/MaxValueValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(MaxValueValidatorDirective, MaxValueValidatorDirective.NAME))
export class MaxValueValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(MaxValueValidatorDirective.NAME);
    }
}
MaxValueValidatorDirective.NAME = 'tiMaxValue';
MaxValueValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiMaxValue][formControlName],[tiMaxValue][formControl],[tiMaxValue][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => MaxValueValidatorDirective),
                        multi: true
                    }]
            },] }
];
MaxValueValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [MaxValueValidatorDirective.NAME,] }]
};
