// @cloud/tiny3/esm2015/directives/validation/validators/directives/MinValueValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(MinValueValidatorDirective, MinValueValidatorDirective.NAME))
export class MinValueValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(MinValueValidatorDirective.NAME);
    }
}
MinValueValidatorDirective.NAME = 'tiMinValue';
MinValueValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiMinValue][formControlName],[tiMinValue][formControl],[tiMinValue][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => MinValueValidatorDirective),
                        multi: true
                    }]
            },] }
];
MinValueValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [MinValueValidatorDirective.NAME,] }]
};
