// @cloud/tiny3/esm2015/directives/validation/validators/directives/NotEqualValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(NotEqualValidatorDirective, NotEqualValidatorDirective.NAME))
export class NotEqualValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(NotEqualValidatorDirective.NAME);
    }
}
NotEqualValidatorDirective.NAME = 'tiNotEqual';
NotEqualValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiNotEqual][formControlName],[tiNotEqual][formControl],[tiNotEqual][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => NotEqualValidatorDirective),
                        multi: true
                    }]
            },] }
];
NotEqualValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [NotEqualValidatorDirective.NAME,] }]
};
