// @cloud/tiny3/esm2015/directives/validation/validators/directives/NotScriptValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(NotScriptValidatorDirective, NotScriptValidatorDirective.NAME))
export class NotScriptValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(NotScriptValidatorDirective.NAME);
    }
}
NotScriptValidatorDirective.NAME = 'tiNotScript';
NotScriptValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiNotScript][formControlName],[tiNotScript][formControl],[tiNotScript][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => NotScriptValidatorDirective),
                        multi: true
                    }]
            },] }
];
NotScriptValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [NotScriptValidatorDirective.NAME,] }]
};
