// @cloud/tiny3/esm2015/directives/validation/validators/directives/NumberValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(NumberValidatorDirective, NumberValidatorDirective.NAME))
export class NumberValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(NumberValidatorDirective.NAME);
    }
}
NumberValidatorDirective.NAME = 'tiNumber';
NumberValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiNumber][formControlName],[tiNumber][formControl],[tiNumber][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => NumberValidatorDirective),
                        multi: true
                    }]
            },] }
];
NumberValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [NumberValidatorDirective.NAME,] }]
};
