// @cloud/tiny3/esm2015/directives/validation/validators/directives/PortValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(PortValidatorDirective, PortValidatorDirective.NAME))
export class PortValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(PortValidatorDirective.NAME);
    }
}
PortValidatorDirective.NAME = 'tiPort';
PortValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiPort][formControlName],[tiPort][formControl],[tiPort][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => PortValidatorDirective),
                        multi: true
                    }]
            },] }
];
PortValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [PortValidatorDirective.NAME,] }]
};
