// @cloud/tiny3/esm2015/directives/validation/validators/directives/RangeValueValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(RangeValueValidatorDirective, RangeValueValidatorDirective.NAME))
export class RangeValueValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(RangeValueValidatorDirective.NAME);
    }
}
RangeValueValidatorDirective.NAME = 'tiRangeValue';
RangeValueValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiRangeValue][formControlName],[tiRangeValue][formControl],[tiRangeValue][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => RangeValueValidatorDirective),
                        multi: true
                    }]
            },] }
];
RangeValueValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [RangeValueValidatorDirective.NAME,] }]
};
