// @cloud/tiny3/esm2015/directives/validation/validators/directives/RangeSizeValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(RangeSizeValidatorDirective, RangeSizeValidatorDirective.NAME))
export class RangeSizeValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(RangeSizeValidatorDirective.NAME);
    }
}
RangeSizeValidatorDirective.NAME = 'tiRangeSize';
RangeSizeValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiRangeSize][formControlName],[tiRangeSize][formControl],[tiRangeSize][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => RangeSizeValidatorDirective),
                        multi: true
                    }]
            },] }
];
RangeSizeValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [RangeSizeValidatorDirective.NAME,] }]
};
