// @cloud/tiny3/esm2015/directives/validation/validators/directives/RegExpValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(RegExpValidatorDirective, RegExpValidatorDirective.NAME))
export class RegExpValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(RegExpValidatorDirective.NAME);
    }
}
RegExpValidatorDirective.NAME = 'tiRegExp';
RegExpValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiRegExp][formControlName],[tiRegExp][formControl],[tiRegExp][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => RegExpValidatorDirective),
                        multi: true
                    }]
            },] }
];
RegExpValidatorDirective.propDecorators = {
    baseValue: [{ type: Input, args: [RegExpValidatorDirective.NAME,] }]
};
