// @cloud/tiny3/esm2015/directives/validation/validators/directives/RequiredValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(RequiredValidatorDirective, RequiredValidatorDirective.NAME))
export class RequiredValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(RequiredValidatorDirective.NAME);
    }
}
RequiredValidatorDirective.NAME = 'tiRequired';
RequiredValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiRequired][formControlName],[tiRequired][formControl],[tiRequired][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => RequiredValidatorDirective),
                        multi: true
                    }]
            },] }
];
RequiredValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [RequiredValidatorDirective.NAME,] }]
};
