// @cloud/tiny3/esm2015/directives/validation/validators/directives/UrlValidatorDirective.js
import { Directive, Input, forwardRef } from '@angular/core';
import { BaseValidator } from './BaseValidator';
import { NG_VALIDATORS } from '@angular/forms';
// @Directive(BaseValidator.getDirectiveConfig(UrlValidatorDirective, UrlValidatorDirective.NAME))
// @Directive(BaseValidator.getDirectiveConfig(RequiredValidatorDirective, RequiredValidatorDirective.NAME))
export class UrlValidatorDirective extends BaseValidator {
    constructor() {
        super(...arguments);
        this.validatorStr = BaseValidator.getValidatorStr(UrlValidatorDirective.NAME);
    }
}
UrlValidatorDirective.NAME = 'tiUrl';
UrlValidatorDirective.decorators = [
    { type: Directive, args: [{
                selector: `[tiUrl][formControlName],[tiUrl][formControl],[tiUrl][ngModel]`,
                providers: [{
                        provide: NG_VALIDATORS,
                        // tslint:disable-next-line: no-forward-ref
                        useExisting: forwardRef(() => UrlValidatorDirective),
                        multi: true
                    }]
            },] }
];
UrlValidatorDirective.propDecorators = {
    enabled: [{ type: Input, args: [UrlValidatorDirective.NAME,] }]
};
