// @cloud/tinyplus3/esm2015/components/foldtext/TpFoldtextComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent, TiLocale } from '@cloud/tiny3';
/**
 * 折叠文本
 *
 * 10.0.1 新增
 *
 * <example-url>../tinyplus3demo/#/foldtext/foldtext-all</example-url>
 */
export class TpFoldtextComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 文本内容
         */
        this.texts = [];
        /**
         * 折叠时显示几行
         */
        this.line = 2;
        /**
         * @ignore
         * “查看全部” 词条
         */
        this.showMore = TiLocale.getLocaleWords().tpFoldtext.showMore;
        /**
         * @ignore
         * “收起” 词条
         */
        this.showLess = TiLocale.getLocaleWords().tpFoldtext.showLess;
        /**
         * @ignore
         * 是否折叠
         */
        this.folded = true;
        // 表格中的默认font-size:12px, 默认line-height:1.5, 所以默认值取18px
        this.lineHeight = '18px';
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngAfterViewInit() {
        super.ngAfterViewInit();
        const CSSStyles = getComputedStyle(this.nativeElement);
        const lineHeight = CSSStyles.getPropertyValue('line-height');
        if (!Number.isNaN(parseFloat(lineHeight))) {
            this.lineHeight = lineHeight;
        }
        // 修改了页面的数据height，消除ExpressionChangedAfterItHasBeenCheckedError报错使用setTimeout
        setTimeout(() => {
            this.height = (this.line * parseFloat(this.lineHeight)) + 'px';
        }, 0);
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        if (changes['texts'] && !changes['texts'].firstChange || changes['line'] && !changes['line'].firstChange) {
            this.height = (this.line * parseFloat(this.lineHeight)) + 'px';
        }
    }
    /**
     * @ignore
     * 切换文本的点击事件
     */
    onClick(event) {
        this.folded = !this.folded;
    }
}
TpFoldtextComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-foldtext',
                template: "<section class=\"tp-foldtext-wraper\" [ngStyle]=\"{'height': folded? height:''}\">\r\n    <section tiOverflow *ngFor=\"let text of texts; index as i\">\r\n        <span [innerHTML]=\"text\" [id]=\"appendId('text_' + i)\"></span>\r\n    </section>\r\n</section>\r\n<section tiOutline class=\"tp-foldtext-toggle\" (click)=\"onClick($event)\" [id]=\"appendId('toggle')\"\r\n[tabindex]=\"0\" (keydown.enter)=\"onClick($event)\">\r\n    {{folded? showMore:showLess}} <span *ngIf=\"folded\">({{texts.length}})</span>\r\n</section>",
                styles: [".tp-link-no-decoration{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-link-no-decoration:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}.tp-foldtext-wraper{overflow-y:hidden}.tp-foldtext-wraper ::ng-deep a{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-foldtext-wraper ::ng-deep a:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}.tp-foldtext-toggle{color:var(--ti-common-color-text-highlight)}.tp-foldtext-toggle:hover{cursor:pointer}"]
            },] }
];
TpFoldtextComponent.propDecorators = {
    texts: [{ type: Input }],
    line: [{ type: Input }]
};
