// @cloud/tinyplus3/esm2015/components/collapsebox/TpCollapseboxComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 折叠框组件 collapsebox
 *
 * collapsebox组件是一个带有关闭操作的容器类组件，服务可自定义内容。
 *
 * <example-url>../tinyplus3demo/#/collapsebox/collapsebox-all</example-url>
 */
export class TpCollapseboxComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * @ignore
         * 设置折叠框默认展开
         */
        this.collapsed = false;
        /**
         * 是否可关闭
         * 10.0.3新增
         */
        this.closeable = true;
        /**
         * 关闭事件
         * 10.0.3新增
         */
        this.close = new EventEmitter();
    }
    /**
     * @ignore
     * 10.0.3新增
     */
    onClickClose(event) {
        if (this.close.observers.length > 0) {
            this.close.emit();
        }
        else {
            this.collapsed = true;
        }
    }
}
TpCollapseboxComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-collapsebox',
                template: "<div class=\"tp-collapsebox-container\" [tiCollapse]=\"collapsed\">\r\n    <tp-icon tiOutline local [hidden]=\"!closeable\" class=\"tp-icon-cloud-close tp-icon tp-collapsebox-close\" name=\"cloud-close\"\r\n    (click)=\"onClickClose($event)\" [id]=\"appendId('collapsed')\" [tabindex]=\"0\" (keydown.enter)=\"onClickClose($event)\"></tp-icon>\r\n    <ng-content></ng-content>\r\n</div>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-collapsebox-close-position:20px}.tp-collapsebox-container{background:var(--ti-common-color-bg-white-normal);box-shadow:var(--ti-common-shadow-1-down);display:block;padding:var(--ti-common-space-6x) var(--ti-common-space-8x);position:relative}.tp-collapsebox-close{cursor:pointer;position:absolute;right:var(--tp-collapsebox-close-position);top:var(--tp-collapsebox-close-position)}.tp-collapsebox-close:hover{color:var(--ti-common-color-icon-hover)}"]
            },] }
];
TpCollapseboxComponent.propDecorators = {
    closeable: [{ type: Input }],
    close: [{ type: Output }]
};
