// @cloud/tinyplus3/esm2015/components/leftmenu/TpLeftmenuComponent.js
import { Component, ContentChild, ElementRef, EventEmitter, Input, Output } from "@angular/core";
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * TpLeftmenu组件主要用于左侧导航，是基于Tiny3中的TiLeftmenu组件进行封装的，详细的数据接口说明可参照Tiny3中的TiLeftmenu。
 *
 * <example-url>../tinyplus3demo/#/leftmenu/leftmenu-all</example-url>
 */
export class TpLeftmenuComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 激活菜单项改变时向外通知
         */
        this.activeChange = new EventEmitter();
    }
    /**
     * @ignore
     * 模板中使用
     */
    getHeaderTextClass() {
        return this.config && TpLeftmenuComponent.CHINESE_REG.test(this.config.serviceName) && this.config.serviceName.length > 8 ?
            'tp-leftmenu-title-long' : 'tp-leftmenu-title';
    }
    /**
     * @ignore
     * 激活菜单项改变时向外通知
     */
    onActiveChange(item) {
        this.activeChange.emit(item);
    }
    /**
     * @ignore
     * 触发点击事件
     */
    triggerClick(item) {
        if (typeof (item.click) === 'function') {
            item.click(item);
        }
    }
}
TpLeftmenuComponent.CHINESE_REG = new RegExp('[\u4e00-\u9fa5]'); // 汉字
TpLeftmenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-leftmenu',
                template: "<ti-leftmenu\r\n    [(active)]='active'\r\n    (activeChange)='onActiveChange($event)'\r\n    [(reloadState)]='config.reloadState'\r\n    [toggleable]='config.toggleable'\r\n    [(collapsed)]='config.collapsed'\r\n    (collapsedChange)='config.collapsedChangeFn($event)'>\r\n    <ti-leftmenu-head *ngIf=\"config.serviceName\">\r\n        <span class=\"{{getHeaderTextClass()}}\" [id]=\"appendId('service_name')\">{{config.serviceName}}</span>\r\n    </ti-leftmenu-head>\r\n    <ng-container *ngFor=\"let group of items; index as index\">\r\n        <ti-leftmenu-group *ngIf=\"group.isGroup && !group.hide\"\r\n                            [expanded]=\"group.expanded\">\r\n            <ti-leftmenu-group-item (click)=\"triggerClick(group)\">\r\n                <div [id]=\"appendId('group_' + index)\">{{group.label}}</div>\r\n            </ti-leftmenu-group-item>\r\n            <ng-container *ngFor=\"let level1 of group.children; index as i\">\r\n                <ti-leftmenu-level1 [item]=\"level1\" *ngIf=\"!level1.hide\">\r\n                    <ti-leftmenu-item (click)=\"triggerClick(level1)\">\r\n                        <div [id]=\"appendId('level1_' + index + '_' + i)\">{{level1.label}}</div>\r\n                    </ti-leftmenu-item>\r\n                    <ng-container *ngFor=\"let level2 of level1.children; index as j\">\r\n                        <ti-leftmenu-level2 [item]=\"level2\" (click)=\"triggerClick(level2)\" *ngIf=\"!level2.hide\">\r\n                            <div [id]=\"appendId('level2_' + index + '_' + i + '_' + j)\">{{level2.label}}</div>\r\n                        </ti-leftmenu-level2>\r\n                    </ng-container>\r\n                </ti-leftmenu-level1>\r\n            </ng-container>\r\n        </ti-leftmenu-group>\r\n\r\n        <ti-leftmenu-level1 *ngIf=\"!group.hide && !group.isGroup\" [item]=\"group\">\r\n            <ti-leftmenu-item (click)=\"triggerClick(group)\">\r\n                <div [id]=\"appendId('group_' + index)\">{{group.label}}</div>\r\n            </ti-leftmenu-item>\r\n            <ng-container *ngFor=\"let level2 of group.children; index as i\">\r\n                <ti-leftmenu-level2 [item]=\"level2\" (click)=\"triggerClick(level2)\"\r\n                                    *ngIf=\"!level2.hide\">\r\n                    <div [id]=\"appendId('level2_' + index + '_' + i)\">{{level2.label}}</div>\r\n                </ti-leftmenu-level2>\r\n            </ng-container>\r\n        </ti-leftmenu-level1>\r\n    </ng-container>\r\n    <ti-leftmenu-foot *ngIf='leftmenuFoot'>\r\n        <ng-container *ngTemplateOutlet=\"leftmenuFoot\">  </ng-container>\r\n    </ti-leftmenu-foot>\r\n</ti-leftmenu>",
                styles: [".tp-leftmenu-title,.tp-leftmenu-title-long{color:var(--ti-leftmenu-head-text-color);font-size:var(--ti-leftmenu-head-font-size);font-weight:400;line-height:20px;margin:0;word-wrap:break-word}.tp-leftmenu-title-long{font-size:var(--ti-common-font-size-3)}"]
            },] }
];
TpLeftmenuComponent.propDecorators = {
    config: [{ type: Input }],
    items: [{ type: Input }],
    active: [{ type: Input }],
    activeChange: [{ type: Output }],
    leftmenuFoot: [{ type: ContentChild, args: ['foot',] }]
};
