// @cloud/tiny3/esm2015/components/base/TiBaseModule.js
import { NgModule } from '@angular/core';
import { TiBaseComponent } from './TiBaseComponent';
import { TiFormComponent } from './TiFormComponent';
import { TiRadioBaseComponent } from './TiRadioBaseComponent';
import { TiWholeComponent } from './TiWholeComponent';
import { TiAutofocusComponent } from './TiAutofocusComponent';
// 在使用tiny npm包生产环境编辑时，要求基类也注册。
/**
 * @ignore
 */
export class TiBaseModule {
}
TiBaseModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                exports: [],
                declarations: [
                    TiBaseComponent,
                    TiFormComponent,
                    TiRadioBaseComponent,
                    TiWholeComponent,
                    TiAutofocusComponent
                ]
            },] }
];
export { TiBaseComponent } from './TiBaseComponent';
export { TiFormComponent } from './TiFormComponent';
export { TiRadioBaseComponent } from './TiRadioBaseComponent';
export { TiWholeComponent } from './TiWholeComponent';
export { TiAutofocusComponent } from './TiAutofocusComponent';
