// @cloud/tiny3/esm2015/components/base/TiWholeComponent.js
import { Component, Input } from '@angular/core';
import { TiFormComponent } from './TiFormComponent';
/**
 * 复原完整model对象基类，基于valueKey时，select/buttongroup实现复原选中值到完整对象
 */
export class TiWholeComponent extends TiFormComponent {
    constructor() {
        super(...arguments);
        /**
         * @ignore
         * 找到value值的函数
         */
        this.valueFn = (item) => {
            return item[this.valueKey];
            // tslint:disable-next-line:semicolon
        }; // 这个分号是书写正确的
    }
    /**
     * @ignore
     */
    get modelWhole() {
        return this._modelWhole;
    }
    /**
     * @ignore
     * 用户点选时modelWhole改变，要同时更改model
     */
    set modelWhole(value) {
        this._modelWhole = value;
        if (this.valueKey && value !== undefined && value !== null) {
            // 存在valueKey，且value不为空
            this.model = this['multiple'] ?
                value.map((item) => this.valueFn(item))
                : this.valueFn(value);
        }
        else {
            this.model = value;
        }
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        // 监听 options/items 变化
        if ((changes['options'] || changes['items']) && this.valueKey && this.model) {
            // 存在 valueKey 时，modelWhole 需通过匹配 valueKey 来重新赋值（解决动态改变options/items时选中项不高亮问题 issues#1332）
            this.writeValue(this.model);
        }
    }
    /**
     * 每次用户主动设置ngModel值时，需要顺便更改modelWhole
     * @param model 设置的ngModel值
     */
    writeValue(model) {
        super.writeValue(model);
        if (this.valueKey && model !== undefined && model !== null) { // 存在valueKey，且value不为空
            // 历史设计缺陷：select总数据集是options，buttongroup总数据集时items
            const options = this['options'] ? this['options'] : this['items'];
            // 直接去写内部值，不触发set函数
            this._modelWhole = this['multiple'] ?
                options.filter((item) => model.includes(this.valueFn(item))) :
                options.find((item) => this.valueFn(item) === model);
        }
        else {
            this._modelWhole = model;
        }
    }
}
TiWholeComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-whole',
                template: ''
            },] }
];
TiWholeComponent.propDecorators = {
    valueKey: [{ type: Input }],
    valueFn: [{ type: Input }]
};
