// @cloud/tinyplus3/esm2015/components/rename/TpRenameModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { TpRenameComponent } from './TpRenameComponent';
import { TpIconModule } from '../icon/TpIconModule';
import { TpRenameModalContainerComponent } from './TpRenameModalContainerComponent';
import { TpRenameModalComponent } from './TpRenameModalComponent';
import { TiButtonModule, TiLocaleModule, TiModalModule, TiPopupModule, TiTableModule, TiTextModule, TiTipModule, TiValidationModule } from '@cloud/tiny3';
export class TpRenameModule {
}
TpRenameModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    TiButtonModule,
                    TiLocaleModule,
                    TiModalModule,
                    TiPopupModule,
                    TiTipModule,
                    TiValidationModule,
                    TiTableModule,
                    TiTextModule,
                    TpIconModule
                ],
                exports: [TpRenameComponent],
                declarations: [TpRenameComponent, TpRenameModalComponent, TpRenameModalContainerComponent],
                entryComponents: [TpRenameModalComponent, TpRenameModalContainerComponent]
            },] }
];
export { TpRenameComponent } from './TpRenameComponent';
export { TpRenameModalComponent } from './TpRenameModalComponent';
export { TpRenameModalContainerComponent } from './TpRenameModalContainerComponent';
