// @cloud/tinyplus3/esm2015/components/rename/TpRenameModalComponent.js
import { Component, Input, ViewChild, ElementRef } from '@angular/core';
import { TiLocale, TiBaseComponent } from '@cloud/tiny3';
/**
 * @ignore
 */
export class TpRenameModalComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        this.okLabel = TiLocale.getLocaleWords().tpRename.okButton;
        this.cancelLabel = TiLocale.getLocaleWords().tiCommon.cancelBtn;
        this.validation = {
            errorMessage: {
                regExp: TiLocale.getLocaleWords().tpRename.inValidMessage
            }
        };
    }
    // 此处的close并不是按钮真正的close()，参数form的类型本应是FormGroup，所以没有引入FormGroup，指定为any
    close(form) {
    }
    dismiss() {
    }
    ngOnInit() {
        super.ngOnInit();
        this.model = this.value;
    }
    ngAfterViewInit() {
        this.input.nativeElement.focus();
        this.input.nativeElement.setSelectionRange(0, 0, 'backward');
    }
}
TpRenameModalComponent.decorators = [
    { type: Component, args: [{
                template: "<div class=\"tp-rename-modal-header\" [id]=\"appendId('header')\">{{modalHeader}}</div>\r\n<form #form=\"ngForm\" >\r\n    <input\r\n        #input type=\"text\"\r\n        class=\"tp-rename-modal-input\"\r\n        [(ngModel)]=\"model\"\r\n        [tiValidation]=\"validation\"\r\n        tiRequired\r\n        tiRegExp=\"^[\\u4e00-\\u9fa5a-zA-Z0-9-_]*$\"\r\n        name=\"input\"\r\n        [tiRangeSize]=\"[1, 64]\"\r\n        [id]=\"appendId('input')\"\r\n        tiText>\r\n</form>\r\n<div>\r\n    <button tiButton size=\"small\" color=\"danger\" (click)=\"close(form.form)\" [id]=\"appendId('ok')\">{{okLabel}}</button>\r\n    <button tiButton size=\"small\" style=\"margin-left: 8px;\" (click)=\"dismiss()\" [id]=\"appendId('dismiss')\">{{cancelLabel}}</button>\r\n</div>\r\n",
                styles: [".tp-rename-modal-header{line-height:var(--ti-common-line-height-number)}.tp-rename-modal-input{margin:var(--ti-common-space-2x) 0;width:200px}"]
            },] }
];
TpRenameModalComponent.propDecorators = {
    value: [{ type: Input }],
    modalHeader: [{ type: Input }],
    input: [{ type: ViewChild, args: ['input', { static: true },] }]
};
