// @cloud/tinyplus3/esm2015/components/helptip/TpHelptipComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 帮助提示组件
 *
 * 由被提示的文本和一个问号图标构成，鼠标悬浮在图标上可以看到对文本内容的解释说明
 *
 * <example-url>../tinyplus3demo/#/helptip/helptip-all</example-url>
 */
export class TpHelptipComponent extends TiBaseComponent {
}
TpHelptipComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-helptip',
                template: "<span class=\"tp-helptip-label\" [id]=\"appendId('label')\">{{label}}</span>\r\n<ng-content *ngIf=\"!label\"></ng-content>\r\n<tp-icon local name=\"cloud-action-tip\" *ngIf=\"iconTip\" [id]=\"appendId('tipicon')\"\r\n    [tiTip]=\"iconTip\"\r\n    [tiTipPosition]=\"iconTipPosition\"\r\n    [tiTipMaxWidth]=\"iconTipMaxWidth\"></tp-icon>\r\n",
                styles: [":host{align-items:center;display:inline-flex}:host .tp-helptip-label{display:inline-block;line-height:var(--ti-common-line-height-number)}:host tp-icon{color:var(--ti-common-color-icon-normal);cursor:help;font-size:16px;margin-left:var(--ti-common-space-3x)}:host tp-icon:hover{color:var(--ti-common-color-icon-hover)}"]
            },] }
];
TpHelptipComponent.propDecorators = {
    label: [{ type: Input }],
    iconTip: [{ type: Input }],
    iconTipPosition: [{ type: Input }],
    iconTipMaxWidth: [{ type: Input }]
};
