// @cloud/tiny3/esm2015/directives/overflow/TiOverflowDirective.js
import { Directive, ElementRef, Input } from '@angular/core';
import { TiOverflowService } from '../../services/overflow/TiOverflowService';
/**
 * 超长情况下文本处理出...并tip提示
 *
 * <example-url>../tiny3demo/#/overflow/overflow-all</example-url>
 */
export class TiOverflowDirective {
    constructor(ele, overflowService) {
        this.ele = ele;
        this.overflowService = overflowService;
        /**
         * 配置文本过长时显示的tip的最大宽度。
         *
         * 3.0.9新增
         */
        this.tiTipMaxWidth = '276px';
    }
    ngOnInit() {
        this.overflowRef = this.overflowService.create(this.ele.nativeElement, {
            tipContent: this.tiTipContent,
            tipPosition: this.tiTipPosition,
            tipHostEleX: this.tiTipHostEleX,
            tipMaxWidth: this.tiTipMaxWidth
        });
    }
    // 服务方式的overflow无法自销毁，因此此处在组件销毁时销毁服务方式生成的组件实例
    ngOnDestroy() {
        if (this.overflowRef) {
            this.overflowRef.destroy();
        }
    }
}
TiOverflowDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiOverflow]:not([maxLine])'
            },] }
];
TiOverflowDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: TiOverflowService }
];
TiOverflowDirective.propDecorators = {
    tiOverflow: [{ type: Input }],
    tiTipContent: [{ type: Input }],
    tiTipPosition: [{ type: Input }],
    tiTipMaxWidth: [{ type: Input }],
    tiTipHostEleX: [{ type: Input }]
};
