// @cloud/tiny3/esm2015/components/accordion/TiAccordionModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiCollapseModule } from '../../directives/collapse/TiCollapseModule';
import { TiOutlineModule } from '../../directives/outline/TiOutlineModule';
import { TiIconModule } from '../icon/TiIconModule';
import { TiAccordionComponent } from './TiAccordionComponent';
import { TiAccordionItemComponent } from './TiAccordionItemComponent';
import { TiAccordionHeadComponent } from './TiAccordionHeadComponent';
export class TiAccordionModule {
}
TiAccordionModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, TiCollapseModule, TiIconModule, TiOutlineModule],
                exports: [TiAccordionComponent, TiAccordionItemComponent, TiAccordionHeadComponent],
                declarations: [TiAccordionComponent, TiAccordionItemComponent, TiAccordionHeadComponent]
            },] }
];
export { TiAccordionComponent } from './TiAccordionComponent';
export { TiAccordionItemComponent } from './TiAccordionItemComponent';
export { TiAccordionHeadComponent } from './TiAccordionHeadComponent';
