// @cloud/tiny3/esm2015/components/accordion/TiAccordionItemComponent.js
import { Component, ElementRef, EventEmitter, Inject, Input, Output, Renderer2 } from '@angular/core';
import { TiAccordionComponent } from './TiAccordionComponent';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * 定义手风琴组件单个折叠面板，该组件包含在ti-accordion标签中
 * <example-url>../tiny3demo/#/accordion/acc-all</example-url>
 */
export class TiAccordionItemComponent extends TiBaseComponent {
    /**
     * @ignore
     * 依赖注入tiAccordion（后续使用其方法）
     */
    constructor(accordion, hostRef, renderer) {
        super(hostRef, renderer);
        this.accordion = accordion;
        /**
         * 单个折叠面板项是否灰化
         */
        this.disabled = false;
        /**
         * 设置用户点击某一折叠面板的head时是否自动实现内容体展示状态切换
         *
         * 头部局部点击可切换场景，需要通过设置该变量实现
         */
        this.clickToggle = true;
        /**
         * 某一折叠面板的头部点击回调
         */
        this.headClick = new EventEmitter();
        /**
         * @ignore
         * 头部class,需要显示声明，否则ngClass会报错
         */
        this.headClass = '';
        /**
         * @ignore
         * body class,需要显示声明，否则ngClass会报错
         */
        this.bodyClass = '';
        // 面板展开状态获取/设置：此处处理涉及到内部和外部修改，因此此处通过set/get进行处理
        this.isOpen = false;
    }
    /**
     * 单个折叠面板是否处于展开状态，用于初始化设置，默认值为false
     *
     */
    get open() {
        return this.isOpen;
    }
    /**
     * @ignore
     */
    set open(value) {
        if (value !== this.open) {
            if (value) { // open属性更新且open值为true时，处理其它item的关闭状态
                this.accordion.closeOtherItems(this);
            }
            this.isOpen = value;
        }
    }
    // 初始化处理：添加当前item到列表中,设置该item的id值
    ngOnInit() {
        super.ngOnInit();
        this.accordion.addItem(this);
        this.headClass = this.accordion.headClass;
        this.bodyClass = this.accordion.bodyClass;
    }
    // item销毁处理：移除item列表中的该项数据
    ngOnDestroy() {
        super.ngOnDestroy();
        this.accordion.removeItem(this);
    }
    /**
     * @ignore
     * item头部点击事件逻辑处理
     */
    toggleOpen($event) {
        if (this.disabled) {
            return;
        }
        if (this.clickToggle) {
            this.open = !this.open;
        }
        this.headClick.emit($event);
    }
}
TiAccordionItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-accordion-item',
                template: "<div [ngClass]=\"[headClass, disabled ? 'disabled': '', 'ti3-accordion-panel-heading']\"\r\n[tabindex]=\"disabled ? -1 : 0\" tiOutline (keydown.enter)=\"toggleOpen($event)\"\r\n     (click)=\"toggleOpen($event)\" [id]=\"appendId('head')\">\r\n    <div class=\"ti3-accordion-panel-title\">\r\n        <div class=\"ti3-icon ti3-icon-angle-right ti3-accordion-panel-title-icon\"\r\n             [ngClass]=\"{'ti3-icon-angle-transform-down': !open,'ti3-icon-angle-transform-up': open,\r\n             'ti3-accordion-panel-title-icon-disabled': disabled}\"></div>\r\n        <span class=\"ti3-accordion-panel-title-content\">\r\n            <span [ngClass]=\"{'ti3-accordion-panel-title-content-disabled': disabled}\">\r\n                <ng-content select=\"ti-accordion-head\"></ng-content>\r\n            </span>\r\n        </span>\r\n    </div>\r\n</div>\r\n<div class=\"ti3-accordion-panel-collapse\" [tiCollapse]=\"!open\">\r\n    <div class=\"ti3-accordion-panel-body\" [ngClass]=\"[bodyClass]\">\r\n        <ng-content></ng-content>\r\n    </div>\r\n</div>",
                host: {
                    '[class.ti3-accordion-panel]': 'true'
                },
                styles: [":host.ti3-accordion-group{display:block}:host.ti3-accordion-panel{background-color:var(--ti-accordion-panel-bg-color);border:var(--ti-accordion-panel-border);border-radius:var(--ti-accordion-panel-border-radius);display:block;margin-bottom:0}.ti3-accordion-panel-heading{-ms-user-select:none;-webkit-user-select:none;background-color:var(--ti-accordion-panel-head-bg-color);color:var(--ti-common-color-text-primary);cursor:pointer;height:var(--ti-accordion-panel-height);line-height:var(--ti-accordion-panel-lineheight);padding:var(--ti-accordion-panel-head-padding);user-select:none}.ti3-accordion-panel-heading.disabled{background-color:var(--ti-accordion-panel-head-bg-color-disabled);cursor:not-allowed;outline:none}.ti3-accordion-panel-title-icon{color:var(--ti-accordion-panel-icon-color);display:inline-block;font-size:var(--ti-accordion-panel-icon-fs);line-height:var(--ti-accordion-panel-icon-fs);vertical-align:middle}.ti3-accordion-panel-title-icon-disabled{color:var(--ti-accordion-panel-icon-color-disabled)}.ti3-icon-angle-transform-down{transform:rotate(90deg)}.ti3-icon-angle-transform-up{transform:rotate(-90deg)}.ti3-accordion-panel-title{align-items:center;color:var(--ti-accordion-panel-title-color);display:flex;font-size:var(--ti-accordion-panel-fs);margin-bottom:0;margin-top:0}.ti3-accordion-panel-title-content{margin-left:var(--ti-accordion-panel-title-margin-left)}.ti3-accordion-panel-title-content-disabled{color:var(--ti-accordion-panel-title-color-disabled)}:host.ti3-accordion-group ::ng-deep .ti3-accordion-panel+.ti3-accordion-panel{margin-top:var(--ti-accordion-panel-space-between)}.ti3-accordion-panel-collapse{border-top:var(--ti-accordion-panel-collapse-border-top)}.ti3-accordion-panel-body{background-color:var(--ti-accordion-panel-body-bg);box-shadow:var(--ti-accordion-panel-body-shadow);padding:var(--ti-accordion-panel-body-padding)}"]
            },] }
];
TiAccordionItemComponent.ctorParameters = () => [
    { type: TiAccordionComponent, decorators: [{ type: Inject, args: [TiAccordionComponent,] }] },
    { type: ElementRef },
    { type: Renderer2 }
];
TiAccordionItemComponent.propDecorators = {
    disabled: [{ type: Input }],
    clickToggle: [{ type: Input }],
    headClick: [{ type: Output }],
    open: [{ type: Input }]
};
