// @cloud/tiny3/esm2015/components/accordion/TiAccordionComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * 手风琴组件，
 * 用于实现多个折叠面板的呈现方式，通过ti-accordion、ti-accordion-item、ti-accordion-head组件嵌套方式实现
 *
 * <p><span style="color: red">使用此组件时需要开发者在项目模块(建议在根模块)
 * 中引入BrowserAnimationsModule。</span>这是因为此组件中使用了Angular动画，需要引入BrowserAnimationsModule，
 * 但是 BrowserAnimationsModule 不能在懒加载模块被重复引入，所以需要开发者来引入BrowserAnimationsModule，保证其引入一次。</p>
 *
 * <example-url>../tiny3demo/#/accordion/acc-all</example-url>
 */
export class TiAccordionComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 展开某一项时，是否关闭其它项
         */
        this.closeOthers = true;
        /**
         * 头部样式类
         */
        this.headClass = '';
        /**
         * 内容体样式类
         */
        this.bodyClass = '';
        this.items = [];
    }
    /**
     * @ignore
     * 关闭其余item
     */
    closeOtherItems(openItem) {
        if (!this.closeOthers) {
            return;
        }
        this.items.forEach((item) => {
            if (item.open && item !== openItem) {
                item.open = false;
            }
        });
    }
    /**
     * @ignore
     * 添加某一accordion条目到items数组中
     */
    addItem(item) {
        this.items.push(item);
    }
    /**
     * @ignore
     * 从items数组中删除某一accordion条目
     */
    removeItem(item) {
        const index = this.items.indexOf(item);
        if (index !== -1) {
            this.items.splice(index, 1);
        }
    }
}
TiAccordionComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-accordion',
                template: "<ng-content></ng-content>\r\n",
                host: {
                    '[class.ti3-accordion-group]': 'true'
                },
                styles: [":host.ti3-accordion-group{display:block}:host.ti3-accordion-panel{background-color:var(--ti-accordion-panel-bg-color);border:var(--ti-accordion-panel-border);border-radius:var(--ti-accordion-panel-border-radius);display:block;margin-bottom:0}.ti3-accordion-panel-heading{-ms-user-select:none;-webkit-user-select:none;background-color:var(--ti-accordion-panel-head-bg-color);color:var(--ti-common-color-text-primary);cursor:pointer;height:var(--ti-accordion-panel-height);line-height:var(--ti-accordion-panel-lineheight);padding:var(--ti-accordion-panel-head-padding);user-select:none}.ti3-accordion-panel-heading.disabled{background-color:var(--ti-accordion-panel-head-bg-color-disabled);cursor:not-allowed;outline:none}.ti3-accordion-panel-title-icon{color:var(--ti-accordion-panel-icon-color);display:inline-block;font-size:var(--ti-accordion-panel-icon-fs);line-height:var(--ti-accordion-panel-icon-fs);vertical-align:middle}.ti3-accordion-panel-title-icon-disabled{color:var(--ti-accordion-panel-icon-color-disabled)}.ti3-icon-angle-transform-down{transform:rotate(90deg)}.ti3-icon-angle-transform-up{transform:rotate(-90deg)}.ti3-accordion-panel-title{align-items:center;color:var(--ti-accordion-panel-title-color);display:flex;font-size:var(--ti-accordion-panel-fs);margin-bottom:0;margin-top:0}.ti3-accordion-panel-title-content{margin-left:var(--ti-accordion-panel-title-margin-left)}.ti3-accordion-panel-title-content-disabled{color:var(--ti-accordion-panel-title-color-disabled)}:host.ti3-accordion-group ::ng-deep .ti3-accordion-panel+.ti3-accordion-panel{margin-top:var(--ti-accordion-panel-space-between)}.ti3-accordion-panel-collapse{border-top:var(--ti-accordion-panel-collapse-border-top)}.ti3-accordion-panel-body{background-color:var(--ti-accordion-panel-body-bg);box-shadow:var(--ti-accordion-panel-body-shadow);padding:var(--ti-accordion-panel-body-padding)}"]
            },] }
];
TiAccordionComponent.propDecorators = {
    closeOthers: [{ type: Input }],
    headClass: [{ type: Input }],
    bodyClass: [{ type: Input }]
};
