// @cloud/tinyplus3/esm2015/components/recommend/TpRecommendComponent.js
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
import { TiBaseComponent, TiLocale } from '@cloud/tiny3';
/**
 * 推荐组件，优先向用户推荐可用区中的某些区域
 *
 * <example-url>../tinyplus3demo/#/recommend/recommend-all</example-url>
 */
export class TpRecommendComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 选中事件，点击某项时向外通知该选中项
         */
        this.select = new EventEmitter();
    }
    /**
     * @ignore 处理资源数量
     */
    getNum(num) {
        return num > 99 ? '99+' : String(num);
    }
    /**
     * @ignore
     */
    clickFn(item) {
        this.select.emit(item);
    }
    ngOnInit() {
        this.label = this.label || TiLocale.getLocaleWords().tpRecommend.label;
    }
}
TpRecommendComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-recommend',
                template: "<tp-smartalert size=\"small\">\r\n    <tp-icon local name=\"cloud-pointer\" #leftIcon style=\"font-size: 12px;\"></tp-icon>\r\n    <div class=\"tp-recommend-container\">\r\n        <span class=\"tp-recommend-label\" [id]=\"appendId('label')\">{{label}}</span>\r\n        <div class=\"tp-recommend-content\">\r\n            <ng-container *ngFor=\"let item of items;index as index; last as last\">\r\n                <span class=\"tp-recommend-item\">\r\n                    <ng-container *ngTemplateOutlet=\"beforeItemTemplate;context:{$implicit: item}\"></ng-container>\r\n                    <span tiOutline class=\"tp-recommend-item-content\" [tabindex]=\"0\" (keydown.enter)=\"clickFn(item)\"\r\n                    (click)=\"clickFn(item)\" [id]=\"appendId(index)\" [tiTip]=\"item.tip\" [tiTipContext]=\"item\">\r\n                    <span class=\"tp-recommend-item-text\" [title]=\"item.displayName\">{{item.displayName}}</span>\r\n                    <span *ngIf=\"item.num && item.showNum !== false\" >({{getNum(item.num)}})</span>\r\n                    </span>\r\n                    <span class=\"tp-recommend-line\" *ngIf=\"!last\"></span>\r\n                </span>\r\n            </ng-container>\r\n        </div>\r\n    </div>\r\n</tp-smartalert>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-recommend-item-text-max-width:108px;--tp-recommend-max-width:685px}:host{display:inline-flex;font-size:var(--ti-common-font-size-base);max-width:var(--tp-recommend-max-width)}:host .tp-recommend-container{-moz-align-items:stretch;-moz-flex-direction:row;-moz-justify-content:flex-start;-ms-align-items:stretch;-ms-justify-content:flex-start;align-items:stretch;display:flex;flex-direction:row;justify-content:flex-start;max-width:calc(var(--tp-recommend-max-width) - var(--ti-common-space-3x)*2)}.tp-recommend-label{color:var(--ti-common-color-text-primary);flex-shrink:0;line-height:var(--ti-common-line-height-number);margin-right:var(--ti-common-space-3x)}.tp-recommend-line{background-color:var(--ti-common-color-bg-light-emphasize);display:inline-block;height:var(--ti-common-space-3x);margin:0 var(--ti-common-space-3x);position:relative;top:2px;width:1px}.tp-recommend-content{display:flex;flex-wrap:wrap}.tp-recommend-item{color:var(--ti-common-color-text-primary);display:inline-block}.tp-recommend-item .tp-recommend-item-content{cursor:pointer}.tp-recommend-item .tp-recommend-item-content:hover{color:var(--ti-common-color-text-highlight)}.tp-recommend-item .tp-recommend-item-text{-ms-box-sizing:border-box;box-sizing:border-box;display:inline-block;margin-right:var(--ti-common-space-base);max-width:var(--tp-recommend-item-text-max-width);overflow:hidden;text-overflow:ellipsis;vertical-align:top;white-space:nowrap}"]
            },] }
];
TpRecommendComponent.propDecorators = {
    items: [{ type: Input }],
    label: [{ type: Input }],
    select: [{ type: Output }],
    beforeItemTemplate: [{ type: ContentChild, args: ['beforeItem', { static: true },] }]
};
