// @cloud/tiny3/esm2015/directives/include/TiIncludeDirective.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { Util } from '../../utils/Util';
/**
 * @ignore
 * 此指令功能:清空宿主元素的子元素，将传入的string或Node放入(append)
 *
 * 类似angular.js的ng-bind-html功能
 */
export class TiIncludeDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.element = this.elementRef.nativeElement;
    }
    ngOnChanges(changes) {
        if (changes['tiInclude'] && !Util.isUndefined(this.tiInclude)) {
            // 清空宿主元素的子节点
            if (this.element.childNodes.length > 0) {
                this.element.innerHTML = '';
            }
            // 如果 tiInclude 为字符串，则当作文本节点处理
            const node = Util.isString(this.tiInclude) ?
                this.renderer.createText(this.tiInclude) : this.tiInclude;
            this.renderer.appendChild(this.element, node);
        }
    }
}
TiIncludeDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiInclude]'
            },] }
];
TiIncludeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TiIncludeDirective.propDecorators = {
    tiInclude: [{ type: Input }]
};
