// @cloud/tiny3/esm2015/components/progressbar/TiProgressbarComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * 进度条组件
 *
 * <example-url>../tiny3demo/#/progressbar/progressbar-all</example-url>
 */
export class TiProgressbarComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         *  设置当前值，当前值/最大值即为进度值
         */
        this.value = 0;
        /**
         *  设置最大值，当前值/最大值即为进度值
         */
        this.max = 100;
        /**
         *  设置进度条在不同当前值下对应的样式，包括动画效果
         */
        this.progressClass = '';
        /**
         *  设置进度条是否开启动画效果
         */
        this.animation = true;
    }
    // 设置合法的数值
    static verifyNum(newVal, defaultValue) {
        return isNaN(parseFloat(newVal)) ? defaultValue : newVal;
    }
    ngOnInit() {
        super.ngOnInit();
        this.max = TiProgressbarComponent.verifyNum(this.max, this.max);
        this.value = TiProgressbarComponent.verifyNum(this.value, this.value);
        this.calcPercentage();
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        // 当前值更新，重新计算进度
        this.setChanges(changes, 'value');
        // 当最大值更新，重新计算进度
        this.setChanges(changes, 'max');
    }
    /**
     * @description 处理用户输入的改变
     * @param: changes 改变的对象
     * @param: changeKey 那个属性改变
     */
    setChanges(changes, changeKey) {
        if (changes[changeKey] && !changes[changeKey].isFirstChange()) {
            const oldVal = changes[changeKey].previousValue;
            const newVal = changes[changeKey].currentValue;
            const _newValue = TiProgressbarComponent.verifyNum(newVal, oldVal);
            this[changeKey] = _newValue;
            this.calcPercentage();
        }
    }
    // 计算当前进度百分比
    calcPercentage() {
        // 如果开发者设置数据不合理，则进度置0
        if (this.max === 0) {
            this.percent = 0;
            return;
        }
        this.percent = Number((this.value * 100 / this.max).toFixed(2));
        if (this.percent > 100) {
            this.percent = 100;
        }
        if (this.percent < 0) {
            this.percent = 0;
        }
    }
}
TiProgressbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-progressbar',
                template: "<div class=\"ti3-progress-bar\" [ngStyle]=\"{'width': percent + '%'}\" [ngClass]=\"progressClass\"\r\n[class.ti3-progress-no-animation]=\"!animation\">\r\n</div>\r\n<div class=\"ti3-progress-label\" [id]=\"appendId('label')\">\r\n\t<ng-content></ng-content>\r\n</div>",
                host: {
                    '[class.ti3-progress]': 'true'
                },
                styles: [":host.ti3-progress{-ms-box-sizing:border-box;background-color:var(--ti-progressbar-progress-bg-color);box-sizing:border-box;display:inline-block;height:var(--ti-progressbar-height);margin-bottom:0;overflow:visible;position:relative;vertical-align:middle;width:100%}.ti3-progress-bar{background-color:var(--ti-progressbar-bg-color);bottom:0;transition:width .6s ease;width:0}.ti3-progress-bar,.ti3-progress-label{-ms-box-sizing:border-box;box-sizing:border-box;left:0;position:absolute;top:0}.ti3-progress-label{background:transparent;color:var(--ti-progressbar-progress-label-color);display:none;height:100%;text-align:center;vertical-align:middle;white-space:nowrap;width:100%}.ti3-progress-label:after{content:\"\";display:inline-block;height:100%;vertical-align:middle;width:0}.ti3-progress-bar.ti3-progress-no-animation{animation:none!important;transition:none!important}"]
            },] }
];
TiProgressbarComponent.propDecorators = {
    value: [{ type: Input }],
    max: [{ type: Input }],
    progressClass: [{ type: Input }],
    animation: [{ type: Input }]
};
