// @cloud/tiny3/esm2015/components/date/TiDateComponent.js
import { Component, ContentChild, TemplateRef } from '@angular/core';
import { TiDateUtil, TiKeymap, Util } from '../../utils/Util';
import { TiLocale, TiLocaleFormat } from '../../locale/TiLocaleModule';
import { TiFormComponent } from '../base/TiBaseModule';
import { TiDateBaseComponent } from '../datebase/TiDateBaseModule';
/**
 * Date日期组件
 *
 * Date组件提供了一种方便的显示和设置日期的方式
 *
 * <example-url>../tiny3demo/#/date/date-all</example-url>
 */
export class TiDateComponent extends TiDateBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * @ignore
         * 标记date/datetime的类型
         */
        this.isDatetime = false;
        /**
         * @ignore
         * 用于标记是不是range
         */
        this.isRange = false;
        /**
         * @ignore
         * placeholder提示文本
         */
        this.placeholder = `${TiLocale.getLocaleWords().tiDate.datePlaceholder}`;
    }
    // tslint:disable-next-line:use-life-cycle-interface
    ngOnChanges(changes) {
        // format支持动态变更
        if (changes['format'] && !changes['format'].firstChange) {
            // 新的format是非法时，format值保持之前值不变
            if (!Util.isString(changes['format'].currentValue)) {
                this.format = changes['format'].previousValue;
                return;
            }
            this.formatValue();
            // 配置面板的接口值
            this.setPickerDate();
        }
        // 验证最大值最小值，为了处理最大值和最小值从合法日期变为undefined 的情景
        if ((changes['max'] && !changes['max'].firstChange) || (changes['min'] && !changes['min'].firstChange)) {
            this.validateMinAndMax(this.config, this.isDatetime);
        }
    }
    /**
     * @ignore
     * model值变化时的回调
     */
    ngOnModelChange() {
        if (!TiDateUtil.isDate(this.model)) {
            this.inputValue = '';
            return;
        }
        this.formatValue();
        this.setPickerDate();
    }
    /**
     * @ignore
     */
    setInputValue(value) {
        let isDisabled = false;
        if (Util.isArray(this.disabledDays) && this.disabledDays.length > 0) {
            this.disabledDays.forEach((item) => {
                if (item.getTime() === value.getTime()) {
                    isDisabled = true;
                }
            });
        }
        if (isDisabled) {
            return;
        }
        if (this.isValidValue(value)) {
            this.model = value;
        }
    }
    /**
     * @ignore
     */
    onKeydownFn(event, val, pos) {
        if (event.keyCode === TiKeymap.KEY_ENTER || event.keyCode === TiKeymap.KEY_NUMPAD_ENTER) {
            this.hideDrop();
        }
    }
    /**
     * @ignore
     * 配置时间日期面板接口
     */
    setPickerDate() {
        this.datePanel = {
            value: {
                begin: this.model,
                // tslint:disable-next-line:no-null-keyword
                end: null
            },
            picker: 'day',
            min: this.min,
            max: this.max,
            format: this.format,
            nowDateTime: this.nowDateTime,
            select: () => {
                // 点击面板：新旧值相同不做处理，不相同，将新值赋给model
                if (!TiDateUtil.isDateEqual(this.model, this.datePanel['value'].begin)) {
                    this.model = this.datePanel['value'].begin;
                }
                this.hideDrop();
            }
        };
        // 如果只显示年
        if (this.format.indexOf('M') === -1 && this.format !== 'mediumDate') {
            this.datePanel['picker'] = 'year';
        }
        else if (this.format.indexOf('d') === -1) { // 如果只显示年月
            this.datePanel['picker'] = 'month';
        }
    }
    /**
     * @ignore
     * 根据model值和format接口，格式化显示时间日期
     */
    formatValue() {
        this.inputValue = (this.model !== null) ? TiLocaleFormat.formatDate(this.model, this.format) : '';
    }
    /**
     * @ignore
     * 判断是不是合法的model
     */
    isValidValue(value) {
        if (TiDateUtil.isDate(value) && TiDateUtil.isBetweenMaxAndmin(value, this.min, this.max)) {
            return true;
        }
        return false;
    }
}
TiDateComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-date',
                template: "<ti-date-dominator #dominator\r\n                   [id]=\"appendId('dominator')\"\r\n                   [clearable]=\"clearIcon\"\r\n                   [(ngModel)]=\"inputValue\"\r\n                   [disabled]=\"disabled\"\r\n                   (clear)=\"onIconClearClick($event)\"\r\n                   (click)=\"onShowClick()\">{{ placeholder }}</ti-date-dominator>\r\n<ti-drop #drop\r\n        [id]=\"appendId('drop')\"\r\n        class=\"ti3-date-picker\"\r\n        [ngClass]=\"{'ti3-date-customize': hasCustomizeTime}\"\r\n        [dominatorElem]=\"dominatorCom.nativeElement\"\r\n        fixMaxHeight=\"true\"\r\n        [panelWidth]=\"datePanelWidth\"\r\n        [dominatorSpace]='dominatorSpace'\r\n        [panelAlign]=\"panelAlign\"\r\n        [browserSpace]=\"browserSpace\"\r\n        [determinPositionFn]=\"determinPositionFn\"\r\n        theme='noborder'>\r\n        <section class='ti3-date-clear'>\r\n            <section *ngIf='hasCustomizeTime' class='ti3-date-customize-container'>\r\n                <ng-container *ngTemplateOutlet=\"customizeTemplate; context:  {$implicit: customizeOptions}\"></ng-container>\r\n            </section>\r\n            <section [ngClass]=\"{'ti3-date-panel-container': hasCustomizeTime}\">\r\n                <ti-date-edit\r\n                [id]=\"appendId('date-edit')\"\r\n                class='ti3-date-edit'\r\n                [(ngModel)] = \"datePanel.value.begin\"\r\n                [min] = \"datePanel.min\"\r\n                [max] = \"datePanel.max\"\r\n                [format]=\"datePanel.format\"\r\n                [disabledDays]='disabledDays'\r\n                (keydown)=\"onKeydownFn($event, datePanel.value.begin, 'begin')\"\r\n                (ngModelChange)=\"setInputValue(datePanel.value.begin)\">\r\n            </ti-date-edit>\r\n            <ti-date-panel\r\n                [id]=\"appendId('panel')\"\r\n                [(value)]=\"datePanel.value\"\r\n                [focusedPosition]='focusedPosition'\r\n                [isRange]='false'\r\n                [(picker)]=\"datePanel.picker\"\r\n                [min]=\"datePanel.min\"\r\n                [max]=\"datePanel.max\"\r\n                [disabledDays]='disabledDays'\r\n                [nowDateTime]=\"datePanel.nowDateTime\"\r\n                [format]=\"datePanel.format\"\r\n                (select)=\"datePanel.select()\"></ti-date-panel>\r\n            </section>\r\n        </section>\r\n</ti-drop>\r\n\r\n<ng-template #customizeTemplate let-options>\r\n    <ul [id]=\"appendId('list')\">\r\n        <li *ngFor='let option of options; index as i;' class='ti3-customize-time-label' (click)='customizeTimeClickFn(option.value.begin)' [id]=\"appendId('label_' + i)\">{{option.label}}</li>\r\n    </ul>\r\n</ng-template>\r\n",
                // tslint:disable-next-line:use-host-property-decorator
                host: {
                    '[class.ti3-date-input-container]': 'true',
                    '(blur)': 'hidePanel()'
                },
                providers: [TiFormComponent.getValueAccessor(TiDateComponent)],
                styles: [".ti3-compnent-container-border,:host.ti3-date-input-container{-ms-box-sizing:border-box;border:1px solid;border-radius:var(--ti-input-border-radius);box-sizing:border-box;display:inline-block}.ti3-compnent-container-border:not([disabled]),:host.ti3-date-input-container:not([disabled]){background-color:var(--ti-input-bg-color);border-color:var(--ti-input-border-color)}.ti3-compnent-container-border:not([disabled]):hover,:host.ti3-date-input-container:not([disabled]):hover{border-color:var(--ti-input-border-color-hover)}.ti3-compnent-container-border:not([disabled])[tiFocused],:host.ti3-date-input-container:not([disabled])[tiFocused]{border-color:var(--ti-input-border-color-focus)}.ti3-compnent-container-border[disabled],:host.ti3-date-input-container[disabled]{background-color:var(--ti-input-bg-color-disabled);border-color:var(--ti-input-border-color-disabled);cursor:not-allowed!important}.ti3-customize-time-label{color:var(--ti-common-color-text-primary);cursor:pointer;font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number);margin-bottom:10px}.ti3-customize-time-label:hover{color:var(--ti-common-color-line-active)}.ti3-date-clear{display:flex;*zoom:1}.ti3-date-clear:after{clear:both}.ti3-date-clear:after,.ti3-date-clear:before{content:\"\";display:table}.ti3-datetime-select-btn{color:var(--ti-common-color-text-highlight);cursor:pointer;float:right;line-height:24px;padding-right:var(--ti-common-space-2x)}.ti3-datetime-select-btn-disabled{color:var(--ti-common-color-text-disabled);cursor:not-allowed}.ti3-date-range-splitline{padding:0 6px}.ti3-tab-input{height:0;left:-9999px;position:absolute;top:-9999px;width:0}::ng-deep :root{--ti-date-picker-padding-bottom:12px;--ti-date-picker-padding-horizon:16px}:host.ti3-date-input-container{width:160px}.ti3-dropdown-container.ti3-date-picker{font-size:var(--ti-datetime-input-font-size);padding:var(--ti-date-picker-padding-bottom) var(--ti-date-picker-padding-horizon);width:calc(var(--ti-datetime-day-width)*7 + var(--ti-date-picker-padding-horizon)*2 + 2px);z-index:var(--ti-datetime-z-index)}.ti3-dropdown-container.ti3-date-picker:focus{outline:0}.ti3-dropdown-container.ti3-date-picker.ti3-date-customize{width:399px!important}.ti3-date-customize-container{border-right:1px solid var(--ti-common-color-line-dividing);height:292px;padding-right:var(--ti-common-space-4x);width:80px}.ti3-date-edit{width:100%}::ng-deep .ti3-date-edit input[tiText]{border-color:var(--ti-input-border-color-focus)!important}.ti3-date-panel-container{padding-left:var(--ti-common-space-4x)}"]
            },] }
];
TiDateComponent.propDecorators = {
    customizeTemplate: [{ type: ContentChild, args: ['customize', { static: true },] }]
};
