// @cloud/tiny3/esm2015/services/renderer/TiRenderer.js
import { Injectable, RendererFactory2 } from '@angular/core';
import { TiRendererModule } from './TiRendererModule';
import * as i0 from "@angular/core";
import * as i1 from "./TiRendererModule";
/**
 * @ignore
 */
export class TiRenderer {
    constructor(rendererFactory) {
        this.renderer = rendererFactory.createRenderer(null, null);
    }
    /**
     * @description: 将节点插入某节点元素之后
     * @param: sourceEle 被插入节点
     * @param: targetEle 节点插入位置
     */
    insertAfter(sourceEle, targetEle) {
        const parent = targetEle.parentNode;
        // 如果最后的节点是目标元素,则直接添加
        if (parent.lastChild === targetEle) {
            this.renderer.appendChild(parent, sourceEle);
        }
        else { // 如果不是,则插在目标元素的下一个兄弟节点之前
            this.renderer.insertBefore(parent, sourceEle, targetEle.nextSibling);
        }
    }
    /**
     * @description: 判读一个元素上是否存在某个样式类名
     * @param: ele 被判断的元素
     * @param: className 样式类名
     */
    hasClass(element, className) {
        const classList = element.classList;
        return classList.contains(className);
    }
    /**
     * @description: 给指定元素设置属性
     * @param: element 被设置的元素
     * @param: attr Object 属性对象
     */
    setAttributes(element, attr) {
        for (const key in attr) {
            if (attr.hasOwnProperty(key)) {
                this.renderer.setAttribute(element, key, String(attr[key]));
            }
        }
    }
    /**
     * @description: 为元素添加多个样式
     * @param: ele 元素对象
     * @param: styles  Object 样式对象，如：{width: 100, height: 200}
     */
    setStyles(ele, styles) {
        for (const key in styles) {
            if (styles.hasOwnProperty(key)) {
                this.renderer.setStyle(ele, key, styles[key]);
            }
        }
    }
    /**
     * @description: 将节点插入某父容器，作为第一个元素
     * @param: parentEle 父节点元素
     * @param: sourceEle 需要插入的节点元素
     */
    prepend(parentEle, sourceEle) {
        if (!parentEle) {
            return;
        }
        this.renderer.insertBefore(parentEle, sourceEle, parentEle.firstElementChild);
    }
    /**
     * @description: 查找一个元素下有某个样式类的子元素
     * @param: element 被查找的元素
     * @param: className 子元素的样式类名
     */
    findChildrenByClassName(element, className) {
        let resultChildren = [];
        resultChildren = Array.from(element.children).filter((child) => {
            return this.hasClass(child, className);
        });
        return resultChildren;
    }
    /**
     * @description: 查找一个元素下有某个样式类的子元素
     * @param: element 被查找的元素
     * @param: className 子元素的样式类名
     */
    findChildrenByTagName(element, tagName) {
        let resultChildren = [];
        resultChildren = Array.from(element.children).filter((child) => {
            return child.tagName === tagName;
        });
        return resultChildren;
    }
}
TiRenderer.ɵprov = i0.ɵɵdefineInjectable({ factory: function TiRenderer_Factory() { return new TiRenderer(i0.ɵɵinject(i0.RendererFactory2)); }, token: TiRenderer, providedIn: i1.TiRendererModule });
TiRenderer.decorators = [
    { type: Injectable, args: [{
                providedIn: TiRendererModule
            },] }
];
TiRenderer.ctorParameters = () => [
    { type: RendererFactory2 }
];
