// @cloud/tinyplus3/esm2015/components/pricecalc/TpPricecalcComponent.js
import { Component, ContentChild, ElementRef, EventEmitter, Input, Output, Renderer2, SecurityContext, TemplateRef, ViewChild } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { TiBaseComponent, TiBrowser, TiLocale, Util } from '@cloud/tiny3';
/**
 * 价格计算器组件
 *
 * <example-url>../tinyplus3demo/#/pricecalc/pricecalc-all</example-url>
 */
export class TpPricecalcComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, domSanitizer) {
        super(elementRef, renderer2);
        this.domSanitizer = domSanitizer;
        /**
         * 图标右侧文本
         *
         * 缺省值 : 价格计算器/price Calculator（国际化）
         */
        this.label = TiLocale.getLocaleWords().tpPricecalc.label;
        /**
         * 查看价格的链接地址
         */
        this.url = 'http://www.helpcenter.com/';
        /**
         * 点击链接前的事件接口
         */
        this.beforeClick = new EventEmitter();
        /**
         * @ignore
         */
        this.detailHrefBeforeText = TiLocale.getLocaleWords().tpPricecalc.detailHrefBeforeText;
        /**
         * @ignore
         */
        this.detailHrefText = TiLocale.getLocaleWords().tpPricecalc.detailHrefText;
        /**
         * @ignore
         */
        this.detailHrefAfterText = TiLocale.getLocaleWords().tpPricecalc.detailHrefAfterText;
    }
    /**
     * @ignore
     */
    onClick() {
        this.beforeClick.emit();
        setTimeout(() => {
            // 问题背景：给 form 标签上设置 rel="noopener noreferrer" 实际是不生效的，还是存在被新打开的页面的恶意带代码重定向原页面的安全漏洞的。
            // 解决方案：先用 window.open 打开一个新的页签窗口(this.blankTab)，然后利用 form 的 target 属性将表单提交到这个窗口中, 那么就可以获取到新窗口
            //          的window对象(window.open 的返回值)，将其 opener 属性置为 null，从而消除该安全漏洞。
            // 注意：该解决方案不适用于IE, IE 下暂未找到解决方案
            if (!TiBrowser.isIE()) {
                this.blankTab = Util.getUniqueId('new_tab'); // 确保是新的页签
                const newWindow = window.open(this.domSanitizer.sanitize(SecurityContext.URL, this.url), this.blankTab);
                if (newWindow) {
                    newWindow.opener = null;
                }
            }
            this.form.nativeElement.target = TiBrowser.isIE() ? '_blank' : this.blankTab;
            this.form.nativeElement.submit();
        }, 0);
    }
    ngOnInit() {
        super.ngOnInit();
        this.isNotDefaultUrl = !!this.values && this.url !== 'http://www.helpcenter.com/';
    }
}
TpPricecalcComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-pricecalc',
                template: "<!-- <form  #myform action=\"{{url}}\" method=\"{{method}}\" target=\"_blank\" rel=\"noopener noreferrer\">-->\r\n<!-- \u7ED9 form \u6807\u7B7E\u4E0A\u8BBE\u7F6E rel=\"noopener noreferrer\" \u5B9E\u9645\u662F\u4E0D\u751F\u6548\u7684\uFF0C\u8FD8\u662F\u5B58\u5728\u88AB\u65B0\u6253\u5F00\u7684\u9875\u9762\u7684\u6076\u610F\u5E26\u4EE3\u7801\u91CD\u5B9A\u5411\u539F\u9875\u9762\u7684\u5B89\u5168\u6F0F\u6D1E\u7684\uFF0C\u6240\u4EE5\u91C7\u7528\u5176\u4ED6\u65B9\u5F0F\u89E3\u51B3 form \u7684\u8BE5\u5B89\u5168\u6F0F\u6D1E\uFF0C\u8BE6\u89C1ts\u4E2D\u7684\u6CE8\u89E3 -->\r\n<form  #myform action=\"{{url}}\" method=\"{{method}}\">\r\n    <input type=\"hidden\" *ngFor=\"let item of values\" name=\"{{item.key}}\" value=\"{{item.value}}\">\r\n    <input type=\"submit\" style=\"display: none;\" value=\"{{label}}\">\r\n</form>\r\n<span>\r\n    <tp-icon local name='cloud-action-calculator' class=\"calculator\"></tp-icon>\r\n    <ng-container *ngIf=\"isNotDefaultUrl\">\r\n        <a tiNavigation (click)=\"onClick()\"  href=\"javascript:void(0)\" oncontextmenu=\"return false\" [id]=\"appendId('label')\">{{label}}</a>\r\n    </ng-container>\r\n    <ng-container *ngIf=\"!isNotDefaultUrl\">\r\n        <a tiNavigation target=\"_blank\" rel=\"noopener noreferrer\" href=\"{{url}}\" [id]=\"appendId('label')\">{{label}}</a>\r\n    </ng-container>\r\n</span>\r\n<!-- \u8BF4\u660E\u6587\u5B57 -->\r\n<ng-container *ngIf=\"detailTemplate else defaultDetailTemplate\">\r\n    <span class=\"detail\" [id]=\"appendId('text')\">\r\n        <ng-container *ngTemplateOutlet=\"detailTemplate\">\r\n        </ng-container>\r\n    </span>\r\n</ng-container>\r\n\r\n<ng-template #defaultDetailTemplate>\r\n    <ng-container *ngIf=\"detailHref\">\r\n        <span class=\"detail\">\r\n            <span>{{detailHrefBeforeText}}</span><!--\r\n         --><a tiNavigation href=\"{{detailHref}}\" target=\"_blank\" rel=\"noopener noreferrer\">{{detailHrefText}}</a><!--\r\n         --><span>{{detailHrefAfterText}}</span>\r\n        </span>\r\n    </ng-container>\r\n</ng-template>",
                styles: [".calculator{color:var(--ti-common-color-text-link);margin-right:5px}.detail{display:inline-block;margin-left:var(--ti-common-space-2x)}"]
            },] }
];
TpPricecalcComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: DomSanitizer }
];
TpPricecalcComponent.propDecorators = {
    label: [{ type: Input }],
    detailHref: [{ type: Input }],
    values: [{ type: Input }],
    url: [{ type: Input }],
    method: [{ type: Input }],
    beforeClick: [{ type: Output }],
    form: [{ type: ViewChild, args: ['myform', { static: true },] }],
    detailTemplate: [{ type: ContentChild, args: ['detail', { static: true },] }]
};
