// @cloud/tinyplus3/esm2015/components/buylayer/TpBuylayerRightComponent.js
import { Component, Input } from '@angular/core';
/**
 * @ignore
 * 购买浮层右侧按钮区域
 * <example-url>../tinyplus3demo/#/buylayer/buylayer-all</example-url>
 */
export class TpBuylayerRightComponent {
    constructor() {
        /**
         * 浮层右侧按钮列表集合
         */
        this.buttons = [];
    }
}
TpBuylayerRightComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-buylayer-right',
                template: "<ng-container *ngIf='buttons && buttons.length'>\r\n  <span *ngFor='let button of buttons' [tiTip]=\"button.tip\" [tiTipPosition]=\"button.tipPosition\">\r\n    <button *ngIf='!button.hide' class='tp-buylayer-button' type=\"button\" tiButton size=\"large\" [disabled]='button.disabled' [id]='button.id' [color]=\"button.color\"\r\n      (click)='button.click && button.click()'>{{button.label}}</button>\r\n  </span>\r\n</ng-container>\r\n",
                host: {
                    '[class.tp-buylayer-right]': 'true'
                },
                styles: [".tp-buylayer-button,:host.tp-buylayer-right{display:inline-block}.tp-buylayer-button{margin-left:10px}"]
            },] }
];
TpBuylayerRightComponent.propDecorators = {
    buttons: [{ type: Input }]
};
