// @cloud/tinyplus3/esm2015/components/buylayer/TpBuylayerLeftComponent.js
import { Component, ContentChild, EventEmitter, Input, Output, TemplateRef } from '@angular/core';
import { getLocaleCurrencySymbol } from '@angular/common';
import { TiBaseComponent, TiLocale, Util } from '@cloud/tiny3';
import { CfUtil } from '../../cfutils/CfUtil';
import { TpUtil } from './../../utils/TpUtil';
/**
 * @ignore
 * 购买浮层左侧价格区域
 * <example-url>../tinyplus3demo/#/buylayer/buylayer-all</example-url>
 */
export class TpBuylayerLeftComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 浮层左侧价格列表集合
         */
        this.costs = [];
        /**
         * 可选优惠项切换选择项时触发
         *
         * 9.0.3 新增接口
         */
        this.discountsChange = new EventEmitter();
        /**
         * @ignore
         */
        this.freechargeDetailLink = '';
        /**
         * @ignore
         * tip中的表头
         */
        this.pricelistHead = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_pricelist;
        /**
         * @ignore
         * 主要价格是否包含数字。如果包含则需要出现计费详情，否则不出现。
         */
        this.hasNumberofMajorprice = false;
    }
    // tslint:disable-next-line:use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        const routeLink = TiLocale.getLocaleWords().tpBuylayer.buyLayer_freeCharge_detailLink;
        this.costsChange();
        this.discountChange();
        if (!Util.isUndefined(this.numFormat) && !Util.isUndefined(this.majorDiscoutsConfig)) {
            // 浮层配置的价格精度同样作用到优惠详情部分，两者保持一致。
            this.majorDiscoutsConfig.numFormat = this.numFormat;
        }
        CfUtil.getLinks()
            .then((linksData) => {
            if (linksData && linksData.links && linksData.links.user_center_common) {
                this.freechargeDetailLink = this.addLinkParams(linksData.links.user_center_common + routeLink);
            }
        });
    }
    // tslint:disable-next-line:use-life-cycle-interface
    ngOnChanges() {
        this.costsChange();
        this.discountChange();
    }
    ngDoCheck() {
        if (!Util.isArray(this.costs) || this.costs.length === 0) {
            return;
        }
        // costs动态hide之后，重新获取主次要价格数组，并且新数组的第一项（头部价格）不带加号连接符
        const currentMajorCosts = this.costs.filter((item) => !item.hide && item.majorPrice);
        const currentMinorCosts = this.costs.filter((item) => !item.hide && !item.majorPrice);
        currentMajorCosts.forEach((item, index) => {
            item.hideConnectSymbol = !index;
        });
        currentMinorCosts.forEach((item, index) => {
            item.hideConnectSymbol = !index;
        });
    }
    // 数据初始化及变更之后都需执行
    costsChange() {
        this.setMajorPrice();
        this.finalCosts = this.costs.filter((item) => !item.hide);
        this.majorPriceNum = this.getMajorPriceNum();
        this.isCostsAllHide();
        if (this.isRefund) {
            this.setRefundParams();
        }
        this.onlyHasMajorprice = this.finalCosts && this.finalCosts.length > 0 && this.majorPriceNum === this.finalCosts.length ? true : false;
    }
    hasDiscountsData(config) {
        const hasSelectableDiscounts = Util.isArray(config.selectableDiscounts) && config.selectableDiscounts.length > 0;
        const hasSelectedDiscounts = Util.isArray(config.selectedDiscounts) && config.selectedDiscounts.length > 0;
        const hasDiscounts = Util.isArray(config.discounts) && config.discounts.length > 0;
        return hasDiscounts || (hasSelectedDiscounts || hasSelectableDiscounts);
    }
    discountChange() {
        if (TpUtil.isObject(this.majorDiscoutsConfig) && this.hasDiscountsData(this.majorDiscoutsConfig) && this.majorPriceNum > 0) {
            this.showMajorDiscount = true;
        }
        else {
            this.showMajorDiscount = false;
        }
        const minorpriceNum = this.costs && this.costs.length - this.majorPriceNum; // 得到次要价格的数目
        if (TpUtil.isObject(this.minorDiscoutsConfig) && this.hasDiscountsData(this.minorDiscoutsConfig) && minorpriceNum > 0) {
            this.showMinorDiscount = true;
        }
        else {
            this.showMinorDiscount = false;
        }
    }
    /**
     * @ignore
     */
    change($event) {
        this.discountsChange.emit($event);
    }
    // 获取主要价格数量
    getMajorPriceNum() {
        if (this.finalCosts && this.finalCosts.length > 0) {
            let num = 0;
            for (const cost of this.finalCosts) {
                if (cost.majorPrice) {
                    num++;
                    if (this.isNumber(cost.value)) {
                        this.hasNumberofMajorprice = true;
                    }
                }
            }
            return num;
        }
        return 0;
    }
    // 判断主/次要价格是否都隐藏
    isCostsAllHide() {
        if (this.costs && this.costs.length > 0) {
            // 获取主要价格集合
            const majorPriceArr = this.costs.filter((item) => item.majorPrice);
            // 获取次要价格集合
            const minorPriceArr = this.costs.filter((item) => !item.majorPrice);
            this.isMinorPriceAllHide = this.isAllHide(minorPriceArr);
            this.isMajorPriceAllHide = this.isAllHide(majorPriceArr);
        }
    }
    isAllHide(arr) {
        let isAllHide = true;
        if (arr.length > 0) {
            for (const cost of arr) {
                if (!cost.hide) {
                    isAllHide = false;
                }
            }
        }
        return isAllHide;
    }
    // 设置价格类型
    setMajorPrice() {
        if (this.costs && this.costs.length > 0) {
            for (const cost of this.costs) {
                if (Util.isUndefined(cost.label) && !this.isRefund) {
                    cost.label = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_label;
                    cost.tipTitle = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_label;
                }
                if (Util.isUndefined(cost.majorPrice)) {
                    cost.majorPrice = true;
                }
            }
        }
    }
    /**
     * 设置退费场景下的相关参数
     */
    setRefundParams() {
        if (this.costs && this.costs.length > 0) {
            for (const cost of this.costs) {
                if (Util.isUndefined(cost.label)) {
                    cost.label = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_label_refund_majorprice;
                    cost.tipTitle = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_label_refund_detail;
                }
                if (cost.pricelist && cost.pricelist.length > 0) {
                    cost.pricelist[0].title = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_pricelist_label_before;
                    cost.pricelist[1].title = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_pricelist_label_after;
                    cost.pricelist[2] = {
                        title: TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_label_refund,
                        priceNum: cost.value
                    };
                    this.pricelistHead = TiLocale.getLocaleWords().tpBuylayer.buyLayer_cost_type;
                }
            }
        }
    }
    /**
     * @ignore
     */
    isNumber(val) {
        return Util.isNumber(val);
    }
    /**
     * @ignore
     */
    isPositiveNumber(val) {
        return Util.isNumber(val) && val > 0;
    }
    /**
     * @ignore
     * 货币国际化
     */
    getLocalSymble() {
        const localeKey = TiLocale.getLocaleWords().tiLocaleKey;
        return getLocaleCurrencySymbol(localeKey);
    }
    // 获取agencyID和locale
    addLinkParams(url) {
        if (url === '') {
            return url;
        }
        const storage = CfUtil.getStorageService();
        if (!(!Util.isUndefined(storage) && !Util.isUndefined(storage.cookieStorage) && Util.isFunction(storage.cookieStorage.getItem))) {
            return url;
        }
        const agencyID = storage.cookieStorage.getItem('agencyID');
        const locale = storage.cookieStorage.getItem('locale');
        let finalUrl = url;
        if (!Util.isEmptyString(agencyID)) {
            finalUrl = finalUrl + '?agencyId=' + agencyID;
            if (!Util.isEmptyString(locale)) {
                finalUrl = finalUrl + '&locale=' + locale;
            }
        }
        else if (!Util.isEmptyString(locale)) {
            finalUrl = finalUrl + '?locale=' + locale;
        }
        return finalUrl;
    }
}
TpBuylayerLeftComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-buylayer-left',
                template: "<div class='tp-buylayer-font-main'>\r\n  <!-- \u4EF7\u683C\u5217\u8868\u90E8\u5206\u5C55\u793A\uFF0C\u8DDF\u4EF7\u683C\u8BA1\u7B97\u5668\u90E8\u5206\u4E0D\u80FD\u540C\u65F6\u5B58\u5728 -->\r\n  <div class='tp-buyLayer-price-container' *ngIf='costs'>\r\n      <div *ngIf='leftZoneEle1'>\r\n          <ng-container *ngTemplateOutlet='leftZoneEle1'></ng-container>\r\n      </div>\r\n      <div>\r\n          <ng-container *ngTemplateOutlet='majorPrice'></ng-container>\r\n          <ng-container *ngTemplateOutlet='minorPrice'></ng-container>\r\n          <!--\u662F\u5426\u663E\u793A\u6298\u6263\u529B\u51CF-->\r\n          <div *ngIf='showFreechargeDetail'>\r\n            <tp-icon local class='tp-buyLayer-freechargeDetail-icon' name='cloud-action-gift'></tp-icon>\r\n            <span class=\"tp-buylayer-freecharge-label\">{{'tpBuylayer.buyLayer_freeCharge_detail' | tiTranslate}}</span>\r\n            <a [href]='freechargeDetailLink' class='tp-buylayer-link' target=\"_blank\" rel=\"noopener noreferrer\">{{'tpBeta.more' | tiTranslate}}</a>\r\n          </div>\r\n      </div>\r\n  </div>\r\n  <!-- \u4EF7\u683C\u8BA1\u7B97\u5668\u5C55\u793A \u53EA\u5728\u7279\u6B8A\u5C40\u70B9\uFF08\u5FB7\u7535\u3001\u6CD5\u7535\uFF09\u624D\u5C55\u793A -->\r\n  <div class='tp-buyLayer-price-cal'>\r\n      <ng-container *ngTemplateOutlet='pricecalc'></ng-container>\r\n  </div>\r\n</div>\r\n\r\n<!-- tip\u5185\u5BB9\u6A21\u677F\uFF0C\u5176\u4E2Dlet-\u5B9A\u4E49\u7684\u662F\u6A21\u677F\u4E2D\u7528\u5230\u7684\u4E0A\u4E0B\u6587 -->\r\n<ng-template #tipContent  let-context='context'>\r\n    <div *ngIf='context.pricelist && context.pricelist.length' class='tp-buylayer-pricelist-tip'>\r\n        <span class=\"tp-buylayer-pricelist-title\">{{context.tipTitle}}</span>\r\n        <ul style=\"padding:0\" [id]=\"appendId('list')\">\r\n            <li class=\"tp-buylayer-pricelist-head\">\r\n                <span>{{pricelistHead}}</span>\r\n                <span style=\"float:right\">{{'tpBuylayer.buyLayer_cost_price' | tiTranslate}}</span>\r\n            </li>\r\n            <li class=\"tp-buylayer-pricelist\" *ngFor='let price of context.pricelist; last as lastItem;'>\r\n              <span>{{price.title}}</span>\r\n              <span class=\"tp-buylayer-pricelist-item\" style='float:right'>\r\n                <span [ngClass]=\"{'tp-buylayer-priceNum-refund': isRefund,\r\n                'tp-buylayer-priceNum-refund-last': lastItem && isRefund}\">{{price.priceNum | tpCurrency: numFormat}}</span>\r\n                <span *ngIf=\"context.measureUnit\">/{{context.measureUnit}}</span>\r\n              </span>\r\n            </li>\r\n        </ul>\r\n      </div>\r\n</ng-template>\r\n<!-- \u95EE\u53F7\u56FE\u6807\u63D0\u793A\u6A21\u677F\uFF084.0\u89C4\u8303\u7684\u8BA1\u8D39\u8BE6\u60C5\uFF09 -->\r\n<ng-template #detailTip>\r\n    <span *ngIf='!costExplain'>{{'tpBuylayer.buyLayer_costExplain_label' | tiTranslate}}</span>\r\n    <span *ngIf='costExplain'>{{costExplain}}</span>\r\n    <!--\u6709\u7684\u4E00\u53E3\u4EF7\u5546\u54C1\uFF0C\u4E0D\u5B58\u5728\u8D85\u94FE\u63A5\uFF0C\u6240\u4EE5\u9700\u8981\u652F\u6301\u9690\u85CF-->\r\n    <a target=\"_blank\" *ngIf=\"detailHref || (costs && costs[0] && costs[0].link)\" [href]='detailHref || (costs && costs[0] && costs[0].link)' rel=\"noopener noreferrer\">{{'tpBuylayer.buyLayer_cost_linklabel' | tiTranslate}}</a>\r\n</ng-template>\r\n<!-- \u4E3B\u8981\u4EF7\u683C\u90E8\u5206\u6A21\u677F -->\r\n<ng-template #majorPrice>\r\n  <div *ngIf=\"!isMajorPriceAllHide\">\r\n    <div style='display: inline-block' *ngFor='let cost of costs | isMajorPrice;'>\r\n      <span *ngIf='!cost.hide'>\r\n        <tp-icon local *ngIf='!cost.hideConnectSymbol' name='cloud-action-add6' class='tp-buylayer-add'></tp-icon>\r\n        <span class='tp-buylayer-label'>{{cost.label}}</span>\r\n        <span *ngIf='cost.isBeta' class='tp-buylayer-beta-label'>{{'tpBuylayer.buyLayer_betaFree_label' | tiTranslate}}</span>\r\n        <span *ngIf='!cost.isBeta'>\r\n          <span *ngIf='!isNumber(cost.value)' [innerHTML]=\"cost.value\"></span>\r\n          <ng-container *ngIf='isNumber(cost.value)'>\r\n            <span *ngIf='cost.pricelist && cost.pricelist.length > 0' [ngClass]=\"{'tp-buyLayer-priceWithTip': cost.pricelist}\" [tiTip]='tipContent' [tiTipContext]='cost'>\r\n              <span class='tp-buylayer-majorprice-value'  [ngClass]=\"{'tp-buylayer-majorprice-refund': isRefund}\">{{cost.value | tpCurrency: numFormat}}</span>\r\n              <span *ngIf='cost.measureUnit'>/{{cost.measureUnit}}</span>\r\n            </span>\r\n            <span  *ngIf='!cost.pricelist || cost.pricelist.length === 0'>\r\n              <span class='tp-buylayer-majorprice-value'  [ngClass]=\"{'tp-buylayer-majorprice-refund': isRefund}\">{{cost.value | tpCurrency: numFormat}}</span>\r\n              <span *ngIf='cost.measureUnit'>/{{cost.measureUnit}}</span>\r\n            </span>\r\n            <span *ngIf='cost.discount && isPositiveNumber(cost.discount)' class='tp-buylayer-discount'>\r\n              <span>({{'tpBuylayer.buyLayer_priceDiscount_label' | tiTranslate}}</span>\r\n              <span>{{cost.discount | tpCurrency: numFormat}}</span>\r\n              <span *ngIf='cost.measureUnit'>/{{cost.measureUnit}}</span>\r\n              <span>)</span>\r\n            </span>\r\n          </ng-container>\r\n        </span>\r\n      </span>\r\n    </div>\r\n    <span *ngIf=\"showMajorDiscount\" class=\"tp-buylayer-discounts-container\">\r\n      <tp-discounts [discoutsConfig]='majorDiscoutsConfig' (discountsChange)='change($event)' [id]=\"appendId('major_discounts')\"></tp-discounts>\r\n    </span>\r\n    <tp-icon local *ngIf='onlyHasMajorprice && hasNumberofMajorprice' [tiTip]='detailTip' [tiTipMaxWidth]='detailTipMaxWidth' name=\"cloud-action-tip\" class='tp-buylayer-help' [id]=\"appendId('major_helptip')\"></tp-icon>\r\n  </div>\r\n</ng-template>\r\n<!-- \u6B21\u8981\u4EF7\u683C\u90E8\u5206\u6A21\u677F -->\r\n<ng-template #minorPrice>\r\n    <div *ngIf=\"!isMinorPriceAllHide\">\r\n        <div style='display: inline-block' *ngFor='let cost of costs | isMinorPrice;'>\r\n          <span *ngIf='!cost.hide'>\r\n            <tp-icon local name='cloud-action-add6' *ngIf='!cost.hideConnectSymbol' class='tp-buylayer-add' [ngClass]=\"{'tp-buylayer-add-first': cost.hideConnectSymbol}\"></tp-icon>\r\n            <span class='tp-buylayer-label'>{{cost.label}}</span>\r\n            <span *ngIf='cost.isBeta' class='tp-buylayer-beta-label'>{{'tpBuylayer.buyLayer_betaFree_label' | tiTranslate}}</span>\r\n            <span *ngIf='!cost.isBeta'>\r\n              <span *ngIf='!isNumber(cost.value)' [innerHTML]=\"cost.value\"></span>\r\n              <ng-container *ngIf='isNumber(cost.value)'>\r\n                <span *ngIf='cost.pricelist && cost.pricelist.length > 0' [ngClass]=\"{'tp-buyLayer-priceWithTip': cost.pricelist}\" [tiTip]='tipContent' [tiTipContext]='cost'>\r\n                  <span class='tp-buylayer-minorprice-value' [ngClass]=\"{'tp-buylayer-minorprice-refund': isRefund}\">{{cost.value | tpCurrency: numFormat}}</span>\r\n                  <span *ngIf='cost.measureUnit' class='tp-buylayer-minorprice-value' [ngClass]=\"{'tp-buylayer-minorprice-refund': isRefund}\">/{{cost.measureUnit}}</span>\r\n                </span>\r\n                <span *ngIf='!cost.pricelist || cost.pricelist.length === 0'>\r\n                  <span class='tp-buylayer-minorprice-value' [ngClass]=\"{'tp-buylayer-minorprice-refund': isRefund}\">{{cost.value | tpCurrency: numFormat}}</span>\r\n                  <span *ngIf='cost.measureUnit' class='tp-buylayer-minorprice-value' [ngClass]=\"{'tp-buylayer-minorprice-refund': isRefund}\">/{{cost.measureUnit}}</span>\r\n                </span>\r\n                <span *ngIf='cost.discount && isPositiveNumber(cost.discount)' class='tp-buylayer-discount'>\r\n                  <span>({{'tpBuylayer.buyLayer_priceDiscount_label' | tiTranslate}}</span>\r\n                  <span>{{cost.discount | tpCurrency: numFormat}}</span>\r\n                  <span *ngIf='cost.measureUnit'>/{{cost.measureUnit}}</span>\r\n                  <span>)</span>\r\n                </span>\r\n              </ng-container>\r\n            </span>\r\n          </span>\r\n        </div>\r\n        <span *ngIf=\"showMinorDiscount\" class=\"tp-buylayer-discounts-container\">\r\n          <tp-discounts [discoutsConfig]='minorDiscoutsConfig' [id]=\"appendId('minor_discounts')\"></tp-discounts>\r\n        </span>\r\n        <tp-icon local *ngIf='!onlyHasMajorprice && finalCosts.length > 0' [tiTip]='detailTip' [tiTipMaxWidth]='detailTipMaxWidth' name=\"cloud-action-tip\" class='tp-buylayer-help' [id]=\"appendId('minor_helptip')\"></tp-icon>\r\n      </div>\r\n  </ng-template>\r\n  <!-- \u4EF7\u683C\u8BA1\u7B97\u5668\u6A21\u677F -->\r\n<ng-template #pricecalc>\r\n    <tp-pricecalc *ngIf='calculator &&calculator.url' [url] = 'calculator.url' [id]=\"appendId('pricecalc')\"\r\n    [values]='calculator.values' [method]='calculator.method' [detailHref]='calculator.detailHref'></tp-pricecalc>\r\n    <div *ngIf='!costExplain && (detailHref || costs && costs[0] && costs[0].link)'>\r\n      <span>{{'tpBuylayer.buyLayer_costDetail_label.text_before_link' | tiTranslate}}</span>\r\n      <a target='_blank' class='tp-buylayer-link' [href]='detailHref || costs && costs[0] && costs[0].link' rel=\"noopener noreferrer\">{{'tpBuylayer.buyLayer_costDetail_label.link_info' | tiTranslate}}</a>\r\n      <span>{{'tpBuylayer.buyLayer_costDetail_label.text_after_link' | tiTranslate}}</span>\r\n    </div>\r\n    <div *ngIf='costExplain'>{{costExplain}}</div>\r\n</ng-template>\r\n\r\n",
                styles: [".tp-buylayer-font-main{color:var(--ti-common-color-text-secondary);font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number)}::ng-deep .tp-buyLayer-price-container{align-items:center;display:flex;justify-content:space-between}.tp-buyLayer-freechargeDetail-icon{color:var(--ti-common-color-warn)}.tp-buylayer-freecharge-label{padding-left:var(--ti-common-space-base)}.tp-buylayer-link{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-buylayer-link:hover{color:var(--ti-common-color-text-link-hover);text-decoration:underline}::ng-deep .tp-buyLayer-price-cal{display:none}.tp-buylayer-pricelist-tip{color:var(--ti-common-color-text-gray);display:inline-block;font-size:var(--ti-common-font-size-base);min-width:240px}.tp-buylayer-pricelist-title{display:inline-block;font-size:var(--ti-common-font-size-1);line-height:var(--ti-common-line-height-number);padding-bottom:var(--ti-common-space-2x)}.tp-buylayer-pricelist-head{background:hsla(0,0%,100%,.05);height:28px;line-height:28px}.tp-buylayer-pricelist,.tp-buylayer-pricelist-head{border-bottom:1px solid hsla(0,0%,100%,.1);padding:0 var(--ti-common-space-2x)}.tp-buylayer-pricelist{height:34px;line-height:34px}.tp-buylayer-pricelist-item{color:var(--ti-common-color-error);font-weight:700}.tp-buylayer-priceNum-refund{color:var(--ti-common-color-text-gray);font-weight:400}.tp-buylayer-priceNum-refund-last{color:var(--ti-common-color-text-success)}.tp-buylayer-add{color:var(--ti-common-color-icon-graybg-normal);padding:0 var(--ti-common-space-base)}.tp-buylayer-add-first{padding-left:0}.tp-buylayer-beta-label{color:var(--ti-common-bg-primary-active);font-size:var(--ti-common-font-size-4);line-height:30px}.tp-buylayer-label{padding-right:var(--ti-common-space-2x)}.tp-buyLayer-priceWithTip{border-bottom:1px dashed var(--ti-common-color-text-secondary)}.tp-buylayer-majorprice-value{color:var(--ti-common-color-error-text);font-size:var(--ti-common-font-size-4);line-height:30px}.tp-buylayer-minorprice-value{color:var(--ti-common-color-text-primary);font-weight:700}.tp-buylayer-discount{padding-left:var(--ti-common-space-base)}.tp-buylayer-majorprice-refund{color:var(--ti-common-color-text-success)}.tp-buylayer-minorprice-refund{color:var(--ti-common-color-error-text)}.tp-buylayer-discounts-container{border-left:1px solid var(--ti-common-color-line-dividing);padding-left:var(--ti-common-space-3x)}.tp-buylayer-discounts-container,.tp-buylayer-help{line-height:normal;margin-left:var(--ti-common-space-3x)}.tp-buylayer-help{display:inline-block;font-size:var(--ti-common-font-size-2);vertical-align:text-bottom}.tp-buylayer-help:hover{color:var(--ti-common-color-icon-hover);cursor:pointer}"]
            },] }
];
TpBuylayerLeftComponent.propDecorators = {
    costs: [{ type: Input }],
    majorDiscoutsConfig: [{ type: Input }],
    minorDiscoutsConfig: [{ type: Input }],
    showFreechargeDetail: [{ type: Input }],
    costExplain: [{ type: Input }],
    calculator: [{ type: Input }],
    numFormat: [{ type: Input }],
    detailHref: [{ type: Input }],
    isRefund: [{ type: Input }],
    detailTipMaxWidth: [{ type: Input }],
    discountsChange: [{ type: Output }],
    leftZoneEle1: [{ type: ContentChild, args: [TemplateRef,] }]
};
