// @cloud/tinyplus3/esm2015/components/buylayer/TpBuylayerModule.js
import { TpPricecalcModule } from '../pricecalc/TpPricecalcModule';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TpBuylayerComponent } from './TpBuylayerComponent';
import { TpBuylayerLeftComponent } from './TpBuylayerLeftComponent';
import { TpBuylayerRightComponent } from './TpBuylayerRightComponent';
import { TiButtonModule, TiLocaleModule, TiTipModule } from '@cloud/tiny3';
import { TpIconModule } from '../icon/TpIconModule';
import { FormsModule } from '@angular/forms';
import { TpDiscountsModule } from '../discounts/TpDiscountsModule';
import { IsMajorPricePipe, IsMinorPricePipe } from './IsMajorPrice';
import { TpLocaleModule } from '../../pipes/locale/TpLocaleModule';
export class TpBuylayerModule {
}
TpBuylayerModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, FormsModule, TiButtonModule,
                    TpDiscountsModule, TpIconModule, TpLocaleModule, TiLocaleModule, TiTipModule, TpPricecalcModule],
                exports: [TpBuylayerComponent, TpBuylayerLeftComponent, TpBuylayerRightComponent],
                declarations: [IsMajorPricePipe, IsMinorPricePipe, TpBuylayerComponent, TpBuylayerLeftComponent, TpBuylayerRightComponent]
            },] }
];
export { TpBuylayerComponent } from './TpBuylayerComponent';
export { TpBuylayerRightComponent } from './TpBuylayerRightComponent';
export { TpBuylayerLeftComponent } from './TpBuylayerLeftComponent';
