// @cloud/tinyplus3/esm2015/components/buylayer/TpBuylayerComponent.js
import { Component, ContentChild, ElementRef, EventEmitter, Input, NgZone, Output, Renderer2, TemplateRef, ViewChild } from '@angular/core';
import { TiBaseComponent, TiBrowser } from '@cloud/tiny3';
/**
 * 购买浮层组件
 *
 * <example-url>../tinyplus3demo/#/buylayer/buylayer-all</example-url>
 */
export class TpBuylayerComponent extends TiBaseComponent {
    constructor(element, renderer2, ngZone) {
        super(element, renderer2);
        this.element = element;
        this.renderer2 = renderer2;
        this.ngZone = ngZone;
        /**
         * 优惠组件可选优惠项切换选择项时触发
         *
         * 9.0.3 新增接口
         */
        this.discountsChange = new EventEmitter();
    }
    // tslint:disable-next-line:use-life-cycle-interface
    ngAfterViewInit() {
        // ie下兼容问题，免费使用场景弹性布局垂直居中
        if (TiBrowser.isIE()) {
            const contentContainerHeight = this.buylayerContentEle.nativeElement.offsetHeight;
            if (contentContainerHeight <= 70) { // 浮层最小高度70px
                this.renderer2.setStyle(this.buylayerContentEle.nativeElement, 'height', 70 + 'px');
            }
        }
        this.ngZone.runOutsideAngular(() => {
            // 在横向滚动条出现时控制fixed定位的购买浮层可以左右移动
            this.unlistenWindowScroll = this.renderer2.listen('window', 'scroll', () => {
                this.renderer2.setStyle(this.element.nativeElement, 'left', `-${document.documentElement.scrollLeft || document.body.scrollLeft}px`);
            });
        });
    }
    ngOnDestroy() {
        if (this.unlistenWindowScroll) {
            this.unlistenWindowScroll();
        }
    }
    /**
     * @ignore
     */
    change($event) {
        this.discountsChange.emit($event);
    }
}
TpBuylayerComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-buylayer',
                template: "\r\n<div class='tp-buylayer-content' #buylayerContent>\r\n    <tp-buylayer-left\r\n    [costs]='buylayerConfig.costs'\r\n    [majorDiscoutsConfig] = 'buylayerConfig.majorDiscoutsConfig'\r\n    [minorDiscoutsConfig] = 'buylayerConfig.minorDiscoutsConfig'\r\n    [showFreechargeDetail] = 'buylayerConfig.showFreechargeDetail'\r\n    [costExplain] = 'buylayerConfig.costExplain'\r\n    [detailHref] = 'buylayerConfig.detailHref'\r\n    [numFormat] = 'buylayerConfig.numFormat'\r\n    [calculator] = 'buylayerConfig.calculator'\r\n    [isRefund] = 'buylayerConfig.isRefund'\r\n    [detailTipMaxWidth] = 'buylayerConfig.detailTipMaxWidth'\r\n    (discountsChange)='change($event)'\r\n    [id]=\"appendId('buylayer_left')\">\r\n    <ng-template>\r\n        <ng-container  *ngTemplateOutlet='leftZoneEle'></ng-container>\r\n    </ng-template>\r\n</tp-buylayer-left>\r\n<tp-buylayer-right [buttons]='buylayerConfig.buttons'></tp-buylayer-right>\r\n</div>\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n\r\n",
                host: {
                    '[class.tp-buylayer-container]': 'true'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-console-left-width:48px}:host.tp-buylayer-container{background:var(--ti-common-color-bg-white-normal);bottom:0;box-shadow:var(--ti-common-shadow-1-up);box-sizing:border-box;left:0;min-height:var(--tp-buylayer-container-min-height);min-width:calc(var(--tp-min-width) + var(--tp-console-left-width) + var(--ti-common-space-5x)*2);padding:10px 0 10px var(--tp-console-left-width);position:fixed;width:100%;z-index:10}.tp-buylayer-content,:host.tp-buylayer-container{align-items:center;display:flex;justify-content:space-between}:host.tp-buylayer-aligncenter .tp-buylayer-content{margin:0 auto;width:var(--tp-min-width)}@media screen and (min-width:1920px){:host.tp-buylayer-container:not(.tp-buylayer-aligncenter) .tp-buylayer-content{margin:0 auto;width:var(--tp-max-width)}}@media screen and (max-width:1920px){:host.tp-buylayer-container:not(.tp-buylayer-aligncenter) .tp-buylayer-content{margin:0 var(--ti-common-space-5x);width:100%}}"]
            },] }
];
TpBuylayerComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: NgZone }
];
TpBuylayerComponent.propDecorators = {
    buylayerConfig: [{ type: Input }],
    discountsChange: [{ type: Output }],
    leftZoneEle: [{ type: ContentChild, args: ['leftZone',] }],
    buylayerContentEle: [{ type: ViewChild, args: ['buylayerContent', { static: false },] }]
};
