// @cloud/tiny3/esm2015/directives/validation/checkHandle/AsyncCheck.js
import { Util } from '../../../utils/Util';
import { CheckHandle } from './CheckHandle';
import { TiPendingStateComponent } from '../TiPendingStateComponent';
/**
 * @ignore
 *
 * 10.1.0新增
 */
export class AsyncCheck extends CheckHandle {
    // 处理异步校验错误提示信息
    onStatusChange(ele, formControl) {
        // 清除先前校验信息
        this.commonHandle.clearValidMsg(ele);
        const errors = formControl.control.errors;
        if (!errors) {
            return;
        }
        const ruleKey = Object.keys(errors)[0];
        const ruleErrors = errors[ruleKey];
        // 添加异步校验错误提示信息
        if (typeof ruleErrors === 'object' && Util.isString(ruleErrors.tiAsyncErrorMessage)) {
            this.commonHandle.addValidMsg(ele, this.validationConf, formControl, true);
        }
    }
    // 处理异步校验pending状态时的loading图标
    onAsyncStatusChange(ele, formControl) {
        if (formControl.pending) {
            this.addPendingIcon(ele, formControl);
        }
        else {
            this.clearPendingIcon(ele);
        }
    }
    addPendingIcon(ele, formControl) {
        if (ele.nativeElement.tiPendingStateRef) {
            return;
        }
        const pendingStateRef = this.getPendingStateRef(ele);
        const pendingDom = pendingStateRef.location.nativeElement;
        if (pendingDom === null || pendingDom.childNodes.length === 0) {
            return;
        }
        this.commonHandle._tiRenderer.insertAfter(pendingDom, ele.nativeElement);
        ele.nativeElement.tiPendingStateRef = pendingStateRef;
    }
    clearPendingIcon(ele) {
        const pendingStateRef = ele.nativeElement.tiPendingStateRef;
        if (pendingStateRef) {
            pendingStateRef.destroy();
            ele.nativeElement.tiPendingStateRef = undefined;
        }
    }
    getPendingStateRef(ele) {
        const pendignComponentRef = this.commonHandle._tiPopupService.createCompoentRef({
            componentType: TiPendingStateComponent,
            context: {
                validElement: ele.nativeElement
            }
        });
        return pendignComponentRef;
    }
}
