// @cloud/tinyplus3/esm2015/components/supportregionlist/TpSupportregionlistComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * @ignore
 * 内部使用组件
 */
export class TpSupportregionlistComponent extends TiBaseComponent {
    /**
     * @ignore
     */
    onClick(item) {
        if (!this.selectedRegionEvent) {
            this.selectedRegionEvent = document.createEvent('HTMLEvents');
            this.selectedRegionEvent.initEvent('cfChangeRegion', true, true);
        }
        this.selectedRegionEvent.detail = item;
        document.dispatchEvent(this.selectedRegionEvent);
    }
    /**
     * @ignore
     */
    onClickIcon(event, item) {
        item.showProjectsList = !item.showProjectsList;
    }
}
TpSupportregionlistComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-supportregionlist',
                template: "<span *ngIf=\"regions && regions.length > 0\" class=\"tp-supportregionlist-label\">{{ 'tpNonsupportregion.supportLabel' | tiTranslate }}</span>\r\n<ul class=\"tp-supportregionlist-regions\" [id]=\"appendId('list')\">\r\n    <ng-container *ngFor=\"let region of regions; index as i\">\r\n        <!-- \u4E00\u7EA7\u533A\u57DF -->\r\n        <ng-container *ngTemplateOutlet=\"regionTemplate; context: {$implicit: region, liId: i, show: true}\"></ng-container>\r\n        <!-- \u4E8C\u7EA7\u533A\u57DF -->\r\n        <ng-container *ngFor=\"let project of region.projects; index as j;\">\r\n            <ng-container *ngTemplateOutlet=\"regionTemplate; context: {$implicit: project, show: region.showProjectsList}\"></ng-container>\r\n        </ng-container>\r\n    </ng-container>\r\n</ul>\r\n\r\n<ng-template #regionTemplate let-item let-liId=liId let-show=show>\r\n        <li *ngIf=\"show\" class=\"tp-supportregionlist-project\" [id]=\"appendId(liId)\">\r\n            <a (click)=\"onClick(item)\" [id]=\"appendId('name' + liId)\">{{item.displayName}}</a>\r\n            <section *ngIf=\"item.projects && item.projects.length!==0\"\r\n                      class=\"tp-supportregionlist-icon\"\r\n                      [ngClass]=\"{'tp-supportregionlist-icon-transform': item.showProjectsList}\" (click)=\"onClickIcon($event, item)\" [id]=\"appendId('icon' + liId)\"></section>\r\n        </li>\r\n</ng-template>",
                styles: [":host{display:flex;justify-content:center;line-height:var(--ti-common-line-height-number);margin-bottom:40px}.tp-supportregionlist-label{margin-right:var(--ti-common-space-2x)}.tp-supportregionlist-regions{text-align:left}.tp-supportregionlist-project{color:var(--ti-common-color-text-link);cursor:pointer;margin-bottom:var(--ti-common-space-base)}.tp-supportregionlist-project:hover{color:var(--ti-common-color-text-link-hover)}.tp-supportregionlist-project:hover section:after{border-top-color:var(--ti-common-color-text-link-hover)}.tp-supportregionlist-icon{display:inline-block;padding-left:var(--ti-common-space-2x)}.tp-supportregionlist-icon:after{border-left:4px solid transparent;border-right:4px solid transparent;border-top:5px solid var(--ti-common-color-text-link);content:\"\";cursor:pointer;display:inline-block;height:0;vertical-align:middle;width:0}.tp-supportregionlist-icon-transform:after{transform:rotate(180deg)}"]
            },] }
];
TpSupportregionlistComponent.propDecorators = {
    regions: [{ type: Input }]
};
