// @cloud/cfdata/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.version = void 0;
// 数据
var AccountData_1 = require("./AccountData");
Object.defineProperty(exports, "AccountData", { enumerable: true, get: function () { return AccountData_1.AccountData; } });
var ConfigData_1 = require("./ConfigData");
Object.defineProperty(exports, "ConfigData", { enumerable: true, get: function () { return ConfigData_1.ConfigData; } });
var CfData_1 = require("./CfData"); // TODO: to delete
Object.defineProperty(exports, "CfData", { enumerable: true, get: function () { return CfData_1.CfData; } });
var EndpointsData_1 = require("./EndpointsData");
Object.defineProperty(exports, "EndpointsData", { enumerable: true, get: function () { return EndpointsData_1.EndpointsData; } });
var FrameData_1 = require("./FrameData");
Object.defineProperty(exports, "FrameData", { enumerable: true, get: function () { return FrameData_1.FrameData; } });
var LinksData_1 = require("./LinksData");
Object.defineProperty(exports, "LinksData", { enumerable: true, get: function () { return LinksData_1.LinksData; } });
var ProjectData_1 = require("./ProjectData");
Object.defineProperty(exports, "ProjectData", { enumerable: true, get: function () { return ProjectData_1.ProjectData; } });
var RegionData_1 = require("./RegionData");
Object.defineProperty(exports, "RegionData", { enumerable: true, get: function () { return RegionData_1.RegionData; } });
var UserData_1 = require("./UserData");
Object.defineProperty(exports, "UserData", { enumerable: true, get: function () { return UserData_1.UserData; } });
// 接口
var CfAcount_1 = require("./interface/CfAcount");
Object.defineProperty(exports, "CfAcount", { enumerable: true, get: function () { return CfAcount_1.CfAcount; } });
var CfEndpoints_1 = require("./interface/CfEndpoints");
Object.defineProperty(exports, "CfEndpoints", { enumerable: true, get: function () { return CfEndpoints_1.CfEndpoints; } });
var CfFrame_1 = require("./interface/CfFrame");
Object.defineProperty(exports, "CfFrame", { enumerable: true, get: function () { return CfFrame_1.CfFrame; } });
var CfLinks_1 = require("./interface/CfLinks");
Object.defineProperty(exports, "CfLinks", { enumerable: true, get: function () { return CfLinks_1.CfLinks; } });
var CfProject_1 = require("./interface/CfProject");
Object.defineProperty(exports, "CfProject", { enumerable: true, get: function () { return CfProject_1.CfProject; } });
var CfUser_1 = require("./interface/CfUser");
Object.defineProperty(exports, "CfUser", { enumerable: true, get: function () { return CfUser_1.CfUser; } });
var HttpRequest_1 = require("./interface/HttpRequest");
Object.defineProperty(exports, "HttpRequest", { enumerable: true, get: function () { return HttpRequest_1.HttpRequest; } });
// service & util
var AccountService_1 = require("./service/AccountService");
Object.defineProperty(exports, "AccountService", { enumerable: true, get: function () { return AccountService_1.AccountService; } });
var ApiService_1 = require("./service/ApiService");
Object.defineProperty(exports, "ApiService", { enumerable: true, get: function () { return ApiService_1.ApiService; } });
var EndpointService_1 = require("./service/EndpointService");
Object.defineProperty(exports, "EndpointService", { enumerable: true, get: function () { return EndpointService_1.EndpointService; } });
var FrameworkService_1 = require("./service/FrameworkService");
Object.defineProperty(exports, "FrameworkService", { enumerable: true, get: function () { return FrameworkService_1.FrameworkService; } });
var IncidentService_1 = require("./service/IncidentService");
Object.defineProperty(exports, "IncidentService", { enumerable: true, get: function () { return IncidentService_1.IncidentService; } });
var MessageService_1 = require("./service/MessageService");
Object.defineProperty(exports, "MessageService", { enumerable: true, get: function () { return MessageService_1.MessageService; } });
var RegionService_1 = require("./service/RegionService");
Object.defineProperty(exports, "RegionService", { enumerable: true, get: function () { return RegionService_1.RegionService; } });
var HttpService_1 = require("./service/HttpService");
Object.defineProperty(exports, "HttpService", { enumerable: true, get: function () { return HttpService_1.HttpService; } });
var StorageService_1 = require("./service/StorageService");
Object.defineProperty(exports, "StorageService", { enumerable: true, get: function () { return StorageService_1.StorageService; } });
var IndexDBService_1 = require("./service/IndexDBService");
Object.defineProperty(exports, "cfApiCacheCtrl", { enumerable: true, get: function () { return IndexDBService_1.cfApiCacheCtrl; } });
var TraceService_1 = require("./service/TraceService");
Object.defineProperty(exports, "TraceService", { enumerable: true, get: function () { return TraceService_1.TraceService; } });
var UtilService_1 = require("./service/UtilService");
Object.defineProperty(exports, "UtilService", { enumerable: true, get: function () { return UtilService_1.UtilService; } });
var UserService_1 = require("./service/UserService");
Object.defineProperty(exports, "UserSerivce", { enumerable: true, get: function () { return UserService_1.UserSerivce; } });
var Analysis_1 = require("./service/Analysis");
Object.defineProperty(exports, "Analysis", { enumerable: true, get: function () { return Analysis_1.Analysis; } });
var TinyReporter_1 = require("./service/TinyReporter");
Object.defineProperty(exports, "TinyReporter", { enumerable: true, get: function () { return TinyReporter_1.TinyReporter; } });
// component
var MaskComponent_1 = require("./component/MaskComponent");
Object.defineProperty(exports, "MaskComponent", { enumerable: true, get: function () { return MaskComponent_1.MaskComponent; } });
Object.defineProperty(exports, "Mask", { enumerable: true, get: function () { return MaskComponent_1.MaskComponent; } });
var MessageComponent_1 = require("./component/MessageComponent");
Object.defineProperty(exports, "MessageComponent", { enumerable: true, get: function () { return MessageComponent_1.MessageComponent; } });
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return MessageComponent_1.MessageComponent; } });
var Version_1 = require("./Version");
Object.defineProperty(exports, "version", { enumerable: true, get: function () { return Version_1.version; } });
window.cfDataVersion = window.cfDataVersion
    ? window.cfDataVersion.includes(Version_1.version)
        ? window.cfDataVersion
        : window.cfDataVersion.concat(Version_1.version)
    : [Version_1.version];
