// @cloud/cfdata/ProjectData.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectData = void 0;
var CfProject_1 = require("./interface/CfProject");
var UserData_1 = require("./UserData");
var ConfigData_1 = require("./ConfigData");
var UserService_1 = require("./service/UserService");
var RegionService_1 = require("./service/RegionService");
var RegionData_1 = require("./RegionData");
var ProjectData = /** @class */ (function () {
    function ProjectData() {
    }
    /**
     * Gets cf project
     * @returns cf project
     */
    ProjectData.getCfProject = function () {
        return __awaiter(this, void 0, void 0, function () {
            var cfUser;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!ProjectData.cfProject) return [3 /*break*/, 2];
                        return [4 /*yield*/, UserData_1.UserData.getCfUser()];
                    case 1:
                        cfUser = _a.sent();
                        ProjectData.cfProject = new CfProject_1.CfProject();
                        // 设置用户数据并初始化
                        ProjectData.initProject(cfUser);
                        _a.label = 2;
                    case 2: return [2 /*return*/, ProjectData.cfProject];
                }
            });
        });
    };
    ProjectData.getProjects = function () {
        return __awaiter(this, void 0, void 0, function () {
            var user, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!!ProjectData.projectList) return [3 /*break*/, 4];
                        if (!!ProjectData.projectPromise) return [3 /*break*/, 2];
                        return [4 /*yield*/, UserData_1.UserData.getCfUser()];
                    case 1:
                        user = _b.sent();
                        ProjectData.projectPromise = UserService_1.UserSerivce.getUserProjects({ userId: user.userId });
                        _b.label = 2;
                    case 2:
                        _a = ProjectData;
                        return [4 /*yield*/, ProjectData.projectPromise];
                    case 3:
                        _a.projectList = ((_b.sent()) || {}).projects || [];
                        _b.label = 4;
                    case 4: return [2 /*return*/, ProjectData.projectList];
                }
            });
        });
    };
    ProjectData.initProject = function (userData) {
        return __awaiter(this, void 0, void 0, function () {
            var options, data, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.initRegion(userData.domainId, userData.projectId);
                        options = {
                            domain_id: userData.domainId,
                            userId: userData.userId,
                        };
                        ProjectData.getMaintenanceRegions();
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, UserService_1.UserSerivce.getUserProjects(options)];
                    case 2:
                        data = _a.sent();
                        ProjectData.cfProject.tmpProjects = data.projects;
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        console.log(error_1);
                        ProjectData.cfProject.tmpProjects = null;
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    ProjectData.initRegion = function (domainId, userRegion) {
        return __awaiter(this, void 0, void 0, function () {
            var param, regions;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        param = {
                            selfDevelop: !!ConfigData_1.ConfigData.getUiSwitchConfig().ctcRegionTiledDisplay,
                            alliance: true,
                            domainId: domainId,
                        };
                        return [4 /*yield*/, RegionData_1.RegionData.getRegions(param)];
                    case 1:
                        regions = _a.sent();
                        regions.forEach(function (item) {
                            var id = item.id || item.regionId;
                            if (id === userRegion || userRegion.indexOf(id + '_') > -1) {
                                ProjectData.cfProject.legalRegionFlag = true;
                            }
                        });
                        if (domainId &&
                            userRegion &&
                            !ProjectData.cfProject.legalRegionFlag &&
                            !ProjectData.cfProject.globalRegionName) {
                            UserData_1.UserData.getCfUserSync().cfStateNoSupportRegionFlag = true;
                        }
                        ProjectData.cfProject.tmpRegions = regions;
                        return [2 /*return*/];
                }
            });
        });
    };
    // maintenance regions 数据表
    ProjectData.getMaintenanceRegions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var promiseMaintenanceRegion, data, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promiseMaintenanceRegion = RegionService_1.RegionService.queryMaintenanceRegions();
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, promiseMaintenanceRegion];
                    case 2:
                        data = _a.sent();
                        ProjectData.cfProject.maintenanceRegions = data;
                        return [3 /*break*/, 4];
                    case 3:
                        error_2 = _a.sent();
                        console.log(error_2);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    return ProjectData;
}());
exports.ProjectData = ProjectData;
