// @cloud/tinyplus3/bundles/cloud-tinyplus3-locale-hws.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@cloud/tiny3'), require('@cloud/tinyplus3/locale/i18n')) :
    typeof define === 'function' && define.amd ? define('@cloud/tinyplus3/locale/hws', ['exports', '@cloud/tiny3', '@cloud/tinyplus3/locale/i18n'], factory) :
    (global = global || self, factory((global.cloud = global.cloud || {}, global.cloud.tinyplus3 = global.cloud.tinyplus3 || {}, global.cloud.tinyplus3.locale = global.cloud.tinyplus3.locale || {}, global.cloud.tinyplus3.locale.hws = {}), global.tiny3, global.cloud.tinyplus3.locale.i18n));
}(this, (function (exports, tiny3, i18n) { 'use strict';

    var siteConfig = {
        tpRegionName: 'hws',
        tpSupportMFA: true,
        tpSupportEmail: true,
        tpRegionSmallItem: false,
        tpRegionSelectStyle: false
    };

    /**
     * 项目构建，需要调用某个局点的初始化
     * import { TpConfig } from "@cloud/tinyplus3/locale/hws/TpConfig";
     * TpConfig.init();
     */
    var TpConfig = /** @class */ (function () {
        function TpConfig() {
        }
        TpConfig.init = function (language) {
            if (language === void 0) { language = 'zh-CN'; }
            var words = tiny3.TiLocale.getWords();
            // 设置所有语言包
            tiny3.TiLocale.setWords({
                'zh-CN': Object.assign(Object.assign(Object.assign(Object.assign({}, words['zh-CN']), i18n.defaultzh_CN), i18n.currencyCNY), siteConfig),
                'en-US': Object.assign(Object.assign(Object.assign(Object.assign({}, words['en-US']), i18n.defaulten_US), i18n.currencyCNY), siteConfig)
            });
            // 设置默认语言
            tiny3.TiLocale.setLocale(language);
        };
        return TpConfig;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.TpConfig = TpConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
