// @cloud/tinyplus3/esm2015/components/guidesteps/TpGuidestepsModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TpGuidestepsComponent } from './TpGuidestepsComponent';
import { TpGuidestepComponent } from './TpGuidestepComponent';
import { TpGuidestepHeaderComponent } from './TpGuidestepHeaderComponent';
import { TpGuidestepContentComponent } from './TpGuidestepContentComponent';
export class TpGuidestepsModule {
}
TpGuidestepsModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [TpGuidestepsComponent, TpGuidestepComponent, TpGuidestepHeaderComponent, TpGuidestepContentComponent],
                declarations: [TpGuidestepsComponent, TpGuidestepComponent, TpGuidestepHeaderComponent, TpGuidestepContentComponent]
            },] }
];
export { TpGuidestepsComponent } from './TpGuidestepsComponent';
export { TpGuidestepComponent } from './TpGuidestepComponent';
export { TpGuidestepHeaderComponent } from './TpGuidestepHeaderComponent';
export { TpGuidestepContentComponent } from './TpGuidestepContentComponent';
