// @cloud/tinyplus3/esm2015/components/guidesteps/TpGuidestepHeaderComponent.js
import { Component, ElementRef, Renderer2 } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
import { TpGuidestepComponent } from './TpGuidestepComponent';
/**
 * TpGuidestepHeaderComponent 是单个小步骤模块组件的标题头
 */
export class TpGuidestepHeaderComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, guidestepComp) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.guidestepComp = guidestepComp;
    }
}
TpGuidestepHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-guidestep-header',
                template: `<span><ng-content></ng-content></span>`,
                host: {
                    '[class.tp-guidestep-dot-complete]': 'guidestepComp.isComplete'
                },
                styles: [":host{display:inline-block;font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number);padding-bottom:8px;padding-left:calc(var(--guidestep-dot-size) + var(--guidestep-icon-space));position:relative}:host:before{background-color:#8a8e99;border:var(--guidestep-dot-border) solid #eef0f5;border-radius:50%;content:\"\";height:var(--guidestep-dot-size);left:0;position:absolute;top:calc((var(--guidestep-header-line-height) - var(--guidestep-dot-outer-size))/2);width:var(--guidestep-dot-size)}[large] :host{font-size:var(--ti-common-font-size-2)}[large] :host:before{background:var(--ti-common-color-icon-normal);top:calc((var(--guidestep-header-large-line-height) - var(--guidestep-dot-outer-size))/2)}:host.tp-guidestep-dot-complete:before{background:var(--ti-common-color-success)}"]
            },] }
];
TpGuidestepHeaderComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TpGuidestepComponent }
];
