// @cloud/tiny3/esm2015/components/upload/TiFileSelectDirective.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { TiUploadUtil } from '../../services/upload/TiUploadUtil';
/**
 * 该指令适用于自定义上传场景，指令定义在input type='file'元素上，传入的是文件上传实例
 * 文件上传实例的生成，请参考[TiUploadService.create]{@link ../injectables/TiUploadService.html#create}
 *
 * 除自定义使用方式外，Tiny还提供了两种已进行设计的上传组件供业务使用，具体见[TiUploadComponent]{@link ../components/TiUploadComponent.html}
 *
 * <example-url>../tiny3demo/#/upload/upload-all</example-url>
 */
export class TiFileSelectDirective {
    constructor(hostElementRef, renderer) {
        this.hostElementRef = hostElementRef;
        this.renderer = renderer;
        this.nativeElement = this.hostElementRef.nativeElement;
        this.fileChangeEvent = this.renderer.listen(this.nativeElement, 'change', () => {
            if (!this.tiFileSelect) { // 配置的上传文件实例不存在情况下，不做后续处理
                return;
            }
            this.onFileChange(this.nativeElement); // 点选文件后的处理
        });
    }
    /**
     * @ignore
     * 选择文件后，根据浏览器差异进行处理
     */
    onFileChange(fileSel) {
        const uploadInst = this.tiFileSelect;
        const files = TiUploadUtil.isHTML5 ? fileSel.files : fileSel; // 获取文件信息
        const addedItems = uploadInst._addToQueue(files); // 文件选择队列
        if (TiUploadUtil.isHTML5) { // H5情况下，重置表单元素值，确保可重复选择文件
            this.nativeElement.value = '';
        }
        else { // 非H5情况下，确保文件下次可继续选择,分两种情况：1.已选文件有效情况下，,重新替换表单元素，确保文件点选元素不会随表单上传消失；
            // 2.文件未加入到队列（文件校验失败情况）情况下，重置点选表单，确保文件可重复选择（校验失败可能是文件队列长度不符，所以为确保用户体验，执行该操作）
            const isRemoveInput = (addedItems.length === 0); // 根据文件是否有效情况确定是否移除input
            this.replaceFileInput(fileSel, isRemoveInput);
        }
        // 自动上传情况下，进行文件上传
        if (uploadInst.config.autoUpload !== false && (addedItems.length !== 0)) {
            uploadInst.uploadItems(addedItems);
        }
    }
    /**
     * @ignore
     * 替换单个文件选择按钮，确保后续文件选择可继续点选
     * 当前文件选择框
     * 文件选择实例对象
     */
    replaceFileInput(fileSel, isRemoveInput) {
        // 清除当前文件输入框选择事件
        this.fileChangeEvent();
        // 新增input,并处理当前input
        const fileSelNew = fileSel.cloneNode(); // 文件元素克隆时,不会复用原有已选文件信息
        if (isRemoveInput) { // 文件选择无效情况下移除input，确保下次可正常选择
            fileSel.parentNode.appendChild(fileSelNew);
            fileSel.remove();
        }
        else {
            fileSel.style.display = 'none'; // 隐藏已选文件，确保点选框页面呈现的唯一性
            fileSel.parentNode.insertBefore(fileSelNew, fileSel.nextSibling); // 点选文件已放入文件队列中,替换新的文件选择按钮，确保下次点选生效
        }
        // 增加当前input事件
        this.fileChangeEvent = this.renderer.listen(fileSelNew, 'change', () => {
            this.onFileChange(fileSelNew);
        });
    }
}
TiFileSelectDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiFileSelect]'
            },] }
];
TiFileSelectDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TiFileSelectDirective.propDecorators = {
    tiFileSelect: [{ type: Input }]
};
