// @cloud/tiny3/esm2015/components/upload/TiUploadModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiFileSelectDirective } from './TiFileSelectDirective';
import { TiUploadComponent } from './TiUploadComponent';
import { TiDisabledDirective } from './TiDisabledDirective';
import { TiTipModule } from '../../directives/tip/TiTipModule';
import { TiProgresspieModule } from '../progresspie/TiProgresspieModule';
import { TiUploadServiceModule } from '../../services/upload/TiUploadServiceModule';
import { TiIconModule } from '../icon/TiIconModule';
import { TiOverflowModule } from '../../directives/overflow/TiOverflowModule';
import { TiValidationModule } from '../../directives/validation/TiValidationModule';
import { TiProgressbarModule } from '../progressbar/TiProgressbarModule';
import { TiUploadbaseComponent } from './TiUploadbaseComponent';
import { TiUploadimageComponent } from './TiUploadimageComponent';
import { TiModalModule } from '../../services/modal/TiModalModule';
import { TiImagepreviewModule } from '../imagepreview/TiImagepreviewModule';
import { TiButtonModule } from '../button/TiButtonModule';
import { TiOutlineModule } from '../../directives/outline/TiOutlineModule';
export class TiUploadModule {
}
TiUploadModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TiIconModule,
                    TiTipModule,
                    TiProgresspieModule,
                    TiUploadServiceModule,
                    TiOverflowModule,
                    TiValidationModule,
                    TiProgressbarModule,
                    TiModalModule,
                    TiImagepreviewModule,
                    TiButtonModule,
                    TiOutlineModule
                ],
                exports: [TiFileSelectDirective, TiUploadComponent, TiDisabledDirective, TiUploadbaseComponent, TiUploadimageComponent],
                declarations: [TiFileSelectDirective, TiUploadComponent, TiDisabledDirective, TiUploadbaseComponent, TiUploadimageComponent]
            },] }
];
export { TiUploadComponent } from './TiUploadComponent';
export { TiFileSelectDirective } from './TiFileSelectDirective';
export { TiDisabledDirective } from './TiDisabledDirective';
export { TiUploadbaseComponent } from './TiUploadbaseComponent';
export { TiUploadimageComponent } from './TiUploadimageComponent';
