// @cloud/tiny3/esm2015/components/upload/TiUploadbaseComponent.js
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
import { TiLocale } from '../../locale/TiLocaleModule';
import { TiBrowser, Util } from '../../utils/Util';
/**
 * 10.0.2版本抽取
 *
 * upload基类，继承于TiBaseComponent组件
 */
export class TiUploadbaseComponent extends TiBaseComponent {
    constructor(hostRef, renderer) {
        super(hostRef, renderer);
        this.hostRef = hostRef;
        this.renderer = renderer;
        /**
         * 上传方式，可选值为：get/post(其他方式IE9不支持，因此暂不开放)
         */
        this.method = 'post';
        /**
         * 向后台发送请求时对应的文件对象name属性，该name属性是后台读取文件的入口值
         */
        this.alias = 'tiFile';
        /**
         * 上传对象信息中，formData是否先于file对象，适用于后台传输时，formData数据先于file对象的场景
         */
        this.formDataFirst = false;
        /**
         * 文件选中后是否自动上传
         */
        this.autoUpload = true;
        /**
         * 文件上传是否 是否一次请求传输多个文件，默认情况下一次请求上传一个文件
         */
        this.batchSend = false;
        /**
         * 是否禁用
         */
        this.disabled = false;
        /**
         * @ignore
         */
        this.title = TiBrowser.isChrome() ? ' ' : ''; // 禁用文件上传默认tip提示,Chrome下设置title为""时会显示默认提示，所以Chrome下需要设置为" ";而IE下设置为" "会显示文本为空的tip
        /**
         * 文件添加失败回调，可使用该回调定义上传错误提示
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * file: [TiFileInfo]{@link ../interfaces/TiFileInfo.html}, // 上传文件信息
         *
         * validResults: Array&lt;string&gt;  // 校验不合法的规则name数组
         *
         * }
         */
        this.addItemFailed = new EventEmitter();
        /**
         * 文件添加成功回调
         *
         * 参数：文件对象
         */
        this.addItemSuccess = new EventEmitter();
        // 以下回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
        /**
         * 上传文件前回调，可在该回调中动态设置formData
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为数组形式，代表上传的文件对象
         */
        this.beforeSendItems = new EventEmitter();
        /**
         * 上传文件进度改变回调
         *
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * fileItems: Array<[TiFileItem]{@link ../interfaces/TiFileItem.html}>, // 上传文件对象数组
         *
         * progress: number  // 进度信息
         *
         * }
         */
        this.progressItems = new EventEmitter();
        /**
         * 文件上传完成回调，成功/失败都会触发
         *
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * fileItems: Array<[TiFileItem]{@link ../interfaces/TiFileItem.html}>, // 上传文件对象数组
         *
         * response: string  // 文件上传响应信息
         *
         * status: number 文件上传响应状态码
         *
         * }
         */
        this.completeItems = new EventEmitter();
        /**
         * 文件上传成功回调
         *
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * fileItems: Array<[TiFileItem]{@link ../interfaces/TiFileItem.html}>, // 上传文件对象数组
         *
         * response: string  // 文件上传响应信息
         *
         * status: number 文件上传响应状态码
         *
         * }
         */
        this.successItems = new EventEmitter();
        /**
         * 文件上传失败回调
         *
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * fileItems: Array<[TiFileItem]{@link ../interfaces/TiFileItem.html}>, // 上传文件对象数组
         *
         * response: string  // 文件上传响应信息
         *
         * status: number 文件上传响应状态码
         *
         * }
         */
        this.errorItems = new EventEmitter();
        /**
         * 文件上传取消回调
         *
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为对象形式，结构如下：
         *
         * {
         *
         * fileItems: Array<[TiFileItem]{@link ../interfaces/TiFileItem.html}>, // 上传文件对象数组
         *
         * response: string  // 文件上传响应信息
         *
         * status: number 文件上传响应状态码
         *
         * }
         */
        this.cancelItems = new EventEmitter();
        /**
         * 上传删除回调
         * 该回调对单个或多个文件在同一url同时上传有效，可使用其进行文件上传过程中的业务处理
         *
         * 参数为数组形式，代表在操作的文件对象
         */
        this.removeItems = new EventEmitter();
        /**
         * 所有文件上传完成回调
         *
         * 参数为数组形式，代表在操作的文件对象
         */
        this.completeAllItems = new EventEmitter();
        /**
         * 上传文件删除前的回调
         *
         * 参数为数组形式，代表在操作的文件对象
         */
        this.beforeRemoveItems = new EventEmitter();
        /**
         * @ignore 词条
         */
        // tslint:disable-next-line:typedef
        this.uploadLan = TiLocale.getLocaleWords().tiUpload;
    }
    ngOnInit() {
        super.ngOnInit();
        this.uploadConfig = {
            url: this.url,
            method: this.method,
            filters: this.filters,
            alias: this.alias,
            formData: this.formData,
            autoUpload: this.autoUpload,
            batchSend: this.batchSend || false,
            formDataFirst: this.formDataFirst || false,
            headers: this.headers,
            onAddItemFailed: (file, validResults) => {
                this.addItemFailed.emit({ file, validResults });
            },
            onAddItemSuccess: (fileItem) => {
                this.addItemSuccess.emit(fileItem);
            },
            onBeforeSendItems: (fileItems) => {
                this.beforeSendItems.emit(fileItems);
            },
            onProgressItems: (fileItems, progress) => {
                this.progressItems.emit({ fileItems, progress });
            },
            onCompleteItems: (fileItems, response, status) => {
                this.completeItems.emit({ fileItems, response, status });
            },
            onSuccessItems: (fileItems, response, status) => {
                this.successItems.emit({ fileItems, response, status });
            },
            onErrorItems: (fileItems, response, status) => {
                this.errorItems.emit({ fileItems, response, status });
            },
            onCancelItems: (fileItems, response, status) => {
                this.cancelItems.emit({ fileItems, response, status });
            },
            onRemoveItems: (fileItems) => {
                this.removeItems.emit(fileItems);
            },
            onBeforeRemoveItems: (fileItems) => {
                this.beforeRemoveItems.emit(fileItems);
            },
            onCompleteAllItems: (fileItems) => {
                this.completeAllItems.emit(fileItems);
            }
        };
        // 根据配置的文件类型过滤属性设置H5下的文件过滤属性,如果accept未定义,则按照type过滤条件进行设置；
        // 如果accept已定义，则按照accept定义设置：此种场景应用于浏览器设置accept后导致文件窗打开很慢的情况
        if (Util.isUndefined(this.accept) && this.filters && this.filters.length) {
            this.filters.forEach((filter) => {
                if (filter.name === 'type') {
                    this.accept = filter.params[0];
                }
                return;
            });
        }
    }
    /**
     * 手动上传，可以通过组件调用
     */
    upload() {
        this.uploadInst.uploadAll();
    }
}
TiUploadbaseComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-uploadbase',
                template: ''
            },] }
];
TiUploadbaseComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TiUploadbaseComponent.propDecorators = {
    url: [{ type: Input }],
    method: [{ type: Input }],
    filters: [{ type: Input }],
    alias: [{ type: Input }],
    formData: [{ type: Input }],
    formDataFirst: [{ type: Input }],
    autoUpload: [{ type: Input }],
    batchSend: [{ type: Input }],
    headers: [{ type: Input }],
    accept: [{ type: Input }],
    disabled: [{ type: Input }],
    addItemFailed: [{ type: Output }],
    addItemSuccess: [{ type: Output }],
    beforeSendItems: [{ type: Output }],
    progressItems: [{ type: Output }],
    completeItems: [{ type: Output }],
    successItems: [{ type: Output }],
    errorItems: [{ type: Output }],
    cancelItems: [{ type: Output }],
    removeItems: [{ type: Output }],
    completeAllItems: [{ type: Output }],
    beforeRemoveItems: [{ type: Output }]
};
