// @cloud/tiny3/esm2015/components/table/TiCellTextComponent.js
import { Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
import { TiRenderer } from '../../services/renderer/TiRenderer';
/**
 * TiCellText 表格单元格中放置文本的容器组件
 *
 * 提供文本溢出时出省略号...,且hover时出tip的功能。
 *
 * 配合TiCellText组件使用可使表格单元格中内容显示不下时，文本出省略号...,
 * 而文本后的放在TiCellIcons图标始终可以显示出来
 *
 */
export class TiCellTextComponent {
    constructor(renderer, elementRef, tiRenderer) {
        this.renderer = renderer;
        this.elementRef = elementRef;
        this.tiRenderer = tiRenderer;
    }
    ngAfterViewInit() {
        setTimeout(() => {
            this.setMaxWidth();
        }, 0);
    }
    setMaxWidth() {
        let otherWidth = 0;
        const parent = this.renderer.parentNode(this.elementRef.nativeElement);
        const iconsContainer = this.tiRenderer.findChildrenByClassName(parent, 'ti3-cell-icons-container')[0];
        const sortContainer = this.tiRenderer.findChildrenByClassName(parent, 'ti3-sort-container')[0];
        const headFilterContainer = this.tiRenderer.findChildrenByClassName(parent, 'ti3-head-filter-container')[0];
        const renameContainer = this.tiRenderer.findChildrenByTagName(parent, 'TP-RENAME')[0];
        // 修复SSR报错：ERROR TypeError: sortContainer.getBoundingClientRect is not a function
        if (typeof parent.getBoundingClientRect !== 'function') {
            return;
        }
        // 修复SSR报错：ERROR TypeError: Cannot read property 'getBoundingClientRect' of undefined
        if (iconsContainer) {
            const iconsContainerWidth = parseFloat(iconsContainer.getBoundingClientRect().width
                .toFixed(1));
            otherWidth += iconsContainerWidth;
        }
        if (sortContainer) {
            const sortContainerWidth = parseFloat(sortContainer.getBoundingClientRect().width
                .toFixed(1));
            otherWidth += sortContainerWidth;
        }
        if (headFilterContainer) {
            const headFilterContainerWidth = parseFloat(headFilterContainer.getBoundingClientRect().width
                .toFixed(1));
            otherWidth += headFilterContainerWidth;
        }
        if (renameContainer) {
            const renameContainerWidth = parseFloat(renameContainer.getBoundingClientRect().width
                .toFixed(1));
            otherWidth += renameContainerWidth;
        }
        this.renderer.setStyle(this.cellTextEle.nativeElement, 'max-width', `calc(100% - ${otherWidth}px)`);
    }
}
TiCellTextComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-cell-text',
                template: "<span #text class=\"ti3-cell-text\" tiOverflow [tiTipContent]=\"tipContent\" [tiTipPosition]=\"tipPosition\">\r\n    <ng-content></ng-content>\r\n</span>",
                host: {
                    '[class.ti3-cell-text-container]': 'true'
                }
            },] }
];
TiCellTextComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: TiRenderer }
];
TiCellTextComponent.propDecorators = {
    tipContent: [{ type: Input }],
    tipPosition: [{ type: Input }],
    cellTextEle: [{ type: ViewChild, args: ['text', { static: true },] }]
};
