// @cloud/tiny3/esm2015/components/table/TiColspanDirective.js
import { Directive, ElementRef, Renderer2 } from '@angular/core';
import { TiTableComponent } from './TiTableComponent';
/**
 * TiColspan 表格跨列数colspan 计算，
 *
 * 根据表头第一行tr中th的数目，同步colspan数。
 *
 * 主要配合表格详情展开功能使用，用于表格详情展示行的的列合并。
 *
 */
export class TiColspanDirective {
    constructor(table, renderer, elementRef) {
        this.table = table;
        this.renderer = renderer;
        this.elementRef = elementRef;
    }
    ngAfterViewInit() {
        // 获取thead中第一行tr
        const tableFirstTr = this.table.nativeElement.querySelector('table>thead>tr');
        if (!tableFirstTr) {
            return;
        }
        // 初始化设置一次colspan
        this.renderer.setAttribute(this.elementRef.nativeElement, 'colspan', String(tableFirstTr.children.length));
        // 修复SSR报错：ERROR ReferenceError: MutationObserver is not defined
        if (typeof MutationObserver === 'undefined') {
            return;
        }
        this.firstTrObserver = new MutationObserver((mutationsList) => {
            for (let mutation of mutationsList) {
                if (mutation.type === 'childList') {
                    this.renderer.setAttribute(this.elementRef.nativeElement, 'colspan', String(tableFirstTr.children.length));
                }
            }
        });
        // 只需要处理tr中th的增删，所以只配置 childList为true,
        this.firstTrObserver.observe(tableFirstTr, { childList: true, });
    }
    ngOnDestroy() {
        if (this.firstTrObserver) {
            this.firstTrObserver.disconnect();
        }
    }
}
TiColspanDirective.decorators = [
    { type: Directive, args: [{
                selector: 'td[tiColspan]'
            },] }
];
TiColspanDirective.ctorParameters = () => [
    { type: TiTableComponent },
    { type: Renderer2 },
    { type: ElementRef }
];
