// @cloud/tiny3/esm2015/components/table/TiDetailsTrDirective.js
import { Directive, Input, Renderer2, TemplateRef, ViewContainerRef } from '@angular/core';
import { TiTableComponent } from './TiTableComponent';
import { TiDetailsIconComponent } from './TiDetailsIconComponent';
/**
 * TiDetailsTr 详情行结构指令
 *
 * 使用时需要在其前面加 * 语法糖，传入当前行数据；
 * 其内部根据row.showDetails的值来控制详情行是否显示，功能类似于ngIf。
 */
export class TiDetailsTrDirective {
    constructor(templateRef, viewContainerRef, renderer, table) {
        this.templateRef = templateRef;
        this.viewContainerRef = viewContainerRef;
        this.renderer = renderer;
        this.table = table;
    }
    ngDoCheck() {
        if (this.row.showDetails !== this.oldShowDetails) {
            this.updateView();
            this.oldShowDetails = this.row.showDetails;
        }
    }
    updateView() {
        if (this.row.showDetails) {
            // 上下文参数是否需要传递
            this.viewContainerRef.createEmbeddedView(this.templateRef);
            // 结构指令的宿主元素是一个 comment(注释) 的Node节点，
            // 生成节点元素插入DOM中的位置跟angularCompilerOptions.enableIvy 配置有关，true时在宿主元素之前，false时在宿主元素之后，
            // 而 ng9开始angularCompilerOptions.enableIvy 默认为true,但一些项目为兼容性，强制设置为了false，所以要做一下兼容处理
            let detailTr;
            const nativeElement = this.templateRef.elementRef.nativeElement;
            if (nativeElement.previousSibling && nativeElement.previousSibling.classList &&
                nativeElement.previousSibling.classList.contains(TiDetailsIconComponent.TABLE_ClASS_DETAIL_BASE)) {
                detailTr = nativeElement.nextSibling;
            }
            else {
                detailTr = nativeElement.previousSibling;
            }
            this.renderer.addClass(detailTr, 'ti3-details-tr');
        }
        else {
            this.viewContainerRef.clear();
            //  TODO: 方案是否可优化
            // 表头锁定时，展开时可在table的AfterViewChecked中获取到表格变化后的高度，
            // 但是收起时，在table的AfterViewChecked中获取不到表格变化后的高度，不能及时处理滚动条的变化
            // 所以此处使用setTimeout使其延时触发table的AfterViewChecked，使其能够处理由高度变化引起的滚动条状态的改变
            // 注：使用ngIf指令也是如此效果
            if (this.table.isFixedHead) {
                setTimeout(() => { }, 0);
            }
        }
    }
}
TiDetailsTrDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiDetailsTr]'
            },] }
];
TiDetailsTrDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef },
    { type: Renderer2 },
    { type: TiTableComponent }
];
TiDetailsTrDirective.propDecorators = {
    row: [{ type: Input, args: ['tiDetailsTr',] }]
};
