// @cloud/tiny3/esm2015/components/table/TiHeadFilterComponent.js
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { TiSelectComponent } from '../select/TiSelectModule';
import { TiHeadFilterDropComponent } from './TiHeadFilterDropComponent';
import { TiFormComponent } from '../base/TiBaseModule';
import { TiBrowser } from '../../utils/Util';
// 下面注释，可以避免编译lib时正则报错。原理未知，副作用未知。
// @dynamic
/**
 * TiHeadFilter 表头过滤(漏斗)组件， 嵌在表头的 th 中使用。
 *
 * ti-head-filter 用来显示过滤操作漏斗图标，点击其可打开下拉框，选中下拉框中的选项时，
 * 可在提供的 select 事件中做对应的表格数据过滤处理。
 *
 * ### 接口说明
 * **Inputs：**
 *
 * |     名称    |  类型   | 默认值   | 功能描述  |
 * | --------   | :-----   | :---- | :---- |
 * | options   | array  | [] | 所有过滤项 |
 * | ngModel   | object  | false | 可设置和获取 headfilter下拉选中值 |
 * | labelKey   |  string  | 'label' | 下拉面板中要显示的字段, 10.0.2版本新增 |
 * | panelMaxHeight   |  string  | 最大显示8条数据 | 下下拉面板的最大显示高度，溢出时则出滚动条。 |
 * | id   | string  | undefined | HTML属性id，自动化测试要求必须给id赋值 |
 * | select   | function  | 无 | 下拉面板选中事件，向外通知选中数据 |
 * | multiple   |  boolean  | false | 下拉面板是否为多选 |
 * | selectAll   |  boolean  | false | 下拉面板是否显示全选框 |
 * | panelAlign   |  string  | 'left' | 下拉面板展开对齐方式，9.0.4版本新增 |
 * | searchable   |  boolean  | false | 是否开启搜索功能，10.0.2版本新增 |
 * | panelWidth   |  'auto' 或 string  | 'auto' | 设置下拉面板的宽度，"auto"表示下拉框的宽度根据下拉选项的内容自动撑开，也可设置固定的下拉框宽度(不小于过滤图标的宽度)，例如："200px", 10.0.2版本新增 |
 * | noDataText   |  string  | '...' | 无数据时的显示文本。10.0.3版本新增 |
 *
 * <p><span style="color: red">以下说明不可用，不做参考。</span>该组件只有以上5个Input接口和1个Output接口
 * 可用，其余方法、Input、输出等(继承于TiSelectComponent,是供select的内部使用的)都不可用</p>
 */
export class TiHeadFilterComponent extends TiSelectComponent {
    constructor() {
        super(...arguments);
        /**
         * 面板对齐方式
         *
         * 9.0.4新增
         */
        this.panelAlign = 'left';
        // 由于headfilter与select中的panelWidth的默认值和类型不一样，所以这里重定义panelWidth接口
        /**
         * 下拉面板的最大显示宽度，该变量与下拉类组件保持一致
         *
         * 1."auto": 下拉框的宽度根据下拉选项的内容自动撑开；
         *
         * 2.表示宽度的字符串: 设置固定的下拉框宽度(不小于Select组件的宽度)。例如："200px"
         *
         * 10.0.2版本新增，默认为'auto'，10.1.2起<b>搜索场景下</b>默认值修改为“180px”，
         */
        // 置为undefined是为了覆盖父类的panelWidth值
        this.panelWidth = undefined;
    }
    ngAfterViewChecked() {
        // 这里不能调用父类方法，因为Select父类对dominator处理，空指针报错。
        // TODO: 看是否能够不继承select
        if (this.searchable !== this.oldSearchable) {
            this.oldSearchable = this.searchable;
            if (this.searchable) {
                this.setFocusableElems([this.filterIconEleRef.nativeElement]
                    .concat(this.dropsearchCom.getFocusableElems()));
            }
            else {
                this.setFocusableElems([this.filterIconEleRef.nativeElement]);
            }
        }
    }
    /**
     * @ignore
     * 失焦情况下，仅关闭面板，不做聚焦等处理
     */
    onBlur() {
        super.onBlur();
        if (TiBrowser.isIE() && this.searchable) {
            if (getComputedStyle(this.filterIconEleRef.nativeElement).outlineStyle === 'none') {
                this.renderer.setStyle(this.filterIconEleRef.nativeElement, 'outlineStyle', '');
            }
        }
    }
}
TiHeadFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-head-filter',
                template: "<a #filterIcon\r\n     href=\"javascript:void(0)\"\r\n     oncontextmenu=\"return false\"\r\n     class=\"ti3-icon ti3-icon-filter ti3-head-filter-icon\"\r\n     (click)=\"onClickDominator()\"\r\n     (keydown)=\"onKeydown($event)\"\r\n     [id]=\"appendId('head-filter')\">\r\n</a>\r\n<ti-head-filter-drop [dominatorElem]=\"filterIconEleRef.nativeElement\"\r\n                     [(ngModel)]=\"model\"\r\n                     [panelWidth]=\"panelWidth?panelWidth:searchable?'180px':'auto'\"\r\n                     [panelAlign]='panelAlign'\r\n                     [searchable]=\"searchable\"\r\n                     [options]=\"options\"\r\n                     [multiple]=\"multiple\"\r\n                     [selectAll]=\"selectAll\"\r\n                     [panelMaxHeight]=\"panelMaxHeight\"\r\n                     [labelKey]=\"labelKey\"\r\n                     [noDataText]=\"noDataText\"\r\n                     (select)=\"onSelect($event)\"\r\n                     [id]=\"appendId('headfilter_drop')\"></ti-head-filter-drop>",
                providers: [TiFormComponent.getValueAccessor(TiHeadFilterComponent)],
                host: {
                    '[class.ti3-head-filter-container]': 'true',
                    '(blur)': 'onBlur()'
                },
                styles: [":host.ti3-head-filter-container{display:inline-block;line-height:16px;padding-left:5px;vertical-align:middle;width:16px}:host.ti3-head-filter-container .ti3-head-filter-icon{color:var(--ti-table-details-icon-color);cursor:pointer;font-size:16px;text-decoration:none}:host.ti3-head-filter-container .ti3-head-filter-icon:hover{color:var(--ti-table-details-icon-color-hover)}:host.ti3-head-filter-container.ti3-head-filtered .ti3-head-filter-icon{color:var(--ti-head-filter-filtered-color)}"]
            },] }
];
TiHeadFilterComponent.propDecorators = {
    panelAlign: [{ type: Input }],
    panelWidth: [{ type: Input }],
    filterIconEleRef: [{ type: ViewChild, args: ['filterIcon', { static: true },] }],
    dropsearchCom: [{ type: ViewChild, args: [TiHeadFilterDropComponent, { static: true },] }]
};
