// @cloud/tiny3/esm2015/components/table/TiHeadSortComponent.js
import { Component, ElementRef, Input } from '@angular/core';
import { TiTableComponent } from './TiTableComponent';
/**
 * TiHeadSort 表头排序(上下箭头)组件
 *
 * 嵌在表头的th中使用，点击该组件时表格数据按该列的排序规则进行升序/降序排序
 *
 * 组件有默认的排序规则，也可支持自定义排序规则
 */
export class TiHeadSortComponent {
    constructor(elementRef, table) {
        this.elementRef = elementRef;
        /**
         * 指定初始时在对应 sortKey下是否为升序排序。默认值为null, 表示无序(不升序也不降序)
         */
        this.asc = null;
        /**
         * @ignore
         */
        this.clickNum = 0;
        this.hostElement = this.elementRef.nativeElement;
        this.table = table;
    }
    ngOnInit() {
        // 更新TiTableComponent的enableSort，表示启用了排序。
        this.table.enableSort = true;
        // 修复SSR报错：ERROR TypeError: Cannot read property 'sort' of undefined
        // 表格记忆
        if (this.table.storageId && this.table.storageConfig && this.table.storageConfig.sort && this.table.sessionStorageState && this.table.sessionStorageState.sort) {
            if (this.sortKey === this.table.sessionStorageState.sort.sortKey) {
                this.asc = this.table.sessionStorageState.sort.asc;
            }
            else {
                this.asc = null;
            }
        }
        if (this.asc !== null) {
            this.clickNum = this.asc ? 0 : 1;
            // 如果有默认排序，即排序的初始化，将排序初始化的各个值赋给 tiTable 的 tableState 对应的 sort 属性
            this.sort(true);
        }
    }
    ngDoCheck() {
        if (this.table.dataState.sort.sortKey !== this.sortKey) {
            this.clickNum = 0;
        }
    }
    /**
     * @ignore
     */
    onClick() {
        if (!this.sortKey) {
            return;
        }
        this.sort(false);
    }
    sort(isDefaultSort) {
        this.clickNum++;
        // this.clickNum整除3余2表示降序，余1表示升序，余0表示无序
        const isDesc = this.clickNum % 3 === 2 ? true : this.clickNum % 3 === 1 ? false : null;
        this.table.updateSort(this.sortKey, isDesc, isDefaultSort, this.compareFn);
    }
}
TiHeadSortComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-head-sort',
                template: "<span tiOutline class=\"ti3-sort ti3-icon\" [attr.tabindex] = '0' (keydown.enter) = 'onClick()'></span>",
                host: {
                    '[class.ti3-sort-container]': 'true',
                    '[class.ti3-sort-ascent]': 'sortKey === table.dataState.sort.sortKey && clickNum % 3 === 1',
                    '[class.ti3-sort-descent]': 'sortKey === table.dataState.sort.sortKey && clickNum % 3 === 2',
                    '(click)': 'onClick()'
                }
            },] }
];
TiHeadSortComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: TiTableComponent }
];
TiHeadSortComponent.propDecorators = {
    sortKey: [{ type: Input }],
    asc: [{ type: Input }],
    compareFn: [{ type: Input }]
};
