// @cloud/tiny3/esm2015/components/table/TiHeadFilterDropComponent.js
import { Component } from '@angular/core';
import { TiDropsearchComponent } from '../dropsearch/TiDropsearchModule';
import { TiFormComponent } from '../base/TiBaseModule';
import { TiBrowser } from '../../utils/Util';
/**
 * @ignore
 * TiHeadFilterDrop 表头过滤漏斗组件的下拉框选择组件
 */
export class TiHeadFilterDropComponent extends TiDropsearchComponent {
    constructor() {
        super(...arguments);
        this.dominatorSpace = TiHeadFilterDropComponent.DOMINATOR_SPACE + 'px';
        this.heightExcludeContent = 10; // 上下边框为2，上下padding为8
        this.buttonHeight = 45; // 确定、取消按钮高度
    }
    ngOnChanges(changes) {
        super.ngOnChanges(changes);
        // 重新设置headfilterdrop的heightExcludeList
        if (changes['searchable']) {
            if (changes['searchable'].currentValue) {
                this.heightExcludeList = this.multiple ? this.heightExcludeContent + this.buttonHeight + TiDropsearchComponent.SEARCHBOX_EXCLUDE_HEIGHT
                    : this.heightExcludeContent + TiDropsearchComponent.SEARCHBOX_EXCLUDE_HEIGHT;
            }
            else {
                this.heightExcludeList = this.multiple ? this.heightExcludeContent + this.buttonHeight : this.heightExcludeContent;
            }
        }
    }
    writeValue(model) {
        super.writeValue(model);
        // 初始化selected选中项
        if (this.multiple) {
            this.selected = [...model ? model : []];
        }
        else {
            this.selected = model ? model : '';
        }
    }
    onSelect(option) {
        // 多选时不处理onSelect事件（在点击确定按钮时触发）
        if (this.multiple) {
            return;
        }
        super.onSelect(option);
        this.model = option;
    }
    hide() {
        // 处理问题：IE 下有搜索框时，在下拉中选择完收起后漏斗图标由于聚焦而有了虚线框
        if (TiBrowser.isIE() && this.searchable) {
            this.render.setStyle(this.dominatorElem, 'outlineStyle', 'none');
            super.hide();
        }
        else {
            super.hide();
        }
        // 还原selected选中项
        if (this.multiple) {
            this.selected = [...this.model ? this.model : []];
        }
    }
    hideWithoutFocus() {
        super.hideWithoutFocus();
        // 还原selected选中项
        if (this.multiple) {
            this.selected = [...this.model ? this.model : []];
        }
    }
    // 按钮面板鼠标按下事件（解决因点击按钮面板导致整个下拉面板隐藏的问题）
    btnContainerMousedown(event) {
        event.preventDefault();
    }
    // 确定按钮
    okClick() {
        // 选中值发生改变时对model值做处理
        if (JSON.stringify(this.selected) !== JSON.stringify(this.model ? this.model : [])) {
            this.model = [...this.selected];
        }
        // 触发onSelect事件
        this.select.emit(this.model);
        this.hide();
    }
    // 取消按钮
    cancelClick() {
        // 还原selected选中项
        this.selected = [...this.model ? this.model : []];
        this.hide();
    }
}
TiHeadFilterDropComponent.DOMINATOR_SPACE = 6;
TiHeadFilterDropComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-head-filter-drop',
                template: "<ti-drop  class=\"ti3-head-filter-drop\"\r\n          [dominatorElem]=\"dominatorElem\"\r\n          [panelWidth]='panelWidth'\r\n          [panelAlign]='panelAlign'\r\n          [dominatorSpace]=\"dominatorSpace\">\r\n    <div *ngIf=\"searchable\">\r\n        <ti-searchbox-notsearch #searchboxCom noBorder\r\n            [placeholder]=\"placeholder\"\r\n            [(ngModel)]=\"searchWord\"\r\n            (ngModelChange)=\"searchWordChange($event)\"\r\n            [id]=\"appendId('searchbox')\">\r\n        </ti-searchbox-notsearch>\r\n    </div>\r\n    <ti-list [multiple]=\"multiple\"\r\n        [selectAll]=\"selectAll\"\r\n        [options]=\"((type !== 'suggestion' && searchable) || type === 'suggestion')?searchResult:options\"\r\n        [(ngModel)]=\"selected\"\r\n        [labelKey]=\"labelKey\"\r\n        [noDataText]=\"noDataText\"\r\n        style=\"display:block;\"\r\n        (select)=\"onSelect($event)\"\r\n        [id]=\"appendId('list')\">\r\n    </ti-list>\r\n    <section *ngIf=\"multiple\" (mousedown)=\"btnContainerMousedown($event)\">\r\n        <section class=\"ti3-head-filter-drop-button-container\">\r\n            <button type=\"button\"\r\n                tiButton\r\n                size=\"small\"\r\n                class=\"ti3-head-filter-drop-button-ok\"\r\n                (click)=\"okClick()\"\r\n                [id]=\"appendId('ok')\">{{ 'tiCommon.okBtn' | tiTranslate }}</button>\r\n            <button type=\"button\"\r\n                tiButton\r\n                size=\"small\"\r\n                (click)=\"cancelClick()\"\r\n                [id]=\"appendId('cancel')\">{{ 'tiCommon.cancelBtn' | tiTranslate }}</button>\r\n        </section>\r\n    </section>\r\n</ti-drop>",
                providers: [TiFormComponent.getValueAccessor(TiHeadFilterDropComponent)],
                styles: [".ti3-head-filter-drop ti-searchbox-notsearch{width:100%}.ti3-head-filter-drop-button-container{border-top:1px solid var(--ti-common-color-line-dividing);margin:0 var(--ti-common-space-10);padding:var(--ti-common-space-10) 0}.ti3-head-filter-drop-button-ok{margin-right:var(--ti-common-space-2x)}"]
            },] }
];
