// @cloud/tiny3/esm2015/components/table/TiTableComponent.js
import { Component, ElementRef, EventEmitter, Input, IterableDiffers, NgZone, Output, Renderer2, ViewEncapsulation } from '@angular/core';
import { TiRenderer } from '../../services/renderer/TiRenderer';
import { Util } from '../../utils/Util';
import { Subject } from 'rxjs';
import { TiTableFixedHeadService } from './TiTableFixedHeadService';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * TiTable 表格组件
 *
 * 支持以表格的形式展示多条数据
 *
 * 支持前后台分页、搜索、排序、详情展开、列拖动（调整列宽）、控制列动态隐藏/显示、行复选、行单选、
 * 表格单元格(th, td)智能tip、树表、列文本对齐方式、表头固定等功能
 *
 * <example-url>../tiny3demo/#/table/table-all</example-url>
 */
export class TiTableComponent extends TiBaseComponent {
    constructor(iterableDiffers, elementRef, fixedHeadService, renderer, zone, tiRenderer) {
        super(elementRef, renderer);
        this.iterableDiffers = iterableDiffers;
        this.fixedHeadService = fixedHeadService;
        this.zone = zone;
        this.tiRenderer = tiRenderer;
        /**
         * 设置表格源数据及其处理数据状态
         *
         * data: 表格源数据。一般情况下，前台分页时传入表格全部数据，后台分页时传入当前页数据。
         *
         * state: 用来标识传入的表格源数据是否已经经过排序(sorted)、过滤(searched)、分页(paginated)等处理。
         *
         * 应用举例：后台分页场景下，传入的data为当前页数据（即已经进行了分页），因此应将paginated设置为true；
         *          如果传入的数据已经进行了排序（如后台排序场景），则应将sorted设置为true。
         */
        this.srcData = TiTableComponent.DEFAULT_SRC_DATA;
        /**
         * 表格记忆各项开关配置
         * 10.0.4 新增
         */
        this.storageConfig = {
            sort: true,
            pagination: true,
            colsWidth: true,
        };
        /**
         * 表格实时呈现的数据改变时向外通知
         */
        this.displayedDataChange = new EventEmitter();
        /**
         * 设置表格的各列信息。开发者可以使用该接口的值在 thead>tr>th 中完成对表头内容的设置；
         *
         * 控件内部根据columns设置完成必要的逻辑处理。 建议对columns数组中的对象设置如下属性：
         *
         * title: 表头列文本内容。当某一列的title为空字符串时，当前列一直在表格中显示，但是不出现在控制列动态
         *         隐藏/显示 设置面板中，例如单选列、多选列、详情展开图标列;
         *
         * width: 设置列宽，支持百分比和px值;
         *
         * show: 此属性只有在表格具有列动态 显示/隐藏 功能时才需要设置，show可以分别设置如下值：
         *
         *       true： 表示该列默认显示，用户可以通过面板操作切换其 显示/隐藏 状态。
         *
         *       false : 表示该列默认隐藏，用户可以通过面板操作切换其 显示/隐藏 状态。
         *
         *       undefined : 表示该列不具备动态 显示/隐藏 功能，固定显示。
         *
         *       注意：设置show属性值时，title为空字符串时，该列固定显示，但不出现在列操作面板中。
         */
        this.columns = [];
        /**
         * 设置表格搜索功能的需被检索的字符串的集合。
         *
         * 此接口需配合searchKeys使用。
         *
         *  1. searchWords长度为1时，在searchKeys指定的字段下搜索searchWords[0]指定的字符串,
         *     在指定字段中的任一字段中搜索到时即满足条件(并集)。如：searchWords: ['b'],
         *     searchKeys: ['firstName', 'lastName']，则在firstName字段包含‘b’，
         *     或在lastName字段包含‘b’时本行数据即满足搜索条件;
         *
         * 2. searchWords长度大于1且searchKeys与searchWords长度相等时，在searchKeys指定的字段下搜索searchWords对应(按顺序)元素指定的字符串,
         *    在指定字段中的所有字段搜索到对应字符串时才满足条件(交集)。如：searchWords: ['b', '18'],
         *    searchKeys: ['firstName', 'age']，则在firstName字段包含‘b’且
         *    在age字段包含‘18’时本行数据才满足搜索条件。
         */
        this.searchWords = null;
        /**
         * 为表格搜索功能指定搜索的字段范围。
         *
         * 此接口需配合searchWords使用。
         */
        this.searchKeys = null; // 是否使用null
        /**
         * 有详情功能时，设置组件是否开启当前页只显示一行详情的功能。
         *
         * 设置为true，表示表格中当前页只展示一行的详情，不设置或设为false时，可展开多行的详情。
         */
        this.closeOtherDetails = false;
        /**
         * 当表格数据状态变化时向外通知
         *
         * 1.功能描述：
         *
         * 状态变化指的是表格的排序、搜索、分页参数发生了变化，触发变化的原因可能是用户操作，也可能是开发者设置。
         * 该回调为后台分页、后台排序和后台过滤提供了处理时机，例如，用户点击第2页时，由于页码的变化，必定会触发本回调函数，
         * 进而开发者可以在该回调中向后台发请求获取第2页的数据，并赋给表格。
         *
         * 注意事项：初始化时不会触发该回调，只有表格状态动态改变时才会触发该回调，所以需要用户自己在初始时向后端发送请求获取表格初始化数据。
         *
         * 2.函数入参：TiTableComponent 组件实例
         *
         */
        this.stateUpdate = new EventEmitter();
        this.promise = Promise.resolve(null);
        /**
         * @ignore 判断是否使用了ti-head-sort组件
         */
        this.enableSort = false;
        /**
         * @ignore
         */
        this.paginationSubject = new Subject();
        /**
         * @ignore
         */
        this.viewInitSubject = new Subject();
        /**
        * @ignore
        */
        this.containerScrollXChangeSubject = new Subject();
        /**
         * @ignore
         */
        this.thResizeSubject = new Subject();
        /**
         * @ignore
         */
        this.updateFixedThLeftSubject = new Subject();
        /**
         * @ignore
         */
        this.updateFixedTdLeftSubject = new Subject();
        /**
         * @ignore
         */
        this.updateColumnsSubject = new Subject();
        /**
         * @ignore
         * 其中包含了sort、search、pagination三个属性，每个属性都描述了当前表格数据的参数状态:
         *
         * sort：object类型，包含了sortKey（类型为string，进行排序的数据属性）、asc（类型为Boolean，是否为升序）属性；
         *
         * search: object类型，包含了searchWords（类型为array<string>，搜索字符串集合）、searchKeys（类型为Array，搜索指定的字段范围）属性；
         *
         * pagination: object类型，包含了currentPage（类型number，表示当前页）、itemsPerPage（类型为Number，每页显示条数）属性。
         */
        this.dataState = {
            search: { searchWords: null, searchKeys: null },
            sort: { sortKey: '', asc: null },
            pagination: { currentPage: -1, itemsPerPage: -1 }
        };
        this.isDataStateChange = false;
        this.oldSearchWords = null;
        this.oldSearchKeys = null;
        this.oldPagination = { currentPage: -1, itemsPerPage: -1 };
        this.oldSort = { sortKey: '', asc: null };
        this.searchedResult = [];
        /**
         * @ignore
         * 表格是否为表头锁定
         */
        this.isFixedHead = false;
        this.oldTbodyHeight = 0;
        /**
         * @ignore
         */
        this.fixedColumnInfo = {};
        this.compareTo = (v1, v2) => {
            let result = 0;
            if (Util.isFunction(this.customCompareFn)) {
                result = this.customCompareFn(v1.value, v2.value, this.dataState.sort.sortKey);
            }
            else {
                result = TiTableComponent.doCompare(v1.predicateValueObj, v2.predicateValueObj);
                if (result === 0) {
                    result = v1.index - v2.index;
                }
            }
            return result;
        };
        this.hostEle = elementRef.nativeElement;
    }
    /**
     * 测试某一字串(testStr)中是否存在子串(key)
     * @param testStr 待检索的字符串或数值
     * @param key 检索的子串
     */
    static isMatched(testStr, keywords) {
        return (Util.isString(testStr) || Util.isNumber(testStr)) ?
            (String(testStr)).toLowerCase()
                .indexOf(keywords.toLowerCase()) >= 0 : false;
    }
    static searchByOneWord(src, searchWords, searchKeys) {
        const output = [];
        src.forEach((item) => {
            if (searchKeys && searchKeys.length > 0) {
                for (const key of searchKeys) {
                    if (TiTableComponent.isMatched(item[key], searchWords[0])) {
                        output.push(item);
                        break;
                    }
                }
            }
            else {
                for (const key in item) {
                    if (TiTableComponent.isMatched(item[key], searchWords[0])) {
                        output.push(item);
                        break;
                    }
                }
            }
        });
        return output;
    }
    static searchByMoreThanOneWord(src, searchWords, searchKeys) {
        const output = [];
        src.forEach((item) => {
            let isMatched = true;
            for (let i = 0; i < searchKeys.length; i++) {
                if (!TiTableComponent.isMatched(item[searchKeys[i]], searchWords[i])) {
                    isMatched = false;
                    break;
                }
            }
            if (isMatched) {
                output.push(item);
            }
        });
        return output;
    }
    static getSearchedData(inputData, searchWords, searchKeys) {
        let outData = [];
        if (searchWords && searchWords.length === 1) {
            outData = TiTableComponent.searchByOneWord(inputData, searchWords, searchKeys);
        }
        else if (searchWords && searchWords.length > 1 && searchKeys &&
            searchKeys.length === searchWords.length) {
            outData = TiTableComponent.searchByMoreThanOneWord(inputData, searchWords, searchKeys);
        }
        return outData;
    }
    static doCompare(a, b) {
        let result = 0;
        if (a.type === b.type) {
            if (a.value !== b.value) {
                result = a.value < b.value ? -1 : 1;
            }
        }
        else {
            result = a.type < b.type ? -1 : 1;
        }
        return result;
    }
    static getPredicateValue(value, predicate) {
        if (!Util.isEmptyString(predicate) && value.hasOwnProperty(predicate)) {
            return value[predicate];
        }
        return value;
    }
    static getPredicateValueObj(value, index) {
        let resultVal = value;
        if (value === null) {
            resultVal = 'null';
        }
        else if (Util.isString(value)) {
            resultVal = value.toLowerCase();
        }
        else if (Util.isArray(value)) {
            resultVal = value.toString();
        }
        else if (typeof value === 'object') {
            resultVal = index;
        }
        else if (Util.isFunction(value)) {
            const resultValue = value();
            return TiTableComponent.getPredicateValueObj(resultValue, index);
        }
        const type = typeof resultVal;
        return { value: resultVal, type };
    }
    static safeCopy(data) {
        return data ? [].concat(data) : [];
    }
    static isValueEqual(n, o) {
        if (!(n instanceof Object) || !(o instanceof Object)) {
            return n === o;
        }
        if (Util.isArray(n) && Util.isArray(o)) {
            if (n.length !== o.length) {
                return false;
            }
            else {
                for (let i = 0; i < o.length; i++) {
                    if (n[i] !== o[i]) {
                        return false;
                    }
                }
            }
            return true;
        }
        if (Object.keys(n).length !== Object.keys(o).length) {
            return false;
        }
        else {
            for (const key in n) {
                if (!o.hasOwnProperty(key) || n[key] !== o[key]) {
                    return false;
                }
            }
        }
        return true;
    }
    ngOnInit() {
        super.ngOnInit();
        if (!this.srcData) {
            this.srcData = TiTableComponent.DEFAULT_SRC_DATA;
        }
        if (!this.srcData.data) {
            this.srcData.data = [];
        }
        this.srcDataDiffer = this.iterableDiffers.find(this.srcData.data)
            .create();
        if (this.searchWords !== null) {
            this.oldSearchWords = TiTableComponent.safeCopy(this.searchWords);
            this.dataState.search.searchWords = this.searchWords;
        }
        if (this.searchKeys !== null) {
            this.oldSearchKeys = TiTableComponent.safeCopy(this.searchKeys);
            this.dataState.search.searchKeys = this.searchKeys;
        }
        // 表格记忆
        if (this.storageId) {
            this.initStorageDataState();
        }
    }
    ngAfterViewInit() {
        this.initFixedHead();
        if (this.fixedColumnInfo.hasFixedColumn) {
            this.addFixedColumnBehavior();
        }
        this.viewInitSubject.next();
        if (this.storageId) {
            // 在ngOnDestroy 无法保存，所有使用beforeunload
            // 为什么不用unload? beforeunload 兼容比unload效果好
            // 参考：https://sinaad.github.io/xfe/2016/06/29/beforeunlod-vs-unload/
            // 修复SSR报错：ERROR ReferenceError: window is not defined
            if (typeof window === 'undefined') {
                return;
            }
            this.unlistenWindowBeforeunload = this.renderer.listen(window, 'beforeunload', () => {
                this.updateStorageDataState();
            });
        }
    }
    // 为了确保表格内嵌组件(比如tiHeadSort,tiPagination等)的数据变化在表格组件中能监听到，排序、搜索、分页
    // 的相关数据监听需要放在表格组件的AfterViewChecked周期中。因为表格组件的AfterViewChecked生命周期晚于
    // 内嵌组件(比如tiHeadSort,tiPagination等)的各个checked生命周期。
    // tslint:disable-next-line:cyclomatic-complexity
    // eslint-disable-next-line complexity
    ngAfterViewChecked() {
        // 监听srcData.data的变化
        const dataArr = this.srcData && this.srcData.data ? this.srcData.data : [];
        const srcDataDiffer = this.srcDataDiffer.diff(dataArr);
        if (srcDataDiffer) {
            this.updateSafeCopy();
            this.data2displayed();
        }
        // 监听search相关接口值变化
        if (this.searchWords !== null &&
            !TiTableComponent.isValueEqual(this.searchWords, this.oldSearchWords)) {
            this.oldSearchWords = TiTableComponent.safeCopy(this.searchWords);
            this.isDataStateChange = true;
            this.triggerEvent = 'search';
            this.dataState.search.searchWords = this.searchWords;
            // 搜索时，如果使用分页时，把用于表格内部计算分页的当前页码置为1
            if (this.dataState.pagination.itemsPerPage !== -1 && this.dataState.pagination.currentPage !== 1) {
                this.dataState.pagination.currentPage = 1;
                // 使用Promise是为了解决使用stateUpdate接口做后台分页+后台搜索时，
                // 分页组件ExpressionChangedAfterItHasBeenCheckedError报错问题。
                // 此处不能使用setTimeout来解决，使用setTimeout会导致stateUpdate多次执行。
                this.promise.then(() => {
                    this.paginationSubject.next({ currentPage: 1 });
                });
            }
        }
        if (this.searchKeys !== null && this.searchWords.length > 0 &&
            !TiTableComponent.isValueEqual(this.searchKeys, this.oldSearchKeys)) {
            this.oldSearchKeys = TiTableComponent.safeCopy(this.searchKeys);
            this.isDataStateChange = true;
            this.triggerEvent = 'search';
            this.dataState.search.searchKeys = this.searchKeys;
            // 搜索时，如果使用分页时，把用于表格内部计算分页的当前页码置为1
            if (this.dataState.pagination.itemsPerPage !== -1 && this.dataState.pagination.currentPage !== 1) {
                this.dataState.pagination.currentPage = 1;
                this.promise.then(() => {
                    this.paginationSubject.next({ currentPage: 1 });
                });
            }
        }
        if (!TiTableComponent.isValueEqual(this.dataState.sort, this.oldSort)) {
            this.oldSort = Object.assign({}, this.dataState.sort);
            this.isDataStateChange = true;
            this.triggerEvent = 'sort';
        }
        if (this.dataState.pagination.currentPage !== -1 &&
            this.dataState.pagination.itemsPerPage !== -1 &&
            !TiTableComponent.isValueEqual(this.dataState.pagination, this.oldPagination)) {
            this.oldPagination = Object.assign({}, this.dataState.pagination);
            this.isDataStateChange = true;
            this.triggerEvent = 'pagination';
        }
        if (this.isDataStateChange) {
            this.updateTable();
            this.isDataStateChange = false;
        }
        if (this.isFixedHead && this.getTbodyHeight() !== this.oldTbodyHeight) {
            // 有些场景下(例如，两页显示数据条数相同，从其中一页跳到另一页时)获取高度不准确，
            // 所以此处添加setTimeout
            setTimeout(() => {
                if (this.hostEle.hasAttribute('tiColsResizable') && this.processYScrollStateChangeWithColsResizable) {
                    this.processYScrollStateChangeWithColsResizable();
                }
                else {
                    this.fixedHeadService.processOverflowY(this.theadContainer, this.tbodyContainer, this);
                }
                this.fixedHeadService.removeTbodyContainerBorderBottom(this.tbodyContainer, this.displayedData);
            }, 0);
            this.oldTbodyHeight = this.getTbodyHeight();
        }
        // 表格容器宽度改变时，列宽和横向滚动条状态都可能发生改变，这时需要处理表格列固定
        if (this.fixedColumnInfo.hasFixedColumn) {
            const containerWidth = this.fixedColumnInfo.container.clientWidth;
            if (containerWidth !== this.fixedColumnInfo.containerWidth) {
                this.fixedColumnInfo.containerWidth = containerWidth;
                this.processFixedColumn();
            }
        }
    }
    ngOnDestroy() {
        if (this.isFixedHead && this.unlistenFixedHeadWindowResize) {
            this.unlistenFixedHeadWindowResize();
        }
        if (this.unlistenFixedColumnWindowResize) {
            this.unlistenFixedColumnWindowResize();
        }
        if (this.storageId) {
            // 路由跳转时，页面没有刷新，beforeunload监听不生效，所以要更新一次记忆数据。
            this.updateStorageDataState();
            // 修复SSR错误：TypeError: this.unlistenWindowBeforeunload is not a function
            (typeof this.unlistenWindowBeforeunload === 'function') && this.unlistenWindowBeforeunload();
        }
    }
    // 初始化表格记忆
    initStorageDataState() {
        // 修复SSR报错：ERROR ReferenceError: window is not defined
        if (typeof window === 'undefined') {
            return;
        }
        // 缓存在sessionStorage中的数据
        let sessionStorageState = window.sessionStorage.getItem(this.storageId);
        // 缓存在localStorage 中的数据
        let localStorageState = window.localStorage.getItem(this.storageId);
        // 版本发布后，用户第一次访问时，sessionStorageState、localStorageState是null
        this.sessionStorageState = sessionStorageState ? JSON.parse(sessionStorageState) : {};
        this.localStorageState = localStorageState ? JSON.parse(localStorageState) : {};
    }
    // 更新本地存储中的
    updateStorageDataState() {
        const localStorageState = Object.assign({}, this.localStorageState);
        const sessionStorageState = Object.assign({}, this.sessionStorageState);
        // 排序
        if (this.storageConfig.sort && this.dataState.sort.sortKey) {
            sessionStorageState['sort'] = this.dataState.sort;
        }
        // 表格分页
        // tslint:disable-next-line: max-line-length
        if ((this.storageConfig.pagination === true || (this.storageConfig.pagination && this.storageConfig.pagination['itemsPerPage'] === true)) &&
            this.dataState.pagination.itemsPerPage !== -1) {
            // 每页条数
            localStorageState['itemsPerPage'] = this.dataState.pagination.itemsPerPage;
        }
        // 表格分页
        // tslint:disable-next-line: max-line-length
        if ((this.storageConfig.pagination === true || (this.storageConfig.pagination && this.storageConfig.pagination['currentPage'] === true)) &&
            this.dataState.pagination.currentPage !== -1) {
            // 当前页
            sessionStorageState['currentPage'] = this.dataState.pagination.currentPage;
        }
        if (this.dataState.search.searchKeys) {
            localStorageState['searchKeys'] = this.dataState.search.searchKeys;
        }
        if (this.dataState.search.searchWords) {
            localStorageState['searchWords'] = this.dataState.search.searchWords;
        }
        if (Object.keys(sessionStorageState).length > 0) {
            // 修复：SSR报错：ReferenceError: window is not defined
            (typeof window !== 'undefined') && window.sessionStorage.setItem(this.storageId, JSON.stringify(sessionStorageState));
        }
        if (Object.keys(localStorageState).length > 0) {
            (typeof window !== 'undefined') && window.localStorage.setItem(this.storageId, JSON.stringify(localStorageState));
        }
    }
    /**
     * @ignore
     */
    updatePagination(currentPage, itemsPerPage) {
        if (this.oldPagination.currentPage === -1 &&
            this.oldPagination.itemsPerPage === -1) {
            this.oldPagination = {
                currentPage,
                itemsPerPage
            };
        }
        this.dataState.pagination.currentPage = currentPage;
        this.dataState.pagination.itemsPerPage = itemsPerPage;
    }
    /**
     * @ignore
     * 排序状态更新及处理
     *
     * @param sortKey - 进行排序的数据属性（如对数据的name属性值进行排序，则传入"name"）
     * @param reverse - 是否为降序，true表示降序，false表示升序，null表示无序
     * @param isDefaultSort - 是否为默认排序
     */
    updateSort(sortKey, reverse, isDefaultSort, compareFn) {
        if (isDefaultSort) {
            this.oldSort = {
                sortKey,
                asc: !reverse
            };
        }
        this.dataState.sort.sortKey = sortKey;
        this.dataState.sort.asc = Util.isNull(reverse) ? reverse : !reverse;
        this.customCompareFn = compareFn;
    }
    /**
     * 获取表格当前的数据状态(sort、search、pagination)
     * @return TiTableDataState
     */
    getDataState() {
        return this.dataState;
    }
    /**
     * 获取当前触发 stateUpdate 的事件(即引起表格数据状态改变的起因事件)
     * @return 有"search", "sort", "pagination"这三个值
     */
    getTriggerEvent() {
        return this.triggerEvent;
    }
    /**
     * 如果使用到了表格组件搜索功能，可通过此方法获取到搜索的数据结果。
     * @return 搜索到的数据集合
     *
     * 3.0.9新增
     */
    getSearchedResult() {
        return this.searchedResult;
    }
    updateTable() {
        if (this.isFrontPagination() || this.isFrontSort() || this.isFrontSearch()) {
            this.data2displayed();
        }
        // 延时的目的是防止stateUpdate 回调中同步修改数据时报错。
        this.promise.then(() => {
            this.stateUpdate.emit(this);
        });
    }
    /**
     * 根据数据状态(search,sort,pagination)处理数据
     */
    dataProcessor(data) {
        let output = data;
        // 如果存在过滤条件且开发者传递的源数据尚未进行过滤处理，则进行前台过滤处理
        if (this.isFrontSearch()) {
            output = TiTableComponent.getSearchedData(output, this.searchWords, this.searchKeys);
            this.searchedResult = [].concat(output);
            const totalNumber = output.length;
            // 如果存在分页且开发者传递的源数据尚未进行分页处理，即进行前台分页,
            if (this.isFrontPagination()) {
                this.promise.then(() => {
                    this.paginationSubject.next({ totalNumber });
                });
            }
        }
        // 如果存在排序且开发者传递的源数据尚未进行排序处理，则进行前台排序处理（字典排序）
        if (this.isFrontSort()) {
            output = this.getSortedData(output, this.dataState.sort.sortKey, this.dataState.sort.asc);
        }
        // 存在分页且开发者传递的源数据尚未进行分页处理，则进行前台分页
        if (this.isFrontPagination()) {
            const pagination = this.dataState.pagination;
            const start = (pagination.currentPage - 1) * pagination.itemsPerPage;
            output = output.slice(start, start + parseInt(pagination.itemsPerPage.toString(), 10));
        }
        return output;
    }
    /**
     * data => displayedData，处理分3步：
     * 1. 复制data数据
     * 2. 根据 分页、过滤、排序 处理数据
     * 3. 通过displayedDataChange 事件更新到displayedData
     */
    data2displayed() {
        // 复制源数据
        let output = TiTableComponent.safeCopy(this.data);
        // 处理数据
        output = this.dataProcessor(output);
        // 更新到displayedData
        this.promise.then(() => {
            // 同步修改内部数据
            this.displayedData = output;
            this.displayedDataChange.emit(output);
        });
    }
    isFrontSearch() {
        const srcState = this.getSrcState();
        return (this.searchWords && this.searchWords.length > 0) && (!srcState || !srcState.searched);
    }
    isFrontSort() {
        const srcState = this.getSrcState();
        return this.enableSort && (!srcState || !srcState.sorted);
    }
    isFrontPagination() {
        const srcState = this.getSrcState();
        const pagination = this.dataState.pagination;
        return (pagination.currentPage !== -1 && pagination.itemsPerPage !== -1 && (!srcState || !srcState.paginated));
    }
    getSortedData(data, predicate, asc) {
        let resultArr = [];
        if (!Util.isArray(data)) {
            return resultArr;
        }
        // 无序状态
        if (!predicate || asc === null) {
            return data;
        }
        const compareArr = data.map((value, index) => {
            const predicateValue = TiTableComponent.getPredicateValue(value, predicate);
            return {
                value,
                index,
                predicateValueObj: TiTableComponent.getPredicateValueObj(predicateValue, index)
            };
        });
        compareArr.sort(this.compareTo);
        resultArr = compareArr.map((item) => {
            return item.value;
        });
        if (!asc) {
            resultArr.reverse();
        }
        return resultArr;
    }
    updateSafeCopy() {
        this.data = TiTableComponent.safeCopy(this.srcData && this.srcData.data ?
            this.srcData.data : []);
    }
    getSrcState() {
        return this.srcData && this.srcData.state ? this.srcData.state : undefined;
    }
    /**
     * @ignore
     */
    closeOtherDetailsFn(currentRow) {
        this.displayedData.forEach((row) => {
            if (row !== currentRow && row['showDetails']) {
                this.promise.then(() => {
                    row['showDetails'] = false;
                });
            }
        });
    }
    initFixedHead() {
        this.theadContainer = this.fixedHeadService.getTheadContainer(this.hostEle);
        if (!this.theadContainer) {
            return;
        }
        this.tbodyContainer = this.fixedHeadService.getTbodyContainer(this.theadContainer);
        this.isFixedHead = !Util.isUndefined(this.theadContainer) && !Util.isUndefined(this.tbodyContainer);
        if (this.isFixedHead) {
            this.renderer.listen(this.tbodyContainer, 'scroll', () => {
                const scrollLeft = this.tbodyContainer.scrollLeft;
                this.renderer.setStyle(this.theadContainer.children[0], 'marginLeft', -scrollLeft + 'px');
                Util.trigger(document, 'tiScroll');
            });
            this.unlistenFixedHeadWindowResize = this.renderer.listen('window', 'resize', () => {
                this.fixedHeadService.processOverflowY(this.theadContainer, this.tbodyContainer, this);
            });
        }
    }
    getTbodyHeight() {
        // 修复SSR报错：ERROR TypeError: this.tbodyContainer.children[0].getBoundingClientRect is not a function
        if (typeof this.tbodyContainer.children[0].getBoundingClientRect !== 'function') {
            return 0;
        }
        return this.fixedHeadService.numRound(this.tbodyContainer.children[0].getBoundingClientRect().height, 0);
    }
    addFixedColumnBehavior() {
        const container = this.tiRenderer.findChildrenByClassName(this.hostEle, 'ti3-table-container')[0];
        if (!container) {
            return;
        }
        this.zone.runOutsideAngular(() => {
            this.renderer.listen(container, 'scroll', () => {
                const scrollLeft = container.scrollLeft;
                const isRightColumnFloat = scrollLeft + container.clientWidth < container.scrollWidth;
                this.containerScrollXChangeSubject.next({
                    scrollLeft,
                    isRightColumnFloat
                });
            });
            this.unlistenFixedColumnWindowResize = this.renderer.listen('window', 'resize', () => {
                const containerWidth = this.fixedColumnInfo.container.clientWidth;
                if (containerWidth !== this.fixedColumnInfo.containerWidth) {
                    this.fixedColumnInfo.containerWidth = containerWidth;
                    this.processFixedColumn();
                }
            });
        });
    }
    /**
     * @ignore
     */
    processFixedColumn() {
        const scrollLeft = this.fixedColumnInfo.container.scrollLeft;
        const isRightColumnFloat = scrollLeft + this.fixedColumnInfo.container.clientWidth < this.fixedColumnInfo.container.scrollWidth;
        this.containerScrollXChangeSubject.next({
            scrollLeft,
            isRightColumnFloat
        });
        this.updateFixedThLeftSubject.next();
        this.updateFixedTdLeftSubject.next();
    }
}
TiTableComponent.DEFAULT_SRC_DATA = {
    data: [],
    state: {
        searched: false,
        sorted: false,
        paginated: false
    }
};
TiTableComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-table',
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.ti3-table]': 'true'
                },
                styles: [".ti3-table-soldout{-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased;font-family:ti3Font!important;font-style:normal;font-variant:normal;font-weight:400;speak:none;text-rendering:auto;text-transform:none}.ti3-table-soldout:before{content:\"\\e92c\"}:root{--ti-table-box-model:border-box;--ti-table-cols-toggle-horizontal-padding:var(--ti-common-space-10);--ti-table-cols-toggle-searchbox-bottom-padding:var(--ti-common-space-base);--ti-table-cols-toggle-searchbox-top-padding:var(--ti-common-space-10);--ti-table-cols-toggle-vertical-padding:var(--ti-common-space-base);--ti-table-nodata-lead-icon-padding:var(--ti-common-space-5x);--ti-table-th-horizontal-padding:var(--ti-common-space-10)}.ti3-table{display:block}.table-nodata-a a{color:var(--ti-table-tr-bg-color-nodata-color);text-decoration:none}.table-nodata-a a:hover{color:var(--ti-table-tr-bg-color-nodata-color-hover);cursor:pointer;text-decoration:underline}.ti3-table>.ti3-resize-wrapper>table,.ti3-table>.ti3-table-container>table,.ti3-table>.ti3-table-fixed-head>table,.ti3-table>table{-ms-box-sizing:var(--ti-table-box-model);border-collapse:separate;border-spacing:0;border-width:0;box-sizing:var(--ti-table-box-model);empty-cells:show;margin-bottom:0;table-layout:fixed;width:100%}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th,.ti3-table>.ti3-table-container>table>thead>tr>th,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th,.ti3-table>table>thead>tr>th{-ms-box-sizing:var(--ti-table-box-model);background-color:var(--ti-table-th-bg-color);border-bottom:1px solid var(--ti-table-border-bottom-color);border-left:none;box-sizing:var(--ti-table-box-model);color:var(--ti-table-th-text-color);cursor:default;font-weight:var(--ti-table-th-text-font-weight);height:var(--ti-table-th-height);line-height:calc(var(--ti-table-th-height) - 1px);outline:0;overflow:hidden;padding:0;padding-left:var(--ti-table-th-horizontal-padding);padding-right:var(--ti-table-th-horizontal-padding);position:relative;text-align:left;text-overflow:ellipsis;vertical-align:middle;white-space:nowrap}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th:not(:last-child):after,.ti3-table>.ti3-table-container>table>thead>tr>th:not(:last-child):after,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th:not(:last-child):after,.ti3-table>table>thead>tr>th:not(:last-child):after{background:var(--ti-table-th-spacing-line-color);color:transparent;content:\"\";height:100%;position:absolute;right:0;top:0;width:1px}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[checkbox-column],.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[radio-column],.ti3-table>.ti3-table-container>table>thead>tr>th[checkbox-column],.ti3-table>.ti3-table-container>table>thead>tr>th[radio-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[radio-column],.ti3-table>table>thead>tr>th[checkbox-column],.ti3-table>table>thead>tr>th[radio-column]{font-size:0;overflow:visible;padding-right:var(--ti-table-column-icon-right-space);width:var(--ti-table-column-select-icon-width)!important}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[checkbox-column]:after,.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[radio-column]:after,.ti3-table>.ti3-table-container>table>thead>tr>th[checkbox-column]:after,.ti3-table>.ti3-table-container>table>thead>tr>th[radio-column]:after,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[checkbox-column]:after,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[radio-column]:after,.ti3-table>table>thead>tr>th[checkbox-column]:after,.ti3-table>table>thead>tr>th[radio-column]:after{width:0}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[checkbox-column],.ti3-table>.ti3-table-container>table>thead>tr>th[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[checkbox-column],.ti3-table>table>thead>tr>th[checkbox-column]{cursor:pointer}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th.ti3-disabled-cell,.ti3-table>.ti3-table-container>table>thead>tr>th.ti3-disabled-cell,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th.ti3-disabled-cell,.ti3-table>table>thead>tr>th.ti3-disabled-cell{cursor:not-allowed}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[details-icon-column],.ti3-table>.ti3-table-container>table>thead>tr>th[details-icon-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[details-icon-column],.ti3-table>table>thead>tr>th[details-icon-column]{overflow:visible;padding-right:var(--ti-table-column-icon-right-space);width:var(--ti-table-column-icon-width)!important}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th[details-icon-column]:after,.ti3-table>.ti3-table-container>table>thead>tr>th[details-icon-column]:after,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th[details-icon-column]:after,.ti3-table>table>thead>tr>th[details-icon-column]:after{width:0}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th:first-child[checkbox-column],.ti3-table>.ti3-resize-wrapper>table>thead>tr>th:first-child[radio-column],.ti3-table>.ti3-table-container>table>thead>tr>th:first-child[checkbox-column],.ti3-table>.ti3-table-container>table>thead>tr>th:first-child[radio-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th:first-child[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>thead>tr>th:first-child[radio-column],.ti3-table>table>thead>tr>th:first-child[checkbox-column],.ti3-table>table>thead>tr>th:first-child[radio-column]{padding-left:var(--ti-table-column-icon-left-space)!important;width:var(--ti-table-column-icon-width)!important}.ti3-table>.ti3-resize-wrapper>table>tbody>tr,.ti3-table>.ti3-table-container>table>tbody>tr,.ti3-table>.ti3-table-fixed-head>table>tbody>tr,.ti3-table>table>tbody>tr{background-color:var(--ti-table-tr-bg-color);border-spacing:0;transition:background-color .15s var(--ti-timing-function-default)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr:hover .cti-rename-edit,.ti3-table>.ti3-table-container>table>tbody>tr:hover .cti-rename-edit,.ti3-table>.ti3-table-fixed-head>table>tbody>tr:hover .cti-rename-edit,.ti3-table>table>tbody>tr:hover .cti-rename-edit{visibility:visible}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-selected-tr,.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-selected-tr:hover,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-selected-tr,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-selected-tr:hover,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-selected-tr,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-selected-tr:hover,.ti3-table>table>tbody>tr.ti3-selected-tr,.ti3-table>table>tbody>tr.ti3-selected-tr:hover{background-color:var(--ti-table-tr-bg-color-selected)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-disabled-tr,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-disabled-tr,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-disabled-tr,.ti3-table>table>tbody>tr.ti3-disabled-tr{background-color:var(--ti-table-tr-bg-color-disabled)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-disabled-tr>td,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-disabled-tr>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-disabled-tr>td,.ti3-table>table>tbody>tr.ti3-disabled-tr>td{color:var(--ti-table-tr-text-color-disabled)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-disabled-tr>td[checkbox-column],.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-disabled-tr>td[radio-column],.ti3-table>.ti3-table-container>table>tbody>tr.ti3-disabled-tr>td[checkbox-column],.ti3-table>.ti3-table-container>table>tbody>tr.ti3-disabled-tr>td[radio-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-disabled-tr>td[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-disabled-tr>td[radio-column],.ti3-table>table>tbody>tr.ti3-disabled-tr>td[checkbox-column],.ti3-table>table>tbody>tr.ti3-disabled-tr>td[radio-column]{cursor:not-allowed}.ti3-table>.ti3-resize-wrapper>table>tbody>tr:not(.ti3-details-tr):not(.ti3-disabled-tr):not(.ti3-table-nodata):not(.ti3-table-nodata-simple):not(.ti3-table-loadfail):not(.ti3-table-nodata-guide):not(.ti3-selected-tr):hover,.ti3-table>.ti3-table-container>table>tbody>tr:not(.ti3-details-tr):not(.ti3-disabled-tr):not(.ti3-table-nodata):not(.ti3-table-nodata-simple):not(.ti3-table-loadfail):not(.ti3-table-nodata-guide):not(.ti3-selected-tr):hover,.ti3-table>.ti3-table-fixed-head>table>tbody>tr:not(.ti3-details-tr):not(.ti3-disabled-tr):not(.ti3-table-nodata):not(.ti3-table-nodata-simple):not(.ti3-table-loadfail):not(.ti3-table-nodata-guide):not(.ti3-selected-tr):hover,.ti3-table>table>tbody>tr:not(.ti3-details-tr):not(.ti3-disabled-tr):not(.ti3-table-nodata):not(.ti3-table-nodata-simple):not(.ti3-table-loadfail):not(.ti3-table-nodata-guide):not(.ti3-selected-tr):hover{background-color:var(--ti-table-tr-bg-color-hover);transition:background-color .2s var(--ti-timing-function-default)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td,.ti3-table>.ti3-table-container>table>tbody>tr>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td,.ti3-table>table>tbody>tr>td{-ms-box-sizing:var(--ti-table-box-model);border-bottom:1px solid var(--ti-table-border-bottom-color);border-left:none;border-top:none;box-sizing:var(--ti-table-box-model);color:var(--ti-table-td-text-color);line-height:var(--ti-table-td-line-height);overflow:hidden;padding:var(--ti-table-td-padding);text-align:left;text-overflow:ellipsis;vertical-align:middle;white-space:nowrap}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td[checkbox-column],.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td[radio-column],.ti3-table>.ti3-table-container>table>tbody>tr>td[checkbox-column],.ti3-table>.ti3-table-container>table>tbody>tr>td[radio-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td[radio-column],.ti3-table>table>tbody>tr>td[checkbox-column],.ti3-table>table>tbody>tr>td[radio-column]{cursor:pointer;font-size:0;overflow:visible;padding-right:var(--ti-table-column-icon-right-space)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td.ti3-disabled-cell,.ti3-table>.ti3-table-container>table>tbody>tr>td.ti3-disabled-cell,.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td.ti3-disabled-cell,.ti3-table>table>tbody>tr>td.ti3-disabled-cell{cursor:not-allowed}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td[details-icon-column],.ti3-table>.ti3-table-container>table>tbody>tr>td[details-icon-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td[details-icon-column],.ti3-table>table>tbody>tr>td[details-icon-column]{overflow:visible;padding-left:var(--ti-table-column-icon-left-space);padding-right:var(--ti-table-column-icon-right-space)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td:first-child[checkbox-column],.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td:first-child[radio-column],.ti3-table>.ti3-table-container>table>tbody>tr>td:first-child[checkbox-column],.ti3-table>.ti3-table-container>table>tbody>tr>td:first-child[radio-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td:first-child[checkbox-column],.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td:first-child[radio-column],.ti3-table>table>tbody>tr>td:first-child[checkbox-column],.ti3-table>table>tbody>tr>td:first-child[radio-column]{padding-left:var(--ti-table-column-icon-left-space)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata,.ti3-table>table>tbody>tr.ti3-table-nodata{background-color:var(--ti-table-tr-bg-color-nodata)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata>td,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata>td,.ti3-table>table>tbody>tr.ti3-table-nodata>td{background:var(--ti-table-nodata-td-bg-img-url) 50% var(--ti-table-nodata-lead-icon-padding) no-repeat;border-bottom:none;color:var(--ti-table-nodata-td-color);font-size:var(--ti-table-nodata-td-font-size);height:var(--ti-table-nodata-height);line-height:20px;overflow:visible!important;padding:calc(var(--ti-table-nodata-lead-icon-height) + var(--ti-table-nodata-lead-icon-padding)*2) 0 0 0;position:relative;text-align:center;vertical-align:top}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail,.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-loadfail,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide,.ti3-table>table>tbody>tr.ti3-table-loadfail,.ti3-table>table>tbody>tr.ti3-table-nodata-guide{background-color:var(--ti-table-tr-bg-color-nodata)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td,.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table>table>tbody>tr.ti3-table-loadfail>td,.ti3-table>table>tbody>tr.ti3-table-nodata-guide>td{background:none;border-bottom:none;color:var(--ti-table-nodata-td-color);font-size:var(--ti-table-nodata-td-font-size);line-height:20px;overflow:visible!important;padding:var(--ti-common-space-5x) 0;position:relative;text-align:center;vertical-align:top}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table>table>tbody>tr.ti3-table-nodata-guide>td a{color:var(--ti-table-tr-bg-color-nodata-color);text-decoration:none}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table>table>tbody>tr.ti3-table-nodata-guide>td a:hover{color:var(--ti-table-tr-bg-color-nodata-color-hover);cursor:pointer;text-decoration:underline}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-simple,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-simple,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-simple,.ti3-table>table>tbody>tr.ti3-table-nodata-simple{background-color:var(--ti-table-tr-bg-color-nodata)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-simple>td,.ti3-table>.ti3-table-container>table>tbody>tr.ti3-table-nodata-simple>td,.ti3-table>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-simple>td,.ti3-table>table>tbody>tr.ti3-table-nodata-simple>td{background:none;border-bottom:none;color:var(--ti-table-nodata-td-color);font-size:var(--ti-table-nodata-td-font-size);line-height:20px;overflow:visible!important;padding:var(--ti-common-space-5x) 0;position:relative;text-align:center;vertical-align:top}.ti3-table-small>.ti3-resize-wrapper>table>thead>tr>th,.ti3-table-small>.ti3-table-container>table>thead>tr>th,.ti3-table-small>.ti3-table-fixed-head>table>thead>tr>th,.ti3-table-small>table>thead>tr>th{height:var(--ti-table-small-th-height);line-height:var(--ti-table-small-th-height)}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr>td,.ti3-table-small>.ti3-table-container>table>tbody>tr>td,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr>td,.ti3-table-small>table>tbody>tr>td{height:var(--ti-table-small-td-height);line-height:calc(var(--ti-table-small-td-height) - var(--ti-table-small-td-vertical-padding)*2);padding:var(--ti-table-small-td-vertical-padding) var(--ti-table-small-td-horizontal-padding)}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-details-tr>td,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-details-tr>td,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-details-tr>td,.ti3-table-small>table>tbody>tr.ti3-details-tr>td{padding:var(--ti-table-small-details-padding)}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata>td,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-nodata>td,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata>td,.ti3-table-small>table>tbody>tr.ti3-table-nodata>td{background:var(--ti-table-nodata-td-bg-img-url) 50% var(--ti-table-nodata-lead-icon-padding) no-repeat;height:var(--ti-table-small-nodata-height);padding:calc(var(--ti-table-nodata-lead-icon-height) + var(--ti-table-nodata-lead-icon-padding)*2) 0 0 0;vertical-align:top}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td,.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-small>table>tbody>tr.ti3-table-loadfail>td,.ti3-table-small>table>tbody>tr.ti3-table-nodata-guide>td{background:none;vertical-align:top}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-small>table>tbody>tr.ti3-table-loadfail>td a,.ti3-table-small>table>tbody>tr.ti3-table-nodata-guide>td a{color:var(--ti-table-tr-bg-color-nodata-color);text-decoration:none}.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table-small>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table-small>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table-small>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-small>table>tbody>tr.ti3-table-loadfail>td a:hover,.ti3-table-small>table>tbody>tr.ti3-table-nodata-guide>td a:hover{color:var(--ti-table-tr-bg-color-nodata-color-hover);cursor:pointer;text-decoration:underline}.ti3-table-nest>.ti3-resize-wrapper>table>thead>tr>th,.ti3-table-nest>.ti3-table-container>table>thead>tr>th,.ti3-table-nest>.ti3-table-fixed-head>table>thead>tr>th,.ti3-table-nest>table>thead>tr>th{height:var(--ti-table-nest-th-height);line-height:var(--ti-table-nest-th-height)}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata>td,.ti3-table-nest>table>tbody>tr.ti3-table-nodata>td{background:var(--ti-table-nodata-td-bg-img-url) 50% var(--ti-table-nodata-lead-icon-padding) no-repeat;height:var(--ti-table-small-nodata-height);padding:calc(var(--ti-table-nodata-lead-icon-height) + var(--ti-table-nodata-lead-icon-padding)*2) 0 0 0;vertical-align:top}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td{background:none;vertical-align:top}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a{color:var(--ti-table-tr-bg-color-nodata-color);text-decoration:none}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-table-nodata-guide>td a:hover,.ti3-table-nest>table>tbody>tr.ti3-table-loadfail .ti3-table-nest>table>tbody>tr.ti3-table-nodata-guide>td a:hover{color:var(--ti-table-tr-bg-color-nodata-color-hover);cursor:pointer;text-decoration:underline}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr>td,.ti3-table-nest>table>tbody>tr>td{border-bottom:1px solid var(--ti-table-nest-td-border-bottom-color);height:var(--ti-table-nest-td-height);line-height:calc(var(--ti-table-nest-td-height) - var(--ti-table-small-td-vertical-padding)*2);padding:var(--ti-table-small-td-vertical-padding) var(--ti-table-small-td-horizontal-padding)}.ti3-table-nest>.ti3-resize-wrapper>table>tbody>tr.ti3-details-tr>td,.ti3-table-nest>.ti3-table-container>table>tbody>tr.ti3-details-tr>td,.ti3-table-nest>.ti3-table-fixed-head>table>tbody>tr.ti3-details-tr>td,.ti3-table-nest>table>tbody>tr.ti3-details-tr>td{padding:var(--ti-table-small-details-padding)}.ti3-resize-wrapper{min-height:0;overflow-x:auto}.ti3-unselectable{-ms-user-select:none;-webkit-user-select:none;user-select:none}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th.col-resize-active,.ti3-table>.ti3-table-container>table>thead>tr>th.col-resize-active,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th.col-resize-active,.ti3-table>table>thead>tr>th.col-resize-active{cursor:col-resize!important}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th.col-resize-active:after,.ti3-table>.ti3-table-container>table>thead>tr>th.col-resize-active:after,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th.col-resize-active:after,.ti3-table>table>thead>tr>th.col-resize-active:after{background:transparent;border-right:var(--ti-table-col-resize-line);bottom:0;content:\"\";height:100%;position:absolute;right:0;top:0;width:2px}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td.col-resize-active,.ti3-table>.ti3-table-container>table>tbody>tr>td.col-resize-active,.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td.col-resize-active,.ti3-table>table>tbody>tr>td.col-resize-active{position:relative}.ti3-table>.ti3-resize-wrapper>table>tbody>tr:not(.ti3-table-nodata)>td.col-resize-active:after,.ti3-table>.ti3-table-container>table>tbody>tr:not(.ti3-table-nodata)>td.col-resize-active:after,.ti3-table>.ti3-table-fixed-head>table>tbody>tr:not(.ti3-table-nodata)>td.col-resize-active:after,.ti3-table>table>tbody>tr:not(.ti3-table-nodata)>td.col-resize-active:after{border-right:var(--ti-table-col-resize-line);bottom:0;content:\"\";height:100%;position:absolute;right:0;top:0;width:2px}.ti3-cols-toggle-container{display:inline-block}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container{background:var(--ti-button-default-normal-bg-color);border:1px solid var(--ti-table-cols-toggle-menu-border-color);border-radius:var(--ti-common-border-radius-normal);box-sizing:border-box;cursor:pointer;display:inline-block;height:var(--ti-table-cols-toggle-menu-size);line-height:var(--ti-table-cols-toggle-menu-line-height);position:relative;text-align:center;width:var(--ti-table-cols-toggle-menu-size)}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container .ti3-cols-toggle-menu{color:var(--ti-table-cols-toggle-menu-color);font-size:var(--ti-table-cols-toggle-menu-font-size)}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container:focus,.ti3-cols-toggle-container .ti3-cols-toggle-menu-container:hover{background-color:var(--ti-button-default-bg-color-hover)}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container:focus .ti3-cols-toggle-menu,.ti3-cols-toggle-container .ti3-cols-toggle-menu-container:hover .ti3-cols-toggle-menu{color:var(--ti-table-cols-toggle-menu-color-hover)}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container.ti3-cols-toggle-menu-disabled{border-color:var(--ti-table-cols-toggle-menu-border-color-disabled);cursor:not-allowed}.ti3-cols-toggle-container .ti3-cols-toggle-menu-container.ti3-cols-toggle-menu-disabled .ti3-cols-toggle-menu{color:var(--ti-table-cols-toggle-menu-color-disabled)}.ti3-cols-toggle-container[tiFocused] .ti3-cols-toggle-menu-container:not(.ti3-cols-toggle-menu-disabled) .ti3-cols-toggle-menu{color:var(--ti-table-cols-toggle-menu-color-hover)}.ti3-cols-toggle-container .ti3-cols-toggle-drop-container{display:inline-block}.ti3-cols-toggle-drop{border-radius:var(--ti-table-cols-toggle-dropdown-border-radius);box-shadow:var(--ti-table-cols-toggle-dropdown-box-shadow);max-width:300px!important;min-width:200px!important}.ti3-cols-toggle-drop>.ti3-cols-toggle-searchbox-wrap{-ms-box-sizing:border-box;box-sizing:border-box;padding:var(--ti-table-cols-toggle-searchbox-top-padding) var(--ti-table-cols-toggle-horizontal-padding) var(--ti-table-cols-toggle-searchbox-bottom-padding)!important}.ti3-cols-toggle-drop>.ti3-cols-toggle-searchbox-wrap .ti3-cols-toggle-searchbox{width:100%!important}.ti3-cols-toggle-drop .ti3-overflow-padding{padding:0 var(--ti-table-cols-toggle-horizontal-padding)!important}.ti3-cols-toggle-drop .ti3-dropdown-no-data{padding-left:var(--ti-table-cols-toggle-horizontal-padding)!important}.ti3-cols-toggle-drop .ti3-dropdown-option-hover{color:var(--ti-common-color-text-primary)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container,.ti3-table>table>thead>tr>th>.ti3-sort-container{cursor:pointer;display:inline-block;outline-offset:-1px}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container .ti3-sort,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container .ti3-sort,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container .ti3-sort,.ti3-table>table>thead>tr>th>.ti3-sort-container .ti3-sort{cursor:pointer;display:inline-block;font-size:16px;height:16px;line-height:16px;margin-left:8px;vertical-align:middle;width:16px}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container .ti3-sort:hover,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container .ti3-sort:hover,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container .ti3-sort:hover,.ti3-table>table>thead>tr>th>.ti3-sort-container .ti3-sort:hover{color:var(--ti-table-sort-color-hover)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container .ti3-sort:before,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container .ti3-sort:before,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container .ti3-sort:before,.ti3-table>table>thead>tr>th>.ti3-sort-container .ti3-sort:before{content:\"\\e917\";display:inline-block;height:16px;line-height:16px;width:16px}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort{color:var(--ti-table-sort-icon-color-active)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:hover,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:hover,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:hover,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:hover{color:var(--ti-table-sort-color-hover)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:before,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:before,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:before,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-ascent .ti3-sort:before{content:\"\\e92e\"}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort{color:var(--ti-table-sort-icon-color-active)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:hover,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:hover,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:hover,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:hover{color:var(--ti-table-sort-color-hover)}.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:before,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:before,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:before,.ti3-table>table>thead>tr>th>.ti3-sort-container.ti3-sort-descent .ti3-sort:before{content:\"\\e92f\"}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>.ti3-table-container>table>tbody>tr>td>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>table>tbody>tr>td>.ti3-cell-text-container>.ti3-cell-text,.ti3-table>table>thead>tr>th>.ti3-cell-text-container>.ti3-cell-text{display:inline-block;vertical-align:top}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td>.ti3-cell-icons-container,.ti3-table>.ti3-resize-wrapper>table>thead>tr>th>.ti3-cell-icons-container,.ti3-table>.ti3-table-container>table>tbody>tr>td>.ti3-cell-icons-container,.ti3-table>.ti3-table-container>table>thead>tr>th>.ti3-cell-icons-container,.ti3-table>.ti3-table-fixed-head>table>tbody>tr>td>.ti3-cell-icons-container,.ti3-table>.ti3-table-fixed-head>table>thead>tr>th>.ti3-cell-icons-container,.ti3-table>table>tbody>tr>td>.ti3-cell-icons-container,.ti3-table>table>thead>tr>th>.ti3-cell-icons-container{display:inline-block}.ti3-table>.ti3-resize-wrapper>table>tbody>.ti3-details-tr,.ti3-table>.ti3-table-container>table>tbody>.ti3-details-tr,.ti3-table>table>tbody>.ti3-details-tr{background:var(--ti-table-details-bg-color)}.ti3-table>.ti3-resize-wrapper>table>tbody>.ti3-details-tr>td,.ti3-table>.ti3-table-container>table>tbody>.ti3-details-tr>td,.ti3-table>table>tbody>.ti3-details-tr>td{border-bottom:1px solid var(--ti-table-border-bottom-color);padding:var(--ti-table-details-padding);white-space:normal}.ti3-table>.ti3-resize-wrapper>table>tbody>.ti3-details-tr>td>.ti3-table-detail-container,.ti3-table>.ti3-table-container>table>tbody>.ti3-details-tr>td>.ti3-table-detail-container,.ti3-table>table>tbody>.ti3-details-tr>td>.ti3-table-detail-container{background-color:var(--ti-common-color-bg-white-normal);padding:var(--ti-common-space-5x)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-toggle-details,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-toggle-details,.ti3-table>table>tbody>tr>td .ti3-toggle-details{color:var(--ti-table-details-icon-color);cursor:pointer}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-toggle-details .ti3-icon,.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-toggle-details .ti3-icon,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout,.ti3-table>table>tbody>tr>td .ti3-toggle-details .ti3-icon,.ti3-table>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout{font-size:15px;vertical-align:middle}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-toggle-details .ti3-icon:hover,.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout:hover,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-toggle-details .ti3-icon:hover,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout:hover,.ti3-table>table>tbody>tr>td .ti3-toggle-details .ti3-icon:hover,.ti3-table>table>tbody>tr>td .ti3-toggle-details .ti3-table-soldout:hover{color:var(--ti-table-details-icon-color-hover)}.ti3-icon-angle-transform-up-thin{display:inline-block;transform:rotate(-90deg)}.ti3-icon-angle-transform-down-thin{display:inline-block;transform:rotate(90deg)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree,.ti3-table>table>tbody>tr>td .ti3-table-tree{cursor:pointer}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square,.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square{font-size:var(--ti-table-tree-square-icon-font-size);line-height:18px;margin-right:var(--ti-table-tree-square-right-space);vertical-align:bottom}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square{color:var(--ti-table-tree-minus-square-color)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square{color:var(--ti-table-tree-plus-square-color)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:hover,.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:hover,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:hover,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:hover,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:hover,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:hover{color:var(--ti-table-tree-minus-square-color-hover)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:active,.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:active,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:active,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:active,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-minus-square:active,.ti3-table>table>tbody>tr>td .ti3-table-tree .ti3-icon-plus-square:active{background:var(--ti-table-tree-square-bg);color:var(--ti-table-tree-minus-square-color)}.ti3-table>.ti3-resize-wrapper>table>tbody>tr>td .ti3-table-tree-no-leaf,.ti3-table>.ti3-table-container>table>tbody>tr>td .ti3-table-tree-no-leaf,.ti3-table>table>tbody>tr>td .ti3-table-tree-no-leaf{margin-left:calc(var(--ti-table-tree-square-icon-font-size) + var(--ti-table-tree-square-right-space))}.ti3-table-cell-clone{-ms-box-sizing:var(--ti-table-box-model);box-sizing:var(--ti-table-box-model);overflow:visible!important;position:absolute!important;visibility:hidden!important;width:auto!important}.ti3-table>.ti3-table-fixed-head{overflow:hidden;position:relative}.ti3-table>.ti3-table-fixed-head>table{margin-left:0}.ti3-table>.ti3-table-fixed-head>.ti3-table-fixed-head-filler{background-color:var(--ti-table-th-bg-color);bottom:0;height:100%;position:absolute;right:0;top:0}.ti3-table>.ti3-table-fixed-head+.ti3-table-container{border-bottom:1px solid var(--ti-table-border-bottom-color)}.ti3-table>.ti3-table-fixed-head+.ti3-table-container>table>thead>tr>th{height:0}.ti3-table>.ti3-table-fixed-head+.ti3-table-container>table>tbody>tr:last-child>td{border-bottom:none!important}.ti3-table>.ti3-table-container{overflow:auto}.ti3-table-column-fixed-left{background-color:inherit;left:0;position:sticky!important;z-index:1}.ti3-table-column-fixed-left.ti3-table-floating-fixed-column{box-shadow:5px 0 5px 0 rgba(0,0,0,.1)}.ti3-table-column-fixed-right{background-color:inherit;position:sticky!important;right:0;z-index:1}.ti3-table-column-fixed-right.ti3-table-floating-fixed-column{box-shadow:-5px 0 5px 0 rgba(0,0,0,.1)}.ti3-table-row-disable{color:#999}.ti3-table-soldout{background:#999;color:#fff;display:inline-block;font-size:12px;height:16px;line-height:16px;text-align:center;width:16px}.ti3-table-row-message,.ti3-table-row-message-new,.ti3-table-row-message-recommended,.ti3-table-row-message-soldout{color:#fff;display:inline-block;font-size:12px;height:16px;line-height:16px;margin-left:5px;padding:0 4px;text-align:center}.ti3-table-row-message-soldout{background:#999}.ti3-table-row-message-new{background:#e41f2b}.ti3-table-row-message-recommended{background:#f57f06}@keyframes ti3-table-tr-entrance-keyframes{0%{transform:scaleX(.98);transform-origin:right center}to{transform:scaleX(1);transform-origin:right center}}.ti3-table:not([ti-table-tr-no-animate]) tbody tr:not(.ti3-details-tr).ng-enter{animation:ti3-table-tr-entrance-keyframes .3s cubic-bezier(.4,0,.2,1),ti3-fade-in-keyframes .3s cubic-bezier(.4,0,.2,1)}.ti3-table:not([ti-table-tr-no-animate]) tbody tr.ng-enter-stagger{animation-delay:.01s}@keyframes ti3-details-tr-open-keyframes{0%{transform:scaleY(0);transform-origin:top center}to{transform:scaleY(1);transform-origin:top center}}.ti3-details-tr.ng-enter{animation:ti3-details-tr-open-keyframes .15s cubic-bezier(.4,0,.2,1) forwards,ti3-fade-in-keyframes .15s cubic-bezier(.4,0,.2,1) forwards}"]
            },] }
];
TiTableComponent.ctorParameters = () => [
    { type: IterableDiffers },
    { type: ElementRef },
    { type: TiTableFixedHeadService },
    { type: Renderer2 },
    { type: NgZone },
    { type: TiRenderer }
];
TiTableComponent.propDecorators = {
    srcData: [{ type: Input }],
    displayedData: [{ type: Input }],
    storageId: [{ type: Input }],
    storageConfig: [{ type: Input }],
    displayedDataChange: [{ type: Output }],
    columns: [{ type: Input }],
    searchWords: [{ type: Input }],
    searchKeys: [{ type: Input }],
    closeOtherDetails: [{ type: Input }],
    stateUpdate: [{ type: Output }]
};
