// @cloud/tinyplus3/esm2015/components/modalbody/TpModalBodyComponent.js
import { Component, ViewEncapsulation } from '@angular/core';
import { TiModalBodyComponent } from '@cloud/tiny3';
/**
 *
 *  业务典型场景用到的弹窗组件
 *
 *  10.1.0新增
 *
 *  提供了UI规范推荐的四种尺寸的弹窗样式，还有带alert提示的弹窗和典型confirm弹窗的样式
 *
 *  10.0.4及之前的版本是以模板的形式提供(相关的样式是全局的）；10.1.0版本为了实现样式按需加载，去掉了之前的模板，将其封装在一个组件模块，所以该组件实际只提供了样式。
 *
 *  <example-url>../tinyplus3demo/#/modalbody/modalbody-all</example-url>
 *
 */
export class TpModalBodyComponent extends TiModalBodyComponent {
}
TpModalBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-modal-body',
                template: '<ng-content></ng-content>',
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.ti3-modal-body]': 'true' // 1. tiny3中的 .ti3-modal-body 有些样式；2. TiModalComponent 中有通过该样式类获取dom的逻辑
                },
                styles: ["tp-modal-body.ti3-modal-body{overflow:auto;position:static!important}.ti3-modal-dialog.tp-modal-xl{width:900px}.ti3-modal-dialog.tp-modal-lg{width:700px}.ti3-modal-dialog.tp-modal-md{width:550px}.ti3-modal-dialog.tp-modal-sm{width:400px}ti-alert.tp-modal-body-tip{left:32px;position:absolute;top:40px;width:calc(100% - 64px)}.tp-modal-confirm{font-size:0}.tp-modal-confirm .tp-modal-confirm-icon{color:#ffd939;display:inline-block;font-size:45px;margin-right:var(--ti-common-space-5x);vertical-align:top}.tp-modal-confirm .tp-modal-confirm-text{display:inline-block;vertical-align:top;width:calc(100% - 68px)}.tp-modal-confirm .tp-modal-confirm-text .tp-modal-confirm-title{color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-3)}.tp-modal-confirm .tp-modal-confirm-text .tp-modal-confirm-detail{color:var(--ti-common-color-text-secondary);font-size:var(--ti-common-font-size-1);margin-top:5px}.tp-modal-confirm.tp-modal-confirm-no-detail .tp-modal-confirm-icon,.tp-modal-confirm.tp-modal-confirm-no-detail .tp-modal-confirm-text{vertical-align:middle}"]
            },] }
];
