// @cloud/tinyplus3/esm2015/components/card/TpCardModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TpIconModule } from '../icon/TpIconModule';
import { TpCardComponent } from './TpCardComponent';
import { TpCardAddComponent } from './TpCardAddComponent';
import { TpCardHeaderComponent } from './TpCardHeaderComponent';
export class TpCardModule {
}
TpCardModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TpIconModule
                ],
                declarations: [
                    TpCardComponent,
                    TpCardAddComponent,
                    TpCardHeaderComponent
                ],
                exports: [
                    TpCardComponent,
                    TpCardAddComponent,
                    TpCardHeaderComponent
                ]
            },] }
];
export { TpCardComponent } from './TpCardComponent';
export { TpCardAddComponent } from './TpCardAddComponent';
export { TpCardHeaderComponent } from './TpCardHeaderComponent';
