// @cloud/tinyplus3/esm2015/components/card/TpCardComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 卡片组件
 *
 * <example-url>../tinyplus3demo/#/card/card-all</example-url>
 */
export class TpCardComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 卡片禁用状态
         */
        this.disabled = false;
        /**
         * 卡片激活选中状态
         */
        this.active = false;
    }
}
TpCardComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-card',
                template: `
      <ng-content ></ng-content>
    `,
                host: {
                    '[class.tp-card-disabled]': 'disabled',
                    '[class.tp-card-active]': 'active',
                    '[attr.tabindex]': 'disabled? "-1":"0"'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host{-ms-box-sizing:border-box;background:var(--ti-common-color-bg-white-normal);border:1px solid var(--ti-common-color-line-normal);border-radius:var(--ti-common-border-radius-normal);box-sizing:border-box;display:block;min-width:200px;padding:var(--ti-common-space-5x);position:relative}:host:not(.tp-card-disabled):active,:host:not(.tp-card-disabled):hover{border-color:var(--ti-common-color-line-active);transition:border-color .2s}:host:not(.tp-card-disabled):focus{outline:none}:host.tp-card-active,:host:not(.tp-card-disabled):focus{border-color:var(--ti-common-color-line-active)}:host.tp-card-disabled{background-color:var(--ti-common-color-bg-disabled);border:1px solid var(--ti-common-color-line-disabled);color:var(--ti-common-color-text-disabled);outline:none}:host.tp-card-disabled:hover{border-color:var(--ti-common-color-line-disabled);cursor:not-allowed}"]
            },] }
];
TpCardComponent.propDecorators = {
    disabled: [{ type: Input }],
    active: [{ type: Input }]
};
