// @cloud/tinyplus3/esm2015/components/card/TpCardAddComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 *  特殊卡片组件，添加卡片组件
 *
 * <example-url>../tinyplus3demo/#/card/card-all</example-url>
 */
export class TpCardAddComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 禁用状态
         */
        this.disabled = false;
        /**
         * 点击事件
         */
        this.add = new EventEmitter();
    }
    /**
     * @ignore
     * 加号点击回调
     */
    onClick($event) {
        if (this.disabled) {
            return;
        }
        this.add.emit($event);
    }
}
TpCardAddComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-card-add',
                template: "<tp-icon local name=\"cloud-action-add\" class=\"tp-card-add-icon\" (click)=\"onClick($event)\" [id]=\"appendId('add')\"></tp-icon>",
                host: {
                    '[class.tp-card-add-disabled]': 'disabled',
                    '[attr.tabindex]': 'disabled? "-1":"0"'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-card-add-icon-width:16px}:host{-ms-box-sizing:border-box;background-color:var(--ti-common-color-bg-white-normal);border:1px dashed var(--ti-common-color-line-normal);border-radius:var(--ti-common-border-radius-normal);box-sizing:border-box;display:block;min-width:200px;outline:none;padding:var(--ti-common-space-5x);position:relative}:host:not(.tp-card-add-disabled):active,:host:not(.tp-card-add-disabled):hover{border-color:var(--ti-common-color-line-active)}:host:not(.tp-card-add-disabled):active .tp-card-add-icon,:host:not(.tp-card-add-disabled):hover .tp-card-add-icon{color:var(--ti-common-color-icon-active)}:host.tp-card-add-disabled{border-color:var(--ti-common-color-line-disabled);color:var(--ti-common-color-text-disabled);cursor:not-allowed}:host.tp-card-add-disabled:active,:host.tp-card-add-disabled:hover{border-color:var(--ti-common-color-line-disabled);outline:none}.tp-card-add-icon{color:var(--ti-common-color-icon-normal);cursor:pointer;font-size:var(--ti-common-font-size-2);left:50%;line-height:1;margin-left:-8px;margin-top:-8px;position:absolute;top:50%}:host.tp-card-add-disabled>.tp-card-add-icon{color:var(--ti-common-color-text-disabled)}:host.tp-card-add-disabled:hover>.tp-card-add-icon{cursor:not-allowed}"]
            },] }
];
TpCardAddComponent.propDecorators = {
    disabled: [{ type: Input }],
    add: [{ type: Output }]
};
