// @cloud/tinyplus3/esm2015/components/beta/TpBetaComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 公测组件
 *
 * <example-url>../tinyplus3demo/#/beta/beta-all</example-url>
 */
export class TpBetaComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 了解更多跳转链接
         */
        this.moreUrl = '';
        /**
         * 服务介绍描述
         */
        this.intro = '';
        /**
         * 服务消息描述
         */
        this.message = '';
        /**
         * 是否实名认证，未实名认证情况会有提示，并且申请按钮灰化
         */
        this.verified = true;
        /**
         * 立即申请跳转链接
         */
        this.applyUrl = '';
    }
    /**
     * @ignore
     */
    onClickFn(event) {
        if (!this.verified) {
            event.preventDefault();
        }
    }
}
TpBetaComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-beta',
                template: "<tp-icon [name]='icon' class='tp-beta-icon'></tp-icon>\r\n<div class='tp-beta-intro'>\r\n    <span class='tp-beta-intro-desc' [id]=\"appendId('intro')\">{{intro}}</span>\r\n    <a [href]='moreUrl' class='tp-link-no-decoration'[id]=\"appendId('more')\">{{'tpBeta.more' | tiTranslate}}</a>\r\n</div>\r\n<span class='tp-beta-message' *ngIf='message' [id]=\"appendId('message')\">{{message}}</span>\r\n<tp-info class=\"tp-beta-info\" *ngIf='!verified'></tp-info>\r\n<a [href]='applyUrl' class='tp-beta-apply' target='_blank' [tabindex]=\"-1\" (click)='onClickFn($event)' [id]=\"appendId('button')\" rel=\"noopener noreferrer\">\r\n    <button tiButton color='danger' size='large' [disabled]='!verified'>{{'tpBeta.apply' | tiTranslate}}</button>\r\n</a>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.tp-link-no-decoration{color:var(--ti-common-color-text-link);cursor:pointer;text-decoration:none}.tp-link-no-decoration:hover{color:var(--tp-link-no-decoration-color-hover);text-decoration:var(--tp-link-no-decoration-text-decoration)}::ng-deep :root{--tp-beta-icon-size:44px}:host{align-items:center;background-color:var(--ti-common-color-bg-white-normal);box-sizing:border-box;display:flex;flex-direction:column;padding-top:150px}.tp-beta-icon{color:var(--ti-common-color-icon-normal);font-size:var(--tp-beta-icon-size);height:var(--tp-beta-icon-size);text-align:center;width:var(--tp-beta-icon-size)}.tp-beta-intro{font-size:var(--ti-common-font-size-2);line-height:var(--ti-common-line-height-number);margin:34px 0 20px}.tp-beta-intro-desc{margin-bottom:var(--ti-common-space-5x)}.tp-beta-apply{margin-top:var(--ti-common-space-5x)}.tp-beta-info{margin-top:var(--ti-common-space-10)}"]
            },] }
];
TpBetaComponent.propDecorators = {
    icon: [{ type: Input }],
    moreUrl: [{ type: Input }],
    intro: [{ type: Input }],
    message: [{ type: Input }],
    verified: [{ type: Input }],
    applyUrl: [{ type: Input }]
};
