// @cloud/tinyplus3/esm2015/components/crumb/TpCrumbComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 *  crumb面包屑组件
 *
 * 面包屑组件除最后一级外，其余几级一般都为可跳转的链接，链接地址可以通过href属性设置;
 * 用户也可通过添加事件，实现业务逻辑。
 *
 * <example-url>../tinyplus3demo/#/crumb/crumb-all</example-url>
 */
export class TpCrumbComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 用户设置href属性且不是最后一级时，触发事件
         */
        this.select = new EventEmitter();
    }
    /**
     * @ignore
     */
    onClick(item) {
        this.select.emit(item);
        // 不设置链接时，阻止跳转（a标签在click事件返回值为false时，会阻止默认行为）
        if (!item.href) {
            return false;
        }
    }
}
TpCrumbComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-crumb',
                template: "<ng-container *ngFor=\"let item of items; index as i let last=last;\">\r\n    <ng-container *ngIf=\"!last\">\r\n        <a class=\"tp-crumb-label\"\r\n            rel=\"noopener noreferrer\"\r\n           *ngIf=\"!item.routerLink\"\r\n           [href]=\"item.href\"\r\n           [target]=\"item.target || '_self'\"\r\n           (click)=\"onClick(item)\"\r\n           tiOverflow\r\n           tiTipPosition=\"bottom-left\"\r\n           [id]=\"appendId('label_' + i)\">{{item.label}}</a>\r\n        <a class=\"tp-crumb-label\"\r\n           *ngIf=\"item.routerLink\"\r\n           [routerLink]=\"item.routerLink\"\r\n           [queryParams]=\"item.queryParams\"\r\n           (click)=\"onClick(item)\"\r\n           tiOverflow\r\n           tiTipPosition=\"bottom-left\"\r\n           [id]=\"appendId('label_' + i)\">{{item.label}}</a>\r\n        <span class=\"tp-crumb-divider\">/</span>\r\n    </ng-container>\r\n    <span *ngIf=\"last\" class=\"tp-crumb-nohref-label\" tiOverflow tiTipPosition=\"bottom-left\" [id]=\"appendId('last_label')\">{{item.label}}</span>\r\n</ng-container>",
                styles: [":host{display:flex;line-height:var(--ti-common-line-height-number)}.tp-crumb-label{color:var(--ti-common-color-text-weaken);flex:0 1 auto!important;max-width:160px;text-decoration:none;vertical-align:bottom}.tp-crumb-label:hover{color:var(--ti-common-color-text-link)}.tp-crumb-nohref-label{flex:1;font-weight:700;text-decoration:none;vertical-align:bottom}.tp-crumb-nohref-label,.tp-crumb-nohref-label:hover{color:var(--ti-common-color-text-primary)}.tp-crumb-divider{color:var(--ti-common-color-text-weaken);display:inline-block;text-align:center;width:16px}"]
            },] }
];
TpCrumbComponent.propDecorators = {
    items: [{ type: Input }],
    select: [{ type: Output }]
};
