// @cloud/tinyplus3/esm2015/services/goandback/TpGoandbackComponent.js
import { Component, Input } from '@angular/core';
/**
 * @ignore
 * goandback模板组件定义
 */
export class TpGoandbackComponent {
}
TpGoandbackComponent.decorators = [
    { type: Component, args: [{
                template: "<ti-alert class=\"tp-goandback-alert\"\r\n          [type]=\"warn.type\"\r\n          [(open)]=\"warn.open\"\r\n          dismissOnTimeout=\"10000\"\r\n          [id]=\"warn.id\">\r\n         <span [innerHtml]=\"words.secondMessageFail\"></span>\r\n          <a [href]=\"href\" target=\"_blank\" rel=\"noopener noreferrer\" tiNavigation>{{words.secondMessageToOperation}}</a>\r\n</ti-alert>\r\n<ti-alert class=\"tp-goandback-alert\"\r\n          [type]=\"error.type\"\r\n          [(open)]=\"error.open\"\r\n          dismissOnTimeout=\"10000\"\r\n          [id]=\"error.id\">{{words.secondMessageSystemBusy}}\r\n</ti-alert>\r\n<span>{{words.secondMessageContent}}</span>",
                styles: ["::ng-deep.tp-goandback-body{width:var(--ti-message-default-width)!important}::ng-deep.tp-goandback-body .ti3-modal-body{position:static!important}.tp-goandback-alert{left:0;margin:0 auto;position:absolute;right:0;top:40px;width:calc(100% - 60px)!important;z-index:9999}.tp-goandback-alert ::ng-deep.ti3-alert-warn-bg{line-height:1}"]
            },] }
];
TpGoandbackComponent.propDecorators = {
    warn: [{ type: Input }],
    error: [{ type: Input }],
    href: [{ type: Input }],
    words: [{ type: Input }]
};
