// @cloud/tiny3/esm2015/components/searchbox/TiSearchboxModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TiSearchboxComponent } from './TiSearchboxComponent';
import { FormsModule } from '@angular/forms';
import { TiOverflowModule } from '../../directives/overflow/TiOverflowModule';
import { TiDroplistModule } from '../droplist/TiDroplistModule';
import { TiTextModule } from '../text/TiTextModule';
import { TiSearchboxNotsearchComponent } from './TiSearchboxNotsearchComponent';
import { TiIconModule } from '../icon/TiIconModule';
import { TiOutlineModule } from '../../directives/outline/TiOutlineModule';
export class TiSearchboxModule {
}
TiSearchboxModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    TiOverflowModule,
                    TiDroplistModule,
                    TiIconModule,
                    TiTextModule,
                    TiOutlineModule
                ],
                exports: [TiSearchboxComponent, TiSearchboxNotsearchComponent],
                declarations: [TiSearchboxComponent, TiSearchboxNotsearchComponent]
            },] }
];
export { TiSearchboxComponent } from './TiSearchboxComponent';
export { TiSearchboxNotsearchComponent } from './TiSearchboxNotsearchComponent';
