// @cloud/tinyplus3/esm2015/components/searchbox/TpSearchboxHelpComponent.js
import { Component, ElementRef, Renderer2, TemplateRef, ViewChild } from '@angular/core';
// tslint:disable-next-line: no-implicit-dependencies
import { TiBaseComponent, TiLocale, TiModalService } from '@cloud/tiny3';
export class TpSearchboxHelpComponent extends TiBaseComponent {
    constructor(hostRef, render, tiModal) {
        super(hostRef, render);
        this.tiModal = tiModal;
        /**
         * @ignore
         * 默认字段
         */
        this.defaultField = 'testName';
        /**
         * @ignore
         * 输入项
         */
        this.items = [];
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        // tslint:disable-next-line: typedef
        const { name, status, running, creating, stopped, restarting, tag } = TiLocale.getLocaleWords().tpSearchbox;
        // 初始化数据
        this.items = [{
                label: name,
                field: 'testName',
                type: 'radio',
                options: [{
                        label: 'ecs-1'
                    }, {
                        label: 'obs-2'
                    }, {
                        label: 'vpc-1'
                    }]
            },
            {
                label: status,
                field: 'testStatus',
                type: 'checkbox',
                options: [{
                        label: running
                    }, {
                        label: creating
                    }, {
                        label: stopped
                    }, {
                        label: restarting
                    }]
            },
            {
                id: '3',
                label: tag,
                field: 'testTag',
                type: 'map',
                options: [
                    {
                        label: 'aaa',
                        options: [
                            {
                                label: 'ddd'
                            },
                            {
                                label: 'eee'
                            },
                            {
                                label: 'fff'
                            }
                        ]
                    },
                    {
                        label: 'bbb',
                        options: [
                            {
                                label: 'ddd'
                            },
                            {
                                label: 'eee'
                            },
                            {
                                label: 'fff'
                            }
                        ]
                    },
                    {
                        label: 'ccc',
                        options: [
                            {
                                label: 'ddd'
                            },
                            {
                                label: 'eee'
                            },
                            {
                                label: 'fff'
                            }
                        ]
                    }
                ]
            }];
        this.tags1 = [
            {
                field: 'testName',
                label: name,
                value: 'ecs-1',
                type: 'radio'
            }
        ];
        this.tags2 = [
            {
                field: 'testName',
                label: name,
                value: 'ecs-1',
                type: 'radio'
            },
            {
                field: 'testStatus',
                label: status,
                value: `${running} | ${creating} | ${stopped}`,
                type: 'checkbox'
            }
        ];
        this.tags3 = [
            {
                field: 'testName',
                label: name,
                value: 'ecs-1',
                type: 'radio'
            },
            {
                field: 'testStatus',
                label: status,
                value: running,
                type: 'checkbox'
            },
            {
                field: 'testTag',
                label: tag,
                value: 'aaa=',
                type: 'map'
            }
        ];
    }
    /**
     * 展示弹窗
     */
    show() {
        this.tiModal.open(this.helpModalTemplateRef, {
            id: 'tp-searchbox-help-modal',
            modalClass: 'tp-searchbox-help-modal'
        });
    }
}
TpSearchboxHelpComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-searchbox-help',
                template: "<ng-template let-context='context'>\r\n    <ti-modal-header>\r\n        {{'tpSearchbox.modalTitle' | tiTranslate }}\r\n    </ti-modal-header>\r\n    <ti-modal-body>\r\n        <section class=\"tp-searchbox-help-sense\">\r\n            <div class=\"tp-searchbox-help-title\">1. {{'tpSearchbox.ruleTitle1' | tiTranslate }}</div>\r\n            <div class=\"tp-searchbox-help-describe\">\r\n                <span>{{'tpSearchbox.format' | tiTranslate }}</span>\r\n                <span class=\"tp-searchbox-help-remark\">{{'tpSearchbox.rulekeyword1' | tiTranslate }}</span>\r\n            </div>\r\n            <div class=\"tp-searchbox-help-detail\">\r\n                <div>\r\n                    {{ 'tpSearchbox.ruleDetail1' | tiTranslate }}\r\n                </div>\r\n                <div>\r\n                    {{ 'tpSearchbox.autoDetection' | tiTranslate }}\r\n                </div>\r\n            </div>\r\n            <tp-searchbox [items]=\"items\" [(ngModel)]=\"tags1\" [defaultField]=\"defaultField\" [showHelp]=\"false\">\r\n            </tp-searchbox>\r\n        </section>\r\n        <section class=\"tp-searchbox-help-sense\">\r\n            <div class=\"tp-searchbox-help-title\">2. {{'tpSearchbox.ruleTitle2' | tiTranslate }}</div>\r\n            <div class=\"tp-searchbox-help-describe\">\r\n                <span>{{'tpSearchbox.format' | tiTranslate }}</span>\r\n                <span class=\"tp-searchbox-help-remark\">{{'tpSearchbox.rulekeyword2' | tiTranslate }}</span>\r\n            </div>\r\n            <div class=\"tp-searchbox-help-detail\">{{ 'tpSearchbox.ruleDetail2' | tiTranslate }}</div>\r\n            <tp-searchbox [items]=\"items\" [(ngModel)]=\"tags2\" [defaultField]=\"defaultField\" [showHelp]=\"false\">\r\n            </tp-searchbox>\r\n        </section>\r\n        <section class=\"tp-searchbox-help-sense\">\r\n            <div class=\"tp-searchbox-help-title\">3. {{'tpSearchbox.ruleTitle3' | tiTranslate }}</div>\r\n            <div class=\"tp-searchbox-help-describe\">\r\n                <span>{{'tpSearchbox.format' | tiTranslate }}</span>\r\n                <span class=\"tp-searchbox-help-remark\">{{'tpSearchbox.rulekeyword3' | tiTranslate }}</span>\r\n                <span class=\"tp-searchbox-help-remark\">{{'tpSearchbox.rulekeyword32' | tiTranslate }}</span>\r\n            </div>\r\n            <div class=\"tp-searchbox-help-detail\">{{ 'tpSearchbox.ruleDetail3' | tiTranslate }}</div>\r\n            <tp-searchbox [items]=\"items\" [(ngModel)]=\"tags3\" [defaultField]=\"defaultField\" [showHelp]=\"false\">\r\n            </tp-searchbox>\r\n        </section>\r\n    </ti-modal-body>\r\n    <ti-modal-footer>\r\n        <button type=\"button\" tiButton color=\"danger\" autofocus\r\n            (click)=\"context.close()\">{{ 'tpSearchbox.buttonOk' | tiTranslate }}</button>\r\n    </ti-modal-footer>\r\n</ng-template>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-searchbox-help-modal-height:600px;--tp-searchbox-help-modal-width:720px}::ng-deep .ti3-modal-dialog.tp-searchbox-help-modal{height:var(--tp-searchbox-help-modal-height);width:var(--tp-searchbox-help-modal-width)}.tp-searchbox-help-sense{color:var(--ti-common-color-text-secondary);font-size:var(--ti-common-font-size-base);margin-bottom:var(--ti-common-space-5x)}.tp-searchbox-help-sense .tp-searchbox-help-title{font-size:var(--ti-common-font-size-1);font-weight:700;margin-bottom:var(--ti-common-space-3x)}.tp-searchbox-help-sense .tp-searchbox-help-describe{margin-bottom:var(--ti-common-space-2x)}.tp-searchbox-help-sense .tp-searchbox-help-describe .tp-searchbox-help-remark{color:var(--ti-common-color-text-primary);margin-left:var(--ti-common-space-3x)}.tp-searchbox-help-sense .tp-searchbox-help-detail{line-height:var(--ti-common-line-height-number);margin-bottom:var(--ti-common-space-3x)}"]
            },] }
];
TpSearchboxHelpComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TiModalService }
];
TpSearchboxHelpComponent.propDecorators = {
    helpModalTemplateRef: [{ type: ViewChild, args: [TemplateRef, { static: false },] }]
};
