// @cloud/tinyplus3/esm2015/components/searchbox/TpSearchboxModule.js
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { TpIconModule } from '../icon/TpIconModule';
import { TiButtonModule, TiDateModule, TiDropModule, TiDropsearchModule, TiLocaleModule, TiModalModule, TiOverflowModule, TiTipModule, TiValidationModule, TiDominatorModule, TiTextModule, TiOutlineModule } from '@cloud/tiny3';
import { TpSearchboxComponent } from './TpSearchboxComponent';
import { TpSearchboxHelpComponent } from './TpSearchboxHelpComponent';
export class TpSearchboxModule {
}
TpSearchboxModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    TiButtonModule,
                    TiModalModule,
                    TiDateModule,
                    TiDropModule,
                    TiDropsearchModule,
                    TiDominatorModule,
                    TiLocaleModule,
                    TiOverflowModule,
                    TiTextModule,
                    TiTipModule,
                    TiValidationModule,
                    TpIconModule,
                    TiOutlineModule
                ],
                exports: [TpSearchboxComponent, TpSearchboxHelpComponent],
                declarations: [TpSearchboxComponent, TpSearchboxHelpComponent]
            },] }
];
export { TpSearchboxComponent, TP_REGEXP_IP, TP_REGEXP_UUID } from './TpSearchboxComponent';
export { TpSearchboxHelpComponent } from './TpSearchboxHelpComponent';
