// @cloud/tiny3/esm2015/components/steps/TiStepsComponent.js
import { Component, ElementRef, EventEmitter, Input, IterableDiffers, NgZone, Output, QueryList, Renderer2, ViewChild, ViewChildren } from '@angular/core';
import { Util } from '../../utils/Util';
import { TiBaseComponent } from '../base/TiBaseModule';
/**
 * steps步骤组件
 *
 * 点击可跳转和不可跳转两种方式（默认点击不跳转）
 *
 * <example-url>../tiny3demo/#/steps/steps-all</example-url>
 */
export class TiStepsComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, iterableDiffers, zone) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.iterableDiffers = iterableDiffers;
        this.zone = zone;
        /**
         * 是否支持点击跳转功能，默认不支持
         */
        this.clickable = false;
        /**
         * 用户可指定要显示的属性值，默认显示label属性
         */
        this.labelKey = 'label';
        /**
         * 宽度是否自动撑满父容器。
         *
         * 不设置时默认各步骤间线宽60px，步骤确定后整体宽度固定；设置为true时，整体宽度会自适应撑满父容器，适用于在弹窗中使用的场景。
         */
        this.adaptive = false;
        /**
         * 当前步骤改变时的事件
         */
        this.activeStepChange = new EventEmitter();
        /**
         * 点击后用户可根据条件判断是否跳转:见测试用例
         */
        this.beforeStep = new EventEmitter();
        this.isInitLabelChange = true;
        this.explainTotalWidth = 0;
        // 设置每个步骤之间的横线的宽度
        this.setLineWidth = (width) => {
            this.explainTotalWidth = Util.isUndefined(width) ? this.getExplainTotalWidth() : width;
            this.lineRef.forEach((stepLine) => {
                this.renderer2.setStyle(stepLine.nativeElement, 'width', `calc((100% - ${this.explainTotalWidth}px) / ${this.steps.length - 1})`);
            });
        };
    }
    ngOnInit() {
        super.ngOnInit();
        if (!this.adaptive) {
            return;
        }
        // this.trackByLabelFn监听对象：可以深度监听
        this.stepsDiffer = this.iterableDiffers.find(this.steps)
            .create(this.trackByLabelFn);
        // 修复错误：ERROR ReferenceError: window is not defined
        if (typeof window === 'undefined') {
            return;
        }
        this.zone.runOutsideAngular(() => {
            this.windowResizeListener = this.renderer2.listen(window, 'resize', this.setLineWidth);
        });
    }
    ngDoCheck() {
        if (this.adaptive) {
            // 处理增删步骤
            const stepsDiffer = this.stepsDiffer.diff(this.steps);
            if (stepsDiffer) {
                this.labelChange = true;
            }
        }
    }
    ngAfterViewChecked() {
        // 需要判断当label变化时才计算线长
        if (this.adaptive && this.labelChange) {
            this.labelChange = false;
            this.setLineWidth();
            // 初始时在前几次的 ngAfterViewChecked 中计算文本宽度，生产环境清缓存场景下计算有点不准确(时机有点早)，但在初始时的后几次ngAfterViewChecked中计算的文本宽度是准确的。
            // TODO: 是否有更合适的时机计算(已验证使用MutationObserver也解决不了，在ngAfterViewChecked中加setTimeout(0)也解决不了)
            if (this.isInitLabelChange) {
                setTimeout(() => {
                    const explainTotalWidth = this.getExplainTotalWidth();
                    if (explainTotalWidth !== this.explainTotalWidth) {
                        this.setLineWidth(explainTotalWidth);
                    }
                }, 50); // 50ms是经验值
                this.isInitLabelChange = false;
            }
        }
    }
    ngOnDestroy() {
        if (this.windowResizeListener) {
            this.windowResizeListener();
        }
    }
    /**
     * @ignore
     * 每步的点击事件处理
     */
    onClick(step) {
        // 不支持点击跳转和当前项灰化时，不处理
        if (!this.clickable || step.disabled || this.activeStep === step) {
            return;
        }
        // 未定义beforeStep事件，直接跳转，定义beforeStep事件将当前点击项索引传出
        if (this.beforeStep.observers.length === 0) {
            this.activeStep = step;
            this.activeStepChange.emit(this.activeStep);
        }
        else {
            this.beforeStep.emit(step);
        }
    }
    getExplainTotalWidth() {
        let explainTotalWidth = 0;
        if (!Util.isUndefined(this.explainRef)) {
            this.explainRef.forEach((step) => {
                // 计算每步选择框和文字的总长度
                explainTotalWidth += step.nativeElement.offsetWidth;
            });
        }
        return explainTotalWidth;
    }
    /**
     * @ignore
     * Diff监听steps中label值的改变
     */
    trackByLabelFn(index, item) {
        return item.label;
    }
    /**
     * @ignore
     * ngFor 使用
     */
    trackByIndexFn(index) {
        return index;
    }
    /**
     * @ignore
     * 判断当前项的状态
     */
    getStepState(index) {
        const activedIndex = this.steps.indexOf(this.activeStep);
        if (index < activedIndex) {
            return 'complete';
        }
        else if (index === activedIndex) {
            return 'active';
        }
        else {
            return 'uncomplete';
        }
    }
}
TiStepsComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-steps',
                template: "<ul class=\"ti3-steps\" #stepRef [id]=\"appendId('list')\">\r\n    <ng-container *ngFor=\"let step of steps; let i=index; trackBy: trackByIndexFn\">\r\n        <li class=\"ti3-steps-line\" #line *ngIf=\"i > 0\"\r\n            [ngClass]=\"{'ti3-steps-complete': getStepState(i) === 'complete',\r\n                    'ti3-steps-active':getStepState(i) === 'active',\r\n                    'ti3-steps-uncomplete': getStepState(i) === 'uncomplete' && clickable && !step.disabled}\">\r\n        </li>\r\n\r\n        <li [id]=\"appendId(i)\" class=\"ti3-steps-explain\" #explain\r\n            [ngClass]=\"{'ti3-steps-complete': getStepState(i) === 'complete',\r\n                    'ti3-steps-active':getStepState(i) === 'active',\r\n                    'ti3-steps-disabled': getStepState(i) === 'uncomplete' && (step.disabled || !clickable),\r\n                    'ti3-steps-uncomplete': getStepState(i) === 'uncomplete' && clickable && !step.disabled,\r\n                    'ti3-steps-explain-clickable': clickable && !step.disabled}\"\r\n            (click)=\"onClick(step)\"\r\n            (keydown.enter)=\"onClick(step)\"\r\n            tiOutline\r\n            [attr.tabindex]=\"clickable && !step.disabled ? 0 : -1\">\r\n            <div class=\"ti3-steps-box\">\r\n                <div class=\"ti3-steps-box-number\">{{i + 1}}</div>\r\n            </div>\r\n            <div class=\"ti3-steps-text\" [ngStyle]=\"{'max-width': maxWidth}\" [id]=\"appendId('i_label')\"\r\n                tiOverflow [textContent]=\"step[labelKey]\" [maxLine]=\"2\">{{step[labelKey]}}</div>\r\n        </li>\r\n    </ng-container>\r\n</ul>\r\n\r\n",
                styles: ["::ng-deep :root{--ti-steps-box-border-width:1px;--ti-steps-box-size:20px;--ti-steps-line-height:1px;--ti-steps-line-width:60px;--ti-steps-transition-time:300ms}:host{display:block}.ti3-steps{padding:var(--ti-common-space-3x) 0;white-space:nowrap}.ti3-steps .ti3-steps-explain{display:inline-block;vertical-align:top}.ti3-steps .ti3-steps-explain .ti3-steps-box{-ms-box-sizing:border-box;border:var(--ti-steps-box-border-width) solid;border-radius:50%;box-sizing:border-box;display:inline-block;font-size:var(--ti-steps-number-font-size);height:var(--ti-steps-box-size);margin-left:var(--ti-common-space-2x);position:relative;width:var(--ti-steps-box-size)}.ti3-steps .ti3-steps-explain .ti3-steps-box:before{background-color:var(--ti-steps-box-bg-color-highlight);border-radius:50%;content:\"\";display:inline-block;height:var(--ti-steps-box-size);opacity:0;position:absolute;right:calc(-1*var(--ti-steps-box-border-width));top:calc(-1*var(--ti-steps-box-border-width));width:var(--ti-steps-box-size)}.ti3-steps .ti3-steps-explain .ti3-steps-box .ti3-steps-box-number{height:100%;line-height:calc(var(--ti-steps-box-size) - var(--ti-steps-box-border-width)*2 - 1px);position:absolute;text-align:center;width:100%}.ti3-steps .ti3-steps-explain .ti3-steps-text{display:inline-block;line-height:var(--ti-steps-box-size);margin:0 var(--ti-common-space-2x) 0 var(--ti-common-space-base);max-width:150px;vertical-align:top;white-space:normal;word-break:break-word}.ti3-steps li:first-child.ti3-steps-explain .ti3-steps-box{margin-left:0}.ti3-steps li:last-child.ti3-steps-explain .ti3-steps-text{margin-right:0}.ti3-steps .ti3-steps-explain-clickable{cursor:pointer}.ti3-steps .ti3-steps-line{background-color:var(--ti-steps-line-color);display:inline-block;font-size:0;height:var(--ti-steps-line-height);line-height:0;margin-top:calc((var(--ti-steps-box-size) - var(--ti-steps-line-height))/2);min-width:20px;vertical-align:top;width:var(--ti-steps-line-width)}.ti3-steps .ti3-steps-line:before{background-color:var(--ti-steps-line-color-highlight);content:\" \";display:inline-block;height:100%;width:0}.ti3-steps .ti3-steps-uncomplete.ti3-steps-line:before{transition:width var(--ti-steps-transition-time);width:0}.ti3-steps .ti3-steps-uncomplete.ti3-steps-explain .ti3-steps-box{border-color:var(--ti-steps-box-uncomplete);color:var(--ti-steps-box-uncomplete)}.ti3-steps .ti3-steps-uncomplete.ti3-steps-explain .ti3-steps-box:before{opacity:0;transform:scale(0);transition:transform var(--ti-steps-transition-time),opacity var(--ti-steps-transition-time)}.ti3-steps .ti3-steps-uncomplete.ti3-steps-explain .ti3-steps-text{-webkit-background-clip:text;background-image:linear-gradient(var(--ti-steps-text-highlight),var(--ti-steps-text-highlight));background-image:-ms-linear-gradient(top,transparent,transparent);background-repeat:no-repeat;background-size:0 100%;color:var(--ti-steps-text-uncomplete);transition:background-size var(--ti-steps-transition-time),color var(--ti-steps-transition-time)}.ti3-steps .ti3-steps-complete.ti3-steps-line:before{transition:width var(--ti-steps-transition-time);width:100%}.ti3-steps .ti3-steps-complete.ti3-steps-explain .ti3-steps-box{border-color:var(--ti-steps-box-highlight);color:var(--ti-steps-box-highlight)}.ti3-steps .ti3-steps-complete.ti3-steps-explain .ti3-steps-box:before{opacity:0;transform:scale(0);transition:transform var(--ti-steps-transition-time),opacity var(--ti-steps-transition-time)}.ti3-steps .ti3-steps-complete.ti3-steps-explain .ti3-steps-text{color:var(--ti-steps-text-highlight)}.ti3-steps .ti3-steps-active.ti3-steps-line:before{transition:width var(--ti-steps-transition-time);width:100%}.ti3-steps .ti3-steps-active.ti3-steps-explain .ti3-steps-box{border-color:var(--ti-steps-box-highlight);color:var(--ti-steps-box-number-active)}.ti3-steps .ti3-steps-active.ti3-steps-explain .ti3-steps-box:before{opacity:1;transform:scale(1);transition:transform var(--ti-steps-transition-time),opacity var(--ti-steps-transition-time)}.ti3-steps .ti3-steps-active.ti3-steps-explain .ti3-steps-text{-webkit-background-clip:text;background-image:linear-gradient(var(--ti-steps-text-highlight),var(--ti-steps-text-highlight));background-image:-ms-linear-gradient(top,transparent,transparent);background-repeat:no-repeat;background-size:100% 100%;color:var(--ti-steps-text-highlight);transition:background-size var(--ti-steps-transition-time),color var(--ti-steps-transition-time)}.ti3-steps .ti3-steps-disabled.ti3-steps-explain .ti3-steps-box{border-color:var(--ti-steps-box-disabled);color:var(--ti-steps-box-disabled)}.ti3-steps .ti3-steps-disabled.ti3-steps-explain .ti3-steps-box:before{display:none!important}.ti3-steps .ti3-steps-disabled.ti3-steps-explain .ti3-steps-text{color:var(--ti-steps-text-disabled);cursor:not-allowed}"]
            },] }
];
TiStepsComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: IterableDiffers },
    { type: NgZone }
];
TiStepsComponent.propDecorators = {
    steps: [{ type: Input }],
    clickable: [{ type: Input }],
    maxWidth: [{ type: Input }],
    labelKey: [{ type: Input }],
    activeStep: [{ type: Input }],
    adaptive: [{ type: Input }],
    activeStepChange: [{ type: Output }],
    beforeStep: [{ type: Output }],
    stepsRef: [{ type: ViewChild, args: ['stepRef',] }],
    lineRef: [{ type: ViewChildren, args: ['line',] }],
    explainRef: [{ type: ViewChildren, args: ['explain',] }]
};
