// @cloud/tiny3/esm2015/components/textarea/TiTextareaComponent.js
import { Component, ElementRef, Input, NgZone, Optional, Renderer2, ViewChild, ViewEncapsulation, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { TiBrowser, Util } from '../../utils/Util';
import { TiAutofocusComponent } from '../base/TiBaseModule';
import { NgControl } from '@angular/forms';
/**
 * Textarea文本框组件
 *
 * 文本框组件基于原生textarea标签进行扩展，原生textarea加tiTextarea属性指令即为textarea组件。
 *
 * <example-url>../tiny3demo/#/textarea/textarea-all</example-url>
 */
export class TiTextareaComponent extends TiAutofocusComponent {
    constructor(elementRef, renderer, zone, formControl, document) {
        super(elementRef, renderer);
        this.zone = zone;
        this.formControl = formControl;
        this.document = document;
        /**
         * 用来调整文本框大小，该属性提供了四个属性值:
         *
         * none(不可调整组件大小):
         *
         * vertical(仅可调整垂直方向的大小，即调整组件的高度)
         *
         * horizontal(仅可调节水平方向的大小，即调整组件的宽度)
         *
         * both(水平和垂直方向均可调节，宽高都可调节)
         */
        this.resize = 'both';
        // 用作拖动文本框大小
        this.options = {
            $target: undefined,
            storeWidth: 0,
            storeHeight: 0,
            mouseXPosition: 0,
            mouseYPosition: 0
        };
        // 默认最大最小宽高
        this.defaultStyle = {
            minWidth: 60,
            minHeight: 64,
            maxWidth: 1280,
            maxHeight: 9999
        };
        this.isFirstFocus = true; // 是否为第一次聚焦
        /**
         * @ignore
         * 绑定在document上的mouseMove事件
         */
        this.mouseMoveHandlerFn = ($event) => {
            if (this.element.disabled) {
                return;
            }
            // 拖动的动作应该是先mousedown，然后mousemove，因此先判断是否已经触发了mousedown
            if (this.options.mouseXPosition === 0 && this.options.mouseYPosition === 0 || !this.options.$target) {
                return;
            }
            this.mouseMove($event);
        };
        /**
         * @ignore
         * 绑定在document上的mouseUp事件
         */
        this.mouseUpHandlerFn = () => {
            if (this.element.disabled) {
                return;
            }
            // 解决问题：当拖动鼠标时，鼠标不在小三角上时，鼠标放开不会移除样式
            // 为确保鼠标弹起时, textarea的边框样式恢复默认:
            if (this.isResizing()) {
                this.renderer.removeClass(this.element, 'ti3-textarea-resize-border');
            }
            try {
                this.stopResize();
                this.toggleTextSelection(false);
            }
            catch (e) { }
        };
        this.element = elementRef.nativeElement; // 获取宿主元素
        this.attrs = this.element.attributes; // 可以根据属性名获取各个属性
        this.container = this.renderer.createElement('div'); // 创建div标签
        this.renderer.addClass(this.container, 'ti3-textarea-container');
    }
    ngOnInit() {
        super.ngOnInit();
        // 初始化maxlength属性
        this.hasMaxlength = Util.isUndefined(this.attrs['maxlength']) ? false : true;
        // 原生不设置maxlength属性时该值为-1;
        this.maxLength = this.hasMaxlength ? this.element.maxLength : -1;
        // 文本框的值发生变化时触发
        // 使用ngModel初始赋值时，触发valueChanges。
        // 使用formControl初始赋值时，不会触发valueChanges。
        if (this.formControl) {
            this.formControl.valueChanges.subscribe((value) => {
                this.setCountLength();
            });
        }
        // 使用formControl初始赋值时，初始化不会触发valueChanges，所以此处额外执行一次。
        // 使用ngModel初始赋值时，此处通过this.element.value拿到的值为空字符串，无法正确显示当前输入内容的长度
        this.setCountLength();
    }
    ngAfterViewInit() {
        // 如果祖先元素中display为none，获取到的组件的宽度为0,元素不可见
        this.isVisibleInit = this.element.offsetWidth !== 0;
        if (this.isVisibleInit) {
            // 设置container的样式
            this.setContainerSize();
            // 初始化创建dom
            this.initFn();
        }
        // 将宿主元素的样式类移到container上
        this.setContainerClass();
        // maxLength接口添加样式
        this.setMaxLengthFn();
        // 处理resize功能
        this.resizeFn();
        // 上面执行完初始化dom后，调用父类的ngAfterViewInit去设置autofocus
        super.ngAfterViewInit();
    }
    ngAfterViewChecked() {
        // 初始祖先元素中display为none，之后第一次由none变成可见时，需要重置silder
        if (!this.isVisibleInit && this.element.offsetWidth !== 0) {
            this.setContainerSize();
            // 初始化创建dom
            this.initFn();
            this.isVisibleInit = true;
        }
        // 处理灰化功能
        if (this.element.disabled) {
            this.renderer.addClass(this.container, 'ti3-textarea-container-disable');
        }
        else {
            this.renderer.removeClass(this.container, 'ti3-textarea-container-disable');
        }
        // 修复SSR报错：ERROR ReferenceError: getComputedStyle is not defined
        if (typeof getComputedStyle === 'undefined') {
            return;
        }
        // 处理通过设置display的方式隐藏组件：内部生成的dom也需要隐藏
        if (getComputedStyle(this.element).display === 'none') {
            this.renderer.setStyle(this.container, 'display', 'none');
        }
        else {
            this.renderer.setStyle(this.container, 'display', 'block');
        }
    }
    // maxlength通过接口传值需触发重新判断，因为init函数中执行较早且只执行一次。
    ngDoCheck() {
        this.hasMaxlength = Util.isUndefined(this.attrs['maxlength']) ? false : true;
        this.maxLength = this.hasMaxlength ? this.element.maxLength : -1;
    }
    // 销毁组件时需要将内部生成dom也移除
    ngOnDestroy() {
        this.renderer.removeChild(this.renderer.parentNode(this.container), this.container);
        if (this.documentMouseMoveListener) {
            this.documentMouseMoveListener();
        }
        if (this.documentMouseUpListener) {
            this.documentMouseUpListener();
        }
    }
    /**
     * @ignore
     */
    onFocus() {
        // IE和火狐浏览器在首次autofoucus聚焦时会将光标移到文字起始位置，用户体验不好
        // 需要手动设置光标位置到行尾
        if (this.isFirstFocus && this.element.autofocus &&
            (TiBrowser.isFirefox() || TiBrowser.isIE())) {
            this.element.setSelectionRange(this.element.value.length, this.element.value.length);
        }
        this.isFirstFocus = false;
    }
    setContainerSize() {
        // 修正SSR错误：ERROR ReferenceError: getComputedStyle is not defined
        if (typeof getComputedStyle === 'undefined') {
            return;
        }
        const containerHeight = getComputedStyle(this.element).height;
        const containerWidth = getComputedStyle(this.element).width;
        // 若服务在textarea上面设置margin外边距,会导致textarea所占的页面空间大于外框的宽高,所以要将textarea上设置的margin转移到外边框
        const containerMargin = this.element.style.margin;
        // 取掉宿主元素的margin属性:
        this.element.style.margin = '';
        // 设置this.container的width、height、margin
        this.renderer.setStyle(this.container, 'height', containerHeight);
        this.renderer.setStyle(this.container, 'width', containerWidth);
        this.renderer.setStyle(this.container, 'margin', containerMargin);
    }
    setContainerClass() {
        // 获取未经过angular解析后的属性需要在onint中处理：若服务在textarea添加class，则把其加到父容器上
        const cNames = this.element.className;
        if (cNames) {
            const cNameArr = this.element.className.split(' '); // 存在多个class
            for (const cName of cNameArr) {
                // 跟校验有关的样式是以"ng-"开头，如果转移到外部容器上会影响校验样式
                if (!/^ng-/.test(cName)) {
                    this.renderer.removeClass(this.element, cName);
                    this.renderer.addClass(this.container, cName);
                }
            }
        }
    }
    // 初始化创建dom
    initFn() {
        const pNode = this.renderer.parentNode(this.element); // 找到宿主元素父元素
        const nextNode = this.renderer.nextSibling(this.element); // 找到宿主元素的兄弟元素
        // 将宿主元素包裹在this.container里边：
        this.renderer.removeChild(pNode, this.element); // 从父元素中移除宿主元素
        this.renderer.appendChild(this.container, this.element); // 将宿主元素放入创建的div标签中
        this.renderer.insertBefore(pNode, this.container, nextNode);
        if (this.resize !== 'none') {
            this.moveNode(this.resizeIconRef.nativeElement);
        }
        if (this.hasMaxlength) {
            this.moveNode(this.eleCharacterRef.nativeElement);
        }
    }
    moveNode(node) {
        this.renderer.removeChild(this.element, node);
        this.renderer.appendChild(this.container, node);
    }
    // maxLength接口
    setMaxLengthFn() {
        // 添加字数限制与当前输入字数显示功能
        if (this.attrs['maxlength'] && !isNaN(this.maxLength)) {
            // 有maxlength限制时给外框div加上标志类 ti3-textarea-container-counter
            this.renderer.addClass(this.container, 'ti3-textarea-container-counter');
        }
        else {
            this.renderer.removeClass(this.container, 'ti3-textarea-container-counter');
        }
    }
    // resize接口
    resizeFn() {
        if (this.resize === 'none') {
            return;
        }
        this.textAreaMinWidth = this.getSizeNumber('minWidth');
        this.textAreaMinHeight = this.getSizeNumber('minHeight');
        this.textAreaMaxWidth = this.getSizeNumber('maxWidth');
        this.textAreaMaxHeight = this.getSizeNumber('maxHeight');
        // 清除textarea上的样式
        this.renderer.setStyle(this.element, 'minWidth', '');
        this.renderer.setStyle(this.element, 'minHeight', '');
        this.renderer.setStyle(this.element, 'maxWidth', '');
        this.renderer.setStyle(this.element, 'maxHeight', '');
        this.zone.runOutsideAngular(() => {
            // resizeIconRef模板的mousedown事件
            this.renderer.listen(this.resizeIconRef.nativeElement, 'mousedown', ($event) => {
                this.renderer.addClass(this.element, 'ti3-textarea-resize-border'); // 拖拽时边框颜色为聚焦时
                this.options.$target = $event.target;
                this.options.mouseXPosition = $event.pageX;
                this.options.mouseYPosition = $event.pageY;
                // 在拖动前获取当前文本框最新的高度、宽度
                this.updateTextAreaSize();
                // 给页面设置不可选样式，避免拖动时页面出现被选中的蓝色区域
                this.toggleTextSelection(true);
            });
            // resizeIconRef模板的mouseup事件
            this.renderer.listen(this.resizeIconRef.nativeElement, 'mouseup', ($event) => {
                this.renderer.removeClass(this.element, 'ti3-textarea-resize-border');
            });
            this.documentMouseMoveListener = this.renderer.listen(this.document, 'mousemove', this.mouseMoveHandlerFn);
            this.documentMouseUpListener = this.renderer.listen(this.document, 'mouseup', this.mouseUpHandlerFn);
        });
    }
    setCountLength() {
        if (this.hasMaxlength) {
            this.countLength = this.element.value.length;
            // safari浏览器回车键是按两个字计算
            if (TiBrowser.isSafari()) {
                this.countLength = this.countLength + this.element.value.split('\n').length - 1;
            }
        }
    }
    // 拖动函数
    mouseMove(event) {
        // 阻止调整文本框大小时文本框聚焦
        event.preventDefault();
        const width = this.options.storeWidth;
        const height = this.options.storeHeight;
        const horizonWidth = event.pageX - this.options.mouseXPosition;
        const verticalHeight = event.pageY - this.options.mouseYPosition;
        if (this.resize !== 'vertical' && (width + horizonWidth) >= this.textAreaMinWidth
            && (width + horizonWidth) < this.textAreaMaxWidth) {
            // 更新textarea宽度
            this.options.storeWidth += horizonWidth;
            // 调整width
            this.renderer.setStyle(this.container, 'width', `${this.options.storeWidth}px`);
        }
        if (this.resize !== 'horizontal' && (height + verticalHeight) >= this.textAreaMinHeight
            && (height + verticalHeight) < this.textAreaMaxHeight) {
            // 更新textarea高度
            this.options.storeHeight += verticalHeight;
            this.renderer.setStyle(this.container, 'height', `${this.options.storeHeight}px`);
        }
        // 更新光标位置
        this.options.mouseXPosition = event.pageX;
        this.options.mouseYPosition = event.pageY;
    }
    stopResize() {
        this.options.mouseXPosition = 0;
        this.options.mouseYPosition = 0;
        this.updateTextAreaSize();
        this.options.$target = null;
    }
    // 获取min-width、min-height等的样式值
    getSizeNumber(value) {
        const val = this.element.style[value]; // 可实现获取样式
        if (val !== 'none' && val !== '') {
            return parseFloat(val.replace(/px/, ''));
        }
        return this.defaultStyle[value];
    }
    // 在拖动前获取当前文本框最新的高度、宽度
    updateTextAreaSize() {
        this.options.storeWidth = this.container.clientWidth;
        this.options.storeHeight = this.container.clientHeight;
    }
    // 给页面设置不可选样式，避免拖动时页面出现被选中的蓝色区域
    toggleTextSelection(toggle) {
        const body = this.document.getElementsByTagName('body');
        if (toggle) {
            this.renderer.addClass(body[0], 'ti3-unselectable');
            this.renderer.setAttribute(body[0], 'unselectable', 'on');
        }
        else {
            this.renderer.removeClass(body[0], 'ti3-unselectable');
            this.renderer.removeAttribute(body[0], 'unselectable');
        }
    }
    isResizing() {
        return this.element.className.indexOf('ti3-textarea-resize-border') > -1;
    }
}
TiTextareaComponent.decorators = [
    { type: Component, args: [{
                // tslint:disable-next-line:component-selector
                selector: '[tiTextarea]',
                template: "<span #resize *ngIf=\"resize !== 'none'\" class=\"ti3-textarea-resize-icon\" ></span>\r\n<span #maxlength *ngIf=\"hasMaxlength\" class=\"ti3-textarea-counter\" [id]=\"appendId('max_length')\">\r\n    <span #charactersCount [ngClass]=\"{'ti3-textarea-counter-count': countLength > 0}\">{{countLength | tiFormatNum}}</span>/{{maxLength | tiFormatNum}}</span>",
                host: {
                    '(focus)': 'onFocus()'
                },
                encapsulation: ViewEncapsulation.None,
                styles: [":root{--ti-textarea-counter-height:16px;--ti-textarea-counter-margin-top:2px}textarea.ng-invalid.ng-dirty:not([tiBlurCheck])~.ti3-textarea-resize-icon:after,textarea.ng-invalid.ng-dirty:not([tiBlurCheck])~.ti3-textarea-resize-icon:before,textarea.ng-invalid.ng-dirty[tiBlurCheck]:not([tiFocused])~.ti3-textarea-resize-icon:after,textarea.ng-invalid.ng-dirty[tiBlurCheck]:not([tiFocused])~.ti3-textarea-resize-icon:before,textarea.ng-invalid.ng-touched:not([tiBlurCheck])~.ti3-textarea-resize-icon:after,textarea.ng-invalid.ng-touched:not([tiBlurCheck])~.ti3-textarea-resize-icon:before,textarea.ng-invalid.ng-touched[tiBlurCheck]:not([tiFocused])~.ti3-textarea-resize-icon:after,textarea.ng-invalid.ng-touched[tiBlurCheck]:not([tiFocused])~.ti3-textarea-resize-icon:before{border-bottom-color:var(--ti-input-border-color-error)!important}.ti3-textarea-container{display:inline-block;line-height:normal;min-height:64px;overflow:hidden;position:relative;vertical-align:middle}.ti3-textarea-container>textarea[tiTextarea]{-ms-box-sizing:border-box;background-color:var(--ti-input-bg-color);border:1px solid var(--ti-input-border-color);border-radius:var(--ti-input-border-radius);box-sizing:border-box;color:var(--ti-input-text-color);height:100%!important;outline:none;overflow-x:hidden;padding:var(--ti-textarea-padding-vertical) var(--ti-input-padding-horizontal);resize:none!important;transition:border-color .15s;vertical-align:middle;width:100%!important}.ti3-textarea-container>textarea[tiTextarea]:disabled{background-color:var(--ti-input-bg-color-disabled);border-color:var(--ti-input-border-color-disabled);color:var(--ti-input-text-color-disabled);cursor:not-allowed!important}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):hover{border-color:var(--ti-input-border-color-hover);transition:border-color .2s}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):hover~.ti3-textarea-resize-icon:after,.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):hover~.ti3-textarea-resize-icon:before{border-bottom-color:var(--ti-input-border-color-hover)}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):focus{border-color:var(--ti-input-border-color-focus);transition:border-color .2s}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):focus~.ti3-textarea-resize-icon:after,.ti3-textarea-container>textarea[tiTextarea]:not(:disabled):focus~.ti3-textarea-resize-icon:before{border-bottom-color:var(--ti-input-border-color-focus)}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled).ti3-textarea-resize-border{border-color:var(--ti-input-border-color-focus);transition:border-color .2s}.ti3-textarea-container>textarea[tiTextarea]:not(:disabled).ti3-textarea-resize-border~.ti3-textarea-resize-icon:after,.ti3-textarea-container>textarea[tiTextarea]:not(:disabled).ti3-textarea-resize-border~.ti3-textarea-resize-icon:before{border-bottom-color:var(--ti-input-border-color-focus)}.ti3-textarea-container>.ti3-textarea-resize-icon{bottom:1px;cursor:nw-resize;height:10px;position:absolute;right:1px;width:10px}.ti3-textarea-container>.ti3-textarea-resize-icon:before{bottom:-2px;height:8.4px;right:-2px;width:8.4px}.ti3-textarea-container>.ti3-textarea-resize-icon:after,.ti3-textarea-container>.ti3-textarea-resize-icon:before{border-bottom:1px solid var(--ti-input-border-color);content:\"\";display:block;position:absolute;transform:rotate(135deg)}.ti3-textarea-container>.ti3-textarea-resize-icon:after{bottom:0;height:4.2px;right:0;width:4.2px}.ti3-textarea-container>.ti3-textarea-counter{-ms-box-sizing:border-box;background-color:var(--ti-input-bg-color);box-sizing:border-box;color:var(--ti-textarea-counter-text-color);display:block;height:var(--ti-textarea-counter-height);line-height:var(--ti-textarea-counter-height);margin-top:var(--ti-textarea-counter-margin-top);position:absolute;right:0;text-align:right}.ti3-textarea-container>.ti3-textarea-counter>.ti3-textarea-counter-count{color:var(--ti-textarea-counter-inputed-text-color)}.ti3-textarea-container.ti3-textarea-container-counter>textarea[tiTextarea]{height:calc(100% - var(--ti-textarea-counter-height) - var(--ti-textarea-counter-margin-top))!important}.ti3-textarea-container.ti3-textarea-container-counter>.ti3-textarea-resize-icon{bottom:calc(var(--ti-textarea-counter-height) + var(--ti-textarea-counter-margin-top))}.ti3-unselectable{-ms-user-select:none;-webkit-user-select:none;user-select:none}.ti3-textarea-container-disable>span.ti3-textarea-resize-icon{cursor:not-allowed}"]
            },] }
];
TiTextareaComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: NgZone },
    { type: NgControl, decorators: [{ type: Optional }] },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
TiTextareaComponent.propDecorators = {
    resize: [{ type: Input }],
    resizeIconRef: [{ type: ViewChild, args: ['resize',] }],
    eleCharacterRef: [{ type: ViewChild, args: ['maxlength',] }],
    charactersCountRef: [{ type: ViewChild, args: ['charactersCount',] }]
};
