// @cloud/tiny3/esm2015/components/tag/TiTagComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '../base/TiBaseModule';
// TODO: 等待规范统一，可能只保留方形tag
// TODO: hover时是否需要高亮，选中是否高亮。增加selectable接口，表示是否可以选中。
// TODO: 禁用时是否显示叉号？规范和tiny2不一致。
// TODO: 禁用时，不发出原生click事件。
/**
 * Tag标签组件，支持显示自定义内容，和删除事件通知。
 *
 * <example-url>../tiny3demo/#/tag/tag-all</example-url>
 */
export class TiTagComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        // @Input() data; // 数据，作为唯一标识
        /**
         * @ignore
         * 样式：有方形和箭头形两种样式，后面可能废弃箭头样式
         */
        // @Input() shape: 'rect'|'arrow'  = 'rect';
        /**
         * 是否禁用
         */
        this.disabled = false;
        /**
         * 选中项文本超出时tip展开方向
         *
         * 9.0.3新增
         */
        this.selectedTipPosition = 'auto';
        /**
         * 删除Tag时的通知事件
         */
        this.delete = new EventEmitter();
    }
    /**
     * @ignore
     * @param $event
     */
    onClickDelete($event) {
        if (this.disabled) {
            return;
        }
        $event.stopPropagation(); // 阻止冒泡，阻止了Tag整体的onClick事件，也阻止Tag之外接收到Click。
        this.delete.emit();
    }
    /**
     * 删除Tag标签element
     */
    remove() {
        this.renderer.removeChild(this.renderer.parentNode(this.nativeElement), this.nativeElement);
    }
}
TiTagComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-tag',
                template: "<!--\u5DE6\u53F3\u4E24\u4E2Adiv\u662Fabsolute\uFF0C\u8131\u79BB\u6587\u6863\u6D41\uFF0C\u6302\u5728\u7236\u5BB9\u5668\u4E4B\u5916\u7684\u3002ti-overflow\u72EC\u5360\u4E00\u884C-->\r\n<!-- <ng-container *ngIf=\"shape ==='arrow'\">\r\n        <div class=\"ti3-tag-selected-triangle\"></div>\r\n        <div tiOverflow><ng-content></ng-content></div>\r\n        <div *ngIf=\"!disabled\" class=\"ti3-tag-clear ti3-icon ti3-icon-close\" (click)=\"onClickDelete($event)\" [id]=\"appendId('_tag')\">\r\n        </div>\r\n</ng-container>\r\n<ng-container *ngIf=\"shape ==='rect'\"> -->\r\n        <section #content class=\"ti3-box-cell-key\" [id]=\"appendId('content')\" [title]=\"content.textContent\"><ng-content></ng-content></section>\r\n        <section class=\"ti3-icon ti3-icon-clear ti3-icon-close-staic\" (click)=\"onClickDelete($event)\" [id]=\"appendId('closeicon')\"></section>\r\n<!-- </ng-container> -->\r\n",
                host: {
                    '[class.ti3-multiselect-box-cell]': 'true',
                    '[class.ti3-multiselect-option-disabled]': 'disabled',
                    '[style.maxWidth]': 'maxWidth'
                }
                // styleUrls: ['./tag-rect.less', './tag-arrow.less'],
                // host: {
                //       '[class.ti3-tag-container]': 'shape ==="arrow"',
                //       '[class.ti3-multiselect-box-cell]': 'shape ==="rect"',
                //       '[class.ti3-multiselect-option-disabled]': 'shape ==="rect"&&disabled'
                // }
                ,
                styles: ["::ng-deep :root{--ti-tag-box-cell-space:1px;--ti-tag-icon-area-width:32px}:host.ti3-multiselect-box-cell{background-color:var(--ti-tag-bg-color);border-radius:var(--ti-common-border-radius-normal);box-sizing:border-box;color:var(--ti-tag-text-color);display:inline-block;float:left;height:calc(var(--ti-input-height) - 6px);line-height:calc(var(--ti-input-height) - 6px);margin:var(--ti-tag-box-cell-space);max-width:calc(100% - var(--ti-tag-box-cell-space)*2);overflow:hidden;width:auto}:host.ti3-multiselect-box-cell:after{content:\"\";display:block;overflow:hidden}:host.ti3-multiselect-box-cell .ti3-box-cell-key{box-sizing:border-box;float:left;margin-left:var(--ti-common-space-10);margin-right:var(--ti-common-space-6);overflow:hidden;text-overflow:ellipsis;white-space:nowrap;width:calc(100% - var(--ti-tag-icon-area-width) - var(--ti-common-space-6) - var(--ti-common-space-10))}:host.ti3-multiselect-box-cell .ti3-box-cell-key:hover{color:var(--ti-tag-text-color-hover)}:host.ti3-multiselect-box-cell .ti3-icon-close-staic{box-sizing:border-box;color:var(--ti-tag-icon-color);float:left;height:100%;text-align:center;width:var(--ti-tag-icon-area-width)}:host.ti3-multiselect-box-cell .ti3-icon-close-staic:hover{color:var(--ti-tag-icon-color-hover)}:host.ti3-multiselect-box-cell.ti3-multiselect-option-disabled{background-color:var(--ti-tag-bg-color-disabled);color:var(--ti-tag-text-color-disabled);cursor:not-allowed}:host.ti3-multiselect-box-cell.ti3-multiselect-option-disabled .ti3-box-cell-key{color:var(--ti-tag-text-color-disabled)}:host.ti3-multiselect-box-cell.ti3-multiselect-option-disabled .ti3-icon-close-staic{color:var(--ti-tag-icon-color-disabled)}"]
            },] }
];
TiTagComponent.propDecorators = {
    disabled: [{ type: Input }],
    selectedTipPosition: [{ type: Input }],
    maxWidth: [{ type: Input }],
    delete: [{ type: Output }]
};
