// @cloud/tinyplus3/esm2015/components/guides/TpGuideComponent.js
import { Component, ElementRef, Renderer2 } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
import { TpGuidesComponent } from './TpGuidesComponent';
/**
 * TiGuideComponent 是单个引导模块组件
 */
export class TpGuideComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, guidesComp) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.guidesComp = guidesComp;
        /**
         * @ignore
         * 下标值
         */
        this.cardIndex = -1;
        /**
         * @ignore
         * 是否含有body
         */
        this.hasBody = false;
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngAfterContentInit() {
        super.ngAfterContentInit();
        this.setCardIndex();
        this.cardIndex = this.nativeElement.cardIndex;
    }
    /**
     * @ignore
     * 设置步骤标志
     */
    setCardIndex() {
        const cardNum = this.guidesComp.nativeElement.children.length;
        for (let i = 0; i < cardNum; i++) {
            this.guidesComp.nativeElement.children[i].cardIndex = i;
            if (cardNum <= 2) {
                this.renderer.setStyle(this.guidesComp.nativeElement.children[i], 'max-width', '445px');
            }
        }
    }
    /**
     * @ignore
     * 是否待办项
     */
    isUndo(index) {
        return this.guidesComp.activeIndex >= 0 && index > this.guidesComp.activeIndex;
    }
    /**
     * @ignore
     * 是否当前项
     */
    isActive(index) {
        return this.guidesComp.activeIndex >= 0 && index === this.guidesComp.activeIndex;
    }
    /**
     * @ignore
     * 是否已完成
     */
    isDone(index) {
        return this.guidesComp.activeIndex >= 0 && index < this.guidesComp.activeIndex;
    }
    /**
     * @ignore
     * 是否含有间距
     */
    hasSpace(index) {
        return this.isActive(index) && this.hasBody;
    }
}
TpGuideComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-guide',
                template: "<!-- \u5361\u7247\u6807\u5FD7\u7EBF -->\r\n<span class=\"tp-guide-line\"></span>\r\n\r\n<!-- \u5361\u7247\u6807\u5FD7 -->\r\n<div class=\"tp-guide-circle\">\r\n    <span class=\"tp-guide-number\" *ngIf=\"!isDone(cardIndex)\">{{cardIndex + 1}}</span>\r\n    <tp-icon local name=\"cloud-action-complete2\"\r\n             class=\"tp-guide-number tp-guide-card-done\"\r\n             *ngIf=\"isDone(cardIndex)\"></tp-icon>\r\n</div>\r\n\r\n<!-- \u5361\u7247\u5185\u5BB9 -->\r\n<div class=\"tp-guide-card\">\r\n    <ng-content></ng-content>\r\n</div>\r\n\r\n<!-- \u6709\u65E0body\u95F4\u8DDD\u6837\u5F0F -->\r\n<div class=\"tp-guide-space\"\r\n     *ngIf=\"hasSpace(cardIndex)\"\r\n     [ngClass]=\"{'tp-guide-space-active': isActive(cardIndex) }\"></div>\r\n",
                host: {
                    '[class.tp-guide-card-container]': 'true',
                    '[class.tp-guide-card-undo]': 'isUndo(cardIndex)',
                    '[class.tp-guide-card-active]': 'isActive(cardIndex)',
                    '[class.tp-guide-card-done]': 'isDone(cardIndex)'
                },
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}:host{background:var(--ti-common-color-bg-white-normal);box-shadow:var(--ti-common-shadow-1-down);display:inline-block;flex:1;margin-right:var(--ti-common-space-2x);position:relative}:host:hover{box-shadow:var(--ti-common-shadow-3-down)}:host:last-child{margin-right:0}::ng-deep .tp-guide-card{-ms-box-sizing:border-box;background:var(--ti-common-color-bg-white-normal);border-top:var(--tp-guide-card-border-top) solid var(--ti-common-color-line-active);box-sizing:border-box;padding:var(--ti-common-space-6x) var(--ti-common-space-6x) var(--ti-common-space-5x)}:host.tp-guide-card-undo ::ng-deep .tp-guide-card{border-top-color:var(--ti-common-color-line-normal)}:host.tp-guide-card-active ::ng-deep .tp-guide-card{border-top-color:var(--ti-common-color-line-active)}[small] :host.tp-guide-card-active ::ng-deep .tp-guide-card{border-top-color:var(--ti-common-color-warn)}:host.tp-guide-card-done ::ng-deep .tp-guide-card{border-top-color:var(--ti-common-color-success)}[small] :host ::ng-deep .tp-guide-card{padding:var(--ti-common-space-3x) 0 var(--ti-common-space-3x) var(--ti-common-space-6x)}::ng-deep .tp-guide-circle{-ms-box-sizing:border-box;background:var(--ti-common-color-bg-white-normal);border:2px solid var(--ti-common-color-line-active);border-radius:50%;box-sizing:border-box;height:var(--tp-guide-circle-size);left:var(--tp-guide-circle-space-left);position:absolute;top:-15px;width:var(--tp-guide-circle-size)}::ng-deep .tp-guide-circle .tp-guide-number{color:var(--ti-common-color-text-highlight);display:inline-block;font-size:var(--ti-common-font-size-4);height:100%;line-height:var(--ti-common-line-height-number);text-align:center;width:100%}:host.tp-guide-card-undo ::ng-deep .tp-guide-circle{border-color:var(--ti-common-color-line-normal)}:host.tp-guide-card-undo ::ng-deep .tp-guide-circle .tp-guide-number{color:var(--ti-common-color-text-primary)}:host.tp-guide-card-active ::ng-deep .tp-guide-circle{background:var(--ti-common-color-icon-graybg-active);border-color:var(--ti-common-color-line-active)}:host.tp-guide-card-active ::ng-deep .tp-guide-circle .tp-guide-number{color:var(--ti-common-color-text-white)}[small] :host.tp-guide-card-active ::ng-deep .tp-guide-circle{background:var(--ti-common-color-warn);border-color:var(--ti-common-color-warn)}[small] :host.tp-guide-card-active ::ng-deep .tp-guide-circle .tp-guide-number{color:var(--ti-common-color-text-white)}:host.tp-guide-card-done ::ng-deep .tp-guide-circle{border-color:var(--ti-common-color-success)}:host.tp-guide-card-done ::ng-deep .tp-guide-circle .tp-guide-number{color:var(--ti-common-color-success);line-height:var(--tp-guide-circle-size)}[small] :host ::ng-deep .tp-guide-circle{height:var(--tp-guide-circle-size-small);top:-10px;width:var(--tp-guide-circle-size-small)}[small] :host ::ng-deep .tp-guide-circle .tp-guide-number{font-size:var(--ti-common-font-size-base);line-height:16px}.tp-guide-line{background:var(--ti-common-color-bg-white-normal);height:var(--tp-guide-card-border-top);left:calc(var(--tp-guide-circle-space-left) - var(--tp-guide-circle-white-space));position:absolute;width:calc(var(--tp-guide-circle-size) + 2*var(--tp-guide-circle-white-space))}[small] :host .tp-guide-line{width:calc(var(--tp-guide-circle-size-small) + 2*var(--tp-guide-circle-white-space))}.tp-guide-space{background:transparent;bottom:calc(-1*var(--tp-guide-body-space));height:var(--tp-guide-body-space);position:absolute;width:100%}.tp-guide-space.tp-guide-space-active{background:var(--ti-common-color-bg-white-normal)}"]
            },] }
];
TpGuideComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TpGuidesComponent }
];
