// @cloud/tinyplus3/esm2015/components/guides/TpGuideBodyComponent.js
import { Component, ElementRef, Renderer2 } from '@angular/core';
import { TiBaseComponent, TiRenderer } from '@cloud/tiny3';
import { TpGuideComponent } from './TpGuideComponent';
import { TpGuidesComponent } from './TpGuidesComponent';
/**
 * TiGuideBodyComponent 是单个引导模块的body区域
 */
export class TpGuideBodyComponent extends TiBaseComponent {
    constructor(elementRef, renderer2, tiRenderer, guideComp, guidesComp) {
        super(elementRef, renderer2);
        this.elementRef = elementRef;
        this.renderer2 = renderer2;
        this.tiRenderer = tiRenderer;
        this.guideComp = guideComp;
        this.guidesComp = guidesComp;
    }
    // tslint:disable-next-line: use-life-cycle-interface
    ngOnInit() {
        super.ngOnInit();
        this.guideComp.hasBody = true;
        this.tiRenderer.insertAfter(this.nativeElement, this.guidesComp.nativeElement);
    }
    /**
     * @ignore
     * 展示激活项body
     */
    showActiveBody() {
        return this.guidesComp.activeIndex === this.guideComp.cardIndex;
    }
}
TpGuideBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-guide-body',
                template: "<div class=\"tp-guide-body\" *ngIf=\"showActiveBody()\" >\r\n    <ng-content></ng-content>\r\n</div>\r\n",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.tp-guide-body{-ms-box-sizing:border-box;background:var(--ti-common-color-bg-white-normal);box-shadow:var(--ti-common-shadow-1-down);box-sizing:border-box;display:block;font-size:var(--ti-common-font-size-base);margin-top:var(--tp-guide-body-space);padding:var(--ti-common-space-6x) var(--ti-common-space-6x) var(--ti-common-space-6x) var(--ti-common-space-6x);width:100%}"]
            },] }
];
TpGuideBodyComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: TiRenderer },
    { type: TpGuideComponent },
    { type: TpGuidesComponent }
];
