// @cloud/tinyplus3/esm2015/components/guides/TpGuidesModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TpIconModule } from '../icon/TpIconModule';
import { TpGuidesComponent } from './TpGuidesComponent';
import { TpGuideComponent } from './TpGuideComponent';
import { TpGuideHeaderComponent } from './TpGuideHeaderComponent';
import { TpGuideContentComponent } from './TpGuideContentComponent';
import { TpGuideBodyComponent } from './TpGuideBodyComponent';
export class TpGuidesModule {
}
TpGuidesModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    TpIconModule
                ],
                exports: [TpGuidesComponent, TpGuideComponent, TpGuideHeaderComponent, TpGuideContentComponent, TpGuideBodyComponent],
                declarations: [TpGuidesComponent, TpGuideComponent, TpGuideHeaderComponent, TpGuideContentComponent, TpGuideBodyComponent]
            },] }
];
export { TpGuidesComponent } from './TpGuidesComponent';
export { TpGuideComponent } from './TpGuideComponent';
export { TpGuideHeaderComponent } from './TpGuideHeaderComponent';
export { TpGuideContentComponent } from './TpGuideContentComponent';
export { TpGuideBodyComponent } from './TpGuideBodyComponent';
