// @cloud/tinyplus3/esm2015/components/productpreview/TpProductpreviewComponent.js
import { Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
import { TiImagepreviewComponent, TiModalService, TiBaseComponent } from '@cloud/tiny3';
/**
 * 商品预览组件
 *
 * 该组件用来展示商品的预览内容，包括上方的主预览图和下方的缩略图
 *
 * 10.0.3 增加
 *
 * <example-url>../tinyplus3demo/#/productpreview/productpreview-all</example-url>
 */
export class TpProductpreviewComponent extends TiBaseComponent {
    constructor(tiModal, render, hostEleRef) {
        super(hostEleRef, render);
        this.tiModal = tiModal;
        this.render = render;
        this.hostEleRef = hostEleRef;
        /**
         * 预览文件列表
         */
        this.files = [];
        /**
         * @ignore
         * 缩略图列表，翻页时重新设置
         */
        this.thumbList = [];
        /**
         * @ignore 记录当前预览的商品索引
         */
        this.currentPreviewIndex = -1;
        /**
         * @ignore 缩略图当前页
         */
        this.currentThumbPage = 1;
        /**
         * @ignore 缩略图总页数
         *
         */
        this.totalThumbPage = 1;
    }
    ngAfterViewInit() {
        this.render.addClass(this.productThumbEle.nativeElement.children[0], 'tp-product-preview-thumb-item-active');
    }
    ngOnChanges(changes) {
        if (changes.files && changes.files.firstChange === false) {
            this.files = changes.files.currentValue;
            this.initThumb();
            if (this.productThumbEle) {
                this.setThumbList();
            }
            // 确保files接口变化时，第一个缩略图的选中样式生效
            setTimeout(() => {
                if (this.productThumbEle) {
                    this.render.addClass(this.productThumbEle.nativeElement.children[0], 'tp-product-preview-thumb-item-active');
                }
            }, 0);
        }
    }
    ngOnInit() {
        super.ngOnInit();
        this.initThumb();
    }
    /**
     * 初始化缩略图相关变量
     */
    initThumb() {
        if (this.files && this.files.length > 0) {
            this.currentPreviewIndex = 0;
            // 计算总页数
            this.totalThumbPage = Math.ceil(this.files.length / TpProductpreviewComponent.ITEM_PER_PAGE);
            this.currentThumbPage = 1;
            this.files.length >= TpProductpreviewComponent.ITEM_PER_PAGE ?
                this.thumbList = this.files.slice(0, TpProductpreviewComponent.ITEM_PER_PAGE) : this.thumbList = this.files;
        }
    }
    /**
     * @ignore 主图查看
     * @param event: 事件
     * @param index: 当前选中项在this.thumbList中的索引,this.thumbList中始终只有4项
     */
    onMouseenterThumb(event, index) {
        this.currentPreviewIndex = (this.currentThumbPage - 1) * TpProductpreviewComponent.ITEM_PER_PAGE + index;
        // 计算最后一页的缩略图个数
        const thumbItemNumber = this.files.length % TpProductpreviewComponent.ITEM_PER_PAGE;
        // 如果是最后一页，且最后一页不满4个
        if (this.totalThumbPage > 1 && this.currentThumbPage === this.totalThumbPage && thumbItemNumber !== 0) {
            this.currentPreviewIndex -= (TpProductpreviewComponent.ITEM_PER_PAGE - thumbItemNumber);
        }
        // 设置选中样式
        const thumbItemEles = this.productThumbEle.nativeElement.children;
        for (const item of thumbItemEles) {
            this.render.removeClass(item, 'tp-product-preview-thumb-item-active');
        }
        this.render.addClass(event.target.parentElement, 'tp-product-preview-thumb-item-active');
    }
    /**
     * @ignore 弹框中预览大图
     */
    onClickMain(event) {
        this.tiModal.open(TiImagepreviewComponent, {
            id: this.id + '_productPreviewModal',
            modalClass: 'tp-product-preview-modal',
            context: {
                index: this.currentPreviewIndex,
                fileList: this.files,
                id: this.id // 传递id
            }
        });
    }
    /**
     * @ignore 缩略图上一页
     */
    prev() {
        if (this.currentThumbPage === 1) {
            return;
        }
        this.currentThumbPage--;
        this.setThumbList();
    }
    /**
     * @ignore 缩略图下一页
     */
    next() {
        if (this.currentThumbPage === this.totalThumbPage) {
            return;
        }
        this.currentThumbPage++;
        this.setThumbList();
    }
    /**
     * 设置缩略图列表
     */
    setThumbList() {
        // undefined、null、false、string情况下，不处理
        if (!this.files || typeof this.files === 'string' || (this.files && this.files.slice === undefined)) {
            return;
        }
        // tslint:disable-next-line: max-line-length
        this.thumbList = this.files.slice((this.currentThumbPage - 1) * TpProductpreviewComponent.ITEM_PER_PAGE, this.currentThumbPage * TpProductpreviewComponent.ITEM_PER_PAGE);
        // 计算最后一页的缩略图个数
        const thumbItemNumber = this.files.length % TpProductpreviewComponent.ITEM_PER_PAGE;
        if (this.totalThumbPage > 1 && this.currentThumbPage === this.totalThumbPage && thumbItemNumber !== 0) {
            this.thumbList = this.files.slice(this.files.length - TpProductpreviewComponent.ITEM_PER_PAGE, this.files.length);
        }
    }
}
/**
 * 每页显示的小图个数
 */
TpProductpreviewComponent.ITEM_PER_PAGE = 4;
TpProductpreviewComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-productpreview',
                template: "<div class=\"tp-product-preview-container\" *ngIf=\"files && files.length\">\r\n    <!-- \u5927\u56FE\u663E\u793A -->\r\n    <div tiOutline class=\"tp-product-preview-main\" (click)=\"onClickMain($event)\" [tabindex]=\"0\" (keydown.enter)=\"onClickMain($event)\" [id]=\"appendId('main')\">\r\n        <img [src]='files[currentPreviewIndex].previewUrl' tiZoom/>\r\n    </div>\r\n    <!-- \u7F29\u7565\u56FE\u663E\u793A\u53CA\u5DE6\u53F3\u5207\u6362 -->\r\n    <div class=\"tp-product-preview-thumb-container\">\r\n        <!-- \u5411\u524D\u7FFB\u9875 -->\r\n        <div tiOutline class=\"tp-product-preview-thumb-left tp-product-preview-thumb-page\"\r\n            [ngClass]=\"{'tp-product-preview-thumb-page-disabled': currentThumbPage === 1}\"\r\n            (click)=\"prev()\"\r\n            [tabindex]=\"currentThumbPage === 1 ? -1 : 0\"\r\n            (keydown.enter)=\"prev()\"\r\n            [id]=\"appendId('prev')\">\r\n            <ti-icon class=\"tp-product-preview-icon\" name='angle-left'></ti-icon>\r\n        </div>\r\n        <!-- \u7F29\u7565\u56FE\u5217\u8868 -->\r\n        <div class=\"tp-product-preview-thumb-wapper\">\r\n            <div class='tp-product-preview-thumb' #productThumb>\r\n                <div class=\"tp-product-preview-thumb-item\" *ngFor='let item of thumbList; index as i;'>\r\n                    <img [src]='item.previewUrl' (mouseenter)='onMouseenterThumb($event, i)' [id]=\"appendId('active_' + i)\"/>\r\n                </div>\r\n            </div>\r\n        </div>\r\n        <!-- \u5411\u540E\u7FFB\u9875 -->\r\n        <div tiOutline class=\"tp-product-preview-thumb-right tp-product-preview-thumb-page\"\r\n            [ngClass]=\"{'tp-product-preview-thumb-page-disabled': currentThumbPage === totalThumbPage}\"\r\n            (click)=\"next()\"\r\n            [tabindex]=\"currentThumbPage === totalThumbPage ? -1 : 0\"\r\n            (keydown.enter)=\"next()\"\r\n            [id]=\"appendId('next')\">\r\n            <ti-icon class='tp-product-preview-icon' name='angle-right'></ti-icon>\r\n        </div>\r\n    </div>\r\n</div>\r\n",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}::ng-deep :root{--tp-product-preview-main-large-size:450px;--tp-product-preview-main-middle-size:380px;--tp-product-preview-main-small-size:300px;--tp-product-preview-modal-width:640px;--tp-product-preview-thumb-large-size:86.5px;--tp-product-preview-thumb-middle-size:69px;--tp-product-preview-thumb-page-border-radius:2px;--tp-product-preview-thumb-page-button-size:calc(var(--tp-product-preview-thumb-page-size) + var(--tp-product-preview-thumb-page-margin));--tp-product-preview-thumb-page-font-size:24px;--tp-product-preview-thumb-page-margin:16px;--tp-product-preview-thumb-page-size:24px;--tp-product-preview-thumb-small-size:49px}:host{display:block}.tp-product-preview-container{width:100%}.tp-product-preview-container .tp-product-preview-main{align-items:center;border:1px solid var(--ti-common-color-line-dividing);box-sizing:border-box;display:flex;justify-content:center}.tp-product-preview-container .tp-product-preview-thumb-container{align-items:center;display:flex;justify-content:space-between;margin-top:var(--ti-common-space-3x)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper{flex:1;overflow:hidden}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb{display:flex;left:0;position:relative}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item{align-items:center;border:1px solid transparent;box-sizing:border-box;display:flex;flex-shrink:0;justify-content:center}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item img{height:auto;width:auto}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item:not(:last-child){margin-right:var(--ti-common-space-2x)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item-active{border:1px solid var(--ti-common-color-line-active);cursor:pointer}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-page{border-radius:var(--tp-product-preview-thumb-page-border-radius);height:var(--tp-product-preview-thumb-page-size);position:relative;width:var(--tp-product-preview-thumb-page-size)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-page:hover:not(.tp-product-preview-thumb-page-disabled){background-color:var(--ti-common-color-bg-light-normal);color:var(--ti-common-color-icon-hover);cursor:pointer}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-page:active:not(.tp-product-preview-thumb-page-disabled){background-color:var(--ti-common-color-icon-active);color:var(--ti-common-bg-minor-active)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-page .tp-product-preview-icon{font-size:var(--tp-product-preview-thumb-page-font-size);line-height:var(--tp-product-preview-thumb-page-size)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-page-disabled{color:var(--ti-common-color-icon-disabled);cursor:not-allowed;outline:none}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-left{margin-right:var(--tp-product-preview-thumb-page-margin)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-right{margin-left:var(--tp-product-preview-thumb-page-margin)}@media screen and (max-width:1440px){:host{width:var(--tp-product-preview-main-small-size)}::ng-deep .ti3-img-zoom-viewer{height:var(--tp-product-preview-main-small-size)!important;width:var(--tp-product-preview-main-small-size)!important}.tp-product-preview-container .tp-product-preview-main{height:var(--tp-product-preview-main-small-size)}.tp-product-preview-container .tp-product-preview-main img{max-height:calc(var(--tp-product-preview-main-small-size) - 2px);max-width:calc(var(--tp-product-preview-main-small-size) - 2px)}.tp-product-preview-container .tp-product-preview-main ::ng-deep .ti3-img-zoom-selector{height:calc(var(--tp-product-preview-main-small-size)/2)!important;width:calc(var(--tp-product-preview-main-small-size)/2)!important}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item{height:var(--tp-product-preview-thumb-small-size);width:var(--tp-product-preview-thumb-small-size)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item img{max-height:calc(var(--tp-product-preview-thumb-small-size) - 2px);max-width:calc(var(--tp-product-preview-thumb-small-size) - 2px)}}@media screen and (min-width:1440px) and (max-width:1600px){:host{width:var(--tp-product-preview-main-middle-size)}::ng-deep .ti3-img-zoom-viewer{height:var(--tp-product-preview-main-middle-size)!important;width:var(--tp-product-preview-main-middle-size)!important}.tp-product-preview-container .tp-product-preview-main{height:var(--tp-product-preview-main-middle-size)}.tp-product-preview-container .tp-product-preview-main img{max-height:calc(var(--tp-product-preview-main-middle-size) - 2px);max-width:calc(var(--tp-product-preview-main-middle-size) - 2px)}.tp-product-preview-container .tp-product-preview-main ::ng-deep .ti3-img-zoom-selector{height:calc(var(--tp-product-preview-main-middle-size)/2)!important;width:calc(var(--tp-product-preview-main-middle-size)/2)!important}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item{height:var(--tp-product-preview-thumb-middle-size);width:var(--tp-product-preview-thumb-middle-size)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item img{max-height:calc(var(--tp-product-preview-thumb-middle-size) - 2px);max-width:calc(var(--tp-product-preview-thumb-middle-size) - 2px)}}@media screen and (min-width:1600px){:host{width:var(--tp-product-preview-main-large-size)}::ng-deep .ti3-img-zoom-viewer{height:var(--tp-product-preview-main-large-size)!important;width:var(--tp-product-preview-main-large-size)!important}.tp-product-preview-container .tp-product-preview-main{height:var(--tp-product-preview-main-large-size)}.tp-product-preview-container .tp-product-preview-main img{max-height:calc(var(--tp-product-preview-main-large-size) - 2px);max-width:calc(var(--tp-product-preview-main-large-size) - 2px)}.tp-product-preview-container .tp-product-preview-main ::ng-deep .ti3-img-zoom-selector{height:calc(var(--tp-product-preview-main-large-size)/2)!important;width:calc(var(--tp-product-preview-main-large-size)/2)!important}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item{height:var(--tp-product-preview-thumb-large-size);width:var(--tp-product-preview-thumb-large-size)}.tp-product-preview-container .tp-product-preview-thumb-container .tp-product-preview-thumb-wapper .tp-product-preview-thumb .tp-product-preview-thumb-item img{max-height:calc(var(--tp-product-preview-thumb-large-size) - 2px);max-width:calc(var(--tp-product-preview-thumb-large-size) - 2px)}}::ng-deep .tp-product-preview-modal{width:var(--tp-product-preview-modal-width)!important}::ng-deep .tp-product-preview-modal .ti3-close{top:0}::ng-deep .tp-product-preview-modal .ti3-image-preview-container{margin-top:var(--ti-common-space-2x)}"]
            },] }
];
TpProductpreviewComponent.ctorParameters = () => [
    { type: TiModalService },
    { type: Renderer2 },
    { type: ElementRef }
];
TpProductpreviewComponent.propDecorators = {
    files: [{ type: Input }],
    productThumbEle: [{ type: ViewChild, args: ['productThumb',] }]
};
