// @cloud/tiny3/esm2015/directives/drag/TiDraggableDirective.js
import { Directive, ElementRef, Input } from '@angular/core';
import { TiDragService } from '../../services/drag/TiDragService';
/**
 * 拖拽指令，除指令方式外，还提供了服务生成方式[TiDragService]{@link ../injectables/TiDragService.html}
 *
 * <example-url>../tiny3demo/#/drag/drag-all</example-url>
 */
export class TiDraggableDirective {
    constructor(hostEle, dragService) {
        this.hostEle = hostEle;
        this.dragService = dragService;
        /**
         * 拖拽是否灰化，可动态修改
         */
        this.tiDragDisabled = false;
    }
    ngOnInit() {
        this.options = Object.assign(Object.assign({}, this.tiDraggable), { helper: this.hostEle.nativeElement, disabled: this.tiDragDisabled });
        this.dragInstance = this.dragService.create(this.options);
    }
    ngOnChanges(changes) {
        // 通过直接修改options的方式修改disabled状态
        if (changes['tiDragDisabled'] && !changes['tiDragDisabled'].firstChange) {
            this.options.disabled = this.tiDragDisabled;
        }
    }
    ngOnDestroy() {
        // 修复SSR错误：TypeError: Cannot read property 'destroy' of undefined
        this.dragInstance && this.dragInstance.destroy();
    }
}
TiDraggableDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiDraggable]'
            },] }
];
TiDraggableDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: TiDragService }
];
TiDraggableDirective.propDecorators = {
    tiDraggable: [{ type: Input }],
    tiDragDisabled: [{ type: Input }]
};
