// @cloud/tiny3/esm2015/directives/validation/TiErrorMsgComponent.js
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { TiBaseComponent } from '../../components/base/TiBaseModule';
/**
 * @ignore
 */
export class TiErrorMsgComponent extends TiBaseComponent {
}
TiErrorMsgComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-error-msg',
                template: `
        <span class="ti3-error-icon-bg">
            <span class='ti3-error-icon ti3-icon ti3-icon-alert-warn'></span>
        </span>
        <span class="ti3-validate-text" [innerHTML]="errorMessage"></span>
    `,
                encapsulation: ViewEncapsulation.None,
                host: {
                    '[class.ti3-unifyvalid-error-container]': 'isAppendBody',
                    '[class.ti3-unifyvalid-tip-error-container]': 'true'
                },
                styles: [".ti3-error-icon-bg{background-color:var(--ti-valid-error-color);border-radius:50%;flex-shrink:0;height:16px;margin-right:var(--ti-common-space-2x);text-align:center;width:16px}.ti3-error-icon{color:var(--ti-common-color-text-white);font-size:var(--ti-common-font-size-base)}.ti3-unifyvalid-error-container{color:var(--ti-valid-error-color);display:block;margin-top:var(--ti-common-space-2x)}.ti3-unifyvalid-tip-error-container{align-content:center;display:flex}.ti3-tooltip-content .ti3-error-icon-bg{line-height:16px;margin-top:1px}"]
            },] }
];
TiErrorMsgComponent.propDecorators = {
    errorMessage: [{ type: Input }],
    isAppendBody: [{ type: Input }]
};
