// @cloud/tiny3/esm2015/directives/validation/TiPendingStateComponent.js
import { Component, Input } from '@angular/core';
import { TiBaseComponent } from '../../components/base/TiBaseModule';
/**
 * @ignore
 */
export class TiPendingStateComponent extends TiBaseComponent {
    ngOnInit() {
        this.renderer.addClass(this.validElement, 'ti3-text-input-show-pending');
    }
    ngOnDestroy() {
        if (this.validElement) {
            this.renderer.removeClass(this.validElement, 'ti3-text-input-show-pending');
        }
    }
}
TiPendingStateComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-pending-state',
                template: "<ti-loading class=\"ti3-validation-pending-loading\"></ti-loading>",
                host: {
                    '[class.ti3-validation-pending-container]': 'true'
                },
                styles: [":host.ti3-validation-pending-container{display:inline-block;height:var(--ti-input-height);position:relative;vertical-align:top;width:0}.ti3-validation-pending-loading{left:-22px;position:absolute;top:calc((var(--ti-input-height) - 14px)/2)}::ng-deep [tiText].ti3-text-input-show-clear+.ti3-validation-pending-container .ti3-validation-pending-loading{left:-38px}::ng-deep [tiText].ti3-text-input-show-pending{padding-right:32px!important}::ng-deep [tiText].ti3-text-input-show-pending.ti3-text-input-show-clear{padding-right:48px!important}"]
            },] }
];
TiPendingStateComponent.propDecorators = {
    validElement: [{ type: Input }]
};
