// @cloud/tiny3/esm2015/directives/outline/TiOutlineDirective.js
import { Directive, ElementRef, Renderer2, NgZone } from '@angular/core';
/**
 * @ignore
 * 点击需要和聚焦区分开。当点击的时候，不需要有outline。
 * 10.1.4新增
 */
export class TiOutlineDirective {
    constructor(renderer, hostEle, zone) {
        this.renderer = renderer;
        this.hostEle = hostEle;
        this.zone = zone;
    }
    ngAfterViewInit() {
        this.zone.runOutsideAngular(() => {
            const outlineHostEle = this.hostEle.nativeElement;
            if (outlineHostEle) {
                this.renderer.listen(outlineHostEle, 'mousedown', () => {
                    this.renderer.setStyle(outlineHostEle, 'outline', 'none');
                });
                this.renderer.listen(outlineHostEle, 'blur', () => {
                    this.renderer.setStyle(outlineHostEle, 'outline', '');
                });
            }
        });
    }
}
TiOutlineDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiOutline]'
            },] }
];
TiOutlineDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef },
    { type: NgZone }
];
