// @cloud/tiny3/esm2015/components/switch/TiSwitchComponent.js
import { Component, ElementRef, EventEmitter, Output, ViewChild } from '@angular/core';
import { TiFormComponent } from '../base/TiBaseModule';
/**
 * Switch开关组件
 *
 * 用于实现页面中的开关操作。
 *
 * <example-url>../tiny3demo/#/switch/switch-all</example-url>
 */
export class TiSwitchComponent extends TiFormComponent {
    constructor() {
        super(...arguments);
        /**
         * 组件开关状态即将切换之前发出事件:。组件状态不改变，后续全权交给组件调用者处理。
         */
        this.beforeChange = new EventEmitter();
        this.initialized = false;
        this.hasAnimation = false;
    }
    ngOnInit() {
        // 基类中做了设置宿主id的操作
        super.ngOnInit();
        this.setFocusableElems([this.aRef.nativeElement]);
    }
    ngOnModelChange(value) {
        super.ngOnModelChange(value);
        if (value === null) { // 以 ngModel 的形式传入值时, writeValue首次传入null
            return;
        }
        if (this.initialized && !this.hasAnimation) { // 保证初始时没有动画
            this.renderer.addClass(this.aRef.nativeElement, 'ti3-switch-animation');
            this.hasAnimation = true;
        }
        this.initialized = true;
    }
    /**
     * 切换开关状态
     */
    toggle() {
        if (this.beforeChange.observers.length === 0) { // 用户未定义beforeChange
            this.model = !this.model;
        }
        else {
            this.beforeChange.emit(this);
        }
    }
    /**
     * @ignore
     * 点击事件
     */
    onClick() {
        if (!this.disabled) {
            this.toggle();
        }
    }
}
TiSwitchComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-switch',
                template: "<a #a\r\n    role=\"radio\"\r\n    class=\"ti3-switch\"\r\n    unselectable=\"on\"\r\n    href=\"javascript:void(0)\"\r\n    oncontextmenu=\"return false\"\r\n    (click)=\"onClick()\"\r\n    [ngClass]=\"{\r\n    'ti3-switch-off': !model, \r\n    'ti3-switch-on': model}\"\r\n    [id]=\"appendId('awrapper')\">\r\n    <div class=\"ti3-switch-track\"></div>\r\n    <span class=\"ti3-switch-pointer\"></span>\r\n</a>",
                host: {
                    '[class.ti3-switch-container]': 'true'
                },
                providers: [TiFormComponent.getValueAccessor(TiSwitchComponent)],
                styles: ["::ng-deep :root{--ti-switch-height:20px;--ti-switch-pointer-off-left:2px;--ti-switch-pointer-on-left:20px;--ti-switch-pointer-top:2px;--ti-switch-pointer-width:16px;--ti-switch-width:38px}.ti3-switch,:host.ti3-switch-container{display:inline-block;height:var(--ti-switch-height)}.ti3-switch{-ms-user-select:none;-webkit-user-select:none;cursor:pointer;position:relative;user-select:none;width:var(--ti-switch-width)}.ti3-switch .ti3-switch-track{border-radius:var(--ti-switch-height);height:100%;width:100%;z-index:0}.ti3-switch .ti3-switch-pointer,.ti3-switch .ti3-switch-track{-ms-box-sizing:border-box;box-sizing:border-box;position:absolute}.ti3-switch .ti3-switch-pointer{background:var(--ti-common-color-bg-white-normal);border-radius:50%;display:inline-block;height:var(--ti-switch-pointer-width);top:var(--ti-switch-pointer-top);width:var(--ti-switch-pointer-width);z-index:8}.ti3-switch.ti3-switch-off .ti3-switch-track{background:var(--ti-switch-track-off-bg-color)}.ti3-switch.ti3-switch-off .ti3-switch-pointer{left:var(--ti-switch-pointer-off-left)}.ti3-switch.ti3-switch-on .ti3-switch-track{background:var(--ti-switch-track-on-bg-color)}.ti3-switch.ti3-switch-on .ti3-switch-pointer{left:var(--ti-switch-pointer-on-left)}.ti3-switch[disabled]{cursor:not-allowed}.ti3-switch[disabled] .ti3-switch-pointer{background:var(--ti-common-color-bg-disabled)}.ti3-switch[disabled].ti3-switch-on .ti3-switch-track{background-color:var(--ti-common-color-bg-light-emphasize)}.ti3-switch[disabled].ti3-switch-off .ti3-switch-track{background-color:var(--ti-switch-track-bg-color-disabled)}.ti3-switch.ti3-switch-animation .ti3-switch-pointer,.ti3-switch.ti3-switch-animation .ti3-switch-track{transition:all .25s linear}"]
            },] }
];
TiSwitchComponent.propDecorators = {
    beforeChange: [{ type: Output }],
    aRef: [{ type: ViewChild, args: ['a', { static: true },] }]
};
