// @cloud/tinyplus3/esm2015/components/purchaselist/TpPurchaseitemComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * <example-url>../tinyplus3demo/#/purchaselist/purchaselist-all</example-url>
 */
export class TpPurchaseitemComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 商品信息是否带有region
         */
        this.hasArea = false;
        /**
         * 单个产品右侧叉号点击事件
         */
        this.close = new EventEmitter();
    }
    /**
     * @ignore
     */
    closeFn() {
        this.close.emit();
    }
}
TpPurchaseitemComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-purchaseheader-item',
                template: "\r\n    <div *ngIf='hasArea' class='tp-purchaselist-item-area'>\r\n        <ng-content select=\"[area]\"></ng-content>\r\n    </div>\r\n    <div class='tp-purchaselist-item-content-container'>\r\n        <div class='tp-purchaselist-item-content' [ngClass]=\"{'tp-purchaselist-item-noneborder': hasArea}\">\r\n            <ng-content></ng-content>\r\n            <tp-icon local tiOutline class='tp-purchaselist-item-clear' name='cloud-close' [tabindex]=\"0\" (keydown.enter)=\"closeFn()\"\r\n            (click)='closeFn()' [id]=\"appendId('close')\"></tp-icon>\r\n        </div>\r\n    </div>\r\n",
                styles: [".tp-purchaselist-item-area{background-color:var(--ti-common-color-bg-white-emphasize);height:32px;line-height:32px}.tp-purchaselist-item-area,.tp-purchaselist-item-content-container{padding:0 var(--ti-common-space-4x)}.tp-purchaselist-item-content{border-top:1px solid var(--ti-common-color-line-dividing);padding:var(--ti-common-space-3x) 0;position:relative}.tp-purchaselist-item-noneborder{border:none}.tp-purchaselist-item-clear{color:var(--ti-common-color-icon-normal);cursor:pointer;position:absolute;right:0;top:16px}.tp-purchaselist-item-clear:hover{color:var(--ti-common-color-prompt)}"]
            },] }
];
TpPurchaseitemComponent.propDecorators = {
    hasArea: [{ type: Input }],
    close: [{ type: Output }]
};
