// @cloud/tinyplus3/esm2015/components/purchaselist/TpPurchaselistComponent.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TiBaseComponent } from '@cloud/tiny3';
/**
 * 购买清单组件
 * <example-url>../tinyplus3demo/#/purchaselist/purchaselist-all</example-url>
 */
export class TpPurchaselistComponent extends TiBaseComponent {
    constructor() {
        super(...arguments);
        /**
         * 是否存在一键清空功能
         */
        this.enableClear = true;
        /**
         * 费用总金额
         */
        this.totalPrice = 0;
        /**
         * 优惠金额
         */
        this.discountPrice = 0;
        /**
         * 用户可根据实际场景决定确认按钮是否禁用（比如未实名认证等）
         *
         * 10.1.1新增接口
         */
        this.buttonDisabled = true;
        /**
         * 采购产品数量统计
         */
        this.total = 0;
        /**
         * 订单确认按钮点击事件
         */
        this.confirm = new EventEmitter();
        /**
         * 头部一键清空按钮点击事件
         */
        this.clear = new EventEmitter();
        /**
         * 可选优惠项切换选择项时触发
         *
         * 10.1.2 新增接口
         */
        this.discountChange = new EventEmitter();
        /**
         * @ignore
         */
        this.emptyText = '--';
        /**
         * @ignore
         * 协议是否勾选
         */
        this.checked = false;
    }
    /**
     * @ignore
     */
    discountsChange($event) {
        this.discountChange.emit($event);
    }
    onCheckboxNgModelChange(model) {
        // Todo 禁用tip
    }
    /**
     * @ignore
     * 确认按钮点击触发
     */
    clickFn() {
        this.confirm.emit();
    }
    /**
     * @ignore
     * 一键清空
     */
    clearFn() {
        this.clear.emit();
    }
}
TpPurchaselistComponent.decorators = [
    { type: Component, args: [{
                selector: 'tp-purchaselist',
                template: "<div class='tp-purchaselist-container'>\r\n    <!-- \u5934\u90E8\u533A\u57DF -->\r\n    <div class='tp-purchaselist-header'>\r\n        <div class='tp-purchaselist-title' [id]=\"appendId('title')\">\r\n            {{purchaseTitle}}<span>(</span>{{total}}<span>)</span>\r\n        </div>\r\n        <div *ngIf='enableClear && total' tiOutline [tabindex]=\"0\" class='tp-purchaselist-clear' (keydown.enter)=\"clearFn()\" (click)='clearFn()' [id]=\"appendId('clear')\">\r\n            {{'tpPurchaselist.purchase_list_clear_label' | tiTranslate}}\r\n        </div>\r\n    </div>\r\n    <div class='tp-purchaselist-body' [ngClass]=\"{'tp-purchaselist-body-align': !total}\">\r\n        <ng-content></ng-content>\r\n        <span *ngIf=\"!total\" class=\"tp-purchaselist-nodata-text\">{{'tpPurchaselist.purchase_list_nodata_label' | tiTranslate}}</span>\r\n    </div>\r\n    <div class='tp-purchaselist-footer'>\r\n        <div class='tp-purchaselist-price-container' [ngClass]=\"{'tp-purchaselist-nodata-price': total === 0}\">\r\n            <span>{{'tpPurchaselist.purchase_list_cost_label' | tiTranslate}}</span>\r\n            <span *ngIf='total === 0'>{{emptyText}}</span>\r\n            <span *ngIf='total > 0' class='tp-purchaselist-totalprice'>{{totalPrice | tpCurrency: numFormat}}</span>\r\n            <span style='display: inline-block;'>\r\n                <div *ngIf='discountPrice !== 0 && total > 0' style='display: inline-block;'>\r\n                    <span class='tp-purchaselist-discount' [id]=\"appendId('discount')\">\r\n                        <span>(</span>\r\n                        <span>{{'tpBuylayer.buyLayer_priceDiscount_label' | tiTranslate}}</span>\r\n                        <span>{{discountPrice | tpCurrency: numFormat}}</span>\r\n                       <span>)</span>\r\n                    </span>\r\n                </div><!-- \u4F18\u60E0\u8BE6\u60C5\r\n                --><tp-discounts *ngIf=\"purchaseDiscoutsConfig !== null && total > 0\" class='tp-purchaselist-discount-tip' (discountsChange)='discountsChange($event)'\r\n                    [discoutsConfig]='purchaseDiscoutsConfig' [id]=\"appendId('discounts')\"></tp-discounts><!--\r\n                --><tp-helptip\r\n                *ngIf=\"helpOptions !== null && total > 0\"\r\n                [iconTip]=\"helpOptions.iconTip\"\r\n                [iconTipPosition]='helpOptions.iconTipPosition'\r\n                [iconTipMaxWidth]='helpOptions.iconTipMaxWidth' class='tp-purchaselist-helptip' [id]=\"appendId('helptip')\"></tp-helptip>\r\n            </span>\r\n\r\n        </div>\r\n        <div *ngIf=\"footerProtocol\" class='tp-footer-protocol' [id]=\"appendId('protocol')\">\r\n            <input id='protocol' type=\"checkbox\" (ngModelChange)=\"onCheckboxNgModelChange($event)\" [(ngModel)]=\"checked\" tiCheckbox >\r\n            <label for='protocol' [innerHTML]='footerProtocol'></label>\r\n        </div>\r\n        <button class='tp-purchaselist-button' type=\"button\"\r\n            [ngClass]=\"{'tp-purchaselist-no-footerPrompt': !footerPrompt}\"\r\n            color=\"danger\"\r\n            [disabled]=\" (footerProtocol ? !checked : false) || !total || buttonDisabled\"\r\n            (click)='clickFn()' [id]=\"appendId('button')\" tiButton>{{buttonText}}</button>\r\n        <div *ngIf=\"footerPrompt\" class='tp--footer-prompt' [innerHTML]='footerPrompt' [id]=\"appendId('prompt')\"></div>\r\n    </div>\r\n</div>",
                styles: [".ellipsis{overflow:hidden;text-overflow:ellipsis;white-space:nowrap}.tp-purchaselist-container{background-color:var(--ti-common-color-bg-white-normal);box-shadow:var(--ti-common-shadow-1-down)}.tp-purchaselist-header{align-items:center;display:flex;height:56px;justify-content:space-between;line-height:var(--ti-common-line-height-number);width:100%}.tp-purchaselist-title{border-left:4px solid var(--ti-common-color-prompt);color:var(--ti-common-color-text-primary);font-size:var(--ti-common-font-size-2);font-weight:700;padding-left:var(--ti-common-space-3x)}.tp-purchaselist-clear{color:var(--ti-common-color-text-link);cursor:pointer;font-size:var(--ti-common-font-size-base);padding-right:var(--ti-common-space-4x)}.tp-purchaselist-clear:hover{color:var(--ti-common-color-text-link-hover)}.tp-purchaselist-footer{padding:0 var(--ti-common-space-4x)}.tp-purchaselist-totalprice{color:var(--ti-common-color-text-important);font-size:var(--ti-common-font-size-3);line-height:var(--ti-common-line-height-number)}.tp-purchaselist-discount{border-right:1px solid var(--ti-common-color-line-dividing);color:var(--ti-common-color-text-secondary);padding-left:var(--ti-common-space-base);padding-right:var(--ti-common-space-3x)}.tp-purchaselist-discount-tip{padding-bottom:var(--ti-common-space-base);padding-left:var(--ti-common-space-3x)}.tp-footer-protocol{display:flex;padding-top:var(--ti-common-space-5x)}.tp-purchaselist-button{margin-top:var(--ti-common-space-2x);width:100%}.tp-purchaselist-no-footerPrompt{margin-bottom:var(--ti-common-space-4x)}.tp--footer-prompt{color:var(--ti-common-color-text-weaken);font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number);padding:var(--ti-common-space-2x) 0}.tp-purchaselist-body-align{text-align:center}.tp-purchaselist-nodata-text{color:var(--ti-common-color-text-weaken);font-size:var(--ti-common-font-size-base);line-height:var(--ti-common-line-height-number);padding-top:var(--ti-common-space-base)}.tp-purchaselist-helptip{vertical-align:super}.tp-purchaselist-price-container{border-top:1px solid var(--ti-common-color-line-dividing);padding-top:12px}.tp-purchaselist-nodata-price{border:none}"]
            },] }
];
TpPurchaselistComponent.propDecorators = {
    purchaseTitle: [{ type: Input }],
    enableClear: [{ type: Input }],
    totalPrice: [{ type: Input }],
    discountPrice: [{ type: Input }],
    numFormat: [{ type: Input }],
    purchaseDiscoutsConfig: [{ type: Input }],
    hasNodata: [{ type: Input }],
    helpOptions: [{ type: Input }],
    footerProtocol: [{ type: Input }],
    footerPrompt: [{ type: Input }],
    buttonText: [{ type: Input }],
    buttonDisabled: [{ type: Input }],
    total: [{ type: Input }],
    confirm: [{ type: Output }],
    clear: [{ type: Output }],
    discountChange: [{ type: Output }]
};
