// @cloud/tinyplus3/esm2015/components/purchaselist/TpPurchaselistModule.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { TiLocaleModule, TiCheckboxModule, TiButtonModule, TiOutlineModule } from '@cloud/tiny3';
import { TpLocaleModule } from '../../pipes/locale/TpLocaleModule';
import { TpPurchaselistComponent } from './TpPurchaselistComponent';
import { TpPurchaseitemComponent } from './TpPurchaseitemComponent';
import { TpPurchaseNodataComponent } from './TpPurchaseNodataComponent';
import { TpDiscountsModule } from '../discounts/TpDiscountsModule';
import { TpHelptipModule } from '../helptip/TpHelptipModule';
import { TpIconModule } from '../icon/TpIconModule';
export class TpPurchaselistModule {
}
TpPurchaselistModule.decorators = [
    { type: NgModule, args: [{
                imports: [TiButtonModule,
                    CommonModule,
                    FormsModule,
                    TpIconModule,
                    TpLocaleModule,
                    TiLocaleModule,
                    TpDiscountsModule,
                    TpHelptipModule,
                    TiOutlineModule,
                    TiCheckboxModule],
                exports: [TpPurchaselistComponent, TpPurchaseitemComponent, TpPurchaseNodataComponent],
                declarations: [TpPurchaselistComponent, TpPurchaseitemComponent, TpPurchaseNodataComponent]
            },] }
];
export { TpPurchaselistComponent } from './TpPurchaselistComponent';
export { TpPurchaseitemComponent } from './TpPurchaseitemComponent';
export { TpPurchaseNodataComponent } from './TpPurchaseNodataComponent';
