// @cloud/tiny3/esm2015/directives/tip/TiTipDirective.js
import { Directive, ElementRef, Input } from '@angular/core';
import { Util } from '../../utils/Util';
import { TiTipService } from '../../services/tip/TiTipService';
// TODO:exportAs: 'tiTip' 可去除？
/**
 * tip提供了两种使用方式:
 * 1.服务方式：[TiTipService]{@link ../injectables/TiTipService.html}
 *
 * 2.指令方式（见如下说明）
 *
 * <example-url>../tiny3demo/#/tip/tip-all</example-url>
 */
export class TiTipDirective {
    constructor(tipService, hostEleRef) {
        this.tipService = tipService;
        /**
         * Tip显示位置属性配置
         */
        this.tiTipPosition = 'auto';
        /**
         * 最大宽度
         *
         */
        this.tiTipMaxWidth = '276px';
        /**
         * 是否带箭头
         */
        this.tiTipHasArrow = true;
        /**
         * tip生成方式，默认支持鼠标移入移出时显示/隐藏
         *
         * 注意：指令形式时才会使用到该参数
         */
        this.tiTipTrigger = 'mouse';
        this.hostEle = hostEleRef.nativeElement;
    }
    static isInValidValue(value) {
        return Util.isUndefined(value) || Util.isNull(value) || value === '';
    }
    ngOnInit() {
        // 初始创建tip实例
        this.tipInstance = this.tipService.create(this.hostEle, {
            position: this.tiTipPosition,
            maxWidth: this.tiTipMaxWidth,
            hasArrow: this.tiTipHasArrow,
            hostEleX: this.tiTipHostEleX,
            trigger: this.tiTipTrigger || 'mouse',
            showFn: () => {
                if (!this.tipInstance || TiTipDirective.isInValidValue(this._tiTip)) {
                    return;
                }
                return { content: this._tiTip, context: this.tiTipContext };
            }
        });
    }
    /**
     * tip显示内容配置
     *
     * 类型：string | TemplateRef<any> | any
     *
     * 传入string类型时，在v3.0.6及之前的版本存在XSS攻击风险；v3.0.7做了安全处理后不存在XSS风险了，把传入的字符串当做纯文本解析，传入什么显示什么；
     * v10.1.4版本支持html字符串片段，也已做安全处理。
     */
    set tiTip(value) {
        this._tiTip = value;
        if (TiTipDirective.isInValidValue(value)) {
            this.hide();
        }
    }
    /**
     * @ignore
     */
    get tiTip() {
        return this._tiTip;
    }
    /**
     * 显示tip方法
     */
    show() {
        if (!this.tipInstance || TiTipDirective.isInValidValue(this._tiTip)) {
            return;
        }
        return this.tipInstance.show(this._tiTip, this.tiTipContext);
    }
    /**
     * 隐藏tip方法
     */
    hide() {
        if (this.tipInstance) {
            this.tipInstance.hide();
        }
    }
    // 宿主销毁，tip连带销毁
    ngOnDestroy() {
        this.tipInstance.hide();
        /* tslint:disable:no-null-keyword */
        this.tipInstance = null;
    }
}
TiTipDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tiTip]',
                exportAs: 'tiTip'
            },] }
];
TiTipDirective.ctorParameters = () => [
    { type: TiTipService },
    { type: ElementRef }
];
TiTipDirective.propDecorators = {
    tiTipPosition: [{ type: Input }],
    tiTipMaxWidth: [{ type: Input }],
    tiTipHasArrow: [{ type: Input }],
    tiTipContext: [{ type: Input }],
    tiTipTrigger: [{ type: Input }],
    tiTipHostEleX: [{ type: Input }],
    tiTip: [{ type: Input }]
};
