// @cloud/tinyplus3/esm2015/components/feedback/TpFeedbackModule.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { TiAlertModule, TiButtonModule, TiCheckboxModule, TiIconModule, TiLocaleModule, TiOutlineModule, TiTextareaModule, TiTextModule, TiValidationModule } from '@cloud/tiny3';
import { TpHalfmodalModule } from '../halfmodal/TpHalfmodalModule';
import { TpScoreModule } from '../score/TpScoreModule';
import { TpIconModule } from '../icon/TpIconModule';
import { TpFeedbackComponent } from './TpFeedbackComponent';
import { TpFeedbackContentComponent } from './TpFeedbackContentComponent';
export class TpFeedbackModule {
}
TpFeedbackModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    TiLocaleModule,
                    TiAlertModule,
                    TiButtonModule,
                    TiCheckboxModule,
                    TiIconModule,
                    TiTextareaModule,
                    TiTextModule,
                    TiValidationModule,
                    TiOutlineModule,
                    TpHalfmodalModule,
                    TpScoreModule,
                    TpIconModule
                ],
                exports: [TpFeedbackComponent, TpFeedbackContentComponent],
                declarations: [TpFeedbackComponent, TpFeedbackContentComponent],
            },] }
];
export { TpFeedbackComponent } from './TpFeedbackComponent';
export { TpFeedbackContentComponent } from './TpFeedbackContentComponent';
