// @cloud/tiny3/esm2015/components/progresspie/TiProgresspieComponent.js
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
/**
 * @ignore
 */
export class TiProgresspieComponent {
    constructor(hostEle) {
        this.hostEle = hostEle;
        this.defaultConfig = {
            value: 0,
            maxValue: 100,
            color: '#3dcca6',
            lineWidth: 2
        };
        this.value = this.defaultConfig.value;
        this.maxValue = this.defaultConfig.maxValue;
        this.nativeElement = this.hostEle.nativeElement;
    }
    ngOnChanges(changes) {
        // value和maxValue的动态修改均需要重绘进度呈现
        if ((changes.value && !changes.value.isFirstChange()) ||
            (changes.maxValue && !changes.maxValue.isFirstChange())) {
            this.draw();
        }
    }
    ngAfterViewInit() {
        this.canvasElement = this.canvasEle.nativeElement;
        // 修复SSR错误：ERROR ReferenceError: getComputedStyle is not defined
        if (typeof getComputedStyle === 'undefined') {
            return;
        }
        this.canvasElement.width = parseFloat(getComputedStyle(this.nativeElement).width);
        this.canvasElement.height = parseFloat(getComputedStyle(this.nativeElement).height);
        this.draw();
    }
    draw() {
        this.calcPercent();
        this.drawProgressPie();
    }
    // 计算百分比，外部需要保证maxVlue和value均为数字类型
    calcPercent() {
        // 计算percent值
        if (this.maxValue === 0) {
            this.percent = 0;
            return;
        }
        let percent = this.value / this.maxValue;
        if (isNaN(percent)) {
            this.percent = 0;
            return;
        }
        if (percent > 1) {
            percent = 1;
        }
        if (percent < 0) {
            percent = 0;
        }
        this.percent = percent;
    }
    drawProgressPie() {
        const canvas = this.canvasElement;
        const ctx = canvas.getContext('2d');
        // 清除先前画布内容
        const width = canvas.width;
        const height = canvas.height;
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        // 圆半径及起始点计算
        const d = Math.min(width, height);
        const cx = width / 2;
        const cy = height / 2;
        const lineWidth = this.defaultConfig.lineWidth;
        const r = d / 2 - lineWidth;
        const startPoint = -Math.PI / 2;
        // 笔触样式设置
        const color = this.defaultConfig.color;
        ctx.fillStyle = color;
        ctx.strokeStyle = color;
        ctx.lineWidth = lineWidth;
        // 画外圆
        ctx.beginPath();
        ctx.arc(cx, cy, r, startPoint, startPoint + Math.PI * 2);
        ctx.stroke();
        // 画扇形
        const endPoint = startPoint + Math.PI * this.percent * 2;
        ctx.beginPath();
        ctx.moveTo(cx, cy); // 移动至圆心
        ctx.arc(cx, cy, r, startPoint, endPoint); // 从外边上的点画曲线
        ctx.lineTo(cx, cy); // 从圆心画直线到计算好的圆外上的点
        ctx.fill();
        if (this.percent === 0) { // 为0的情况下，绘制圆心值边缘竖线
            ctx.stroke();
        }
        ctx.closePath();
    }
}
TiProgresspieComponent.decorators = [
    { type: Component, args: [{
                selector: 'ti-progresspie',
                template: '<canvas #canvas></canvas>'
            },] }
];
TiProgresspieComponent.ctorParameters = () => [
    { type: ElementRef }
];
TiProgresspieComponent.propDecorators = {
    value: [{ type: Input }],
    maxValue: [{ type: Input }],
    canvasEle: [{ type: ViewChild, args: ['canvas', { static: true },] }]
};
